<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Metacell Tutorial</title>
  <meta name="description" content="<p>This is a tutorial about metacell construction and analysis.</p>" />
  <meta name="generator" content="bookdown #bookdown:version# and GitBook 2.6.7" />

  <meta property="og:title" content="Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>This is a tutorial about metacell construction and analysis.</p>" />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Metacell Tutorial" />
  
  <meta name="twitter:description" content="<p>This is a tutorial about metacell construction and analysis.</p>" />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-07-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



<!--bookdown:title:start-->
<div id="header">
<h1 class="title">Metacell Tutorial</h1>
<p class="author"><em>Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller</em></p>
<p class="date"><em>2023-07-12</em></p>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#about" id="toc-about"><span class="toc-section-number">1</span> About</a>
<ul>
<li><a href="#book-structure" id="toc-book-structure"><span class="toc-section-number">1.1</span> Book structure</a></li>
<li><a href="#installation-and-requirements" id="toc-installation-and-requirements"><span class="toc-section-number">1.2</span> Installation and requirements</a></li>
<li><a href="#render-book" id="toc-render-book"><span class="toc-section-number">1.3</span> Render book</a></li>
</ul></li>
<li><a href="#the-metacell-concept" id="toc-the-metacell-concept"><span class="toc-section-number">2</span> The metacell concept</a>
<ul>
<li><a href="#MC2" id="toc-MC2"><span class="toc-section-number">2.1</span> Metacell (MC2)</a></li>
<li><a href="#SuperCell" id="toc-SuperCell"><span class="toc-section-number">2.2</span> SuperCell</a></li>
<li><a href="#SEACells" id="toc-SEACells"><span class="toc-section-number">2.3</span> SEACells</a></li>
</ul></li>
<li><a href="#constructing-metacells-for-a-discrete-data" id="toc-constructing-metacells-for-a-discrete-data"><span class="toc-section-number">3</span> Constructing metacells (for a ‘discrete’ data)</a>
<ul>
<li><a href="#mc2-python" id="toc-mc2-python"><span class="toc-section-number">3.1</span> MC2 (Python)</a>
<ul>
<li><a href="#running-mc2" id="toc-running-mc2">Running MC2</a></li>
<li><a href="#compute-latent-space-for-metacell-qc" id="toc-compute-latent-space-for-metacell-qc">Compute latent space for metacell QC</a></li>
<li><a href="#metacell-qc" id="toc-metacell-qc">Metacell QC</a></li>
</ul></li>
<li><a href="#supercell-r" id="toc-supercell-r"><span class="toc-section-number">3.2</span> SuperCell (R)</a></li>
<li><a href="#seacells-python" id="toc-seacells-python"><span class="toc-section-number">3.3</span> SEACells (Python)</a>
<ul>
<li><a href="#running-seacells" id="toc-running-seacells">Running SEACells</a></li>
<li><a href="#metacell-qc-1" id="toc-metacell-qc-1">Metacell QC</a></li>
<li><a href="#save-metacell-object-for-further-analysis" id="toc-save-metacell-object-for-further-analysis">Save metacell object for further analysis</a></li>
</ul></li>
</ul></li>
<li><a href="#downstream-analysis-of-metacells-for-a-discrete-dataset" id="toc-downstream-analysis-of-metacells-for-a-discrete-dataset"><span class="toc-section-number">4</span> Downstream analysis of metacells (for a discrete dataset)</a>
<ul>
<li><a href="#standard-analysis-r" id="toc-standard-analysis-r"><span class="toc-section-number">4.1</span> Standard analysis (R)</a>
<ul>
<li><a href="#dimensionality-reduction" id="toc-dimensionality-reduction"><span class="toc-section-number">4.1.1</span> Dimensionality reduction</a></li>
<li><a href="#clustering" id="toc-clustering"><span class="toc-section-number">4.1.2</span> Clustering</a></li>
<li><a href="#differential-expression-analysis" id="toc-differential-expression-analysis"><span class="toc-section-number">4.1.3</span> Differential expression analysis</a></li>
</ul></li>
<li><a href="#standard-analysis-Py" id="toc-standard-analysis-Py"><span class="toc-section-number">4.2</span> Standard analysis (Python)</a>
<ul>
<li><a href="#dimensionality-reduction-1" id="toc-dimensionality-reduction-1"><span class="toc-section-number">4.2.1</span> Dimensionality reduction</a></li>
<li><a href="#clustering-1" id="toc-clustering-1"><span class="toc-section-number">4.2.2</span> Clustering</a></li>
<li><a href="#differential-expression-analysis-1" id="toc-differential-expression-analysis-1"><span class="toc-section-number">4.2.3</span> Differential expression analysis</a></li>
</ul></li>
<li><a href="#advanced-analysis" id="toc-advanced-analysis"><span class="toc-section-number">4.3</span> Advanced analysis</a>
<ul>
<li><a href="#grn" id="toc-grn"><span class="toc-section-number">4.3.1</span> GRN</a></li>
</ul></li>
<li><a href="#weighted-analysis" id="toc-weighted-analysis"><span class="toc-section-number">4.4</span> Sample-weighted analysis</a></li>
</ul></li>
<li><a href="#MC-continuous" id="toc-MC-continuous"><span class="toc-section-number">5</span> Construction of metacells (for ‘continuous’ data)</a>
<ul>
<li><a href="#mc2-python-1" id="toc-mc2-python-1"><span class="toc-section-number">5.1</span> MC2 (Python)</a>
<ul>
<li><a href="#running-mc2-1" id="toc-running-mc2-1">Running MC2</a></li>
<li><a href="#compute-latent-space-for-metacell-qc-1" id="toc-compute-latent-space-for-metacell-qc-1">Compute latent space for metacell QC</a></li>
<li><a href="#metacell-qc-2" id="toc-metacell-qc-2">Metacell QC</a></li>
</ul></li>
<li><a href="#supercell-r-1" id="toc-supercell-r-1"><span class="toc-section-number">5.2</span> SuperCell (R)</a></li>
<li><a href="#seacells-python-1" id="toc-seacells-python-1"><span class="toc-section-number">5.3</span> SEACells (Python)</a></li>
</ul></li>
<li><a href="#downstream-analysis-of-metacells-for-a-continuous-dataset" id="toc-downstream-analysis-of-metacells-for-a-continuous-dataset"><span class="toc-section-number">6</span> Downstream analysis of metacells (for a continuous dataset)</a>
<ul>
<li><a href="#standard-analysis-r-1" id="toc-standard-analysis-r-1"><span class="toc-section-number">6.1</span> Standard analysis (R)</a>
<ul>
<li><a href="#dimensionality-reduction-2" id="toc-dimensionality-reduction-2"><span class="toc-section-number">6.1.1</span> Dimensionality reduction</a></li>
<li><a href="#clustering-2" id="toc-clustering-2"><span class="toc-section-number">6.1.2</span> Clustering</a></li>
<li><a href="#differential-expression-analysis-2" id="toc-differential-expression-analysis-2"><span class="toc-section-number">6.1.3</span> Differential expression analysis</a></li>
</ul></li>
<li><a href="#standard-analysis-Py" id="toc-standard-analysis-Py"><span class="toc-section-number">6.2</span> Standard analysis (Python)</a>
<ul>
<li><a href="#dimensionality-reduction-3" id="toc-dimensionality-reduction-3"><span class="toc-section-number">6.2.1</span> Dimensionality reduction</a></li>
<li><a href="#clustering-3" id="toc-clustering-3"><span class="toc-section-number">6.2.2</span> Clustering</a></li>
<li><a href="#differential-expression-analysis-3" id="toc-differential-expression-analysis-3"><span class="toc-section-number">6.2.3</span> Differential expression analysis</a></li>
</ul></li>
<li><a href="#weighted-analysis-cont" id="toc-weighted-analysis-cont"><span class="toc-section-number">6.3</span> Sample-weighted analysis</a></li>
</ul></li>
<li><a href="#comparison-of-metacells-usage-for-a-discrete-and-continuous-data" id="toc-comparison-of-metacells-usage-for-a-discrete-and-continuous-data"><span class="toc-section-number">7</span> Comparison of metacells usage for a discrete and continuous data</a>
<ul>
<li><a href="#continuous-metacells-have-lower-purity" id="toc-continuous-metacells-have-lower-purity"><span class="toc-section-number">7.1</span> Continuous metacells have lower purity</a></li>
</ul></li>
<li><a href="#data-integration-with-metacells" id="toc-data-integration-with-metacells"><span class="toc-section-number">8</span> Data integration with metacells</a></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<div id="about" class="section level1 hasAnchor" number="1">
<h1 class="hasAnchor"><span class="header-section-number">1</span> About<a href="#about" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The structure of this tutorial.</p>
<div id="book-structure" class="section level2 hasAnchor" number="1.1">
<h2 class="hasAnchor"><span class="header-section-number">1.1</span> Book structure<a href="#book-structure" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Book consists of several Chapters (i.e., first-level headings). Each chapter is in separate .Rmd file in the root folder, with a name in XY_text.Rmd format, with <code>XY</code> being numbers.</p>
<p>Each Chapter consists of sections and sup-sections (i.e., second-level and lower heading), files for which are located in <code>./sub_pages</code>. Sub-pages and chapters may also call <em>functional_chunks</em>, which are located in <code>./functional_chunks</code> and represent parts of code that can be repetitively run (e.g., <code>load_anndata</code>, <code>save_mc_anndata</code> etc). When the book is rendered, the included <em>sub_pages</em> and <em>functional_chunks</em> are basically inserted in the Chapter as inline code. The only challenge is the relative path of the files and resulting outputs, such as plots. To resolve this issue, currently, I manually set up the project folder as a knitting root directory in each sub-file (i.e., sub_pages and functional_chunks) as <code>knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())</code> . Also, in my RStudio settings, I have the following setting <code>Tools -&gt; Global Options... -&gt; R Markdown -&gt; Evaluate chunk in directory -&gt; Project</code>.</p>
<p><strong>Note:</strong> each chapters runs in a new R session and they do not share the environment, thus, we need to provide global knit options for each chapter, otherwise they are lost. I do it with a <code>source('./R/config.R')</code> in the beginning of each chapter.</p>
</div>
<div id="installation-and-requirements" class="section level2 hasAnchor" number="1.2">
<h2 class="hasAnchor"><span class="header-section-number">1.2</span> Installation and requirements<a href="#installation-and-requirements" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>R requirements</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&#39;rprojroot&#39;</span>) <span class="co"># to reset work directory to the Project root</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&#39;bookdown&#39;</span>) <span class="co"># to render book</span></span></code></pre></div>
<p>To run <strong>MC2</strong> and <strong>SEACells</strong> in RStudio, we need</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&#39;reticulate&#39;</span>) <span class="co"># to run Python</span></span></code></pre></div>
<p>Then, we need to setup virtual environment</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install virtualenv</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="op">&lt;</span>Path_to_Metacell_tutorial<span class="op">&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">virtualenv</span> my_env</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> my_env/bin/activate</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Installing SEACells, pip install installs old version, that does not work for me, thus install from git</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/dpeerlab/SEACells.git</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> SEACells</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> setup.py install</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install <span class="at">-r</span> SEACells_requirements.txt <span class="co"># here some packages have wrong/non-existing vision, so I manually changed their versions </span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install ipywidgets</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install jupyter</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install metacells</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># in project dir</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">&#39;RETICULATE_PYTHON=my_env/bin/python&#39;</span> <span class="op">&gt;</span> <span class="st">&#39;.Renviron&#39;</span> </span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># restart RStudio and open &#39;Metacell_tutorial.Rproj&#39; </span></span></code></pre></div>
</div>
<div id="render-book" class="section level2 hasAnchor" number="1.3">
<h2 class="hasAnchor"><span class="header-section-number">1.3</span> Render book<a href="#render-book" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The function to render book is <code>bookdown::render_book()</code>, this will take some time, as it will execute all the chunks in the book, there is an option to cache some chunks, but we have to make sure that cached chunks do not share variables with non-cached chunks (it will raise an error anyway).</p>
<p><code>bookdown::preview_chapter()</code> renders a chapter.</p>
<!--chapter:end:index.Rmd-->
</div>
</div>
<div id="the-metacell-concept" class="section level1 hasAnchor" number="2">
<h1 class="hasAnchor"><span class="header-section-number">2</span> The metacell concept<a href="#the-metacell-concept" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>See chapters @ref(MC2), @ref(SuperCell), @ref(SEACells)</p>
<div id="MC2" class="section level2 hasAnchor" number="2.1">
<h2 class="hasAnchor"><span class="header-section-number">2.1</span> Metacell (MC2)<a href="#MC2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
</div>
<div id="SuperCell" class="section level2 hasAnchor" number="2.2">
<h2 class="hasAnchor"><span class="header-section-number">2.2</span> SuperCell<a href="#SuperCell" class="anchor-section" aria-label="Anchor link to header"></a></h2>
</div>
<div id="SEACells" class="section level2 hasAnchor" number="2.3">
<h2 class="hasAnchor"><span class="header-section-number">2.3</span> SEACells<a href="#SEACells" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<!--chapter:end:10-intro.Rmd-->
</div>
</div>
<div id="constructing-metacells-for-a-discrete-data" class="section level1 hasAnchor" number="3">
<h1 class="hasAnchor"><span class="header-section-number">3</span> Constructing metacells (for a ‘discrete’ data)<a href="#constructing-metacells-for-a-discrete-data" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this chapter, we will demonstrate metacell construction using three different methods. MetaCell-2 (MC2) and SEACells in Pyhton and SuperCell in R.</p>
<p>For this, we will use a dataset of PBMCs from study. This dataset contains 30K cells and … This is an example of a complex dataset with well defined cells types. For an example of more continuous data, see chapter @ref(MC-continuous)</p>
<pre><code>#&gt; findfont: Font family [&#39;Raleway&#39;] not found. Falling back to DejaVu Sans.
#&gt; findfont: Font family [&#39;Lato&#39;] not found. Falling back to DejaVu Sans.</code></pre>
<div id="mc2-python" class="section level2 hasAnchor" number="3.1">
<h2 class="hasAnchor"><span class="header-section-number">3.1</span> MC2 (Python)<a href="#mc2-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here we construct metacells using <a href="https://github.com/tanaylab/metacells">Metacell-2 (MC2)</a>. The code is adapted from the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">author’s tutorial</a>.</p>
<div id="imports" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Imports<a href="#imports" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>) </span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span></code></pre></div>
</div>
<div id="parameters" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Parameters<a href="#parameters" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;MC2&quot;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>gamma   <span class="op">=</span> <span class="dv">50</span> <span class="co"># graining level</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we can modify dataset </span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> [<span class="st">&quot;cell_lines&quot;</span>, <span class="st">&quot;3k_pbmc&quot;</span>][<span class="dv">1</span>]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>annotation_label <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span>:<span class="st">&#39;cell_line&#39;</span>,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&#39;3k_pbmc&#39;</span>:<span class="st">&#39;louvain&#39;</span>}[proj_name] <span class="co"># name of annotation field (obs)</span></span></code></pre></div>
</div>
<div id="load-data" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Load data<a href="#load-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">## here code to download dataset and store it at f&#39;Metacell_tutorial/data/{proj_name}/singlecell_anndata_filtered.h5ad&#39;</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load pre-filtered data</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>data_folder <span class="op">=</span> os.path.join(<span class="st">&quot;./data/&quot;</span>, proj_name) <span class="co"># here path to folder with adata</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(data_folder, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span></code></pre></div>
<p><strong>Setup MC object</strong></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(ad, proj_name)</span></code></pre></div>
</div>
<div id="gene-filtering-according-to-mc2" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Gene filtering according to MC2<a href="#gene-filtering-according-to-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>excluded_gene_names <span class="op">=</span> [] <span class="co"># for example, [&#39;IGHMBP2&#39;, &#39;IGLL1&#39;, &#39;IGLL5&#39;, &#39;IGLON5&#39;, &#39;NEAT1&#39;, &#39;TMSB10&#39;, &#39;TMSB4X&#39;]</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>excluded_gene_patterns <span class="op">=</span> [<span class="st">&#39;MT-.*&#39;</span>]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_genes(ad,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_names<span class="op">=</span>excluded_gene_names,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_patterns<span class="op">=</span>excluded_gene_patterns,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>                          random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[properly_sampled_gene]: 16579 true (50.64%) out of 32738 bools</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[excluded_gene]: 13 true (0.03971%) out of 32738 bools</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[noisy_lonely_gene]: 0 true (0%) out of 32738 bools</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_genes(ad)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[clean_gene]: 16566 true (50.6%) out of 32738 bools</span></span></code></pre></div>
</div>
<div id="cell-fintering" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Cell fintering<a href="#cell-fintering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Since our data is pre-processed and low-quality cells have been already filtered out, this step is not applicable to our data, but we keep this chunk so that you can apply cell filtering to a newly generated dataset.</p>
<p>The first round of cell cleaning usually implies filltering out cell with very low and very hight UMI content. The second round includes cell filtering based on mitochondrial and/or ribosomal content. We will skip both steps as our data have been pre-filtered and will use very lenient cutoffs (<code>properly_sampled_min_cell_total</code>, <code>properly_sampled_max_cell_total</code> and <code>properly_sampled_max_excluded_genes_fraction</code>) such that all the cells are kept for the metacell construction.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">### The first round (high/low UMIs)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_min_cell_total <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span> : <span class="dv">5000</span>, <span class="st">&#39;3k_pbmc&#39;</span>: <span class="dv">200</span>}[proj_name] <span class="co"># setup for the dataset that will be used </span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_cell_total <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span> : <span class="dv">110000</span>, <span class="st">&#39;3k_pbmc&#39;</span>: <span class="dv">10000</span>}[proj_name] <span class="co"># setup for the dataset that will be used </span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>total_umis_of_cells <span class="op">=</span> mc.ut.get_o_numpy(ad, name<span class="op">=</span><span class="st">&#39;__x__&#39;</span>, <span class="bu">sum</span><span class="op">=</span><span class="va">True</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>df_umis <span class="op">=</span> pd.DataFrame(total_umis_of_cells, columns <span class="op">=</span> [<span class="st">&#39;total_UMIs&#39;</span>])</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>sns.histplot(data<span class="op">=</span>df_umis, x<span class="op">=</span><span class="st">&quot;total_UMIs&quot;</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">&#39;UMIs&#39;</span>)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">&#39;Density&#39;</span>)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>plt.axvline(x<span class="op">=</span>properly_sampled_min_cell_total, color<span class="op">=</span><span class="st">&#39;darkgreen&#39;</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>plt.axvline(x<span class="op">=</span>properly_sampled_max_cell_total, color<span class="op">=</span><span class="st">&#39;crimson&#39;</span>)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc-filtering-run-first-1.png" width="672" /></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>too_small_cells_count <span class="op">=</span> <span class="bu">sum</span>(total_umis_of_cells <span class="op">&lt;</span> properly_sampled_min_cell_total)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>too_large_cells_count <span class="op">=</span> <span class="bu">sum</span>(total_umis_of_cells <span class="op">&gt;</span> properly_sampled_max_cell_total)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>too_small_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_small_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>too_large_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_large_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Will exclude %s (%.2f%%) cells with less than %s UMIs&quot;</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_small_cells_count,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>         too_small_cells_percent,</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>         properly_sampled_min_cell_total))</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Will exclude 0 (0.00%) cells with less than 200 UMIs</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Will exclude %s (%.2f%%) cells with more than %s UMIs&quot;</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_large_cells_count,</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>         too_large_cells_percent,</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>         properly_sampled_max_cell_total))</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Will exclude 0 (0.00%) cells with more than 10000 UMIs</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The second round (content of non-clean genes, e.g., mito-genes)</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_excluded_genes_fraction <span class="op">=</span> <span class="fl">0.25</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>excluded_genes_data <span class="op">=</span> mc.tl.filter_data(ad, var_masks<span class="op">=</span>[<span class="st">&#39;~clean_gene&#39;</span>])[<span class="dv">0</span>]</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>excluded_umis_of_cells <span class="op">=</span> mc.ut.get_o_numpy(excluded_genes_data, name<span class="op">=</span><span class="st">&#39;__x__&#39;</span>, <span class="bu">sum</span><span class="op">=</span><span class="va">True</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>excluded_fraction_of_umis_of_cells <span class="op">=</span> excluded_umis_of_cells <span class="op">/</span> total_umis_of_cells</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>df_umis <span class="op">=</span> pd.DataFrame(excluded_fraction_of_umis_of_cells, columns <span class="op">=</span> [<span class="st">&#39;frac_excl_genes_UMIs&#39;</span>])</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>sns.histplot(df_umis, x <span class="op">=</span> <span class="st">&#39;frac_excl_genes_UMIs&#39;</span>)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">&#39;Fraction of excluded gene UMIs&#39;</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">&#39;Density&#39;</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>plt.axvline(x<span class="op">=</span>properly_sampled_max_excluded_genes_fraction, color<span class="op">=</span><span class="st">&#39;crimson&#39;</span>)</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc-cell-filtering-run-second-1.png" width="672" /></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>too_excluded_cells_count <span class="op">=</span> <span class="bu">sum</span>(excluded_fraction_of_umis_of_cells <span class="op">&gt;</span> properly_sampled_max_excluded_genes_fraction)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>too_excluded_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_excluded_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Will exclude %s (%.2f%%) cells with more than %.2f%% excluded gene UMIs&quot;</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_excluded_cells_count,</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>         too_excluded_cells_percent,</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>         <span class="fl">100.0</span> <span class="op">*</span> properly_sampled_max_excluded_genes_fraction))</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Will exclude 0 (0.00%) cells with more than 25.00% excluded gene UMIs</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_cells(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    properly_sampled_min_cell_total<span class="op">=</span>properly_sampled_min_cell_total,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_cell_total<span class="op">=</span>properly_sampled_max_cell_total,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_excluded_genes_fraction<span class="op">=</span>properly_sampled_max_excluded_genes_fraction</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[properly_sampled_cell]: 2638 true (100%) out of 2638 bools</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_cells(ad)</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[clean_cell]: 2638 true (100%) out of 2638 bools</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract clean dataset (with filtered cells and genes)</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> mc.pl.extract_clean_data(ad)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[full_cell_index]: 2638 int64s</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[full_gene_index]: 16566 int64s</span></span></code></pre></div>
</div>
<div id="running-mc2" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Running MC2<a href="#running-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Metacell-2 uses its own feature selection approach (i.e., selection of genes used to build metacells). Additionally, we can explicitly specify which features to use by providing two arguments: feature_gene_names - genes that have to be used forbidden_gene_names - genes to exclude.
In contrast to the SuperCell and SEACells, Metacell-2 does not allow to explicitly obtain metacell data at a user-defined graining level. Instead, to vary graining level, we have to vary a <code>target_metacell_size</code> parameter, that is <code>160000</code> by default. Here we provide a chunk to calibrate this value to get a desired graining level. Please, increase or decrease scale if the obtained graining level <code>gamma_obtained</code> is lower or larger than the requested one (gamma).</p>
<p><strong>Estimate target_metacell_size (gamma)</strong></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&#39;The requested graining level is </span><span class="sc">{</span>gamma<span class="sc">}</span><span class="ss">, lets estimate the target_metacell_size that should result in such graining level.&#39;</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The requested graining level is 50, lets estimate the target_metacell_size that should result in such graining level.</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>scale <span class="op">=</span> <span class="dv">2</span> <span class="co"># increase or decrease if the obtained graining level (`gamma_obtained`) is significantly &gt; or &lt; then the requested one `gamma`</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>N_c <span class="op">=</span> ad.shape[<span class="dv">0</span>]</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># estimated mean UMI content in downsampled data</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>est_downsample_UMI <span class="op">=</span> np.quantile(np.array(total_umis_of_cells), <span class="fl">0.05</span>)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>target_metacell_size <span class="op">=</span> <span class="bu">int</span>(est_downsample_UMI <span class="op">*</span> gamma <span class="op">*</span> scale)</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>target_metacell_size</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 101455</span></span></code></pre></div>
<div id="aggregate-mc2" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Aggregate metacells<a href="#aggregate-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb19"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>mc.pl.divide_and_conquer_pipeline(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#feature_gene_names   = feature_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#forbidden_gene_names = forbidden_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    target_metacell_size <span class="op">=</span> target_metacell_size,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene_module]: 16566 int32 elements with all outliers (100%)</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cells_rare_gene_module]: 2638 int32 elements with all outliers (100%)</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[rare_cell]: 0 true (0%) out of 2638 bools</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.layers[downsampled]: csr_matrix 2638 X 16566 float32s (1218892 &gt; 0)</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[downsample_samples]: 989</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_top3_gene]: 552 true (3.332%) out of 16566 bools</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_total_gene]: 4519 true (27.28%) out of 16566 bools</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_relative_variance_gene]: 3027 true (18.27%) out of 16566 bools</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[feature_gene]: 293 true (1.769%) out of 16566 bools</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obsp[obs_similarity]: ndarray 2638 X 2638 float32s</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obsp[obs_outgoing_weights]: csr_matrix 2638 X 2638 float32s (119842 &gt; 0)</span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[seed]: 491 outliers (18.61%) out of 2638 int32 elements with 60 groups with mean size 35.78</span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[candidate]: 0 outliers (0%) out of 2638 int32 elements with 63 groups with mean size 41.87</span></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[gene_deviant_votes]: 936 positive (5.65%) out of 16566 int32s</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cell_deviant_votes]: 899 positive (34.08%) out of 2638 int32s</span></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[dissolved]: 10 true (0.3791%) out of 2638 bools</span></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell]: 909 outliers (34.46%) out of 2638 int64 elements with 62 groups with mean size 27.89</span></span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[outlier]: 909 true (34.46%) out of 2638 bools</span></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[pre_directs]: 0</span></span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[directs]: 1</span></span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_high_total_gene]: * &lt;- 0</span></span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_total_gene]: 4519 positive (27.28%) out of 16566 int32s</span></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_high_relative_variance_gene]: * &lt;- 0</span></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_relative_variance_gene]: 3027 positive (18.27%) out of 16566 int32s</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[forbidden_gene]: * &lt;- False</span></span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_feature_gene]: * &lt;- 0</span></span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[feature_gene]: 293 positive (1.769%) out of 16566 int32s</span></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_gene_deviant_votes]: * &lt;- 0</span></span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_cell_directs]: * &lt;- 0</span></span>
<span id="cb19-37"><a href="#cb19-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cell_directs]: * &lt;- 0</span></span>
<span id="cb19-38"><a href="#cb19-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_pile]: * &lt;- -1</span></span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pile]: * &lt;- 0</span></span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_candidate]: * &lt;- -1</span></span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_cell_deviant_votes]: * &lt;- 0</span></span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_dissolved]: * &lt;- False</span></span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_metacell]: * &lt;- -1</span></span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a><span class="co">## make anndata of metacells</span></span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> mc.pl.collect_metacells(ad, name<span class="op">=</span><span class="st">&#39;metacells&#39;</span>)</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[excluded_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[clean_gene]: 16566 true (100%) out of 16566 bools</span></span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[forbidden_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[pre_feature_gene]: 0 positive (0%) out of 16566 int32s</span></span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[feature_gene]: 293 positive (1.769%) out of 16566 int32s</span></span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[pile]: [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ]</span></span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[candidate]: [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 15 ]</span></span></code></pre></div>
<p>Here we estimate whether a deviation of the obtained gamma is acceptable, and if not, suggest to increase or decrease <code>scale</code> parameter to get better graining level.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>gamma_obtained <span class="op">=</span> ad.shape[<span class="dv">0</span>]<span class="op">/</span>mc_ad.shape[<span class="dv">0</span>]</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gamma_obtained)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 42.54838709677419</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>gamma_dev <span class="op">=</span> (gamma_obtained <span class="op">-</span> gamma)<span class="op">/</span>gamma</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">abs</span>(gamma_dev) <span class="op">&lt;</span> <span class="fl">0.3</span>: </span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    gamma_dev <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> gamma_dev <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Increase `target_metacell_size` parameter by increasing `scale` and re-run metacell divide_and_conquer_pipeline() to get larger graining level&quot;</span>)</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Deacrease `target_metacell_size` parameter by decreasing `scale` and re-run metacell divide_and_conquer_pipeline() to get smaller graining level&quot;</span>)</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size` &quot;</span>)</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size`</span></span></code></pre></div>
<p>If the obtained graining level is not acceptable and you updated <code>scale</code> parameter according to suggestion, do not forget to re-run chunk @ref(chunk:mc2-divide-n-conquer) @ref(aggregate-mc2)</p>
</div>
<div id="visualize-metacells" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Visualize metacells<a href="#visualize-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb21"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>mc.pl.compute_umap_by_features(mc_ad, max_top_feature_genes<span class="op">=</span><span class="dv">1000</span>,</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>                               min_dist<span class="op">=</span><span class="fl">2.0</span>, random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[top_feature_gene]: 293 true (1.769%) out of 16566 bools</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[obs_balanced_ranks]: 762 nonzero (19.82%) out of 3844 elements</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[obs_pruned_ranks]: 257 nonzero (6.686%) out of 3844 elements</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[obs_outgoing_weights]: 257 nonzero (6.686%) out of 3844 elements</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[umap_distances]: csr_matrix 62 X 62 float32s (3782 &gt; 0)</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /Users/mariiabilous/Documents/PhD/UNIL/R/Metacell_tutorial/my_env_mc2/lib/python3.8/site-packages/umap/umap_.py:1356: RuntimeWarning: divide by zero encountered in power</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   return 1.0 / (1.0 + a * x ** (2 * b))</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /Users/mariiabilous/Documents/PhD/UNIL/R/Metacell_tutorial/my_env_mc2/lib/python3.8/site-packages/umap/umap_.py:1780: UserWarning: using precomputed metric; inverse_transform will be unavailable</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   warn(&quot;using precomputed metric; inverse_transform will be unavailable&quot;)</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[umap_x]: [ 1.6564819, 10.523616, 0.29960603, 12.346935, -18.181883, 10.824747, 9.3235655, 11.606001, 8.378801, 11.439771, -18.370312, 13.707805, -16.662771, 13.427194, 10.35691, 0.69431365, -16.371231, 8.772759, 8.827683, -15.460641, 10.793309, 4.4988446, -17.57337, 7.6907706, 14.141677, 8.847828, 10.136032, 8.403874, 9.716653, -19.049778, 3.3346436, 8.916034, -17.734402, -19.183643, 2.4466102, 9.458938, 9.972965, 14.90282, 14.809836, 8.051339, 11.39458, 13.204389, 9.030105, 13.247052, -19.935547, 12.671885, 13.710647, 11.188181, -16.07956, 12.613102, 12.006663, 2.273494, 10.232938, -16.919983, 12.528366, 14.898915, 11.630459, -13.0429125, -15.82163, -14.578497, -14.365959, -13.944745 ]</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[umap_y]: [ 7.753808, 5.679683, 8.05517, 3.7745268, 17.191807, 12.818527, 6.2553945, 17.103827, 4.138062, 11.021334, 20.70879, 13.585729, 20.566952, 12.627911, 18.061642, 6.2272234, 15.916855, 18.031816, 7.2213125, 19.455956, 14.636021, 8.450331, 15.314596, 6.2872787, 6.581571, 14.994367, 3.1421099, 8.773476, 4.545281, 15.859472, 7.0959816, 13.212819, 19.351728, 18.600061, 6.080823, 10.117547, 16.1381, 11.50319, 7.828978, 16.442448, 6.698089, 11.034152, 11.458855, 5.4966006, 17.60186, 14.4407425, 8.617028, 8.6199465, 18.102512, 9.883782, 12.338453, 9.131331, 8.351595, 17.561296, 7.718665, 10.21264, 4.989835, 14.793287, 13.923492, 15.697431, 13.76865, 16.49492 ]</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>                               </span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>umap_x <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_x&#39;</span>)</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>umap_y <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_y&#39;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(x<span class="op">=</span>umap_x, y<span class="op">=</span>umap_y)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc2-plot-umap-1.png" width="672" /></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a membership -- index of metacell single cell belongs to </span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&#39;membership&#39;</span>] <span class="op">=</span> [<span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">0</span> <span class="cf">else</span> np.nan <span class="cf">for</span> i <span class="kw">in</span> ad.obs.metacell] </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Save single-cell metadata (i.e., `raw.obs` dataframe) in the metacell adata object</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>mc_ad.uns <span class="op">=</span> ad.uns.copy()</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.obs&#39;</span>] <span class="op">=</span> ad.obs.copy()</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the requested gamma</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;gamma&#39;</span>] <span class="op">=</span> gamma</span></code></pre></div>
</div>
</div>
<div id="compute-latent-space-for-metacell-qc" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Compute latent space for metacell QC<a href="#compute-latent-space-for-metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>MC2 builds metacells from gene expression data, not from latent space. Some of QC metrics (e.g., <strong>compactness</strong> and <strong>separation</strong>) are computed from the latent space. Thus, to compute those metrics, we need to compute latent space.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save count as a separate layer</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>ad.layers[<span class="st">&#39;counts&#39;</span>] <span class="op">=</span> ad.X</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the counts to &quot;.raw&quot; attribute of the anndata since it is necessary for downstream analysis</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This step should be performed after filtering </span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(ad.layers[<span class="st">&#39;counts&#39;</span>])</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> ad.obs_names, ad.var_names</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>ad.raw <span class="op">=</span> raw_ad</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize cells, log transform and compute highly variable genes</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(ad)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ad)</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(ad, n_top_genes<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute principal components - </span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>n_comp    <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(ad, n_comps<span class="op">=</span>n_comp, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute UMAP for visualization </span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(ad, n_neighbors<span class="op">=</span><span class="dv">10</span>, n_pcs<span class="op">=</span>n_comp)</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(ad)</span></code></pre></div>
</div>
<div id="metacell-qc" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Metacell QC<a href="#metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
<div id="supercell-r" class="section level2 hasAnchor" number="3.2">
<h2 class="hasAnchor"><span class="header-section-number">3.2</span> SuperCell (R)<a href="#supercell-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
</div>
<div id="seacells-python" class="section level2 hasAnchor" number="3.3">
<h2 class="hasAnchor"><span class="header-section-number">3.3</span> SEACells (Python)<a href="#seacells-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Replace the following code with <code>./sub_pages/21-seacells.Rmd</code></p>
<p>Constructing metacells with <a href="https://github.com/dpeerlab/SEACells">SEACells</a>. The code is adapted from <a href="https://github.com/dpeerlab/SEACells/blob/main/notebooks/SEACell_computation.ipynb">the authors’ tutorial</a>.</p>
<p>SEACells work with <a href="https://anndata.readthedocs.io/en/latest/#">Anndata</a> objects and used <a href="https://scanpy.readthedocs.io/en/stable/">Scanpy</a> for pre-processing single-cell data and for the analysis of data at the metacell level (for the analysis, see section @ref(standard-analysis-Py)).</p>
<div id="imports-1" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Imports<a href="#imports-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb25"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> SEACells</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>) </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span></code></pre></div>
</div>
<div id="global-parameters" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Global parameters<a href="#global-parameters" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb26"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Parameters</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;SEACell&quot;</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>gamma <span class="op">=</span> <span class="dv">50</span>   <span class="co"># the requested graining level</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co">## Here we can modify dataset </span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> [<span class="st">&quot;cell_lines&quot;</span>, <span class="st">&quot;3k_pbmc&quot;</span>][<span class="dv">1</span>]</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>annotation_label <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span>:<span class="st">&#39;cell_line&#39;</span>,</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&#39;3k_pbmc&#39;</span>:<span class="st">&#39;louvain&#39;</span>}[proj_name]</span></code></pre></div>
</div>
<div id="load-data-1" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Load data<a href="#load-data-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">## here code to download dataset and store it at f&#39;Metacell_tutorial/data/{proj_name}/singlecell_anndata_filtered.h5ad&#39;</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load pre-filtered data</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>data_folder <span class="op">=</span> os.path.join(<span class="st">&quot;./data/&quot;</span>, proj_name) <span class="co"># here path to folder with adata</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(data_folder, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span></code></pre></div>
<p>Saving count layers to a raw attribute as raw counts will be used for metacell aggregation.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save count as a separate layer</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>ad.layers[<span class="st">&#39;counts&#39;</span>] <span class="op">=</span> ad.X</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the counts to &quot;.raw&quot; attribute of the anndata since it is necessary for downstream analysis</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This step should be performed after filtering </span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(ad.layers[<span class="st">&#39;counts&#39;</span>])</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> ad.obs_names, ad.var_names</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>ad.raw <span class="op">=</span> raw_ad</span></code></pre></div>
<p>Standard pre-processing for single-cell RNA-seq data with Scanpy, for more information, see <a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">the Scanpy tutorial</a>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize cells, log transform and compute highly variable genes</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(ad)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ad)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(ad, n_top_genes<span class="op">=</span><span class="dv">1000</span>)</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute principal components - </span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we use 10 components to be consistent with out main tutorial, but fill free to explore other number of principal components to use </span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>n_comp    <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(ad, n_comps<span class="op">=</span>n_comp, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute UMAP for visualization </span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(ad, n_neighbors<span class="op">=</span><span class="dv">10</span>, n_pcs<span class="op">=</span>n_comp)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(ad)</span></code></pre></div>
</div>
<div id="running-seacells" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Running SEACells<a href="#running-seacells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="setting-up-seacells-parameters" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Setting up SEACells parameters<a href="#setting-up-seacells-parameters" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb32"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co">## User defined parameters</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>build_kernel_on <span class="op">=</span> <span class="st">&#39;X_pca&#39;</span> <span class="co"># key in ad.obsm to use for computing metacells</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>                          <span class="co"># This would be replaced by &#39;X_svd&#39; for ATAC data</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Additional parameters</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>n_waypoint_eigs <span class="op">=</span> <span class="dv">10</span>      <span class="co"># Number of eigenvalues to consider when initializing metacells</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>n_iter <span class="op">=</span> <span class="dv">50</span></span></code></pre></div>
</div>
<div id="initialize-the-seacell-model" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Initialize the SEACell model<a href="#initialize-the-seacell-model" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb33"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed for reproducibility </span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>random.seed(<span class="dv">123</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># The number of SEACells is computed as a ratio between the number of single cells and the desired graining level</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>n_SEACells <span class="op">=</span> <span class="bu">int</span>(ad.shape[<span class="dv">0</span>]<span class="op">/</span>gamma) </span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> SEACells.core.SEACells(ad, </span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>                  build_kernel_on<span class="op">=</span>build_kernel_on, </span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>                  n_SEACells<span class="op">=</span>n_SEACells, </span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>                  n_waypoint_eigs<span class="op">=</span>n_waypoint_eigs,</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>                  convergence_epsilon <span class="op">=</span> <span class="fl">1e-5</span>,</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>                  verbose <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Welcome to SEACells!</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>model.construct_kernel_matrix()</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>M <span class="op">=</span> model.kernel_matrix</span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize archetypes</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>model.initialize_archetypes()</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Building kernel on X_pca</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing diffusion components from X_pca for waypoint initialization ... </span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Determing nearest neighbor graph...</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Sampling waypoints ...</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Selecting 42 cells from waypoint initialization.</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Initializing residual matrix using greedy column selection</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Initializing f and g...</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Selecting 10 cells from greedy initialization.</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span><span class="op">%|</span>          <span class="op">|</span> <span class="dv">0</span><span class="op">/</span><span class="dv">20</span> [<span class="dv">00</span>:<span class="dv">00</span><span class="op">&lt;</span>?, ?it<span class="op">/</span>s]</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a><span class="dv">100</span><span class="op">%|</span><span class="co">##########| 20/20 [00:00&lt;00:00, 463.72it/s]</span></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the initialization to ensure they are evenly spread</span></span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>SEACells.plot.plot_initialization(ad, model, plot_basis<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>) <span class="co">## error, mb missing some modules</span></span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/seacells-model-init-1.png" width="672" /></p>
</div>
<div id="fitting-model" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Fitting model<a href="#fitting-model" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb36"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co">## fit model</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>model.fit(min_iter <span class="op">=</span> <span class="dv">10</span>, max_iter <span class="op">=</span> n_iter)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Randomly initialized A matrix.</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Setting convergence threshold at 0.00088</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 1.</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 1.</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 10.</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 10.</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 20.</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 20.</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 30.</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 30.</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Converged after 37 iterations.</span></span></code></pre></div>
</div>
<div id="check-model-convergence" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Check model convergence<a href="#check-model-convergence" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb37"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check for convergence </span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>model.plot_convergence()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/seacells-convergence-3.png" width="672" /></p>
</div>
<div id="aggregate-metacells" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Aggregate metacells<a href="#aggregate-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><code>membership</code></p>
<div class="sourceCode" id="cb38"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>membership <span class="op">=</span> model.get_hard_assignments()</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co"># aggregate metacells    </span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> SEACells.core.summarize_by_SEACell(ad, SEACells_label<span class="op">=</span><span class="st">&#39;SEACell&#39;</span>, summarize_layer<span class="op">=</span><span class="st">&#39;raw&#39;</span>)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span><span class="op">%|</span>          <span class="op">|</span> <span class="dv">0</span><span class="op">/</span><span class="dv">52</span> [<span class="dv">00</span>:<span class="dv">00</span><span class="op">&lt;</span>?, ?it<span class="op">/</span>s]</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a> <span class="dv">79</span><span class="op">%|</span><span class="co">#######8  | 41/52 [00:00&lt;00:00, 406.97it/s]</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="dv">100</span><span class="op">%|</span><span class="co">##########| 52/52 [00:00&lt;00:00, 420.10it/s]</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make `membership` numeric</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>d <span class="op">=</span> {x: <span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">for</span> i, x <span class="kw">in</span> <span class="bu">enumerate</span>(mc_ad.obs_names)}</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>ad.obs.merge(membership)</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         n_genes  percent_mito  n_counts          louvain     SEACell</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0           781      0.030178    2421.0      CD4 T cells  SEACell-15</span></span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1           781      0.030178    2421.0      CD4 T cells  SEACell-15</span></span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2           781      0.030178    2421.0      CD4 T cells  SEACell-15</span></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3           781      0.030178    2421.0      CD4 T cells  SEACell-15</span></span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4           781      0.030178    2421.0      CD4 T cells  SEACell-15</span></span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ...         ...           ...       ...              ...         ...</span></span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 189387     1567      0.021160    5678.0  Dendritic cells  SEACell-48</span></span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 189388     1567      0.021160    5678.0  Dendritic cells  SEACell-48</span></span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 189389     1567      0.021160    5678.0  Dendritic cells  SEACell-48</span></span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 189390     1567      0.021160    5678.0  Dendritic cells  SEACell-48</span></span>
<span id="cb38-24"><a href="#cb38-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 189391     1567      0.021160    5678.0  Dendritic cells  SEACell-48</span></span>
<span id="cb38-25"><a href="#cb38-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb38-26"><a href="#cb38-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [189392 rows x 5 columns]</span></span>
<span id="cb38-27"><a href="#cb38-27" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&#39;membership&#39;</span>] <span class="op">=</span> [d[x] <span class="cf">for</span> x <span class="kw">in</span> membership.SEACell]</span></code></pre></div>
</div>
<div id="project-metacells-on-the-single-cell-umap" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Project metacells on the single-cell umap<a href="#project-metacells-on-the-single-cell-umap" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb39"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>SEACells.plot.plot_2D(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>)</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/seacells-umap-5.png" width="480" /></p>
</div>
</div>
<div id="metacell-qc-1" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Metacell QC<a href="#metacell-qc-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Size distribution</strong></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_size = SEACells.plot.plot_SEACell_sizes(ad, bins=20)</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs = pd.merge(mc_ad.obs, mc_size, left_index=True, right_index=True)</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs</span></span></code></pre></div>
<p><strong>Purity</strong> of metacells is a proportion of the most abundant cell type within metacell [ref SuperCell]</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>mc_purity <span class="op">=</span> mc_QC.purity(ad, annotation_label, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>mc_purity.head()</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     louvain  louvain_purity</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; membership                                 </span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1               CD4 T cells        0.553571</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2                   B cells        0.977778</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3               CD4 T cells        0.955882</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4           CD14+ Monocytes        0.943396</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5                  NK cells        1.000000</span></span></code></pre></div>
<p><strong>Compactness</strong> of metacells is variance of components within metacells [ref SEACells]</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>compactness <span class="op">=</span> mc_QC.compactness(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Compactness_PCA&#39;</span>, n_comp<span class="op">=</span>n_comp)[<span class="st">&#39;Compactness_PCA&#39;</span>]</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add compactness to metadata</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs <span class="op">=</span> mc_ad.obs.join(compactness)</span></code></pre></div>
<p><strong>Separation</strong> of metacells is a distance to a closest metacell [ref SEACells]</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>separation <span class="op">=</span> mc_QC.separation(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Separation_PCA&#39;</span>, n_comp<span class="op">=</span>n_comp)[<span class="st">&#39;Separation_PCA&#39;</span>]</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add separation to metadata</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs <span class="op">=</span> mc_ad.obs.join(separation)</span></code></pre></div>
<p><strong>Inner normalized variance (INV)</strong> of metacells is mean-normalized variance of gene expression within metacells [ref MC-2]</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>mc_INV <span class="op">=</span> mc_QC.mc_inner_normalized_var(ad, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>mc_INV_val <span class="op">=</span> mc_INV.quantile(<span class="fl">0.95</span>, axis<span class="op">=</span><span class="dv">1</span>, numeric_only<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>mc_INV_val <span class="op">=</span> pd.DataFrame(mc_INV_val.transpose()).set_axis([<span class="st">&#39;INV&#39;</span>], axis<span class="op">=</span><span class="dv">1</span>, inplace<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add INV to metadata</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>mc_ad.obs <span class="op">=</span> mc_ad.obs.join(mc_INV_val)</span></code></pre></div>
</div>
<div id="save-metacell-object-for-further-analysis" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Save metacell object for further analysis<a href="#save-metacell-object-for-further-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb45"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>mc_ad.write_h5ad(os.path.join(<span class="st">&#39;..&#39;</span>, <span class="st">&#39;data&#39;</span>, proj_name, <span class="ss">f&#39;metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad&#39;</span>))</span></code></pre></div>
<!--chapter:end:21-MC_construction_discrete.Rmd-->
</div>
</div>
</div>
<div id="downstream-analysis-of-metacells-for-a-discrete-dataset" class="section level1 hasAnchor" number="4">
<h1 class="hasAnchor"><span class="header-section-number">4</span> Downstream analysis of metacells (for a discrete dataset)<a href="#downstream-analysis-of-metacells-for-a-discrete-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Here we use the obtained metacell to run the downstream analysis on them instead of single-cell data. In this analysis, we treat metacell as single cell, neglecting information about their size (i.e., number of containing single cells). If you are interested in sample-weighted analysis, where metacell size is taken into account, see section @ref(weighted-analysis).</p>
<div id="standard-analysis-r" class="section level2 hasAnchor" number="4.1">
<h2 class="hasAnchor"><span class="header-section-number">4.1</span> Standard analysis (R)<a href="#standard-analysis-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Standard analysis includes dimensionality reduction, clustering, differential expression etc using Seurat [ref] framework.</p>
<div id="dimensionality-reduction" class="section level3 hasAnchor" number="4.1.1">
<h3 class="hasAnchor"><span class="header-section-number">4.1.1</span> Dimensionality reduction<a href="#dimensionality-reduction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="clustering" class="section level3 hasAnchor" number="4.1.2">
<h3 class="hasAnchor"><span class="header-section-number">4.1.2</span> Clustering<a href="#clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="differential-expression-analysis" class="section level3 hasAnchor" number="4.1.3">
<h3 class="hasAnchor"><span class="header-section-number">4.1.3</span> Differential expression analysis<a href="#differential-expression-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
<div id="standard-analysis-Py" class="section level2 hasAnchor" number="4.2">
<h2 class="hasAnchor"><span class="header-section-number">4.2</span> Standard analysis (Python)<a href="#standard-analysis-Py" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Standard analysis includes dimensionality reduction, clustering, differential expression etc using <a href="https://scanpy-tutorials.readthedocs.io/en/latest/#">Scanpy</a> framework.</p>
<div id="dimensionality-reduction-1" class="section level3 hasAnchor" number="4.2.1">
<h3 class="hasAnchor"><span class="header-section-number">4.2.1</span> Dimensionality reduction<a href="#dimensionality-reduction-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="clustering-1" class="section level3 hasAnchor" number="4.2.2">
<h3 class="hasAnchor"><span class="header-section-number">4.2.2</span> Clustering<a href="#clustering-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="differential-expression-analysis-1" class="section level3 hasAnchor" number="4.2.3">
<h3 class="hasAnchor"><span class="header-section-number">4.2.3</span> Differential expression analysis<a href="#differential-expression-analysis-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
<div id="advanced-analysis" class="section level2 hasAnchor" number="4.3">
<h2 class="hasAnchor"><span class="header-section-number">4.3</span> Advanced analysis<a href="#advanced-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="grn" class="section level3 hasAnchor" number="4.3.1">
<h3 class="hasAnchor"><span class="header-section-number">4.3.1</span> GRN<a href="#grn" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Gene correlation analysis suffers from large dropout rate of single-cell data and at the same time is very time and memory demanding. Metacells simultaneously adress both issues and thus are beneficial for gene co-expression and gene regulation analysis. Here we demonstrate usage of metacells for GRN analysis using SCENIC [ref].</p>
</div>
</div>
<div id="weighted-analysis" class="section level2 hasAnchor" number="4.4">
<h2 class="hasAnchor"><span class="header-section-number">4.4</span> Sample-weighted analysis<a href="#weighted-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the features of metacells are their size, which is a number of single cell it contains. Since metacells aggregate different number of cells, they also carry different amount of information. And thus, to better reproduce single-cell analysis, bigger metacells should have larger impact on the results than smaller metacells. For this, a sample-weighted analysis can be applied. Sample-weighted analysis is impleented in the SuperCell package.</p>
<!--chapter:end:22-MC_analysis_discrete.Rmd-->
</div>
</div>
<div id="MC-continuous" class="section level1 hasAnchor" number="5">
<h1 class="hasAnchor"><span class="header-section-number">5</span> Construction of metacells (for ‘continuous’ data)<a href="#MC-continuous" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<pre><code>#&gt; findfont: Font family [&#39;Raleway&#39;] not found. Falling back to DejaVu Sans.
#&gt; findfont: Font family [&#39;Lato&#39;] not found. Falling back to DejaVu Sans.</code></pre>
<div id="mc2-python-1" class="section level2 hasAnchor" number="5.1">
<h2 class="hasAnchor"><span class="header-section-number">5.1</span> MC2 (Python)<a href="#mc2-python-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>This is a template, not working code</strong></p>
<p>Here we construct metacells using <a href="https://github.com/tanaylab/metacells">Metacell-2 (MC2)</a>. The code is adapted from the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">author’s tutorial</a>.</p>
<div id="imports-2" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Imports<a href="#imports-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb47"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>) </span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span></code></pre></div>
</div>
<div id="parameters-1" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Parameters<a href="#parameters-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb48"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;MC2&quot;</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>gamma   <span class="op">=</span> <span class="dv">50</span> <span class="co"># graining level</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we can modify dataset </span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">&quot;CD34&quot;</span></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>annotation_label <span class="op">=</span> <span class="st">&quot;cell_type&quot;</span> <span class="co"># name of annotation field (obs)</span></span></code></pre></div>
</div>
<div id="load-data-2" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Load data<a href="#load-data-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Setup MC object</strong></p>
<div class="sourceCode" id="cb49"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(ad, proj_name)</span></code></pre></div>
</div>
<div id="gene-filtering-according-to-mc2-1" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Gene filtering according to MC2<a href="#gene-filtering-according-to-mc2-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb50"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>excluded_gene_names <span class="op">=</span> [] <span class="co"># for example, [&#39;IGHMBP2&#39;, &#39;IGLL1&#39;, &#39;IGLL5&#39;, &#39;IGLON5&#39;, &#39;NEAT1&#39;, &#39;TMSB10&#39;, &#39;TMSB4X&#39;]</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>excluded_gene_patterns <span class="op">=</span> [<span class="st">&#39;MT-.*&#39;</span>]</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_genes(ad,</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_names<span class="op">=</span>excluded_gene_names,</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_patterns<span class="op">=</span>excluded_gene_patterns,</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>                          random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_genes(ad)</span></code></pre></div>
</div>
<div id="cell-fintering-1" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Cell fintering<a href="#cell-fintering-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Since our data is pre-processed and low-quality cells have been already filtered out, this step is not applicable to our data, but we keep this chunk so that you can apply cell filtering to a newly generated dataset.</p>
<p>The first round of cell cleaning usually implies filltering out cell with very low and very hight UMI content. The second round includes cell filtering based on mitochondrial and/or ribosomal content. We will skip both steps as our data have been pre-filtered and will use very lenient cutoffs (<code>properly_sampled_min_cell_total</code>, <code>properly_sampled_max_cell_total</code> and <code>properly_sampled_max_excluded_genes_fraction</code>) such that all the cells are kept for the metacell construction.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">### The first round (high/low UMIs)</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_min_cell_total <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span> : <span class="dv">5000</span>, <span class="st">&#39;3k_pbmc&#39;</span>: <span class="dv">200</span>}[proj_name] <span class="co"># setup for the dataset that will be used </span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_cell_total <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span> : <span class="dv">110000</span>, <span class="st">&#39;3k_pbmc&#39;</span>: <span class="dv">10000</span>}[proj_name] <span class="co"># setup for the dataset that will be used </span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The second round (content of non-clean genes, e.g., mito-genes)</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_excluded_genes_fraction <span class="op">=</span> <span class="fl">0.25</span></span></code></pre></div>
</div>
<div id="running-mc2-1" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Running MC2<a href="#running-mc2-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Metacell-2 uses its own feature selection approach (i.e., selection of genes used to build metacells). Additionally, we can explicitly specify which features to use by providing two arguments: feature_gene_names - genes that have to be used forbidden_gene_names - genes to exclude.
In contrast to the SuperCell and SEACells, Metacell-2 does not allow to explicitly obtain metacell data at a user-defined graining level. Instead, to vary graining level, we have to vary a <code>target_metacell_size</code> parameter, that is <code>160000</code> by default. Here we provide a chunk to calibrate this value to get a desired graining level. Please, increase or decrease scale if the obtained graining level <code>gamma_obtained</code> is lower or larger than the requested one (gamma).</p>
<p><strong>Estimate target_metacell_size (gamma)</strong></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&#39;The requested graining level is </span><span class="sc">{</span>gamma<span class="sc">}</span><span class="ss">, lets estimate the target_metacell_size that should result in such graining level.&#39;</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>scale <span class="op">=</span> <span class="dv">2</span> <span class="co"># increase or decrease if the obtained graining level (`gamma_obtained`) is significantly &gt; or &lt; then the requested one `gamma`</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>N_c <span class="op">=</span> ad.shape[<span class="dv">0</span>]</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co"># estimated mean UMI content in downsampled data</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>est_downsample_UMI <span class="op">=</span> np.quantile(np.array(total_umis_of_cells), <span class="fl">0.05</span>)</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>target_metacell_size <span class="op">=</span> <span class="bu">int</span>(est_downsample_UMI <span class="op">*</span> gamma <span class="op">*</span> scale)</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>target_metacell_size</span></code></pre></div>
<div id="aggregate-mc2" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Aggregate metacells<a href="#aggregate-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb54"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>mc.pl.divide_and_conquer_pipeline(</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#feature_gene_names   = feature_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#forbidden_gene_names = forbidden_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>    target_metacell_size <span class="op">=</span> target_metacell_size,</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>    random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="co">## make anndata of metacells</span></span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> mc.pl.collect_metacells(ad, name<span class="op">=</span><span class="st">&#39;metacells&#39;</span>)</span></code></pre></div>
<p>Here we estimate whether a deviation of the obtained gamma is acceptable, and if not, suggest to increase or decrease <code>scale</code> parameter to get better graining level.</p>
<p>If the obtained graining level is not acceptable and you updated <code>scale</code> parameter according to suggestion, do not forget to re-run chunk @ref(chunk:mc2-divide-n-conquer) @ref(aggregate-mc2)</p>
</div>
<div id="visualize-metacells-1" class="section level4 unnumbered hasAnchor">
<h4 class="unnumbered hasAnchor">Visualize metacells<a href="#visualize-metacells-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb55"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>mc.pl.compute_umap_by_features(mc_ad, max_top_feature_genes<span class="op">=</span><span class="dv">1000</span>,</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>                               min_dist<span class="op">=</span><span class="fl">2.0</span>, random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>                               </span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>umap_x <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_x&#39;</span>)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>umap_y <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_y&#39;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(x<span class="op">=</span>umap_x, y<span class="op">=</span>umap_y)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a membership -- index of metacell single cell belongs to </span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&#39;membership&#39;</span>] <span class="op">=</span> [<span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">0</span> <span class="cf">else</span> np.nan <span class="cf">for</span> i <span class="kw">in</span> ad.obs.metacell] </span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Save single-cell metadata (i.e., `raw.obs` dataframe) in the metacell adata object</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>mc_ad.uns <span class="op">=</span> ad.uns.copy()</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.obs&#39;</span>] <span class="op">=</span> ad.obs.copy()</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the requested gamma</span></span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;gamma&#39;</span>] <span class="op">=</span> gamma</span></code></pre></div>
</div>
</div>
<div id="compute-latent-space-for-metacell-qc-1" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Compute latent space for metacell QC<a href="#compute-latent-space-for-metacell-qc-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="metacell-qc-2" class="section level3 unnumbered hasAnchor">
<h3 class="unnumbered hasAnchor">Metacell QC<a href="#metacell-qc-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
<div id="supercell-r-1" class="section level2 hasAnchor" number="5.2">
<h2 class="hasAnchor"><span class="header-section-number">5.2</span> SuperCell (R)<a href="#supercell-r-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
</div>
<div id="seacells-python-1" class="section level2 hasAnchor" number="5.3">
<h2 class="hasAnchor"><span class="header-section-number">5.3</span> SEACells (Python)<a href="#seacells-python-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<!--chapter:end:31-MC_construction_continuous.Rmd-->
</div>
</div>
<div id="downstream-analysis-of-metacells-for-a-continuous-dataset" class="section level1 hasAnchor" number="6">
<h1 class="hasAnchor"><span class="header-section-number">6</span> Downstream analysis of metacells (for a continuous dataset)<a href="#downstream-analysis-of-metacells-for-a-continuous-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Here we use the obtained metacell to run the downstream analysis on them instead of single-cell data. In this analysis, we treat metacell as single cell, neglecting information about their size (i.e., number of containing single cells). If you are interested in sample-weighted analysis, where metacell size is taken into account, see section @ref(weighted-analysis-cont).</p>
<div id="standard-analysis-r-1" class="section level2 hasAnchor" number="6.1">
<h2 class="hasAnchor"><span class="header-section-number">6.1</span> Standard analysis (R)<a href="#standard-analysis-r-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Standard analysis includes dimensionality reduction, clustering, differential expression etc using Seurat [ref] framework.</p>
<div id="dimensionality-reduction-2" class="section level3 hasAnchor" number="6.1.1">
<h3 class="hasAnchor"><span class="header-section-number">6.1.1</span> Dimensionality reduction<a href="#dimensionality-reduction-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="clustering-2" class="section level3 hasAnchor" number="6.1.2">
<h3 class="hasAnchor"><span class="header-section-number">6.1.2</span> Clustering<a href="#clustering-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="differential-expression-analysis-2" class="section level3 hasAnchor" number="6.1.3">
<h3 class="hasAnchor"><span class="header-section-number">6.1.3</span> Differential expression analysis<a href="#differential-expression-analysis-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
<div id="standard-analysis-Py" class="section level2 hasAnchor" number="6.2">
<h2 class="hasAnchor"><span class="header-section-number">6.2</span> Standard analysis (Python)<a href="#standard-analysis-Py" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Standard analysis includes dimensionality reduction, clustering, differential expression etc using <a href="https://scanpy-tutorials.readthedocs.io/en/latest/#">Scanpy</a> framework.</p>
<div id="dimensionality-reduction-3" class="section level3 hasAnchor" number="6.2.1">
<h3 class="hasAnchor"><span class="header-section-number">6.2.1</span> Dimensionality reduction<a href="#dimensionality-reduction-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="clustering-3" class="section level3 hasAnchor" number="6.2.2">
<h3 class="hasAnchor"><span class="header-section-number">6.2.2</span> Clustering<a href="#clustering-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="differential-expression-analysis-3" class="section level3 hasAnchor" number="6.2.3">
<h3 class="hasAnchor"><span class="header-section-number">6.2.3</span> Differential expression analysis<a href="#differential-expression-analysis-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
<div id="weighted-analysis-cont" class="section level2 hasAnchor" number="6.3">
<h2 class="hasAnchor"><span class="header-section-number">6.3</span> Sample-weighted analysis<a href="#weighted-analysis-cont" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the features of metacells are their size, which is a number of single cell it contains. Since metacells aggregate different number of cells, they also carry different amount of information. And thus, to better reproduce single-cell analysis, bigger metacells should have larger impact on the results than smaller metacells. For this, a sample-weighted analysis can be applied. Sample-weighted analysis is impleented in the SuperCell package.</p>
<!--chapter:end:32-MC_analysis_continuous.Rmd-->
</div>
</div>
<div id="comparison-of-metacells-usage-for-a-discrete-and-continuous-data" class="section level1 hasAnchor" number="7">
<h1 class="hasAnchor"><span class="header-section-number">7</span> Comparison of metacells usage for a discrete and continuous data<a href="#comparison-of-metacells-usage-for-a-discrete-and-continuous-data" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="continuous-metacells-have-lower-purity" class="section level2 hasAnchor" number="7.1">
<h2 class="hasAnchor"><span class="header-section-number">7.1</span> Continuous metacells have lower purity<a href="#continuous-metacells-have-lower-purity" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<!--chapter:end:40-Comparison_discrete_continuous_MC.Rmd-->
</div>
</div>
<div id="data-integration-with-metacells" class="section level1 hasAnchor" number="8">
<h1 class="hasAnchor"><span class="header-section-number">8</span> Data integration with metacells<a href="#data-integration-with-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<!--chapter:end:50-data_integration.Rmd-->
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->
    </div>
  </div>
<!--bookdown:config-->

</body>

</html>
