
[//]: # (Weighted downstream analysis of metacells using R (Seurat))

```{r, include=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

Code has been modified but text should be updated in this section!
```{r r-libraries-weighted, eval=TRUE}
library(Seurat) 
library(dplyr)
library(ggplot2)
library(SuperCell)
```

### Load metacell Seurat object

We will use Seurat objects containing the metacells counts data and their annotation (*e.g.* and cell-type annotation) and
proceed with standard Seurat downstream analyses.
Seurat objects containing metacells counts data and their annotation were generated at the end of sections .....
These objects can also be generated using the command line described in chapter ...

```{r r-mc-loading-weighted, eval=TRUE}
MC_tool = "SuperCell"
proj_name = "3k_pbmc"
annotation_column = "louvain"
MC.seurat = readRDS(paste0('./data/', proj_name, '/metacell_', MC_tool,'.rds'))
```


### Dimensionality reduction

As for single-cells, we normalize the raw counts (here aggregated raw counts) and we identify the most variable features in the metacells gene expression data.
Based on these features, we run PCA and use the first principal components to obtain a two dimensionnal representation of the data using UMAP.

```{r r-mc-dim-reduc-weighted, eval=TRUE}
MC.seurat <- NormalizeData(MC.seurat, normalization.method = "LogNormalize")

MC_list <- list(N.SC = ncol(MC.seurat),
                supercell_size = MC.seurat$size)
MC_list$PCA <- supercell_prcomp(
  Matrix::t(GetAssayData(MC.seurat, slot = "data")),
  genes.use = MC.seurat@misc$var_features,  # or a new set of HVG can be computed
  supercell_size = MC_list$supercell_size, # provide this parameter to run sample-weighted version of PCA,
  k = 10
)

MC_list$UMAP <- supercell_UMAP(
  SC = MC_list,
  PCA_name = "PCA",
  n_neighbors = 50 # large number to repel cells 
)

supercell_DimPlot(SC = MC_list, 
  groups = MC.seurat@meta.data[, annotation_column],
  dim.name = "UMAP", 
  title = paste0("UMAP of metacells colored by cell type assignment")
)

```

### Clustering

We cluster the metacells using Seurat clustering steps and visualize these clusters using UMAP:
```{r r-mc-clustering-weighted, eval=TRUE}
# compute distance among metacells
D  <- dist(MC_list$PCA$x)

# cluster metacells
MC_list$clustering  <- supercell_cluster(D = D, k = 8, supercell_size = MC_list$supercell_size)

# Plot clustering result
supercell_DimPlot(
  MC_list, 
  groups = factor(MC_list$clustering$clustering),
  dim.name = "UMAP", 
  title = paste0("UMAP of metacells colored by metacell clustering")
)
```

### Differential expression analysis

```{r r-mc-DEG-weighted, eval=TRUE}
# Compute upregulated genes in each cell line (versus other cells)
MC.all.markers <- supercell_FindAllMarkers(
  ge = GetAssayData(MC.seurat, slot = "data"), 
  clusters = MC.seurat@meta.data[, annotation_column], 
  supercell_size = MC_list$supercell_size,
  only.pos = TRUE, 
  min.pct = 0, 
  logfc.threshold = 0.2
)


```

We select the top markers for each cell-type: 
```{r r-mc-topMarkers-weighted, eval=TRUE}
# Transform the output of `supercell_FindAllMarkers()` to be in the format of the `Seurat::FindAllMarkers()`
MC.all.markers.df <- data.frame()
for(cl in names(MC.all.markers)){
  cur <- MC.all.markers[[cl]]
  
  if(is.data.frame(cur)){
    cur$cluster <- cl
    cur$gene <- rownames(cur)
    cur$avg_log2FC <- cur$logFC
    MC.all.markers.df <- rbind(MC.all.markers.df, cur)
  } 
}

# Top markers (select top markers of each cell line)
MC.top.markers <- MC.all.markers.df %>%
   group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```


We visualize the top 5 markers for the XX cells.
```{r r-mc-plot-genes-weighted, eval=TRUE}
Idents(MC.seurat) <- annotation_column
# genes.to.plot <- MC.seurat.top.markers$gene[MC.seurat.top.markers$cluster == unique(MC.seurat@meta.data[,annotation_column])[1]]
genes.to.plot <- MC.top.markers$gene[c(seq(1, 20, 5))]
VlnPlot(MC.seurat, features = genes.to.plot, ncol = 4, pt.size = 0.0)  
```


