[//]: # (This file runs standard preprocessing steps with scanpy)


To pre-process the single-cell data, we are using standard pre-processing for single-cell RNA-seq data using Scanpy. For more information, see [the Scanpy tutorial](https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html).

```{python data-normalization, message = FALSE, cache = TO_CACHE}
# Normalize cells, log transform and compute highly variable genes
sc.pp.normalize_per_cell(ad, 10000)
sc.pp.log1p(ad)
sc.pp.highly_variable_genes(ad, n_top_genes=2000)
```

```{python dim-reduction, cache = TO_CACHE, message = FALSE, warning = FALSE}
# Compute principal components - 

n_comp    = 50
sc.tl.pca(ad, n_comps=n_comp, use_highly_variable=True)

# Compute UMAP for visualization
# Here we use 30 components to be consistent with our main tutorial, but fill free to explore other number of principal components to use 

sc.pp.neighbors(ad, n_neighbors=30, n_pcs=50)
sc.tl.umap(ad)
```