<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6.4 Advanced analysis | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="6.4 Advanced analysis | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6.4 Advanced analysis | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-11-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="weighted-analysis.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> This tutorial</a></li>
<li class="chapter" data-level="2" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>2</b> Requirements</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>2.1</b> Installations</a></li>
<li class="chapter" data-level="2.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>2.2</b> Retrieve the discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="2.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>2.3</b> Retrieve the continuous dataset (CD34 dataset)</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>3</b> The metacell concept</a></li>
<li class="chapter" data-level="4" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>4</b> Constructing metacells (for ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="4.1" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>4.1</b> Command line tool</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>5</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>5.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>5.1.1</b> Purity</a></li>
<li class="chapter" data-level="5.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>5.1.2</b> Compactness</a></li>
<li class="chapter" data-level="5.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>5.1.3</b> Separation</a></li>
<li class="chapter" data-level="5.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>5.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>5.2</b> Size distribution</a></li>
<li class="chapter" data-level="5.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>5.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>6</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>6.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>6.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>6.1.3</b> Clustering</a></li>
<li class="chapter" data-level="6.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="6.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>6.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>6.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>6.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering-1"><i class="fa fa-check"></i><b>6.2.2</b> Clustering</a></li>
<li class="chapter" data-level="6.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>6.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>6.3</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>6.3.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.3.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>6.3.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.3.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-2"><i class="fa fa-check"></i><b>6.3.3</b> Clustering</a></li>
<li class="chapter" data-level="6.3.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-2"><i class="fa fa-check"></i><b>6.3.4</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>6.4</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#GRN-analysis"><i class="fa fa-check"></i><b>6.4.1</b> GRN</a></li>
<li class="chapter" data-level="6.4.2" data-path="advanced-analysis.html"><a href="advanced-analysis.html#integration"><i class="fa fa-check"></i><b>6.4.2</b> Integration of metacells</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="advanced-analysis" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> Advanced analysis<a href="advanced-analysis.html#advanced-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="GRN-analysis" class="section level3 hasAnchor" number="6.4.1">
<h3><span class="header-section-number">6.4.1</span> GRN<a href="advanced-analysis.html#GRN-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<!-- ```{r, child='./sub_pages/30-GRN.Rmd'} -->
<!-- ``` -->
</div>
<div id="integration" class="section level3 hasAnchor" number="6.4.2">
<h3><span class="header-section-number">6.4.2</span> Integration of metacells<a href="advanced-analysis.html#integration" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<!-- Under construction.... -->

<!-- # Comparison of metacells usage for a discrete and continuous data -->
<!-- Under construction... -->
<!-- ## Continuous metacells have lower purity -->
<!-- ```{r, child='./sub_pages/40-purity.Rmd'} -->
<!-- ``` -->

<!-- ```{r, eval = FALSE} -->
<!-- install.packages('reticulate') # to run Python -->
<!-- ``` -->
<!-- Then, we need to setup virtual environment -->
<!-- ```{bash, eval = FALSE} -->
<!-- pip install virtualenv -->
<!-- cd <Path_to_Metacell_tutorial> -->
<!-- virtualenv my_env -->
<!-- source my_env/bin/activate -->
<!-- # Installing SEACells, pip install installs old version, that does not work for me, thus install from git -->
<!-- git clone https://github.com/dpeerlab/SEACells.git -->
<!-- cd SEACells -->
<!-- python setup.py install -->
<!-- cd .. -->
<!-- pip install -r SEACells_requirements.txt  -->
<!-- pip install ipywidgets -->
<!-- pip install jupyter -->
<!-- # or pip install git+https://github.com/dpeerlab/SEACells -->
<!-- # Install MC2 -->
<!-- pip install git+https://github.com/tanaylab/metacells -->
<!-- # in project dir -->
<!-- echo 'RETICULATE_PYTHON=my_env/bin/python' > '.Renviron' -->
<!-- # restart RStudio and open 'Metacell_tutorial.Rproj' -->
<!-- ``` -->
<!-- ## Render book  -->
<!-- The function to render book is `bookdown::render_book()`, this will take some time, as it will execute all the chunks in the book, there is an option to cache some chunks, but we have to make sure that cached chunks do not share variables with non-cached chunks (it will raise an error anyway).  -->
<!-- `bookdown::preview_chapter()` renders a chapter. -->
<!-- ```{r, eval = T} -->
<!-- # library('reticulate') # to run Python -->
<!-- # use_python("/home/gabriela/miniconda3/envs/seacells/bin/python3.8") -->
<!-- ``` -->
<!-- ## Discrete data normalization  -->
<!-- We renormalize the data and and identify the most variable features. -->
<!-- ```{python, message = FALSE, cache = TO_CACHE} -->
<!-- # Normalize cells, log transform and compute highly variable genes -->
<!-- sc.pp.normalize_per_cell(adata) -->
<!-- sc.pp.log1p(adata) -->
<!-- sc.pp.highly_variable_genes(adata, n_top_genes=2000) -->
<!-- variable_genes=adata.var_names[adata.var.highly_variable].tolist() -->
<!-- ``` -->
<!-- We run PCA and UMAP to visualize the data. -->
<!-- Here we use 30 components to be consistent with the number of components used in the entire tutorial,  -->
<!-- but feel free to explore other number of principal components to use -->
<!-- ```{python dim-reduction, cache = TO_CACHE, message = FALSE, warning = FALSE} -->
<!-- n_comp    = 30 -->
<!-- sc.tl.pca(adata, n_comps=n_comp, use_highly_variable=True) -->
<!-- # Compute UMAP for visualization -->
<!-- sc.pp.neighbors(adata, n_neighbors=15, n_pcs=30) -->
<!-- sc.tl.umap(adata) -->
<!-- sc.pl.embedding(adata, 'X_umap', color='louvain') -->
<!-- ``` -->
<!-- ## Other options to get PBMC dataset with more cells: -->
<!-- ```{r, eval = T, collapse = TO_CACHE} -->
<!-- library(reticulate) -->
<!-- library(Seurat) -->
<!-- raw_counts <- Matrix::t(as(py$adata$raw$X, "CsparseMatrix")) -->
<!-- colnames(raw_counts) <- rownames(py$adata$obs) -->
<!-- rownames(raw_counts) <- rownames(py$adata$var) -->
<!-- # norm_counts <- Matrix::t(as(py$adata$X, "CsparseMatrix")) -->
<!-- # colnames(norm_counts) <- rownames(py$adata$obs) -->
<!-- # rownames(norm_counts) <- rownames(py$adata$var) -->
<!-- pbmc <- CreateSeuratObject(counts = raw_counts, meta.data = py$adata$obs) -->
<!-- # pbmc@assays$RNA@data <- norm_counts -->
<!-- # pbmc@misc[["sc.pca"]] <- py$adata$obsm$X_pca -->
<!-- # pbmc@misc[["sc.umap"]] <- py$adata$obsm$X_umap -->
<!-- # pbmc@misc[["variable_genes"]] <- py$variable_genes -->
<!-- saveRDS(pbmc, file = paste0("data/3k_pbmc/singlecell_seurat_filtered.rds")) -->
<!-- ``` -->
<!-- ## 10k PBMC -->
<!-- ```{bash, eval = FALSE, cache = TO_CACHE} -->
<!-- # Creat data folder and download demo data from 10X genomics -->
<!-- mkdir data/10k_pbmc/ -->
<!-- wget http://cf.10xgenomics.com/samples/cell-exp/3.0.0/pbmc_10k_v3/pbmc_10k_v3_filtered_feature_bc_matrix.tar.gz -O data/10k_pbmc/pbmc_10k_v3_filtered_feature_bc_matrix.tar.gz -->
<!-- cd data/10k_pbmc/; tar -xzf pbmc_10k_v3_filtered_feature_bc_matrix.tar.gz -->
<!-- ``` -->
<!-- ```{python, eval = T, collapse = T, cache = TO_CACHE} -->
<!-- import scanpy as sc  -->
<!-- import os -->
<!-- adata = sc.read_10x_mtx(os.path.join("data", "10k_pbmc", "filtered_feature_bc_matrix")) -->
<!-- adata.var_names_make_unique() -->
<!-- adata -->
<!-- adata.X.sort_indices() -->
<!-- # # Filter cells based on min nb of genes and filter genes baased on min nb of cells -->
<!-- # sc.pp.filter_cells(adata, min_genes=200) -->
<!-- # sc.pp.filter_genes(adata, min_cells=3) -->
<!-- # Filter out doublets  -->
<!-- # sc.external.pp.scrublet(adata) #estimates doublets -->
<!-- # adata = adata[adata.obs['predicted_doublet'] == False] #do the actual filtering -->
<!-- # adata -->
<!-- raw_ad = sc.AnnData(adata.X.copy()) -->
<!-- raw_ad.obs_names, raw_ad.var_names = adata.obs_names, adata.var_names -->
<!-- adata.raw = raw_ad -->
<!-- sc.pl.embedding(adata, 'X_umap', color='celltype') -->
<!-- directory = os.path.join("data", "10k_pbmc") -->
<!-- if not os.path.exists(directory): -->
<!--     os.makedirs(directory) -->
<!-- adata.write_h5ad(os.path.join("data", "10k_pbmc", "singlecell_anndata_filtered.h5ad")) -->
<!-- ``` -->
<!-- ```{r pbmc10k, cache = TO_CACHE, message = FALSE, warning = FALSE} -->
<!-- library(SeuratData) -->
<!-- # InstallData("pbmcMultiome") -->
<!-- pbmc.rna <- LoadData("pbmcMultiome", "pbmc.rna") -->
<!-- pbmc.rna <- pbmc.rna[,pbmc.rna$seurat_annotations != "filtered"]  -->
<!-- pbmc.rna@meta.data$seurat_annotations <- plyr::revalue(pbmc.rna@meta.data$seurat_annotations , -->
<!--                                                        c("CD8 TEM_1" = "CD8", -->
<!--                                                          "CD8 TEM_2" = "CD8", -->
<!--                                                          "Intermediate B" = "B", -->
<!--                                                          "Naive B" = "B", -->
<!--                                                          "Memory B" = "B", -->
<!--                                                          "CD4 Naive" = "CD4", "CD4 TCM" = "CD4",  -->
<!--                                                          "CD4 TEM" = "CD4", "Treg" = "CD4", -->
<!--                                                          "CD8 Naive" = "CD8" -->
<!--                                                          )) -->
<!-- # library(Seurat) -->
<!-- # pbmc.rna <- NormalizeData(pbmc.rna, normalization.method = "LogNormalize") -->
<!-- # pbmc.rna <- FindVariableFeatures(pbmc.rna, nfeatures = 2000) -->
<!-- # pbmc.rna <- ScaleData(pbmc.rna) -->
<!-- # pbmc.rna <- RunPCA(pbmc.rna, npcs = 30, verbose = F) -->
<!-- # pbmc.rna <- RunUMAP(pbmc.rna, reduction = "pca", dims = c(1:30), n.neighbors = 15, verbose = F) -->
<!-- # UMAPPlot(pbmc.rna, group.by = "seurat_annotations") -->
<!-- saveRDS(pbmc.rna, file = paste0("data/10k_pbmc/singlecell_seurat_filtered.rds")) -->
<!-- # create single-cell anndata to compute separation using the python function -->
<!-- pbmc.rna.ad <- anndata::AnnData( -->
<!--   X = Matrix::t(Seurat::GetAssayData(pbmc.rna, slot = "counts")), -->
<!--   obs = pbmc.rna@meta.data  -->
<!-- ) -->
<!-- anndata::write_h5ad(anndata = pbmc.rna.ad, filename = "data/10k_pbmc/singlecell_anndata_filtered.h5ad") -->
<!-- rm(pbmc.rna, pbmc.rna.ad) -->
<!-- ``` -->

<div id="refs" class="references csl-bib-body hanging-indent">
<div class="csl-entry">
Ben-Kiki, O., A. Bercovich, A. Lifshitz, et al. 2022. <span>“Metacell-2: A Divide-and-Conquer Metacell Algorithm for Scalable scRNA-Seq Analysis.”</span> <em>Genome Biology</em> 23: 100. <a href="https://doi.org/10.1186/s13059-022-02667-1">https://doi.org/10.1186/s13059-022-02667-1</a>.
</div>
<div class="csl-entry">
Bilous, M., L. Tran, C. Cianciaruso, et al. 2022. <span>“Metacells Untangle Large and Complex Single-Cell Transcriptome Networks.”</span> <em>BMC Bioinformatics</em> 23: 336. <a href="https://doi.org/10.1186/s12859-022-04861-1">https://doi.org/10.1186/s12859-022-04861-1</a>.
</div>
<div class="csl-entry">
Persad, Sitara, Zi-Ning Choo, Christine Dien, Ignas Masilionis, Ronan Chaligné, Tal Nawy, Chrysothemis C Brown, Itsik Pe’er, Manu Setty, and Dana Pe’er. 2022. <span>“SEACells: Inference of Transcriptional and Epigenomic Cellular States from Single-Cell Genomics Data.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2022.04.02.486748">https://doi.org/10.1101/2022.04.02.486748</a>.
</div>
</div>
</div>
</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="weighted-analysis.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/30-MC_analysis_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/30-MC_analysis_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
