<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>1.1 Retrieve the PBMCs data | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="1.1 Retrieve the PBMCs data | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="1.1 Retrieve the PBMCs data | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-09-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="the-metacell-concept.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="data.html"><a href="data.html"><i class="fa fa-check"></i><b>1.1</b> Retrieve the PBMCs data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a>
<ul>
<li class="chapter" data-level="2.1" data-path="SuperCell.html"><a href="SuperCell.html"><i class="fa fa-check"></i><b>2.1</b> SuperCell</a></li>
<li class="chapter" data-level="2.2" data-path="MC2.html"><a href="MC2.html"><i class="fa fa-check"></i><b>2.2</b> Metacell (MC2)</a></li>
<li class="chapter" data-level="2.3" data-path="SEACells.html"><a href="SEACells.html"><i class="fa fa-check"></i><b>2.3</b> SEACells</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells (for ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>3.1.1</b> Data loading</a></li>
<li class="chapter" data-level="3.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>3.1.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#visualize-the-supercell-graph"><i class="fa fa-check"></i><b>3.1.4</b> Visualize the SuperCell graph</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a></li>
<li class="chapter" data-level="3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="metacells-qcs.html"><a href="metacells-qcs.html"><i class="fa fa-check"></i><b>4</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="4.1" data-path="metrics.html"><a href="metrics.html"><i class="fa fa-check"></i><b>4.1</b> Metrics</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="metrics.html"><a href="metrics.html#purity"><i class="fa fa-check"></i><b>4.1.1</b> Purity</a></li>
<li class="chapter" data-level="4.1.2" data-path="metrics.html"><a href="metrics.html#compactness"><i class="fa fa-check"></i><b>4.1.2</b> Compactness</a></li>
<li class="chapter" data-level="4.1.3" data-path="metrics.html"><a href="metrics.html#separation"><i class="fa fa-check"></i><b>4.1.3</b> Separation</a></li>
<li class="chapter" data-level="4.1.4" data-path="metrics.html"><a href="metrics.html#inv"><i class="fa fa-check"></i><b>4.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="visualize-metacells-in-the-single-cell-space.html"><a href="visualize-metacells-in-the-single-cell-space.html"><i class="fa fa-check"></i><b>4.2</b> Visualize metacells in the single-cell space</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>5</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="5.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>5.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>5.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>5.1.3</b> Clustering</a></li>
<li class="chapter" data-level="5.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>5.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="5.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#plot-the-expression-of-some-markers"><i class="fa fa-check"></i><b>5.1.5</b> Plot the expression of some markers</a></li>
<li class="chapter" data-level="5.1.6" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#plot-gene-gene-correlation-at-single-cell-and-metacell-levels"><i class="fa fa-check"></i><b>5.1.6</b> Plot gene-gene correlation at single-cell and metacell levels</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>5.2</b> Standard analysis (Python)</a></li>
<li class="chapter" data-level="5.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>5.3</b> Sample-weighted analysis</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data" class="section level2 hasAnchor" number="1.1">
<h2><span class="header-section-number">1.1</span> Retrieve the PBMCs data<a href="data.html#data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<!-- ```{r, eval = T} -->
<!-- # library('reticulate') # to run Python -->
<!-- # use_python("/home/gabriela/miniconda3/envs/seacells/bin/python3.8") -->
<!-- ``` -->
<p>We retrieve the 3k PBMCs from scanpy datasets as follows :</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb1-1"><a href="data.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc </span>
<span id="cb1-2"><a href="data.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-3"><a href="data.html#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="data.html#cb1-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.datasets.pbmc3k()</span>
<span id="cb1-5"><a href="data.html#cb1-5" aria-hidden="true" tabindex="-1"></a>adata_proc <span class="op">=</span> sc.datasets.pbmc3k_processed()</span>
<span id="cb1-6"><a href="data.html#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="data.html#cb1-7" aria-hidden="true" tabindex="-1"></a>adata       <span class="op">=</span> adata[adata_proc.obs_names].copy()</span>
<span id="cb1-8"><a href="data.html#cb1-8" aria-hidden="true" tabindex="-1"></a>adata.obs   <span class="op">=</span> adata_proc.obs.copy()</span>
<span id="cb1-9"><a href="data.html#cb1-9" aria-hidden="true" tabindex="-1"></a>adata.uns   <span class="op">=</span> adata_proc.uns.copy()</span>
<span id="cb1-10"><a href="data.html#cb1-10" aria-hidden="true" tabindex="-1"></a>adata.obsm  <span class="op">=</span> adata_proc.obsm.copy()</span>
<span id="cb1-11"><a href="data.html#cb1-11" aria-hidden="true" tabindex="-1"></a>adata.obsp  <span class="op">=</span> adata_proc.obsp.copy()</span>
<span id="cb1-12"><a href="data.html#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="data.html#cb1-13" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(adata.X.copy())</span>
<span id="cb1-14"><a href="data.html#cb1-14" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> adata.obs_names, adata.var_names</span>
<span id="cb1-15"><a href="data.html#cb1-15" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> raw_ad</span>
<span id="cb1-16"><a href="data.html#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="data.html#cb1-17" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata, <span class="st">&#39;X_umap&#39;</span>, color<span class="op">=</span><span class="st">&#39;louvain&#39;</span>)</span>
<span id="cb1-18"><a href="data.html#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /mnt/c/Aurelie/postdoc_UNIL/Metacell_review/Metacell_tutorial/my_env/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored</span></span>
<span id="cb1-19"><a href="data.html#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   cax = scatter(</span></span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>The data are saved in the following file for future analyses in python: “data/3k_pbmc/singlecell_anndata_filtered.h5ad”.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb2-1"><a href="data.html#cb2-1" aria-hidden="true" tabindex="-1"></a>directory <span class="op">=</span> os.path.join(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;3k_pbmc&quot;</span>)</span>
<span id="cb2-2"><a href="data.html#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="data.html#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(directory):</span>
<span id="cb2-4"><a href="data.html#cb2-4" aria-hidden="true" tabindex="-1"></a>    os.makedirs(directory)</span>
<span id="cb2-5"><a href="data.html#cb2-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-6"><a href="data.html#cb2-6" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(os.path.join(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;3k_pbmc&quot;</span>, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span></code></pre></div>
<p>The data are saved in the following file for future analyses in R: “data/3k_pbmc/singlecell_seurat_filtered.rds”.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="data.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb3-2"><a href="data.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb3-3"><a href="data.html#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Attaching SeuratObject</span></span>
<span id="cb3-4"><a href="data.html#cb3-4" aria-hidden="true" tabindex="-1"></a>raw_counts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">t</span>(<span class="fu">as</span>(py<span class="sc">$</span>adata<span class="sc">$</span>raw<span class="sc">$</span>X, <span class="st">&quot;CsparseMatrix&quot;</span>))</span>
<span id="cb3-5"><a href="data.html#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(raw_counts) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(py<span class="sc">$</span>adata<span class="sc">$</span>obs)</span>
<span id="cb3-6"><a href="data.html#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(raw_counts) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(py<span class="sc">$</span>adata<span class="sc">$</span>var)</span>
<span id="cb3-7"><a href="data.html#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="data.html#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># norm_counts &lt;- Matrix::t(as(py$ad$X, &quot;CsparseMatrix&quot;))</span></span>
<span id="cb3-9"><a href="data.html#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># colnames(norm_counts) &lt;- rownames(py$ad$obs)</span></span>
<span id="cb3-10"><a href="data.html#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># rownames(norm_counts) &lt;- rownames(py$ad$var)</span></span>
<span id="cb3-11"><a href="data.html#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="data.html#cb3-12" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> raw_counts, <span class="at">meta.data =</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs)</span>
<span id="cb3-13"><a href="data.html#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes</span></span>
<span id="cb3-14"><a href="data.html#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; (&#39;-&#39;)</span></span>
<span id="cb3-15"><a href="data.html#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc@assays$RNA@data &lt;- norm_counts</span></span>
<span id="cb3-16"><a href="data.html#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pbmc, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&quot;data/3k_pbmc/singlecell_seurat_filtered.rds&quot;</span>))</span></code></pre></div>
<p>Save the CD34 dataset:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="data.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc </span>
<span id="cb4-2"><a href="data.html#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb4-3"><a href="data.html#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="data.html#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># MC_tool = &quot;MC2&quot;</span></span>
<span id="cb4-5"><a href="data.html#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># proj_name = &quot;CD34&quot;</span></span>
<span id="cb4-6"><a href="data.html#cb4-6" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read(os.path.join(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;CD34&quot;</span>, <span class="st">&quot;cd34_multiome_rna.h5ad&quot;</span>))</span>
<span id="cb4-7"><a href="data.html#cb4-7" aria-hidden="true" tabindex="-1"></a>adata.X.sort_indices()</span>
<span id="cb4-8"><a href="data.html#cb4-8" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(adata.X.copy())</span>
<span id="cb4-9"><a href="data.html#cb4-9" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> adata.obs_names, adata.var_names</span>
<span id="cb4-10"><a href="data.html#cb4-10" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> raw_ad</span>
<span id="cb4-11"><a href="data.html#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="data.html#cb4-12" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata, <span class="st">&#39;X_umap&#39;</span>, color<span class="op">=</span><span class="st">&#39;celltype&#39;</span>)</span>
<span id="cb4-13"><a href="data.html#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /mnt/c/Aurelie/postdoc_UNIL/Metacell_review/Metacell_tutorial/my_env/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored</span></span>
<span id="cb4-14"><a href="data.html#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   cax = scatter(</span></span></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="data.html#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="data.html#cb5-2" aria-hidden="true" tabindex="-1"></a>directory <span class="op">=</span> os.path.join(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;cd34_multiome&quot;</span>)</span>
<span id="cb5-3"><a href="data.html#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="data.html#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(directory):</span>
<span id="cb5-5"><a href="data.html#cb5-5" aria-hidden="true" tabindex="-1"></a>    os.makedirs(directory)</span>
<span id="cb5-6"><a href="data.html#cb5-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-7"><a href="data.html#cb5-7" aria-hidden="true" tabindex="-1"></a>adata.write_h5ad(os.path.join(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;cd34_multiome&quot;</span>, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="data.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb6-2"><a href="data.html#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb6-3"><a href="data.html#cb6-3" aria-hidden="true" tabindex="-1"></a>raw_counts <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">t</span>(<span class="fu">as</span>(py<span class="sc">$</span>adata<span class="sc">$</span>raw<span class="sc">$</span>X, <span class="st">&quot;CsparseMatrix&quot;</span>))</span>
<span id="cb6-4"><a href="data.html#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(raw_counts) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(py<span class="sc">$</span>adata<span class="sc">$</span>obs)</span>
<span id="cb6-5"><a href="data.html#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(raw_counts) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(py<span class="sc">$</span>adata<span class="sc">$</span>var)</span>
<span id="cb6-6"><a href="data.html#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="data.html#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co"># norm_counts &lt;- Matrix::t(as(py$ad$X, &quot;CsparseMatrix&quot;))</span></span>
<span id="cb6-8"><a href="data.html#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="co"># colnames(norm_counts) &lt;- rownames(py$ad$obs)</span></span>
<span id="cb6-9"><a href="data.html#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co"># rownames(norm_counts) &lt;- rownames(py$ad$var)</span></span>
<span id="cb6-10"><a href="data.html#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="data.html#cb6-11" aria-hidden="true" tabindex="-1"></a>cd34 <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> raw_counts, <span class="at">meta.data =</span> py<span class="sc">$</span>adata<span class="sc">$</span>obs)</span>
<span id="cb6-12"><a href="data.html#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># cd34@assays$RNA@data &lt;- norm_counts</span></span>
<span id="cb6-13"><a href="data.html#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(cd34, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&quot;data/cd34_multiome/singlecell_seurat_filtered.rds&quot;</span>))</span></code></pre></div>

</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="the-metacell-concept.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/index.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/index.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
