<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 3 Downstream analysis of metacells | Metacell Analysis Tutorial</title>
<meta name="author" content="Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller">
<meta name="description" content="In this chapter, we run standard and advanced downstream analyses on metacells instead of single-cell data. We will treat each metacell as a single cell, neglecting information about the size of...">
<meta name="generator" content="bookdown 0.37 with bs4_book()">
<meta property="og:title" content="Chapter 3 Downstream analysis of metacells | Metacell Analysis Tutorial">
<meta property="og:type" content="book">
<meta property="og:url" content="https://GfellerLab.github.io/MetacellAnalysisTutorial/downstream-analysis.html">
<meta property="og:description" content="In this chapter, we run standard and advanced downstream analyses on metacells instead of single-cell data. We will treat each metacell as a single cell, neglecting information about the size of...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 3 Downstream analysis of metacells | Metacell Analysis Tutorial">
<meta name="twitter:description" content="In this chapter, we run standard and advanced downstream analyses on metacells instead of single-cell data. We will treat each metacell as a single cell, neglecting information about the size of...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Metacell Analysis Tutorial</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">About this tutorial</a></li>
<li><a class="" href="requirements.html">Requirements</a></li>
<li><a class="" href="introduction-to-the-metacell-concept.html">Introduction to the metacell concept</a></li>
<li><a class="" href="Metacell-construction-chapter.html"><span class="header-section-number">1</span> Constructing metacells</a></li>
<li><a class="" href="QCs.html"><span class="header-section-number">2</span> Metacells’ QCs</a></li>
<li><a class="active" href="downstream-analysis.html"><span class="header-section-number">3</span> Downstream analysis of metacells</a></li>
<li><a class="" href="command-line.html"><span class="header-section-number">4</span> Metacell Analysis Toolkit (MATK)</a></li>
<li><a class="" href="integration.html"><span class="header-section-number">5</span> Integration of metacells</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/GfellerLab/MetacellAnalysisTutorial">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="downstream-analysis" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> Downstream analysis of metacells<a class="anchor" aria-label="anchor" href="#downstream-analysis"><i class="fas fa-link"></i></a>
</h1>
<p>In this chapter, we run standard and advanced downstream analyses on metacells instead of single-cell data.
We will treat each metacell as a single cell, neglecting information about the size of the metacell (i.e., number of containing single cells).
If you are interested in sample-weighted analysis, where metacell size is taken into account, see section <a href="downstream-analysis.html#weighted-analysis">3.3</a>.</p>
<div id="standard-analysis-R" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> Standard analysis (R)<a class="anchor" aria-label="anchor" href="#standard-analysis-R"><i class="fas fa-link"></i></a>
</h2>
<p>In this tutorial, standard analyses include dimensionality reduction, clustering and differential expression using the <a href="https://satijalab.org/seurat/">Seurat</a> framework.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning this package</span></span>
<span><span class="co">#&gt; will retire shortly. Please refer to R-spatial evolution reports on</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html for details.</span></span>
<span><span class="co">#&gt; This package is now running under evolution status 0</span></span>
<span><span class="co">#&gt; Attaching SeuratObject</span></span>
<span><span class="co"># If you have Seurat V5 installed, specify that you want to analyze Seurat V4 objects</span></span>
<span><span class="va">wilcox.test</span> <span class="op">&lt;-</span> <span class="st">"wilcox"</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span><span class="op">(</span><span class="st">"Seurat"</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>Seurat.object.assay.version <span class="op">=</span> <span class="st">"v4"</span><span class="op">)</span> </span>
<span>  <span class="va">wilcox.test</span> <span class="op">&lt;-</span> <span class="st">"wilcox_limma"</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"you are using seurat v5 with assay option v4"</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div id="load-metacell-seurat-object" class="section level3" number="3.1.1">
<h3>
<span class="header-section-number">3.1.1</span> Load metacell Seurat object<a class="anchor" aria-label="anchor" href="#load-metacell-seurat-object"><i class="fas fa-link"></i></a>
</h3>
<p>We will use Seurat objects containing the metacells counts data and their annotation (<em>e.g.</em> cell-type annotation) and
proceed with standard Seurat downstream analyses.
Seurat objects containing metacells counts data and their annotation were generated at the end of sections <a href="Metacell-construction-chapter.html#SuperCell-construction">1.1</a>
These objects can also be generated using the command line described in chapter <a href="command-line.html#command-line">4</a></p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC_tool</span> <span class="op">=</span> <span class="st">"SuperCell"</span></span>
<span><span class="va">proj_name</span> <span class="op">=</span> <span class="st">"bmcite"</span></span>
<span><span class="va">annotation_column</span> <span class="op">=</span> <span class="st">"celltype_simplified"</span></span>
<span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Prog_RBC"</span>, <span class="st">"Unconventional T"</span>, <span class="st">"Naive CD4 cell"</span>, <span class="st">"Non-Naive CD4 cell"</span>,</span>
<span>                <span class="st">"CD14 Mono"</span>, <span class="st">"B cell"</span>, <span class="st">"Naive CD8 cell"</span>, <span class="st">"Non-Naive CD8 cell"</span>,</span>
<span>                <span class="st">"NK"</span>, <span class="st">"GMP"</span>, <span class="st">"CD16 Mono"</span>, <span class="st">"pDC"</span>, <span class="st">"cDC2"</span>, <span class="st">"Prog_B 2"</span>,</span>
<span>                <span class="st">"Prog_Mk"</span>, <span class="st">"Plasmablast"</span>, <span class="st">"HSC"</span>, <span class="st">"LMPP"</span>, <span class="st">"Prog_DC"</span>, <span class="st">"Prog_B 1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">celltype_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#7E57C2"</span>, <span class="st">"#1E88E5"</span>, <span class="st">"#FFC107"</span>, <span class="st">"#004D40"</span>, <span class="st">"#9E9D24"</span>,</span>
<span>                 <span class="st">"#F06292"</span>, <span class="st">"#546E7A"</span>, <span class="st">"#D4E157"</span>, <span class="st">"#76FF03"</span>, <span class="st">"#6D4C41"</span>,</span>
<span>                 <span class="st">"#26A69A"</span>, <span class="st">"#AB47BC"</span>, <span class="st">"#EC407A"</span>, <span class="st">"#D81B60"</span>, <span class="st">"#42A5F5"</span>,</span>
<span>                 <span class="st">"#2E7D32"</span>, <span class="st">"#FFA726"</span>, <span class="st">"#5E35B1"</span>, <span class="st">"#EF5350"</span>, <span class="st">"#3949AB"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">celltype_colors</span><span class="op">)</span> <span class="op">&lt;-</span><span class="va">cell_types</span></span>
<span><span class="va">MC.seurat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">proj_name</span>, <span class="st">'/metacell_'</span>, <span class="va">MC_tool</span>,<span class="st">'.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="dimensionality-reduction" class="section level3" number="3.1.2">
<h3>
<span class="header-section-number">3.1.2</span> Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"><i class="fas fa-link"></i></a>
</h3>
<p>As for single-cells, we normalize the raw counts (here aggregated raw counts) and we identify the most variable features in the metacells gene expression data.
Based on these features, we run PCA and use the first principal components to obtain a two dimensional representation of the data using UMAP.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">annotation_column</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">MC.seurat</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">MC.seurat</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">MC.seurat</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">F</span>, min.dist <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Embeddings.html">Embeddings</a></span><span class="op">(</span><span class="va">MC.seurat</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">$</span><span class="va">size</span>,</span>
<span>                         cell_type <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_column</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"umap_1"</span>, <span class="st">"umap_2"</span><span class="op">)</span></span>
<span><span class="va">p_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span> <span class="va">umap_1</span>, y<span class="op">=</span><span class="va">umap_2</span>, color <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size<span class="op">=</span><span class="va">size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>,  <span class="fl">0.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">celltype_colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_annot</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/r-mc-dim-reduc-1.png" width="672"></div>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#DimPlot(MC.seurat, reduction = "umap", cols = celltype_colors, pt.size = log1p(MC.seurat$size))</span></span></code></pre></div>
</div>
<div id="clustering" class="section level3" number="3.1.3">
<h3>
<span class="header-section-number">3.1.3</span> Clustering<a class="anchor" aria-label="anchor" href="#clustering"><i class="fas fa-link"></i></a>
</h3>
<p>We cluster the metacells using Seurat clustering steps and visualize these clusters using UMAP:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC.seurat</span><span class="op">$</span><span class="va">SCclustering</span>  <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_cluster.html">supercell_cluster</a></span><span class="op">(</span>D <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span>  <span class="op">)</span>, k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">$</span><span class="va">clustering</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Embeddings.html">Embeddings</a></span><span class="op">(</span><span class="va">MC.seurat</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">$</span><span class="va">size</span>,</span>
<span>                         cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"C_"</span>,<span class="va">MC.seurat</span><span class="op">$</span><span class="va">SCclustering</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"umap_1"</span>, <span class="st">"umap_2"</span><span class="op">)</span></span>
<span><span class="va">p_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span> <span class="va">umap_1</span>, y<span class="op">=</span><span class="va">umap_2</span>, color <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size<span class="op">=</span><span class="va">size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span><span class="op">(</span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_cluster</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/r-mc-clustering-1.png" width="672"></div>
</div>
<div id="differential-expression-analysis" class="section level3" number="3.1.4">
<h3>
<span class="header-section-number">3.1.4</span> Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>We perform differential analysis to identify the markers of our cluster 11 as an example using the <code>FindMarkers</code> function.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set idents to metacell annotation</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCclustering"</span></span>
<span></span>
<span><span class="va">cells_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">MC.seurat</span>, ident.1 <span class="op">=</span> <span class="st">"11"</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, min.pct <span class="op">=</span> <span class="fl">0.1</span>, test.use <span class="op">=</span> <span class="va">wilcox.test</span><span class="op">)</span></span>
<span><span class="va">cells_markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">$</span><span class="va">avg_log2FC</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;              p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span><span class="co">#&gt; GNLY  4.624070e-27   4.620624 1.000 0.898 7.865080e-23</span></span>
<span><span class="co">#&gt; NKG7  3.623582e-26   3.796975 1.000 0.878 6.163350e-22</span></span>
<span><span class="co">#&gt; KLRF1 2.903180e-32   3.154920 1.000 0.406 4.938018e-28</span></span>
<span><span class="co">#&gt; KLRD1 7.406286e-32   3.011319 1.000 0.407 1.259735e-27</span></span>
<span><span class="co">#&gt; GZMB  1.220131e-24   3.010468 0.976 0.504 2.075321e-20</span></span>
<span><span class="co">#&gt; KLRB1 1.374561e-25   2.989229 1.000 0.698 2.337991e-21</span></span>
<span><span class="co">#&gt; CMC1  5.065132e-24   2.918483 1.000 0.846 8.615283e-20</span></span>
<span><span class="co">#&gt; PRF1  1.881760e-31   2.869718 1.000 0.398 3.200685e-27</span></span>
<span><span class="co">#&gt; CD7   2.407873e-27   2.806730 1.000 0.735 4.095552e-23</span></span>
<span><span class="co">#&gt; SPON2 4.903245e-24   2.763682 0.929 0.469 8.339929e-20</span></span></code></pre></div>
<p>We see that the top marker genes for this cluster contain the killer cell lectin-like receptor (KLR) family,
which is a group of transmembrane proteins preferentially expressed in NK cells.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"KLRF1"</span>, <span class="st">"KLRD1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">MC.seurat</span>, <span class="va">genes</span>, ncol <span class="op">=</span> <span class="fl">2</span>, pt.size <span class="op">=</span> <span class="fl">0.0</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/r-mc-plot-genes-1.png" width="672"></div>
<p>We can verify the identification of the NK cell cluster by comparing the metacell annotation and the metacell clustering.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_cluster</span> <span class="op">+</span> <span class="va">p_annot</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/unnamed-chunk-5-1.png" width="960"></div>
</div>
<div id="visualize-gene-gene-correlation" class="section level3" number="3.1.5">
<h3>
<span class="header-section-number">3.1.5</span> Visualize gene-gene correlation<a class="anchor" aria-label="anchor" href="#visualize-gene-gene-correlation"><i class="fas fa-link"></i></a>
</h3>
<p>We can use the <code>supercell_GeneGenePlot</code> function from the SuperCell package to visualize the correlation between marker genes of a cell-type:
(i) at the single-cell level and
(ii) at the metacell level.</p>
<p>For that, we load the single-cell data from which the metacells were derived from.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">proj_name</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "bmcite"</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/"</span>, <span class="va">proj_name</span>, <span class="st">"/singlecell_seurat_filtered.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">sc_data</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span></code></pre></div>
<p>We visualize gene-gene correlation at the single-cell level:</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells_markers</span> <span class="op">&lt;-</span> <span class="va">cells_markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">$</span><span class="va">avg_log2FC</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">gene_x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>  </span>
<span><span class="va">gene_y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">)</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.7</span></span>
<span></span>
<span><span class="va">p.sc</span> <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_GeneGenePlot.html">supercell_GeneGenePlot</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">sc_data</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>  gene_x <span class="op">=</span> <span class="va">gene_x</span>,</span>
<span>  gene_y <span class="op">=</span> <span class="va">gene_y</span>,</span>
<span>  clusters <span class="op">=</span> <span class="va">sc_data</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_column</span><span class="op">]</span>,</span>
<span>  sort.by.corr <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  alpha <span class="op">=</span> <span class="va">alpha</span>,</span>
<span>  color.use <span class="op">=</span> <span class="va">celltype_colors</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p.sc</span><span class="op">$</span><span class="va">p</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/r-sc-gene_gene_cor-1.png" width="768"></div>
<p>We visualize gene-gene correlation at the metacell level:</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p.MC</span> <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_GeneGenePlot.html">supercell_GeneGenePlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">MC.seurat</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>                                          gene_x <span class="op">=</span> <span class="va">gene_x</span>,</span>
<span>                                          gene_y <span class="op">=</span> <span class="va">gene_y</span>,</span>
<span>                                          clusters <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_column</span><span class="op">]</span>,</span>
<span>                                          sort.by.corr <span class="op">=</span> <span class="cn">F</span>, supercell_size <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">$</span><span class="va">size</span>,</span>
<span>                                          alpha <span class="op">=</span> <span class="va">alpha</span>,</span>
<span>                                          color.use <span class="op">=</span> <span class="va">celltype_colors</span><span class="op">)</span></span>
<span><span class="va">p.MC</span><span class="op">$</span><span class="va">p</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/r-mc-gene_gene_cors-1.png" width="672"></div>
</div>
</div>
<div id="standard-analysis-Py" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> Standard analysis (Python)<a class="anchor" aria-label="anchor" href="#standard-analysis-Py"><i class="fas fa-link"></i></a>
</h2>
<p>In this section, standard analysis includes dimensionality reduction, clustering, differential expression etc using <a href="https://scanpy-tutorials.readthedocs.io/en/latest/#">Scanpy</a> framework.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb86-1"><a href="downstream-analysis.html#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb86-2"><a href="downstream-analysis.html#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb86-3"><a href="downstream-analysis.html#cb86-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb86-4"><a href="downstream-analysis.html#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb86-5"><a href="downstream-analysis.html#cb86-5" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">1</span> </span></code></pre></div>
<div class="sourceCode" id="cb87"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb87-1"><a href="downstream-analysis.html#cb87-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">"SEACells"</span></span>
<span id="cb87-2"><a href="downstream-analysis.html#cb87-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">"bmcite"</span></span>
<span id="cb87-3"><a href="downstream-analysis.html#cb87-3" aria-hidden="true" tabindex="-1"></a>annotation_column <span class="op">=</span> <span class="st">"celltype_simplified"</span></span>
<span id="cb87-4"><a href="downstream-analysis.html#cb87-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read(os.path.join(<span class="st">'./data'</span>, proj_name, <span class="ss">f'metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad'</span>))</span></code></pre></div>
<div id="dimensionality-reduction-1" class="section level3" number="3.2.1">
<h3>
<span class="header-section-number">3.2.1</span> Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb88"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb88-1"><a href="downstream-analysis.html#cb88-1" aria-hidden="true" tabindex="-1"></a>adata.var_names_make_unique()  <span class="co"># this is unnecessary if using `var_names='gene_ids'` in `sc.read_10x_mtx`</span></span>
<span id="cb88-2"><a href="downstream-analysis.html#cb88-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-3"><a href="downstream-analysis.html#cb88-3" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(adata)</span>
<span id="cb88-4"><a href="downstream-analysis.html#cb88-4" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span>
<span id="cb88-5"><a href="downstream-analysis.html#cb88-5" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata, n_top_genes<span class="op">=</span><span class="dv">2000</span>)</span>
<span id="cb88-6"><a href="downstream-analysis.html#cb88-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-7"><a href="downstream-analysis.html#cb88-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Freeze the state of the AnnData object for later use in differential testing and visualizations of gene expression</span></span>
<span id="cb88-8"><a href="downstream-analysis.html#cb88-8" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> adata <span class="co"># step needd only if I use regress_out steps</span></span>
<span id="cb88-9"><a href="downstream-analysis.html#cb88-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-10"><a href="downstream-analysis.html#cb88-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute PCA (highly variable genes will be used)</span></span>
<span id="cb88-11"><a href="downstream-analysis.html#cb88-11" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(adata, svd_solver<span class="op">=</span><span class="st">'arpack'</span>)</span>
<span id="cb88-12"><a href="downstream-analysis.html#cb88-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-13"><a href="downstream-analysis.html#cb88-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the neighbor graph</span></span>
<span id="cb88-14"><a href="downstream-analysis.html#cb88-14" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, n_neighbors<span class="op">=</span><span class="dv">15</span>, n_pcs<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb88-15"><a href="downstream-analysis.html#cb88-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /opt/conda/envs/MetacellAnalysisToolkit/lib/python3.9/site-packages/umap/distances.py:1063: NumbaDeprecationWarning:  [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details. [0m</span></span>
<span id="cb88-16"><a href="downstream-analysis.html#cb88-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   @numba.jit()</span></span>
<span id="cb88-17"><a href="downstream-analysis.html#cb88-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /opt/conda/envs/MetacellAnalysisToolkit/lib/python3.9/site-packages/umap/distances.py:1071: NumbaDeprecationWarning:  [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details. [0m</span></span>
<span id="cb88-18"><a href="downstream-analysis.html#cb88-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   @numba.jit()</span></span>
<span id="cb88-19"><a href="downstream-analysis.html#cb88-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /opt/conda/envs/MetacellAnalysisToolkit/lib/python3.9/site-packages/umap/distances.py:1086: NumbaDeprecationWarning:  [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details. [0m</span></span>
<span id="cb88-20"><a href="downstream-analysis.html#cb88-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   @numba.jit()</span></span>
<span id="cb88-21"><a href="downstream-analysis.html#cb88-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /opt/conda/envs/MetacellAnalysisToolkit/lib/python3.9/site-packages/umap/umap_.py:660: NumbaDeprecationWarning:  [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details. [0m</span></span>
<span id="cb88-22"><a href="downstream-analysis.html#cb88-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   @numba.jit()</span></span>
<span id="cb88-23"><a href="downstream-analysis.html#cb88-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-24"><a href="downstream-analysis.html#cb88-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Run umap</span></span>
<span id="cb88-25"><a href="downstream-analysis.html#cb88-25" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata)</span>
<span id="cb88-26"><a href="downstream-analysis.html#cb88-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-27"><a href="downstream-analysis.html#cb88-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot metacells in the UMAP space</span></span>
<span id="cb88-28"><a href="downstream-analysis.html#cb88-28" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[annotation_column], size <span class="op">=</span> <span class="dv">100</span>)</span>
<span id="cb88-29"><a href="downstream-analysis.html#cb88-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /opt/conda/envs/MetacellAnalysisToolkit/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored</span></span>
<span id="cb88-30"><a href="downstream-analysis.html#cb88-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   cax = scatter(</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/py-mc-dim-reduc-1.png" width="672"></div>
</div>
<div id="clustering-1" class="section level3" number="3.2.2">
<h3>
<span class="header-section-number">3.2.2</span> Clustering<a class="anchor" aria-label="anchor" href="#clustering-1"><i class="fas fa-link"></i></a>
</h3>
<p>Perform clustering on the metacell data.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb89-1"><a href="downstream-analysis.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run laiden graph-clustering</span></span>
<span id="cb89-2"><a href="downstream-analysis.html#cb89-2" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, neighbors_key <span class="op">=</span> <span class="st">"neighbors"</span>, resolution <span class="op">=</span> <span class="dv">2</span>)</span>
<span id="cb89-3"><a href="downstream-analysis.html#cb89-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-4"><a href="downstream-analysis.html#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the metacells in the UMAP space and color by cluster</span></span>
<span id="cb89-5"><a href="downstream-analysis.html#cb89-5" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[<span class="st">'leiden'</span>], size <span class="op">=</span> <span class="dv">100</span>)</span>
<span id="cb89-6"><a href="downstream-analysis.html#cb89-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /opt/conda/envs/MetacellAnalysisToolkit/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored</span></span>
<span id="cb89-7"><a href="downstream-analysis.html#cb89-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   cax = scatter(</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/py-mc-clustering-3.png" width="672"></div>
</div>
<div id="differential-expression-analysis-1" class="section level3" number="3.2.3">
<h3>
<span class="header-section-number">3.2.3</span> Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-1"><i class="fas fa-link"></i></a>
</h3>
<p>Identify marker genes for each group of metacells.
We visualize top markers for NK Tcells (cluster 6).</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb90-1"><a href="downstream-analysis.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify marker genes</span></span>
<span id="cb90-2"><a href="downstream-analysis.html#cb90-2" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, <span class="st">'leiden'</span>, method<span class="op">=</span><span class="st">'wilcoxon'</span>)</span>
<span id="cb90-3"><a href="downstream-analysis.html#cb90-3" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups(adata, n_genes<span class="op">=</span><span class="dv">25</span>, sharey<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb90-4"><a href="downstream-analysis.html#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /opt/conda/envs/MetacellAnalysisToolkit/lib/python3.9/site-packages/scanpy/plotting/_tools/__init__.py:397: UserWarning: Attempting to set identical low and high ylims makes transformation singular; automatically expanding.</span></span>
<span id="cb90-5"><a href="downstream-analysis.html#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   ax.set_ylim(ymin, ymax)</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/py-mc-markers-5.png" width="2688"></div>
<div class="sourceCode" id="cb91"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb91-1"><a href="downstream-analysis.html#cb91-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-2"><a href="downstream-analysis.html#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the top marker genes</span></span>
<span id="cb91-3"><a href="downstream-analysis.html#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pd.DataFrame(adata.uns[<span class="st">'rank_genes_groups'</span>][<span class="st">'names'</span>]).head(<span class="dv">5</span>))</span>
<span id="cb91-4"><a href="downstream-analysis.html#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;          0        1        2      3  ...        7               8         9      10</span></span>
<span id="cb91-5"><a href="downstream-analysis.html#cb91-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0    TPD52  PIK3IP1   S100A8   LAG3  ...     PKIB           TRADD     PILRA    SPIB</span></span>
<span id="cb91-6"><a href="downstream-analysis.html#cb91-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1    CD79A     LEF1     MNDA  TRGC2  ...    NDRG2           CLIC5    STXBP2    CLN8</span></span>
<span id="cb91-7"><a href="downstream-analysis.html#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2  POU2AF1    SARAF  C14orf2   LYAR  ...     ENHO  RP11-1399P15.1    CDKN1C   PTPRS</span></span>
<span id="cb91-8"><a href="downstream-analysis.html#cb91-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3     LCN8     CCR7    APLP2  DUSP2  ...    GPAT3            AQP3  C19orf38  BCL11A</span></span>
<span id="cb91-9"><a href="downstream-analysis.html#cb91-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4  RASGRP3  FAM134B    MARC1   IL32  ...  CLEC10A         TNFRSF4     COTL1    IRF8</span></span>
<span id="cb91-10"><a href="downstream-analysis.html#cb91-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb91-11"><a href="downstream-analysis.html#cb91-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5 rows x 11 columns]</span></span>
<span id="cb91-12"><a href="downstream-analysis.html#cb91-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-13"><a href="downstream-analysis.html#cb91-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize marker genes</span></span>
<span id="cb91-14"><a href="downstream-analysis.html#cb91-14" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'KLRF1'</span>, <span class="st">'IL2RB'</span>, <span class="st">'GNLY'</span>], groupby<span class="op">=</span>annotation_column, size <span class="op">=</span> <span class="dv">2</span>, rotation <span class="op">=</span> <span class="dv">90</span>)</span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/py-mc-markers-6.png" width="2564"></div>
<div class="sourceCode" id="cb92"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb92-1"><a href="downstream-analysis.html#cb92-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">'KLRF1'</span>, <span class="st">'IL2RB'</span>, <span class="st">'GNLY'</span>], groupby<span class="op">=</span><span class="st">'leiden'</span>, size <span class="op">=</span> <span class="dv">2</span>, rotation <span class="op">=</span> <span class="dv">90</span>)</span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/py-mc-markers-7.png" width="2564"></div>
</div>
</div>
<div id="weighted-analysis" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> Sample-weighted analysis<a class="anchor" aria-label="anchor" href="#weighted-analysis"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="co"># If you have Seurat V5 installed, specify that you want to analyze Seurat V4 objects</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span><span class="op">(</span><span class="st">"Seurat"</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>Seurat.object.assay.version <span class="op">=</span> <span class="st">"v4"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"you are using seurat v5 with assay option v4"</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SuperCell</span><span class="op">)</span></span></code></pre></div>
<div id="load-metacell-seurat-object-1" class="section level3" number="3.3.1">
<h3>
<span class="header-section-number">3.3.1</span> Load metacell Seurat object<a class="anchor" aria-label="anchor" href="#load-metacell-seurat-object-1"><i class="fas fa-link"></i></a>
</h3>
<p>We will use Seurat objects containing the metacells counts data and their annotation (<em>e.g.</em> and cell-type annotation) and
proceed with downstream analyses considering the size of each metacells.
Seurat objects containing metacells counts data and their annotation were generated at the end of sections <a href="Metacell-construction-chapter.html#SuperCell-construction">1.1</a>
These objects can also be generated using the command line described in chapter <a href="command-line.html#command-line">4</a></p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC_tool</span> <span class="op">=</span> <span class="st">"SuperCell"</span></span>
<span><span class="va">proj_name</span> <span class="op">=</span> <span class="st">"bmcite"</span></span>
<span><span class="va">annotation_column</span> <span class="op">=</span> <span class="st">"celltype_simplified"</span></span>
<span></span>
<span><span class="va">cell_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Prog_RBC"</span>, <span class="st">"Unconventional T"</span>, <span class="st">"Naive CD4 cell"</span>, <span class="st">"Non-Naive CD4 cell"</span>,</span>
<span>                <span class="st">"CD14 Mono"</span>, <span class="st">"B cell"</span>, <span class="st">"Naive CD8 cell"</span>, <span class="st">"Non-Naive CD8 cell"</span>,</span>
<span>                <span class="st">"NK"</span>, <span class="st">"GMP"</span>, <span class="st">"CD16 Mono"</span>, <span class="st">"pDC"</span>, <span class="st">"cDC2"</span>, <span class="st">"Prog_B 2"</span>,</span>
<span>                <span class="st">"Prog_Mk"</span>, <span class="st">"Plasmablast"</span>, <span class="st">"HSC"</span>, <span class="st">"LMPP"</span>, <span class="st">"Prog_DC"</span>, <span class="st">"Prog_B 1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">celltype_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#7E57C2"</span>, <span class="st">"#1E88E5"</span>, <span class="st">"#FFC107"</span>, <span class="st">"#004D40"</span>, <span class="st">"#9E9D24"</span>,</span>
<span>                 <span class="st">"#F06292"</span>, <span class="st">"#546E7A"</span>, <span class="st">"#D4E157"</span>, <span class="st">"#76FF03"</span>, <span class="st">"#6D4C41"</span>,</span>
<span>                 <span class="st">"#26A69A"</span>, <span class="st">"#AB47BC"</span>, <span class="st">"#EC407A"</span>, <span class="st">"#D81B60"</span>, <span class="st">"#42A5F5"</span>,</span>
<span>                 <span class="st">"#2E7D32"</span>, <span class="st">"#FFA726"</span>, <span class="st">"#5E35B1"</span>, <span class="st">"#EF5350"</span>, <span class="st">"#3949AB"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">celltype_colors</span><span class="op">)</span> <span class="op">&lt;-</span><span class="va">cell_types</span></span>
<span><span class="va">MC.seurat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">proj_name</span>, <span class="st">'/metacell_'</span>, <span class="va">MC_tool</span>,<span class="st">'.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="dimensionality-reduction-2" class="section level3" number="3.3.2">
<h3>
<span class="header-section-number">3.3.2</span> Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-2"><i class="fas fa-link"></i></a>
</h3>
<p>As for single-cells, we normalize the raw counts (here aggregated raw counts) and we identify the most variable features in the metacells gene expression data.
Based on these features, we run a sample weighted PCA using the function <code>supercell_prcomp</code> from the SuperCell R package
and use the first principal components to obtain a two dimensionnal representation of the data using UMAP.
Using the <code>supercell_DimPlot</code> function from the the SuperCell R package we can visualize the metacells and their sized in UMAP space.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">MC.seurat</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">MC_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>N.SC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span>,</span>
<span>                supercell_size <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">$</span><span class="va">size</span><span class="op">)</span></span>
<span><span class="va">MC_list</span><span class="op">$</span><span class="va">PCA</span> <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_prcomp.html">supercell_prcomp</a></span><span class="op">(</span></span>
<span>  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">MC.seurat</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  genes.use <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">var_features</span>,  <span class="co"># or a new set of HVG can be computed</span></span>
<span>  supercell_size <span class="op">=</span> <span class="va">MC_list</span><span class="op">$</span><span class="va">supercell_size</span>, <span class="co"># provide this parameter to run sample-weighted version of PCA,</span></span>
<span>  k <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">MC_list</span><span class="op">$</span><span class="va">UMAP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_UMAP.html">supercell_UMAP</a></span><span class="op">(</span></span>
<span>  SC <span class="op">=</span> <span class="va">MC_list</span>,</span>
<span>  PCA_name <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>  n.comp <span class="op">=</span> <span class="fl">30</span>, n_neighbors <span class="op">=</span> <span class="fl">15</span>, min_dist<span class="op">=</span><span class="fl">0.5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_DimPlot.html">supercell_DimPlot</a></span><span class="op">(</span>SC <span class="op">=</span> <span class="va">MC_list</span>,</span>
<span>  groups <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_column</span><span class="op">]</span>,</span>
<span>  dim.name <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"UMAP of metacells colored by cell type assignment"</span><span class="op">)</span>, color.use <span class="op">=</span> <span class="va">celltype_colors</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-dim-reduc-weighted-1.png" width="672"><img src="30-MC_analysis_discrete_files/figure-html/r-mc-dim-reduc-weighted-2.png" width="672"></p>
</div>
<div id="clustering-2" class="section level3" number="3.3.3">
<h3>
<span class="header-section-number">3.3.3</span> Clustering<a class="anchor" aria-label="anchor" href="#clustering-2"><i class="fas fa-link"></i></a>
</h3>
<p>We cluster the metacells using the function <code>supercell_cluster</code> from SuperCell R package to perform the clustering step and visualize these clusters in the UMAP space:</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute distance among metacells</span></span>
<span><span class="va">D</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="va">MC_list</span><span class="op">$</span><span class="va">PCA</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># cluster metacells</span></span>
<span><span class="va">MC_list</span><span class="op">$</span><span class="va">SCclustering</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_cluster.html">supercell_cluster</a></span><span class="op">(</span>D <span class="op">=</span> <span class="va">D</span>, k <span class="op">=</span> <span class="fl">15</span>, supercell_size <span class="op">=</span> <span class="va">MC_list</span><span class="op">$</span><span class="va">supercell_size</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span><span class="op">$</span><span class="va">SCclustering</span> <span class="op">&lt;-</span> <span class="va">MC_list</span><span class="op">$</span><span class="va">SCclustering</span><span class="op">$</span><span class="va">clustering</span></span>
<span></span>
<span><span class="co"># Plot clustering result</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_DimPlot.html">supercell_DimPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">MC_list</span>,</span>
<span>  groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">MC_list</span><span class="op">$</span><span class="va">SCclustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span>,</span>
<span>  dim.name <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"UMAP of metacells colored by metacell clustering"</span><span class="op">)</span></span>
<span><span class="op">)</span>  <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-clustering-weighted-1.png" width="672"><img src="30-MC_analysis_discrete_files/figure-html/r-mc-clustering-weighted-2.png" width="672"></p>
</div>
<div id="differential-expression-analysis-2" class="section level3" number="3.3.4">
<h3>
<span class="header-section-number">3.3.4</span> Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-2"><i class="fas fa-link"></i></a>
</h3>
<p>We perform diffrential analysis to identify the markers of our clusters using the <code>supercell_FindAllMarkers</code> function from the SuperCell package.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute upregulated genes in each cell line (versus other cells)</span></span>
<span><span class="va">MC.all.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_FindAllMarkers.html">supercell_FindAllMarkers</a></span><span class="op">(</span></span>
<span>  ge <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">MC.seurat</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>  clusters <span class="op">=</span> <span class="va">MC_list</span><span class="op">$</span><span class="va">SCclustering</span><span class="op">$</span><span class="va">clustering</span>,</span>
<span>  supercell_size <span class="op">=</span> <span class="va">MC_list</span><span class="op">$</span><span class="va">supercell_size</span>,</span>
<span>  only.pos <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min.pct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  logfc.threshold <span class="op">=</span> <span class="fl">0.2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We select the markers for cluster 9:</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_markers</span> <span class="op">&lt;-</span> <span class="va">MC.all.markers</span><span class="op">[[</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">MC.top.markers</span> <span class="op">&lt;-</span> <span class="va">cluster_markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">cluster_markers</span><span class="op">$</span><span class="va">logFC</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">MC.top.markers</span><span class="op">)</span></span>
<span><span class="co">#&gt;       p.value adj.p.value     pct.1     pct.2    logFC w.mean.1   w.mean.2</span></span>
<span><span class="co">#&gt; GNLY        0           0 1.0000000 0.9692845 3.444786 4.629992 0.54215411</span></span>
<span><span class="co">#&gt; NKG7        0           0 1.0000000 0.9536207 2.853815 4.023960 0.51434288</span></span>
<span><span class="co">#&gt; GZMB        0           0 0.9916981 0.6401282 2.529441 2.771860 0.15034752</span></span>
<span><span class="co">#&gt; KLRF1       0           0 1.0000000 0.5681106 2.225002 2.353682 0.09508412</span></span>
<span><span class="co">#&gt; KLRD1       0           0 1.0000000 0.5480537 2.184094 2.508815 0.17047908</span></span>
<span><span class="co">#&gt; CST7        0           0 1.0000000 0.7280256 2.081451 2.706358 0.30405936</span></span></code></pre></div>
<p>We visualize the top 5 markers for the cluster 9 and see that the top marker genes for this cluster contain marker genes of natural killer cells such as GZMB and GNLY.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCclustering"</span></span>
<span><span class="co"># genes.to.plot &lt;- MC.seurat.top.markers$gene[MC.seurat.top.markers$cluster == unique(MC.seurat@meta.data[,annotation_column])[1]]</span></span>
<span><span class="co"># genes.to.plot &lt;- MC.top.markers$gene[c(seq(1, 20, 5))]</span></span>
<span><span class="va">genes.to.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">MC.top.markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">MC.seurat</span>, features <span class="op">=</span> <span class="va">genes.to.plot</span>, ncol <span class="op">=</span> <span class="fl">5</span>, pt.size <span class="op">=</span> <span class="fl">0.0</span><span class="op">)</span>  </span>
<span><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span></code></pre></div>
<div class="inline-figure"><img src="30-MC_analysis_discrete_files/figure-html/r-mc-plot-genes-weighted-1.png" width="960"></div>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="QCs.html"><span class="header-section-number">2</span> Metacells’ QCs</a></div>
<div class="next"><a href="command-line.html"><span class="header-section-number">4</span> Metacell Analysis Toolkit (MATK)</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#downstream-analysis"><span class="header-section-number">3</span> Downstream analysis of metacells</a></li>
<li>
<a class="nav-link" href="#standard-analysis-R"><span class="header-section-number">3.1</span> Standard analysis (R)</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#load-metacell-seurat-object"><span class="header-section-number">3.1.1</span> Load metacell Seurat object</a></li>
<li><a class="nav-link" href="#dimensionality-reduction"><span class="header-section-number">3.1.2</span> Dimensionality reduction</a></li>
<li><a class="nav-link" href="#clustering"><span class="header-section-number">3.1.3</span> Clustering</a></li>
<li><a class="nav-link" href="#differential-expression-analysis"><span class="header-section-number">3.1.4</span> Differential expression analysis</a></li>
<li><a class="nav-link" href="#visualize-gene-gene-correlation"><span class="header-section-number">3.1.5</span> Visualize gene-gene correlation</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#standard-analysis-Py"><span class="header-section-number">3.2</span> Standard analysis (Python)</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#dimensionality-reduction-1"><span class="header-section-number">3.2.1</span> Dimensionality reduction</a></li>
<li><a class="nav-link" href="#clustering-1"><span class="header-section-number">3.2.2</span> Clustering</a></li>
<li><a class="nav-link" href="#differential-expression-analysis-1"><span class="header-section-number">3.2.3</span> Differential expression analysis</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#weighted-analysis"><span class="header-section-number">3.3</span> Sample-weighted analysis</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#load-metacell-seurat-object-1"><span class="header-section-number">3.3.1</span> Load metacell Seurat object</a></li>
<li><a class="nav-link" href="#dimensionality-reduction-2"><span class="header-section-number">3.3.2</span> Dimensionality reduction</a></li>
<li><a class="nav-link" href="#clustering-2"><span class="header-section-number">3.3.3</span> Clustering</a></li>
<li><a class="nav-link" href="#differential-expression-analysis-2"><span class="header-section-number">3.3.4</span> Differential expression analysis</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/GfellerLab/MetacellAnalysisTutorial/blob/main/30-MC_analysis_discrete.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/GfellerLab/MetacellAnalysisTutorial/edit/main/30-MC_analysis_discrete.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Metacell Analysis Tutorial</strong>" was written by Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller. It was last built on 2023-12-21.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
