<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6.2 Supervised integration | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="6.2 Supervised integration | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6.2 Supervised integration | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller" />


<meta name="date" content="2023-11-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="integration_unsupervised.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>This tutorial</a></li>
<li class="chapter" data-level="1" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>1</b> Requirements</a>
<ul>
<li class="chapter" data-level="1.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>1.1</b> Installations</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="installations.html"><a href="installations.html#using-conda-recommended"><i class="fa fa-check"></i><b>1.1.1</b> Using conda (recommended)</a></li>
<li class="chapter" data-level="1.1.2" data-path="installations.html"><a href="installations.html#without-conda"><i class="fa fa-check"></i><b>1.1.2</b> Without conda</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>1.2</b> Retrieve a discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="1.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>1.3</b> Retrieve a continuous dataset (CD34 dataset)</a></li>
<li class="chapter" data-level="1.4" data-path="HLCA-data.html"><a href="HLCA-data.html"><i class="fa fa-check"></i><b>1.4</b> Retrieve the lung atlas dataset</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="HLCA-data.html"><a href="HLCA-data.html#downloading-the-atlas"><i class="fa fa-check"></i><b>1.4.1</b> Downloading the atlas</a></li>
<li class="chapter" data-level="1.4.2" data-path="HLCA-data.html"><a href="HLCA-data.html#splitting-atlas-by-datasets"><i class="fa fa-check"></i><b>1.4.2</b> Splitting atlas by datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a></li>
<li class="chapter" data-level="3" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells</a>
<ul>
<li class="chapter" data-level="3.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#method"><i class="fa fa-check"></i><b>3.1.1</b> Method</a></li>
<li class="chapter" data-level="3.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>3.1.2</b> Data loading</a></li>
<li class="chapter" data-level="3.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>3.1.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.5" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#annotate-metacells-using-available-annotations"><i class="fa fa-check"></i><b>3.1.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.1.6" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#save-output"><i class="fa fa-check"></i><b>3.1.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="MC2-construction.html"><a href="MC2-construction.html#method-1"><i class="fa fa-check"></i><b>3.2.1</b> Method</a></li>
<li class="chapter" data-level="3.2.2" data-path="MC2-construction.html"><a href="MC2-construction.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.2</b> Data loading</a></li>
<li class="chapter" data-level="3.2.3" data-path="MC2-construction.html"><a href="MC2-construction.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.4" data-path="MC2-construction.html"><a href="MC2-construction.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="MC2-construction.html"><a href="MC2-construction.html#annotate-metacells-using-available-annotations-1"><i class="fa fa-check"></i><b>3.2.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.2.6" data-path="MC2-construction.html"><a href="MC2-construction.html#save-output-1"><i class="fa fa-check"></i><b>3.2.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="SEACells-construction.html"><a href="SEACells-construction.html#method-2"><i class="fa fa-check"></i><b>3.3.1</b> Method</a></li>
<li class="chapter" data-level="3.3.2" data-path="SEACells-construction.html"><a href="SEACells-construction.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.2</b> Data loading</a></li>
<li class="chapter" data-level="3.3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.4" data-path="SEACells-construction.html"><a href="SEACells-construction.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="SEACells-construction.html"><a href="SEACells-construction.html#visualize-metacells"><i class="fa fa-check"></i><b>3.3.5</b> Visualize metacells</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>3.4</b> Metacell Analysis Toolkit (MCAT)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>4</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="4.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>4.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>4.1.1</b> Purity</a></li>
<li class="chapter" data-level="4.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>4.1.2</b> Compactness</a></li>
<li class="chapter" data-level="4.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>4.1.3</b> Separation</a></li>
<li class="chapter" data-level="4.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>4.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>4.2</b> Size distribution</a></li>
<li class="chapter" data-level="4.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>4.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>5</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="5.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>5.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>5.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>5.1.3</b> Clustering</a></li>
<li class="chapter" data-level="5.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>5.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="5.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>5.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>5.2</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>5.2.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.2.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>5.2.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.2.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-1"><i class="fa fa-check"></i><b>5.2.3</b> Clustering</a></li>
<li class="chapter" data-level="5.2.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>5.2.4</b> Differential expression analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="integration.html"><a href="integration.html"><i class="fa fa-check"></i><b>6</b> Integration of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html"><i class="fa fa-check"></i><b>6.1</b> Unsupervised integration</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#data-loading-3"><i class="fa fa-check"></i><b>6.1.1</b> Data loading</a></li>
<li class="chapter" data-level="6.1.2" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#setting-up-the-environment"><i class="fa fa-check"></i><b>6.1.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="6.1.3" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#building-metacell"><i class="fa fa-check"></i><b>6.1.3</b> Building metacell</a></li>
<li class="chapter" data-level="6.1.4" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#loading-metacell-objects"><i class="fa fa-check"></i><b>6.1.4</b> Loading metacell objects</a></li>
<li class="chapter" data-level="6.1.5" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#merging-objects-and-basic-quality-control"><i class="fa fa-check"></i><b>6.1.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="6.1.6" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#unintegrated-analysis"><i class="fa fa-check"></i><b>6.1.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="6.1.7" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#seurat-integration"><i class="fa fa-check"></i><b>6.1.7</b> Seurat integration</a></li>
<li class="chapter" data-level="6.1.8" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#downstream-analysis"><i class="fa fa-check"></i><b>6.1.8</b> Downstream analysis</a></li>
<li class="chapter" data-level="6.1.9" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#conclusion"><i class="fa fa-check"></i><b>6.1.9</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="integration_supervised.html"><a href="integration_supervised.html"><i class="fa fa-check"></i><b>6.2</b> Supervised integration</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="integration_supervised.html"><a href="integration_supervised.html#data-loading-4"><i class="fa fa-check"></i><b>6.2.1</b> Data loading</a></li>
<li class="chapter" data-level="6.2.2" data-path="integration_supervised.html"><a href="integration_supervised.html#setting-up-the-environment-1"><i class="fa fa-check"></i><b>6.2.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="6.2.3" data-path="integration_supervised.html"><a href="integration_supervised.html#constructing-supervised-metacell"><i class="fa fa-check"></i><b>6.2.3</b> Constructing supervised metacell</a></li>
<li class="chapter" data-level="6.2.4" data-path="integration_supervised.html"><a href="integration_supervised.html#load-metacell-objects"><i class="fa fa-check"></i><b>6.2.4</b> Load metacell objects</a></li>
<li class="chapter" data-level="6.2.5" data-path="integration_supervised.html"><a href="integration_supervised.html#merging-objects-and-basic-quality-control-1"><i class="fa fa-check"></i><b>6.2.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="6.2.6" data-path="integration_supervised.html"><a href="integration_supervised.html#unintegrated-analysis-1"><i class="fa fa-check"></i><b>6.2.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="6.2.7" data-path="integration_supervised.html"><a href="integration_supervised.html#stacas-integration"><i class="fa fa-check"></i><b>6.2.7</b> STACAS integration</a></li>
<li class="chapter" data-level="6.2.8" data-path="integration_supervised.html"><a href="integration_supervised.html#comparison-with-unsupervised-analysis"><i class="fa fa-check"></i><b>6.2.8</b> Comparison with unsupervised analysis</a></li>
<li class="chapter" data-level="6.2.9" data-path="integration_supervised.html"><a href="integration_supervised.html#downstream-analysis-1"><i class="fa fa-check"></i><b>6.2.9</b> Downstream analysis</a></li>
<li class="chapter" data-level="6.2.10" data-path="integration_supervised.html"><a href="integration_supervised.html#conclusion-1"><i class="fa fa-check"></i><b>6.2.10</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="integration_supervised" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> Supervised integration<a href="integration_supervised.html#integration_supervised" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>As for the unsupervised integration example, we will work with the Human Cell Lung Atlas core <a href="%5Bhttps://www.nature.com/articles/s41591-023-02327-2">HLCA</a>
gathering around 580,000 cells from 107 individuals distributed in 166 samples.</p>
<p>Taking advantage of the single-cell annotation of the original study we will build metacells for each cell type in each sample and
guide the integration with the cell type label using <a href="https://github.com/carmonalab/STACAS">STACAS</a>.</p>
<div id="data-loading-4" class="section level3 hasAnchor" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Data loading<a href="integration_supervised.html#data-loading-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Please follow the section <a href="HLCA-data.html#HLCA-data">1.4</a> to retrieve the HLCA atlas, divide the atlas by dataset and save the splitted data in the following folder: “data/HLCA/”.</p>
</div>
<div id="setting-up-the-environment-1" class="section level3 hasAnchor" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Setting up the environment<a href="integration_supervised.html#setting-up-the-environment-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First we need to specify that we will work with the MetacellAnalysisToolkit conda environment (needed for anndata relying on reticulate and the MCAT tool).
To build the conda environment please follow the instructions on our MetacellAnalysisToolkit <a href="https://github.com/GfellerLab/MetacellToolkit">github repository</a>.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="integration_supervised.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb120-2"><a href="integration_supervised.html#cb120-2" aria-hidden="true" tabindex="-1"></a>conda_env <span class="ot">&lt;-</span>  <span class="fu">conda_list</span>()[reticulate<span class="sc">::</span><span class="fu">conda_list</span>()<span class="sc">$</span>name <span class="sc">==</span> <span class="st">&quot;MetacellAnalysisToolkit&quot;</span>,<span class="st">&quot;python&quot;</span>]</span>
<span id="cb120-3"><a href="integration_supervised.html#cb120-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-4"><a href="integration_supervised.html#cb120-4" aria-hidden="true" tabindex="-1"></a><span class="fu">use_condaenv</span>(conda_env)</span></code></pre></div>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="integration_supervised.html#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb121-2"><a href="integration_supervised.html#cb121-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(anndata)</span>
<span id="cb121-3"><a href="integration_supervised.html#cb121-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SuperCell)</span>
<span id="cb121-4"><a href="integration_supervised.html#cb121-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb121-5"><a href="integration_supervised.html#cb121-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-6"><a href="integration_supervised.html#cb121-6" aria-hidden="true" tabindex="-1"></a>color.celltypes  <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;#E5D2DD&#39;</span>, <span class="st">&#39;#53A85F&#39;</span>, <span class="st">&#39;#F1BB72&#39;</span>, <span class="st">&#39;#F3B1A0&#39;</span>, <span class="st">&#39;#D6E7A3&#39;</span>, <span class="st">&#39;#57C3F3&#39;</span>, <span class="st">&#39;#476D87&#39;</span>,</span>
<span id="cb121-7"><a href="integration_supervised.html#cb121-7" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&#39;#E95C59&#39;</span>, <span class="st">&#39;#E59CC4&#39;</span>, <span class="st">&#39;#AB3282&#39;</span>, <span class="st">&#39;#23452F&#39;</span>, <span class="st">&#39;#BD956A&#39;</span>, <span class="st">&#39;#8C549C&#39;</span>, <span class="st">&#39;#585658&#39;</span>,</span>
<span id="cb121-8"><a href="integration_supervised.html#cb121-8" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&#39;#9FA3A8&#39;</span>, <span class="st">&#39;#E0D4CA&#39;</span>, <span class="st">&#39;#5F3D69&#39;</span>, <span class="st">&#39;#58A4C3&#39;</span>, <span class="st">&quot;#b20000&quot;</span>,<span class="st">&#39;#E4C755&#39;</span>, <span class="st">&#39;#F7F398&#39;</span>,</span>
<span id="cb121-9"><a href="integration_supervised.html#cb121-9" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&#39;#AA9A59&#39;</span>, <span class="st">&#39;#E63863&#39;</span>, <span class="st">&#39;#E39A35&#39;</span>, <span class="st">&#39;#C1E6F3&#39;</span>, <span class="st">&#39;#6778AE&#39;</span>, <span class="st">&#39;#91D0BE&#39;</span>, <span class="st">&#39;#B53E2B&#39;</span>,</span>
<span id="cb121-10"><a href="integration_supervised.html#cb121-10" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&#39;#712820&#39;</span>, <span class="st">&#39;#DCC1DD&#39;</span>, <span class="st">&#39;#CCE0F5&#39;</span>, <span class="st">&#39;#CCC9E6&#39;</span>, <span class="st">&#39;#625D9E&#39;</span>, <span class="st">&#39;#68A180&#39;</span>, <span class="st">&#39;#3A6963&#39;</span>,</span>
<span id="cb121-11"><a href="integration_supervised.html#cb121-11" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&#39;#968175&#39;</span>)</span></code></pre></div>
</div>
<div id="constructing-supervised-metacell" class="section level3 hasAnchor" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> Constructing supervised metacell<a href="integration_supervised.html#constructing-supervised-metacell" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Sikkema et al. made a remarkable job in finely annotating hundreds thousands of cells.
Within the framework of this re-analysis, let’s now try to use this prior knowledge to obtain slightly better results using a supervised workflow.</p>
<p>We added in section <a href="HLCA-data.html#HLCA-data">1.4</a> a <code>ann_sample</code> column in the metadata of the single cell object.
We now can use it to build metacell for each cell type in each sample.</p>
<p>If you are limited in memory you should still be able to process the samples by reducing the number of cores (e.g. <code>-l 3</code>) or
by sequentially processing the samples (just remove the <code>-l</code>) in a slightly longer time</p>
<p>This should take around 30 minutes.</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb122-1"><a href="integration_supervised.html#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> d <span class="kw">in</span> data/HLCA/datasets/<span class="pp">*</span><span class="kw">;</span></span>
<span id="cb122-2"><a href="integration_supervised.html#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> <span class="ex">cli/MCAT</span> <span class="at">-t</span> SuperCell <span class="at">-i</span> <span class="va">$d</span>/sc_adata.h5ad <span class="at">-o</span> <span class="va">$d</span>/sup_mc <span class="at">-a</span> ann_sample <span class="at">-l</span> 3 <span class="at">-n</span> 50 <span class="at">-f</span> 2000 <span class="at">-k</span> 30 <span class="at">-g</span> 50 <span class="at">-s</span> adata</span>
<span id="cb122-3"><a href="integration_supervised.html#cb122-3" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
</div>
<div id="load-metacell-objects" class="section level3 hasAnchor" number="6.2.4">
<h3><span class="header-section-number">6.2.4</span> Load metacell objects<a href="integration_supervised.html#load-metacell-objects" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We load the .h5ad objects and directly convert them in Seurat objects to benefit from all the functions of this framework.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="integration_supervised.html#cb123-1" aria-hidden="true" tabindex="-1"></a>datasets <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="st">&quot;data/HLCA/datasets/&quot;</span>, <span class="at">full.names =</span> F, <span class="at">recursive =</span> F)</span>
<span id="cb123-2"><a href="integration_supervised.html#cb123-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb123-3"><a href="integration_supervised.html#cb123-3" aria-hidden="true" tabindex="-1"></a>metacell.files <span class="ot">&lt;-</span> <span class="fu">sapply</span>(datasets[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], <span class="at">FUN =</span> <span class="cf">function</span>(x){<span class="fu">paste0</span>(<span class="st">&quot;data/HLCA/datasets/&quot;</span>,x,<span class="st">&quot;/sup_mc/mc_adata.h5ad&quot;</span>)})</span>
<span id="cb123-4"><a href="integration_supervised.html#cb123-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb123-5"><a href="integration_supervised.html#cb123-5" aria-hidden="true" tabindex="-1"></a>metacell.objs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> metacell.files, <span class="cf">function</span>(X){</span>
<span id="cb123-6"><a href="integration_supervised.html#cb123-6" aria-hidden="true" tabindex="-1"></a>  adata <span class="ot">&lt;-</span> <span class="fu">read_h5ad</span>(X)</span>
<span id="cb123-7"><a href="integration_supervised.html#cb123-7" aria-hidden="true" tabindex="-1"></a>  countMatrix <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">t</span>(adata<span class="sc">$</span>X)</span>
<span id="cb123-8"><a href="integration_supervised.html#cb123-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(countMatrix) <span class="ot">&lt;-</span> adata<span class="sc">$</span>obs_names</span>
<span id="cb123-9"><a href="integration_supervised.html#cb123-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(countMatrix) <span class="ot">&lt;-</span> adata<span class="sc">$</span>var_names</span>
<span id="cb123-10"><a href="integration_supervised.html#cb123-10" aria-hidden="true" tabindex="-1"></a>  sobj <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> countMatrix,<span class="at">meta.data =</span> adata<span class="sc">$</span>obs)</span>
<span id="cb123-11"><a href="integration_supervised.html#cb123-11" aria-hidden="true" tabindex="-1"></a>  sobj <span class="ot">&lt;-</span> <span class="fu">RenameCells</span>(sobj, <span class="at">add.cell.id =</span> <span class="fu">unique</span>(sobj<span class="sc">$</span>sample)) <span class="co"># we give unique name to metacells</span></span>
<span id="cb123-12"><a href="integration_supervised.html#cb123-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sobj)</span>
<span id="cb123-13"><a href="integration_supervised.html#cb123-13" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
</div>
<div id="merging-objects-and-basic-quality-control-1" class="section level3 hasAnchor" number="6.2.5">
<h3><span class="header-section-number">6.2.5</span> Merging objects and basic quality control<a href="integration_supervised.html#merging-objects-and-basic-quality-control-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Given the single-cell metadata, the MCAT tool automatically assigns annotations to metacells and
computes purities for all the categorical variables present in the metadata of the input single-cell object.</p>
<p>Thus, let’s check the purity of our metacells at different level of annotations, as well as their size (number of single cells they contain).</p>
<p>To do so we merge the objects together and use Seurat <code>VlnPlot</code> function.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="integration_supervised.html#cb124-1" aria-hidden="true" tabindex="-1"></a>unintegrated.mc <span class="ot">&lt;-</span> <span class="fu">merge</span>(metacell.objs[[<span class="dv">1</span>]], metacell.objs[<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb124-2"><a href="integration_supervised.html#cb124-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb124-3"><a href="integration_supervised.html#cb124-3" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(unintegrated.mc[, unintegrated.mc<span class="sc">$</span>ann_level_3 <span class="sc">!=</span> <span class="st">&quot;None&quot;</span>], <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;size&quot;</span>, <span class="st">&quot;ann_level_2_purity&quot;</span>), <span class="at">group.by =</span> <span class="st">&#39;dataset&#39;</span>, <span class="at">pt.size =</span> <span class="fl">0.001</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb124-4"><a href="integration_supervised.html#cb124-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in SingleExIPlot(type = type, data = data[, x, drop = FALSE], idents =</span></span>
<span id="cb124-5"><a href="integration_supervised.html#cb124-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; idents, : All cells have the same value of ann_level_2_purity.</span></span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="integration_supervised.html#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(unintegrated.mc[, unintegrated.mc<span class="sc">$</span>ann_level_3 <span class="sc">!=</span> <span class="st">&quot;None&quot;</span>], <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;ann_level_3_purity&quot;</span>, <span class="st">&quot;ann_level_4_purity&quot;</span>), <span class="at">group.by =</span> <span class="st">&#39;dataset&#39;</span>, <span class="at">pt.size =</span> <span class="fl">0.001</span>,  <span class="at">ncol=</span><span class="dv">2</span>)</span>
<span id="cb125-2"><a href="integration_supervised.html#cb125-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning in SingleExIPlot(type = type, data = data[, x, drop = FALSE], idents =</span></span>
<span id="cb125-3"><a href="integration_supervised.html#cb125-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; idents, : All cells have the same value of ann_level_3_purity.</span></span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-28-2.png" width="672" /></p>
<p>We can also use box plots.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="integration_supervised.html#cb126-1" aria-hidden="true" tabindex="-1"></a>p_4 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(unintegrated.mc<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x =</span> dataset, <span class="at">y =</span> ann_level_4_purity, <span class="at">fill =</span> dataset)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb126-2"><a href="integration_supervised.html#cb126-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">guide =</span> <span class="fu">guide_axis</span>(<span class="at">angle =</span> <span class="dv">45</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;sup metacells level 4 purity&quot;</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb126-3"><a href="integration_supervised.html#cb126-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-4"><a href="integration_supervised.html#cb126-4" aria-hidden="true" tabindex="-1"></a>p_finest <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(unintegrated.mc<span class="sc">@</span>meta.data, <span class="fu">aes</span>(<span class="at">x =</span> dataset, <span class="at">y =</span> ann_finest_level_purity, <span class="at">fill =</span> dataset)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb126-5"><a href="integration_supervised.html#cb126-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">guide =</span> <span class="fu">guide_axis</span>(<span class="at">angle =</span> <span class="dv">45</span>))  <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;sup metacells finest level purity&quot;</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb126-6"><a href="integration_supervised.html#cb126-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb126-7"><a href="integration_supervised.html#cb126-7" aria-hidden="true" tabindex="-1"></a>p_4 <span class="sc">+</span> p_finest</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-29-1.png" width="672" /></p>
<p>Overall using supervised metacells construction we obtain pure metacells until the 3rd level of annotation and
improve metacell purities for finer levels compared to the unsupervised approach (see previous section @ref(integration_unsupervised)).</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="integration_supervised.html#cb127-1" aria-hidden="true" tabindex="-1"></a>meta.data.unsup <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/HLCA/combined.mc.unsup.rds&quot;</span>)<span class="sc">@</span>meta.data</span>
<span id="cb127-2"><a href="integration_supervised.html#cb127-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-3"><a href="integration_supervised.html#cb127-3" aria-hidden="true" tabindex="-1"></a>p_4_unsup <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(meta.data.unsup, <span class="fu">aes</span>(<span class="at">x =</span> dataset, <span class="at">y =</span> ann_level_4_purity, <span class="at">fill =</span> dataset)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb127-4"><a href="integration_supervised.html#cb127-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_discrete</span>(<span class="at">guide =</span> <span class="fu">guide_axis</span>(<span class="at">angle =</span> <span class="dv">45</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;unsup metacells level 4 purity&quot;</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb127-5"><a href="integration_supervised.html#cb127-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-6"><a href="integration_supervised.html#cb127-6" aria-hidden="true" tabindex="-1"></a>p_finest_unsup <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(meta.data.unsup, <span class="fu">aes</span>(<span class="at">x =</span> dataset, <span class="at">y =</span> ann_finest_level_purity, <span class="at">fill =</span> dataset)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb127-7"><a href="integration_supervised.html#cb127-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">guide =</span> <span class="fu">guide_axis</span>(<span class="at">angle =</span> <span class="dv">45</span>))  <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;unsup metacells finest level purity&quot;</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb127-8"><a href="integration_supervised.html#cb127-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-9"><a href="integration_supervised.html#cb127-9" aria-hidden="true" tabindex="-1"></a>p_4_unsup <span class="sc">|</span> p_4</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-30-1.png" width="1344" /></p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="integration_supervised.html#cb128-1" aria-hidden="true" tabindex="-1"></a>p_finest_unsup <span class="sc">+</span> p_finest</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-30-2.png" width="1344" /></p>
</div>
<div id="unintegrated-analysis-1" class="section level3 hasAnchor" number="6.2.6">
<h3><span class="header-section-number">6.2.6</span> Unintegrated analysis<a href="integration_supervised.html#unintegrated-analysis-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s first do a standard dimensionality reduction without batch correction.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="integration_supervised.html#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(unintegrated.mc) <span class="ot">&lt;-</span> <span class="st">&quot;RNA&quot;</span></span>
<span id="cb129-2"><a href="integration_supervised.html#cb129-2" aria-hidden="true" tabindex="-1"></a>unintegrated.mc <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(unintegrated.mc)</span>
<span id="cb129-3"><a href="integration_supervised.html#cb129-3" aria-hidden="true" tabindex="-1"></a>unintegrated.mc <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(unintegrated.mc)</span>
<span id="cb129-4"><a href="integration_supervised.html#cb129-4" aria-hidden="true" tabindex="-1"></a>unintegrated.mc <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(unintegrated.mc)</span>
<span id="cb129-5"><a href="integration_supervised.html#cb129-5" aria-hidden="true" tabindex="-1"></a>unintegrated.mc <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(unintegrated.mc)</span>
<span id="cb129-6"><a href="integration_supervised.html#cb129-6" aria-hidden="true" tabindex="-1"></a>unintegrated.mc <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(unintegrated.mc,<span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb129-7"><a href="integration_supervised.html#cb129-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-8"><a href="integration_supervised.html#cb129-8" aria-hidden="true" tabindex="-1"></a>umap.unintegrated.datasets <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(unintegrated.mc,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">group.by =</span> <span class="st">&quot;dataset&quot;</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;unintegrated datasets&quot;</span>)</span>
<span id="cb129-9"><a href="integration_supervised.html#cb129-9" aria-hidden="true" tabindex="-1"></a>umap.unintegrated.types <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(unintegrated.mc,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">group.by =</span> <span class="st">&quot;ann_level_2&quot;</span>,<span class="at">label =</span> T,<span class="at">repel =</span> T,<span class="at">cols =</span> color.celltypes)<span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;unintegrated cell types&quot;</span>)</span>
<span id="cb129-10"><a href="integration_supervised.html#cb129-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-11"><a href="integration_supervised.html#cb129-11" aria-hidden="true" tabindex="-1"></a>umap.unintegrated.datasets <span class="sc">+</span> umap.unintegrated.types</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-31-1.png" width="672" /></p>
<p>You can see on the plots that a batch effect is clearly present at the metacell level. Let’s correct it using a supervised approach.</p>
</div>
<div id="stacas-integration" class="section level3 hasAnchor" number="6.2.7">
<h3><span class="header-section-number">6.2.7</span> STACAS integration<a href="integration_supervised.html#stacas-integration" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the original study, datasets were integrated using SCANVI semi-supervised integration using partial annotation obtained for each dataset prior integration.
Here in this second example we propose to use a similar approach in R using <a href="https://github.com/carmonalab/STACAS">STACAS</a>. We will use the “<code>ann</code>” labels we used to construct the metacells (3rd level of annotation if available for the cell, otherwise 2nd level).</p>
<p>To be noted that, as in the original study, we use the dataset rather than the donor as the batch parameter. See method section <a href="https://www.nature.com/articles/s41591-023-02327-2">Data integration benchmarking</a> of the original study for more details.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="integration_supervised.html#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install package if needed</span></span>
<span id="cb130-2"><a href="integration_supervised.html#cb130-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;STACAS&quot;</span>)) remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;carmonalab/STACAS&quot;</span>, <span class="at">upgrade =</span> <span class="st">&quot;never&quot;</span>)</span>
<span id="cb130-3"><a href="integration_supervised.html#cb130-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(STACAS)</span>
<span id="cb130-4"><a href="integration_supervised.html#cb130-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-5"><a href="integration_supervised.html#cb130-5" aria-hidden="true" tabindex="-1"></a>t0_integration <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb130-6"><a href="integration_supervised.html#cb130-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-7"><a href="integration_supervised.html#cb130-7" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize and identify variable features for each dataset independently</span></span>
<span id="cb130-8"><a href="integration_supervised.html#cb130-8" aria-hidden="true" tabindex="-1"></a>metacell.objs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="at">X =</span> metacell.objs, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb130-9"><a href="integration_supervised.html#cb130-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">DefaultAssay</span>(x) <span class="ot">&lt;-</span> <span class="st">&quot;RNA&quot;</span>;</span>
<span id="cb130-10"><a href="integration_supervised.html#cb130-10" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">RenameCells</span>(x, <span class="at">add.cell.id =</span> <span class="fu">unique</span>(x<span class="sc">$</span>sample)) <span class="co"># we give unique name to metacells</span></span>
<span id="cb130-11"><a href="integration_supervised.html#cb130-11" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(x)</span>
<span id="cb130-12"><a href="integration_supervised.html#cb130-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(x)})</span>
<span id="cb130-13"><a href="integration_supervised.html#cb130-13" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb130-14"><a href="integration_supervised.html#cb130-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-15"><a href="integration_supervised.html#cb130-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-16"><a href="integration_supervised.html#cb130-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform a supervised integration of the dataset using STACAS</span></span>
<span id="cb130-17"><a href="integration_supervised.html#cb130-17" aria-hidden="true" tabindex="-1"></a>combined.mc <span class="ot">&lt;-</span> <span class="fu">Run.STACAS</span>(<span class="at">object.list =</span> metacell.objs,</span>
<span id="cb130-18"><a href="integration_supervised.html#cb130-18" aria-hidden="true" tabindex="-1"></a>                          <span class="at">anchor.features =</span> <span class="dv">2000</span>,</span>
<span id="cb130-19"><a href="integration_supervised.html#cb130-19" aria-hidden="true" tabindex="-1"></a>                          <span class="at">min.sample.size =</span> <span class="dv">80</span>,</span>
<span id="cb130-20"><a href="integration_supervised.html#cb130-20" aria-hidden="true" tabindex="-1"></a>                          <span class="at">k.weight =</span> <span class="dv">80</span>, <span class="co">#smallest dataset contains 86 metacells</span></span>
<span id="cb130-21"><a href="integration_supervised.html#cb130-21" aria-hidden="true" tabindex="-1"></a>                          <span class="at">cell.labels =</span> <span class="st">&quot;ann&quot;</span>, <span class="co"># Note that by not you can use STACAS in its unsupervised mode</span></span>
<span id="cb130-22"><a href="integration_supervised.html#cb130-22" aria-hidden="true" tabindex="-1"></a>                          <span class="at">reference =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>), <span class="co"># ,5,9,11 the 5 biggest datasets are used as reference</span></span>
<span id="cb130-23"><a href="integration_supervised.html#cb130-23" aria-hidden="true" tabindex="-1"></a>                          <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb130-24"><a href="integration_supervised.html#cb130-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-25"><a href="integration_supervised.html#cb130-25" aria-hidden="true" tabindex="-1"></a>tf_integration <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb130-26"><a href="integration_supervised.html#cb130-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-27"><a href="integration_supervised.html#cb130-27" aria-hidden="true" tabindex="-1"></a>tf_integration <span class="sc">-</span> t0_integration</span>
<span id="cb130-28"><a href="integration_supervised.html#cb130-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-29"><a href="integration_supervised.html#cb130-29" aria-hidden="true" tabindex="-1"></a><span class="fu">remove</span>(metacell.objs) <span class="co"># We don&#39;t need the object list anymore</span></span>
<span id="cb130-30"><a href="integration_supervised.html#cb130-30" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code></pre></div>
<p>Check the obtained object:</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="integration_supervised.html#cb131-1" aria-hidden="true" tabindex="-1"></a>combined.mc</span>
<span id="cb131-2"><a href="integration_supervised.html#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb131-3"><a href="integration_supervised.html#cb131-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 30024 features across 6719 samples within 2 assays </span></span>
<span id="cb131-4"><a href="integration_supervised.html#cb131-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Active assay: integrated (2000 features, 2000 variable features)</span></span>
<span id="cb131-5"><a href="integration_supervised.html#cb131-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 other assay present: RNA</span></span>
<span id="cb131-6"><a href="integration_supervised.html#cb131-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 dimensional reduction calculated: pca</span></span></code></pre></div>
<p>We can verify that the sum of metacell sizes correspond to the original number of single-cells</p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="integration_supervised.html#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(combined.mc<span class="sc">$</span>size)</span>
<span id="cb132-2"><a href="integration_supervised.html#cb132-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 302920</span></span></code></pre></div>
<p>STACAS directly returns a pca for the slot <code>"integrated"</code> that we can use to make a UMAP of the corrected data.</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="integration_supervised.html#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(combined.mc) <span class="ot">=</span> <span class="st">&quot;integrated&quot;</span></span>
<span id="cb133-2"><a href="integration_supervised.html#cb133-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-3"><a href="integration_supervised.html#cb133-3" aria-hidden="true" tabindex="-1"></a>combined.mc <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(combined.mc, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">reduction =</span>  <span class="st">&quot;pca&quot;</span>, <span class="at">reduction.name =</span> <span class="st">&quot;umap&quot;</span>)</span></code></pre></div>
<p>Now we can make the plots and visually compare the results with the unintegrated analysis.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="integration_supervised.html#cb134-1" aria-hidden="true" tabindex="-1"></a>umap.stacas.datasets <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(combined.mc,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">group.by =</span> <span class="st">&quot;dataset&quot;</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;integrated datasets&quot;</span>)</span>
<span id="cb134-2"><a href="integration_supervised.html#cb134-2" aria-hidden="true" tabindex="-1"></a>umap.stacas.celltypes <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(combined.mc,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">group.by =</span> <span class="st">&quot;ann_level_2&quot;</span>,<span class="at">label =</span> T,<span class="at">repel =</span> T,<span class="at">cols =</span> color.celltypes) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;integrated cell types&quot;</span>)</span>
<span id="cb134-3"><a href="integration_supervised.html#cb134-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-4"><a href="integration_supervised.html#cb134-4" aria-hidden="true" tabindex="-1"></a>umap.stacas.datasets <span class="sc">+</span> umap.stacas.celltypes <span class="sc">+</span> umap.unintegrated.datasets <span class="sc">+</span> umap.unintegrated.types</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-36-1.png" width="1152" /></p>
<p>STACAS efficiently corrected the batch effect in the data while keeping the cell type separated.</p>
<p>We can navigate in the different annotation levels.</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="integration_supervised.html#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb135-2"><a href="integration_supervised.html#cb135-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-3"><a href="integration_supervised.html#cb135-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(combined.mc,<span class="at">group.by =</span> <span class="st">&quot;ann_level_1&quot;</span>,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">cols=</span> color.celltypes)</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-37-1.png" width="672" /></p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="integration_supervised.html#cb136-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-2"><a href="integration_supervised.html#cb136-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(combined.mc,<span class="at">group.by =</span> <span class="st">&quot;ann_level_2&quot;</span>,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">label =</span> T,<span class="at">repel =</span> T,<span class="at">cols=</span> color.celltypes)</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-37-2.png" width="672" /></p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="integration_supervised.html#cb137-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-2"><a href="integration_supervised.html#cb137-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(combined.mc,<span class="at">group.by =</span> <span class="st">&quot;ann_level_3&quot;</span>,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">label =</span> T, <span class="at">repel =</span> T,<span class="at">cols=</span> color.celltypes) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-37-3.png" width="672" /></p>
</div>
<div id="comparison-with-unsupervised-analysis" class="section level3 hasAnchor" number="6.2.8">
<h3><span class="header-section-number">6.2.8</span> Comparison with unsupervised analysis<a href="integration_supervised.html#comparison-with-unsupervised-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>we can quickly visually compare these results with the unsupervised integration obtained with Seurat:</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="integration_supervised.html#cb138-1" aria-hidden="true" tabindex="-1"></a>combined.mc.unsup <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/HLCA/combined.mc.unsup.rds&quot;</span>)</span>
<span id="cb138-2"><a href="integration_supervised.html#cb138-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-3"><a href="integration_supervised.html#cb138-3" aria-hidden="true" tabindex="-1"></a>combined.mc<span class="sc">$</span>ann_level_3 <span class="ot">&lt;-</span> <span class="fu">factor</span>(combined.mc<span class="sc">$</span>ann_level_3)</span>
<span id="cb138-4"><a href="integration_supervised.html#cb138-4" aria-hidden="true" tabindex="-1"></a>matched.color.celltypes <span class="ot">&lt;-</span> color.celltypes[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">levels</span>(combined.mc<span class="sc">$</span>ann_level_3))]</span>
<span id="cb138-5"><a href="integration_supervised.html#cb138-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(matched.color.celltypes) <span class="ot">&lt;-</span> <span class="fu">levels</span>(combined.mc<span class="sc">$</span>ann_level_3)</span>
<span id="cb138-6"><a href="integration_supervised.html#cb138-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-7"><a href="integration_supervised.html#cb138-7" aria-hidden="true" tabindex="-1"></a>level3_sup <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(combined.mc,<span class="at">group.by =</span> <span class="st">&quot;ann_level_3&quot;</span>,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">label =</span> T, <span class="at">repel =</span> T,<span class="at">cols=</span> matched.color.celltypes) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Sup workflow&quot;</span>)</span>
<span id="cb138-8"><a href="integration_supervised.html#cb138-8" aria-hidden="true" tabindex="-1"></a>level3_unsup <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(combined.mc.unsup,<span class="at">group.by =</span> <span class="st">&quot;ann_level_3&quot;</span>,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">label =</span> T, <span class="at">repel =</span> T,<span class="at">cols=</span> matched.color.celltypes) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Unsup workflow&quot;</span>)</span>
<span id="cb138-9"><a href="integration_supervised.html#cb138-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-10"><a href="integration_supervised.html#cb138-10" aria-hidden="true" tabindex="-1"></a>level3_sup <span class="sc">+</span> level3_unsup</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-38-1.png" width="1344" /></p>
<p>Look at epithelial cells in particular</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="integration_supervised.html#cb139-1" aria-hidden="true" tabindex="-1"></a>level3_sup <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(combined.mc[,combined.mc<span class="sc">$</span>ann_level_1 <span class="sc">==</span> <span class="st">&quot;Epithelial&quot;</span>],<span class="at">group.by =</span> <span class="st">&quot;ann_level_3&quot;</span>,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">label =</span> T, <span class="at">repel =</span> T,<span class="at">cols=</span> matched.color.celltypes) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Sup workflow&quot;</span>)</span>
<span id="cb139-2"><a href="integration_supervised.html#cb139-2" aria-hidden="true" tabindex="-1"></a>level3_unsup <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(combined.mc.unsup[,combined.mc.unsup<span class="sc">$</span>ann_level_1 <span class="sc">==</span> <span class="st">&quot;Epithelial&quot;</span>],<span class="at">group.by =</span> <span class="st">&quot;ann_level_3&quot;</span>,<span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>,<span class="at">label =</span> T, <span class="at">repel =</span> T,<span class="at">cols=</span> matched.color.celltypes) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&quot;Unsup workflow&quot;</span>)</span>
<span id="cb139-3"><a href="integration_supervised.html#cb139-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-4"><a href="integration_supervised.html#cb139-4" aria-hidden="true" tabindex="-1"></a>level3_sup <span class="sc">+</span> level3_unsup</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-39-1.png" width="1344" /></p>
</div>
<div id="downstream-analysis-1" class="section level3 hasAnchor" number="6.2.9">
<h3><span class="header-section-number">6.2.9</span> Downstream analysis<a href="integration_supervised.html#downstream-analysis-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You can try conduce the same downstream analyses as in the previous example @ref(integration_unsupervised) (clustering, cell type abundances, DEG …).</p>
<p>Here to show you the interest of supervised workflow with pure metacell we can zoom on the smooth muscle sub types. Despite the low metacell number for each cell type these different subtypes are separated on the UMAP, especially the rare FAM83D+ smooth muscles that were discovered in the original study.</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="integration_supervised.html#cb140-1" aria-hidden="true" tabindex="-1"></a>combined.mc<span class="sc">$</span>ann <span class="ot">&lt;-</span> <span class="fu">factor</span>(combined.mc<span class="sc">$</span>ann)</span>
<span id="cb140-2"><a href="integration_supervised.html#cb140-2" aria-hidden="true" tabindex="-1"></a>color.celltypes.ann <span class="ot">&lt;-</span> color.celltypes[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(<span class="fu">levels</span>(combined.mc<span class="sc">$</span>ann)))]</span>
<span id="cb140-3"><a href="integration_supervised.html#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(color.celltypes.ann) <span class="ot">&lt;-</span> <span class="fu">levels</span>(combined.mc<span class="sc">$</span>ann)</span>
<span id="cb140-4"><a href="integration_supervised.html#cb140-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-5"><a href="integration_supervised.html#cb140-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(combined.mc[,combined.mc<span class="sc">$</span>ann_level_2 <span class="sc">==</span> <span class="st">&quot;Smooth muscle&quot;</span>],<span class="at">group.by =</span> <span class="st">&quot;ann&quot;</span>,<span class="at">cols =</span> color.celltypes.ann)</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-40-1.png" width="672" /></p>
<p>Using a DEG analysis we can check if we retrieve their markers. MYH11 and CNN1 genes are canonical smooth muscle markers while FAM83D was found uniquely and consistently expressed by this rare cell type in the original study</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="integration_supervised.html#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(combined.mc) <span class="ot">&lt;-</span> <span class="st">&quot;RNA&quot;</span></span>
<span id="cb141-2"><a href="integration_supervised.html#cb141-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(combined.mc) <span class="ot">&lt;-</span> <span class="st">&quot;ann&quot;</span></span>
<span id="cb141-3"><a href="integration_supervised.html#cb141-3" aria-hidden="true" tabindex="-1"></a>markersSmoothMuscle <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(combined.mc,<span class="at">ident.1 =</span> <span class="st">&quot;Smooth muscle FAM83D+&quot;</span>,<span class="at">only.pos =</span> T)</span>
<span id="cb141-4"><a href="integration_supervised.html#cb141-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-5"><a href="integration_supervised.html#cb141-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(markersSmoothMuscle)</span>
<span id="cb141-6"><a href="integration_supervised.html#cb141-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                 p_val avg_log2FC pct.1 pct.2     p_val_adj</span></span>
<span id="cb141-7"><a href="integration_supervised.html#cb141-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; NMRK2   8.716260e-183  0.6673490 0.421 0.003 2.442645e-178</span></span>
<span id="cb141-8"><a href="integration_supervised.html#cb141-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; SLITRK3 2.001021e-154  0.3900334 0.421 0.004 5.607660e-150</span></span>
<span id="cb141-9"><a href="integration_supervised.html#cb141-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ASB5    3.578963e-108  0.4001160 0.368 0.004 1.002969e-103</span></span>
<span id="cb141-10"><a href="integration_supervised.html#cb141-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; HSPB3   1.156804e-107  0.5994731 0.579 0.011 3.241826e-103</span></span>
<span id="cb141-11"><a href="integration_supervised.html#cb141-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; BRINP3  1.402803e-107  0.8195490 0.526 0.009 3.931215e-103</span></span>
<span id="cb141-12"><a href="integration_supervised.html#cb141-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; MYOCD   3.863561e-101  1.3110302 0.684 0.017  1.082724e-96</span></span>
<span id="cb141-13"><a href="integration_supervised.html#cb141-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-14"><a href="integration_supervised.html#cb141-14" aria-hidden="true" tabindex="-1"></a>markersSmoothMuscle[<span class="fu">c</span>(<span class="st">&quot;MYH11&quot;</span>,<span class="st">&quot;CNN1&quot;</span>,<span class="st">&quot;FAM83D&quot;</span>),]</span>
<span id="cb141-15"><a href="integration_supervised.html#cb141-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;               p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span id="cb141-16"><a href="integration_supervised.html#cb141-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; MYH11  2.875397e-24   4.254605 1.000 0.227 8.058012e-20</span></span>
<span id="cb141-17"><a href="integration_supervised.html#cb141-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; CNN1   9.983682e-56   4.901585 1.000 0.080 2.797827e-51</span></span>
<span id="cb141-18"><a href="integration_supervised.html#cb141-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; FAM83D 3.289285e-13   2.580875 0.789 0.249 9.217893e-09</span></span>
<span id="cb141-19"><a href="integration_supervised.html#cb141-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-20"><a href="integration_supervised.html#cb141-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Many classical smooth muscles cells are not annotated at the 3rd level of annotation (labelled None)</span></span>
<span id="cb141-21"><a href="integration_supervised.html#cb141-21" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(combined.mc,<span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;MYH11&quot;</span>,<span class="st">&quot;CNN1&quot;</span>,<span class="st">&quot;FAM83D&quot;</span>),<span class="at">group.by =</span> <span class="st">&quot;ann&quot;</span>,<span class="at">ncol =</span> <span class="dv">2</span>,<span class="at">cols =</span> color.celltypes.ann)</span></code></pre></div>
<p><img src="40-integration_analysis_files/figure-html/unnamed-chunk-41-1.png" width="1152" /></p>
</div>
<div id="conclusion-1" class="section level3 hasAnchor" number="6.2.10">
<h3><span class="header-section-number">6.2.10</span> Conclusion<a href="integration_supervised.html#conclusion-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Taking advantage of the single cell annotation in a supervised workflow we could improve the precision of our metacell re-analysis.
When cell annotations are given and of good quality, which is far from being the case every time, building metacells accordingly and
use a supervised integration workflow should be preferred.</p>
<p>To be noted that we used an intermediary level of annotation to supervise our analysis, using a finer level for this data would have resulted
in a longer time for metacell building. PLus, we would have obtained to few metacells per cell type in the different sample to be able to
make an efficient supervised batch correction with STACAS.</p>
<p>To be more precise at the cost of computational efficiency one could also try to reduce the graining level of the analysis (using a graining level of 20 for instance),</p>
<p>To conclude, keep in mind that in one hand, for certain analysis such as rare cell type analysis,
we will never achieve the same level of sensitivity with metacells compared to single-cells.
On the other hand, you certainly won’t be able to analyze so many single-cells so easily, and you may not need extremely fine cell-type resolution for many analyses.</p>

</div>
</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="integration_unsupervised.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/40-integration_analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/40-integration_analysis.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
