<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.8 Loading single-cell data | Metacell analysis on a continuous dataset</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="5.8 Loading single-cell data | Metacell analysis on a continuous dataset" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.8 Loading single-cell data | Metacell analysis on a continuous dataset" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Leonard Herault, Aurelie Gabriel, Mariia Bilous, David Gfeller" />


<meta name="date" content="2023-11-21" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="quality-control-with-the-metacellanalysistoolkit-package.html"/>
<link rel="next" href="downstream-analysis-of-metacells.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>This tutorial</a></li>
<li class="chapter" data-level="1" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>1</b> Requirements</a>
<ul>
<li class="chapter" data-level="1.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>1.1</b> Installations</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="installations.html"><a href="installations.html#using-conda-recommended"><i class="fa fa-check"></i><b>1.1.1</b> Using conda (recommended)</a></li>
<li class="chapter" data-level="1.1.2" data-path="installations.html"><a href="installations.html#without-conda"><i class="fa fa-check"></i><b>1.1.2</b> Without conda</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>1.2</b> Retrieve a discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="1.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>1.3</b> Retrieve a continuous dataset (CD34 dataset)</a></li>
<li class="chapter" data-level="1.4" data-path="HLCA-data.html"><a href="HLCA-data.html"><i class="fa fa-check"></i><b>1.4</b> Retrieve the lung atlas dataset</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="HLCA-data.html"><a href="HLCA-data.html#downloading-the-atlas"><i class="fa fa-check"></i><b>1.4.1</b> Downloading the atlas</a></li>
<li class="chapter" data-level="1.4.2" data-path="HLCA-data.html"><a href="HLCA-data.html#splitting-atlas-by-datasets"><i class="fa fa-check"></i><b>1.4.2</b> Splitting atlas by datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a></li>
<li class="chapter" data-level="3" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells</a>
<ul>
<li class="chapter" data-level="3.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#method"><i class="fa fa-check"></i><b>3.1.1</b> Method</a></li>
<li class="chapter" data-level="3.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>3.1.2</b> Data loading</a></li>
<li class="chapter" data-level="3.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>3.1.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.5" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#annotate-metacells-using-available-annotations"><i class="fa fa-check"></i><b>3.1.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.1.6" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#save-output"><i class="fa fa-check"></i><b>3.1.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="MC2-construction.html"><a href="MC2-construction.html#method-1"><i class="fa fa-check"></i><b>3.2.1</b> Method</a></li>
<li class="chapter" data-level="3.2.2" data-path="MC2-construction.html"><a href="MC2-construction.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.2</b> Data loading</a></li>
<li class="chapter" data-level="3.2.3" data-path="MC2-construction.html"><a href="MC2-construction.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.4" data-path="MC2-construction.html"><a href="MC2-construction.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="MC2-construction.html"><a href="MC2-construction.html#annotate-metacells-using-available-annotations-1"><i class="fa fa-check"></i><b>3.2.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.2.6" data-path="MC2-construction.html"><a href="MC2-construction.html#save-output-1"><i class="fa fa-check"></i><b>3.2.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="SEACells-construction.html"><a href="SEACells-construction.html#method-2"><i class="fa fa-check"></i><b>3.3.1</b> Method</a></li>
<li class="chapter" data-level="3.3.2" data-path="SEACells-construction.html"><a href="SEACells-construction.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.2</b> Data loading</a></li>
<li class="chapter" data-level="3.3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.4" data-path="SEACells-construction.html"><a href="SEACells-construction.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="SEACells-construction.html"><a href="SEACells-construction.html#visualize-metacells"><i class="fa fa-check"></i><b>3.3.5</b> Visualize metacells</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>4</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="4.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>4.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>4.1.1</b> Purity</a></li>
<li class="chapter" data-level="4.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>4.1.2</b> Compactness</a></li>
<li class="chapter" data-level="4.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>4.1.3</b> Separation</a></li>
<li class="chapter" data-level="4.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>4.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>4.2</b> Size distribution</a></li>
<li class="chapter" data-level="4.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>4.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>5</b> Metacell Analysis Toolkit (MCAT)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="setting-up-the-environment.html"><a href="setting-up-the-environment.html"><i class="fa fa-check"></i><b>5.1</b> Setting up the environment</a></li>
<li class="chapter" data-level="5.2" data-path="metacell-building.html"><a href="metacell-building.html"><i class="fa fa-check"></i><b>5.2</b> Metacell building</a></li>
<li class="chapter" data-level="5.3" data-path="building-metacell-with-metacell2-mc2.html"><a href="building-metacell-with-metacell2-mc2.html"><i class="fa fa-check"></i><b>5.3</b> Building metacell with MetaCell2 (MC2)</a></li>
<li class="chapter" data-level="5.4" data-path="building-metacell-with-supercell.html"><a href="building-metacell-with-supercell.html"><i class="fa fa-check"></i><b>5.4</b> Building metacell with SuperCell</a></li>
<li class="chapter" data-level="5.5" data-path="short-downstream-analysis-of-the-metacells.html"><a href="short-downstream-analysis-of-the-metacells.html"><i class="fa fa-check"></i><b>5.5</b> Short downstream analysis of the metacells</a></li>
<li class="chapter" data-level="5.6" data-path="metacell2-metacells.html"><a href="metacell2-metacells.html"><i class="fa fa-check"></i><b>5.6</b> MetaCell2 metacells</a></li>
<li class="chapter" data-level="5.7" data-path="quality-control-with-the-metacellanalysistoolkit-package.html"><a href="quality-control-with-the-metacellanalysistoolkit-package.html"><i class="fa fa-check"></i><b>5.7</b> Quality control with the MetacellAnalysisToolkit package</a></li>
<li class="chapter" data-level="5.8" data-path="loading-single-cell-data.html"><a href="loading-single-cell-data.html"><i class="fa fa-check"></i><b>5.8</b> Loading single-cell data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>6</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>6.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>6.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>6.1.3</b> Clustering</a></li>
<li class="chapter" data-level="6.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="6.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>6.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>6.2</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>6.2.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.2.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>6.2.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.2.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-1"><i class="fa fa-check"></i><b>6.2.3</b> Clustering</a></li>
<li class="chapter" data-level="6.2.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>6.2.4</b> Differential expression analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="integration.html"><a href="integration.html"><i class="fa fa-check"></i><b>7</b> Integration of metacells</a>
<ul>
<li class="chapter" data-level="7.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html"><i class="fa fa-check"></i><b>7.1</b> Unsupervised integration</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#data-loading-3"><i class="fa fa-check"></i><b>7.1.1</b> Data loading</a></li>
<li class="chapter" data-level="7.1.2" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#setting-up-the-environment-1"><i class="fa fa-check"></i><b>7.1.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="7.1.3" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#building-metacell"><i class="fa fa-check"></i><b>7.1.3</b> Building metacell</a></li>
<li class="chapter" data-level="7.1.4" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#loading-metacell-objects"><i class="fa fa-check"></i><b>7.1.4</b> Loading metacell objects</a></li>
<li class="chapter" data-level="7.1.5" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#merging-objects-and-basic-quality-control"><i class="fa fa-check"></i><b>7.1.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="7.1.6" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#unintegrated-analysis"><i class="fa fa-check"></i><b>7.1.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="7.1.7" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#seurat-integration"><i class="fa fa-check"></i><b>7.1.7</b> Seurat integration</a></li>
<li class="chapter" data-level="7.1.8" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#downstream-analysis"><i class="fa fa-check"></i><b>7.1.8</b> Downstream analysis</a></li>
<li class="chapter" data-level="7.1.9" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#conclusion"><i class="fa fa-check"></i><b>7.1.9</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="integration_supervised.html"><a href="integration_supervised.html"><i class="fa fa-check"></i><b>7.2</b> Supervised integration</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="integration_supervised.html"><a href="integration_supervised.html#data-loading-4"><i class="fa fa-check"></i><b>7.2.1</b> Data loading</a></li>
<li class="chapter" data-level="7.2.2" data-path="integration_supervised.html"><a href="integration_supervised.html#setting-up-the-environment-2"><i class="fa fa-check"></i><b>7.2.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="7.2.3" data-path="integration_supervised.html"><a href="integration_supervised.html#constructing-supervised-metacell"><i class="fa fa-check"></i><b>7.2.3</b> Constructing supervised metacell</a></li>
<li class="chapter" data-level="7.2.4" data-path="integration_supervised.html"><a href="integration_supervised.html#load-metacell-objects"><i class="fa fa-check"></i><b>7.2.4</b> Load metacell objects</a></li>
<li class="chapter" data-level="7.2.5" data-path="integration_supervised.html"><a href="integration_supervised.html#merging-objects-and-basic-quality-control-1"><i class="fa fa-check"></i><b>7.2.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="7.2.6" data-path="integration_supervised.html"><a href="integration_supervised.html#unintegrated-analysis-1"><i class="fa fa-check"></i><b>7.2.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="7.2.7" data-path="integration_supervised.html"><a href="integration_supervised.html#stacas-integration"><i class="fa fa-check"></i><b>7.2.7</b> STACAS integration</a></li>
<li class="chapter" data-level="7.2.8" data-path="integration_supervised.html"><a href="integration_supervised.html#comparison-with-unsupervised-analysis"><i class="fa fa-check"></i><b>7.2.8</b> Comparison with unsupervised analysis</a></li>
<li class="chapter" data-level="7.2.9" data-path="integration_supervised.html"><a href="integration_supervised.html#downstream-analysis-1"><i class="fa fa-check"></i><b>7.2.9</b> Downstream analysis</a></li>
<li class="chapter" data-level="7.2.10" data-path="integration_supervised.html"><a href="integration_supervised.html#conclusion-1"><i class="fa fa-check"></i><b>7.2.10</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell analysis on a continuous dataset</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="loading-single-cell-data" class="section level2 hasAnchor" number="5.8">
<h2><span class="header-section-number">5.8</span> Loading single-cell data<a href="loading-single-cell-data.html#loading-single-cell-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>First we need to load the single cell data and make a classical dimentionality reduction analysis.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="loading-single-cell-data.html#cb102-1" aria-hidden="true" tabindex="-1"></a>cd34.singlecells <span class="ot">&lt;-</span> <span class="fu">read_h5ad</span>(<span class="st">&quot;data/CD34/cd34_multiome_rna.h5ad&quot;</span>)</span>
<span id="cb102-2"><a href="loading-single-cell-data.html#cb102-2" aria-hidden="true" tabindex="-1"></a>cd34.singlecells <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(cd34.singlecells<span class="sc">$</span>X),<span class="at">meta.data =</span> cd34.singlecells<span class="sc">$</span>obs)</span>
<span id="cb102-3"><a href="loading-single-cell-data.html#cb102-3" aria-hidden="true" tabindex="-1"></a>cd34.singlecells <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(cd34.singlecells)</span>
<span id="cb102-4"><a href="loading-single-cell-data.html#cb102-4" aria-hidden="true" tabindex="-1"></a>cd34.singlecells <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cd34.singlecells)</span>
<span id="cb102-5"><a href="loading-single-cell-data.html#cb102-5" aria-hidden="true" tabindex="-1"></a>cd34.singlecells <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cd34.singlecells)</span>
<span id="cb102-6"><a href="loading-single-cell-data.html#cb102-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span id="cb102-7"><a href="loading-single-cell-data.html#cb102-7" aria-hidden="true" tabindex="-1"></a>cd34.singlecells <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cd34.singlecells)</span>
<span id="cb102-8"><a href="loading-single-cell-data.html#cb102-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 1 </span></span>
<span id="cb102-9"><a href="loading-single-cell-data.html#cb102-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  NKAIN2, MEIS1, CRHBP, INPP4B, CALN1, CHRM3, AVP, PCDH9, GPC5, PRKG1 </span></span>
<span id="cb102-10"><a href="loading-single-cell-data.html#cb102-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ANGPT1, MECOM, MIR99AHG, PRKG2, PBX1, PRKCH, SKAP1, PREX2, ZNF385D, THRB </span></span>
<span id="cb102-11"><a href="loading-single-cell-data.html#cb102-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     IL12A-AS1, RBPMS, HLF, LIMCH1, LINC01122, ST8SIA6, TMEM163, MED12L, CABLES1, PDZD2 </span></span>
<span id="cb102-12"><a href="loading-single-cell-data.html#cb102-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  DIAPH3, POLQ, RRM2, ASPM, TUBB, MKI67, TOP2A, CIT, TUBA1B, BRIP1 </span></span>
<span id="cb102-13"><a href="loading-single-cell-data.html#cb102-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     NUSAP1, CLSPN, ATAD2, SHCBP1, H2AFZ, STMN1, TMPO, KIF15, NCAPG2, DTL </span></span>
<span id="cb102-14"><a href="loading-single-cell-data.html#cb102-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CENPF, GTSE1, KNL1, PCLAF, HMGN2, ACTB, UHRF1, ZNF367, MELK, AURKB </span></span>
<span id="cb102-15"><a href="loading-single-cell-data.html#cb102-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 2 </span></span>
<span id="cb102-16"><a href="loading-single-cell-data.html#cb102-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  PLCB1, NEGR1, PRKCE, AFF3, ATP8B4, SAMHD1, CD44, UBE2E2, MPO, PTPRE </span></span>
<span id="cb102-17"><a href="loading-single-cell-data.html#cb102-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     TTC7A, FLT3, ARHGAP26, GAS7, VIM, RELL1, AKAP13, RAB11FIP1, RUNX2, HDAC9 </span></span>
<span id="cb102-18"><a href="loading-single-cell-data.html#cb102-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PLD4, ADGRE5, SLC8A1, KLF6, IRF8, SFMBT2, CD74, RBM47, GAB2, SDK2 </span></span>
<span id="cb102-19"><a href="loading-single-cell-data.html#cb102-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  ANK1, RYR3, XACT, ITGA2B, APOC1, EEF1A1, TFR2, CPB1, KLF1, RPS3 </span></span>
<span id="cb102-20"><a href="loading-single-cell-data.html#cb102-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     APOE, PTH2R, RPL7A, KCNH2, RPS6, BLVRB, FAM178B, KEL, ZNF385D, GATA1 </span></span>
<span id="cb102-21"><a href="loading-single-cell-data.html#cb102-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ADAMTS3, SLC40A1, RPS5, EEF1B2, RPS3A, RPL13A, CSF1, RPL8, CNRIP1, HBD </span></span>
<span id="cb102-22"><a href="loading-single-cell-data.html#cb102-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 3 </span></span>
<span id="cb102-23"><a href="loading-single-cell-data.html#cb102-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  CD74, HLA-DRA, HLA-DRB1, FAM160A1, ACTG1, JCHAIN, EEF1A1, HLA-DPA1, ARL4C, TGFBI </span></span>
<span id="cb102-24"><a href="loading-single-cell-data.html#cb102-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     BLNK, TMSB4X, SPIB, PLXNA4, RPL13A, LGMN, CORO1A, RPS5, SCT, RPS3 </span></span>
<span id="cb102-25"><a href="loading-single-cell-data.html#cb102-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     RNASE6, RPS3A, LIME1, CUX2, ADAM19, PLD4, CST3, IRF7, CARD11, RPL28 </span></span>
<span id="cb102-26"><a href="loading-single-cell-data.html#cb102-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  AZU1, FNDC3B, ELANE, MPO, LRMDA, PRTN3, KCNQ5, LYST, AFF2, SLC22A15 </span></span>
<span id="cb102-27"><a href="loading-single-cell-data.html#cb102-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     F13A1, EREG, CFD, PLPPR3, ATP8B4, ZEB2, PCBP3, PDE3B, IL1RAP, ACSM3 </span></span>
<span id="cb102-28"><a href="loading-single-cell-data.html#cb102-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ZNF804A, CSF3R, SLC39A11, PLD1, DENND4A, ANXA1, CTSG, RREB1, TFRC, DGKG </span></span>
<span id="cb102-29"><a href="loading-single-cell-data.html#cb102-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 4 </span></span>
<span id="cb102-30"><a href="loading-single-cell-data.html#cb102-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  SPINK2, CALN1, PLCB1, INPP4B, NKAIN2, ANGPT1, ATP8B4, CHRM3, MEIS1, DIAPH3 </span></span>
<span id="cb102-31"><a href="loading-single-cell-data.html#cb102-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PHGDH, MIR924HG, SMIM24, MDK, TOP2A, CSF3R, ASPM, C1QTNF4, KIF15, TYMS </span></span>
<span id="cb102-32"><a href="loading-single-cell-data.html#cb102-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PRKCH, NUSAP1, TUBA1B, RRM2, CIT, TK1, CDCA5, POLQ, NDC80, AURKB </span></span>
<span id="cb102-33"><a href="loading-single-cell-data.html#cb102-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  SAMHD1, CD36, SULF2, ANK1, FAM160A1, TGFBI, CST3, RYR3, RAB11FIP1, S100A6 </span></span>
<span id="cb102-34"><a href="loading-single-cell-data.html#cb102-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     APOC1, XACT, IRF8, RNASE6, FCER1G, PLXNA4, LYZ, CPB1, APOE, UGCG </span></span>
<span id="cb102-35"><a href="loading-single-cell-data.html#cb102-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     FAM178B, GPR183, CSF2RB, RBM47, PIK3R5, MPEG1, CUX2, MICAL2, SLC8A1, CYBB </span></span>
<span id="cb102-36"><a href="loading-single-cell-data.html#cb102-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 5 </span></span>
<span id="cb102-37"><a href="loading-single-cell-data.html#cb102-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  MECOM, PCDH9, PRKG1, ASPM, PLXDC2, THRB, SKAP1, PHTF1, ZNF385D, PRKG2 </span></span>
<span id="cb102-38"><a href="loading-single-cell-data.html#cb102-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     SETBP1, PBX1, GPC5, LIMCH1, TOP2A, MED12L, MKI67, GTSE1, ELL2, PTK2 </span></span>
<span id="cb102-39"><a href="loading-single-cell-data.html#cb102-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PREX2, CENPE, AURKB, CRHBP, DNM3, KIF18B, NUSAP1, HLF, CIT, DIAPH3 </span></span>
<span id="cb102-40"><a href="loading-single-cell-data.html#cb102-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  GAPDH, RPS19, IGLL1, RPLP0, C1QTNF4, RPL28, RPL13A, RPL18A, RPS24, EEF1A1 </span></span>
<span id="cb102-41"><a href="loading-single-cell-data.html#cb102-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     GYPC, PRSS57, RPLP2, SMIM24, RPS8, RPL35, MIR181A1HG, RPS3, RPL27A, RPL14 </span></span>
<span id="cb102-42"><a href="loading-single-cell-data.html#cb102-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     RPL7, RPS6, RPL6, RPL8, RPL32, RACK1, ACTG1, RPL29, RPS3A, RPL7A</span></span>
<span id="cb102-43"><a href="loading-single-cell-data.html#cb102-43" aria-hidden="true" tabindex="-1"></a>cd34.singlecells <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(cd34.singlecells,<span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>))</span>
<span id="cb102-44"><a href="loading-single-cell-data.html#cb102-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:21 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span id="cb102-45"><a href="loading-single-cell-data.html#cb102-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:21 Read 6881 rows and found 50 numeric columns</span></span>
<span id="cb102-46"><a href="loading-single-cell-data.html#cb102-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:21 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span id="cb102-47"><a href="loading-single-cell-data.html#cb102-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:21 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span id="cb102-48"><a href="loading-single-cell-data.html#cb102-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span id="cb102-49"><a href="loading-single-cell-data.html#cb102-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span id="cb102-50"><a href="loading-single-cell-data.html#cb102-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; **************************************************|</span></span>
<span id="cb102-51"><a href="loading-single-cell-data.html#cb102-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:21 Writing NN index file to temp file /tmp/35395114/RtmptFIlEg/file14f2014f1bb643</span></span>
<span id="cb102-52"><a href="loading-single-cell-data.html#cb102-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:21 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span id="cb102-53"><a href="loading-single-cell-data.html#cb102-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:23 Annoy recall = 100%</span></span>
<span id="cb102-54"><a href="loading-single-cell-data.html#cb102-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:23 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span id="cb102-55"><a href="loading-single-cell-data.html#cb102-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:24 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span id="cb102-56"><a href="loading-single-cell-data.html#cb102-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:24 Commencing optimization for 500 epochs, with 293954 positive edges</span></span>
<span id="cb102-57"><a href="loading-single-cell-data.html#cb102-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:32 Optimization finished</span></span></code></pre></div>
<p>Plot single cell data.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="loading-single-cell-data.html#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">UMAPPlot</span>(cd34.singlecells,<span class="at">group.by =</span> <span class="st">&quot;celltype&quot;</span>,<span class="at">label =</span> T)</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-21-1.png" width="672" />
## Visualization of metacells in single-cell space</p>
<p>Now we can plot metacells from MetaCell2 in the single-cell space.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="loading-single-cell-data.html#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mc_projection</span>(</span>
<span id="cb104-2"><a href="loading-single-cell-data.html#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">sc.obj =</span> cd34.singlecells,</span>
<span id="cb104-3"><a href="loading-single-cell-data.html#cb104-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mc.obj =</span> cd34.metacell,</span>
<span id="cb104-4"><a href="loading-single-cell-data.html#cb104-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.membership =</span> cd34.metacell<span class="sc">@</span>misc<span class="sc">$</span>membership,</span>
<span id="cb104-5"><a href="loading-single-cell-data.html#cb104-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sc.reduction =</span> <span class="st">&quot;umap&quot;</span>,</span>
<span id="cb104-6"><a href="loading-single-cell-data.html#cb104-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">sc.label =</span> <span class="st">&quot;celltype&quot;</span>, <span class="co"># single cells will be colored according the sc.label</span></span>
<span id="cb104-7"><a href="loading-single-cell-data.html#cb104-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">metacell.label =</span> <span class="st">&quot;celltype&quot;</span> <span class="co"># metacells cell will be colored according the metacell.label</span></span>
<span id="cb104-8"><a href="loading-single-cell-data.html#cb104-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<p>And same for metacells obtained with SuperCell.</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="loading-single-cell-data.html#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mc_projection</span>(</span>
<span id="cb105-2"><a href="loading-single-cell-data.html#cb105-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">sc.obj =</span> cd34.singlecells,</span>
<span id="cb105-3"><a href="loading-single-cell-data.html#cb105-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mc.obj =</span> cd34.supercell,</span>
<span id="cb105-4"><a href="loading-single-cell-data.html#cb105-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell.membership =</span> cd34.supercell<span class="sc">@</span>misc<span class="sc">$</span>cell_membership,</span>
<span id="cb105-5"><a href="loading-single-cell-data.html#cb105-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">sc.reduction =</span> <span class="st">&quot;umap&quot;</span>,</span>
<span id="cb105-6"><a href="loading-single-cell-data.html#cb105-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">sc.label =</span> <span class="st">&quot;celltype&quot;</span>, <span class="co"># single cells will be colored according the sc.label</span></span>
<span id="cb105-7"><a href="loading-single-cell-data.html#cb105-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">metacell.label =</span> <span class="st">&quot;celltype&quot;</span> <span class="co"># metacells cell will be colored according the metacell.label</span></span>
<span id="cb105-8"><a href="loading-single-cell-data.html#cb105-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-23-1.png" width="672" />
## Compactness and separation</p>
<p>We can compute the compactness and separation of the metacells from the single cell pca.
With the <code>mc_compactnness</code> we can use the approach of SEACells by computing these metrics in a diffusion map obtained from the pca.</p>
<p>First for MetaCell2 metacells</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="loading-single-cell-data.html#cb106-1" aria-hidden="true" tabindex="-1"></a>cd34.metacell<span class="sc">$</span>compactness <span class="ot">&lt;-</span> <span class="fu">mc_compactness</span>(<span class="at">cell.membership =</span> cd34.metacell<span class="sc">@</span>misc<span class="sc">$</span>cell_membership, <span class="at">sc.obj =</span> cd34.singlecells,</span>
<span id="cb106-2"><a href="loading-single-cell-data.html#cb106-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">sc.reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">n.components =</span> <span class="dv">50</span>, <span class="at">diffusion.components =</span> T)</span>
<span id="cb106-3"><a href="loading-single-cell-data.html#cb106-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing compactness ...</span></span>
<span id="cb106-4"><a href="loading-single-cell-data.html#cb106-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb106-5"><a href="loading-single-cell-data.html#cb106-5" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> cd34.metacell, <span class="at">qc.metrics =</span> <span class="st">&quot;compactness&quot;</span>)</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-24-1.png" width="672" /></p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="loading-single-cell-data.html#cb107-1" aria-hidden="true" tabindex="-1"></a>cd34.metacell<span class="sc">$</span>separation <span class="ot">&lt;-</span> <span class="fu">mc_separation</span>(<span class="at">cell.membership =</span> cd34.metacell<span class="sc">@</span>misc<span class="sc">$</span>cell_membership, <span class="at">sc.obj =</span> cd34.singlecells,</span>
<span id="cb107-2"><a href="loading-single-cell-data.html#cb107-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">sc.reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">n.components =</span> <span class="dv">50</span>, <span class="at">diffusion.components =</span> T)</span>
<span id="cb107-3"><a href="loading-single-cell-data.html#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing separation ...</span></span>
<span id="cb107-4"><a href="loading-single-cell-data.html#cb107-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-5"><a href="loading-single-cell-data.html#cb107-5" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> cd34.metacell, <span class="at">qc.metrics =</span> <span class="st">&quot;separation&quot;</span>)</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-25-1.png" width="672" />
Same for metacell obtained with SuperCell.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="loading-single-cell-data.html#cb108-1" aria-hidden="true" tabindex="-1"></a>cd34.supercell<span class="sc">$</span>compactness <span class="ot">&lt;-</span> <span class="fu">mc_compactness</span>(<span class="at">cell.membership =</span> cd34.supercell<span class="sc">@</span>misc<span class="sc">$</span>cell_membership, <span class="at">sc.obj =</span> cd34.singlecells,</span>
<span id="cb108-2"><a href="loading-single-cell-data.html#cb108-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">sc.reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">n.components =</span> <span class="dv">50</span>, <span class="at">diffusion.components =</span> T)</span>
<span id="cb108-3"><a href="loading-single-cell-data.html#cb108-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing compactness ...</span></span>
<span id="cb108-4"><a href="loading-single-cell-data.html#cb108-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-5"><a href="loading-single-cell-data.html#cb108-5" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> cd34.supercell, <span class="at">qc.metrics =</span> <span class="st">&quot;compactness&quot;</span>)</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-26-1.png" width="672" /></p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="loading-single-cell-data.html#cb109-1" aria-hidden="true" tabindex="-1"></a>cd34.supercell<span class="sc">$</span>separation <span class="ot">&lt;-</span> <span class="fu">mc_separation</span>(<span class="at">cell.membership =</span> cd34.supercell<span class="sc">@</span>misc<span class="sc">$</span>cell_membership, <span class="at">sc.obj =</span> cd34.singlecells,</span>
<span id="cb109-2"><a href="loading-single-cell-data.html#cb109-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">sc.reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">n.components =</span> <span class="dv">50</span>, <span class="at">diffusion.components =</span> T)</span>
<span id="cb109-3"><a href="loading-single-cell-data.html#cb109-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing separation ...</span></span>
<span id="cb109-4"><a href="loading-single-cell-data.html#cb109-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-5"><a href="loading-single-cell-data.html#cb109-5" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> cd34.supercell, <span class="at">qc.metrics =</span> <span class="st">&quot;separation&quot;</span>)</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-27-1.png" width="672" />
## Inner normalized variance (INV)
We can compute the INV for the MetaCell2 and SuperCell metacells as defined in MetaCell paper.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="loading-single-cell-data.html#cb110-1" aria-hidden="true" tabindex="-1"></a>cd34.supercell<span class="sc">$</span>INV <span class="ot">&lt;-</span> <span class="fu">mc_INV</span>(<span class="at">cell.membership =</span> cd34.supercell<span class="sc">@</span>misc<span class="sc">$</span>cell_membership, <span class="at">sc.obj =</span> cd34.singlecells)</span>
<span id="cb110-2"><a href="loading-single-cell-data.html#cb110-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing INV ...</span></span>
<span id="cb110-3"><a href="loading-single-cell-data.html#cb110-3" aria-hidden="true" tabindex="-1"></a>cd34.metacell<span class="sc">$</span>INV <span class="ot">&lt;-</span> <span class="fu">mc_INV</span>(<span class="at">cell.membership =</span> cd34.metacell<span class="sc">@</span>misc<span class="sc">$</span>cell_membership, <span class="at">sc.obj =</span> cd34.singlecells)</span>
<span id="cb110-4"><a href="loading-single-cell-data.html#cb110-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing INV ...</span></span></code></pre></div>
<p>Comparison of MetaCell2 and SuperCell for compactness (lower the better).</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="loading-single-cell-data.html#cb111-1" aria-hidden="true" tabindex="-1"></a>cd34.metacell<span class="sc">$</span>tool <span class="ot">&lt;-</span> <span class="st">&quot;MetaCell2&quot;</span></span>
<span id="cb111-2"><a href="loading-single-cell-data.html#cb111-2" aria-hidden="true" tabindex="-1"></a>cd34.supercell<span class="sc">$</span>tool <span class="ot">&lt;-</span> <span class="st">&quot;SuperCell&quot;</span></span>
<span id="cb111-3"><a href="loading-single-cell-data.html#cb111-3" aria-hidden="true" tabindex="-1"></a>data.metrics <span class="ot">&lt;-</span> <span class="fu">rbind</span>(cd34.metacell<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="st">&quot;tool&quot;</span>,<span class="st">&quot;INV&quot;</span>,<span class="st">&quot;compactness&quot;</span>,<span class="st">&quot;separation&quot;</span>)],</span>
<span id="cb111-4"><a href="loading-single-cell-data.html#cb111-4" aria-hidden="true" tabindex="-1"></a>                      cd34.supercell<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="st">&quot;tool&quot;</span>,<span class="st">&quot;INV&quot;</span>,<span class="st">&quot;compactness&quot;</span>,<span class="st">&quot;separation&quot;</span>)])</span>
<span id="cb111-5"><a href="loading-single-cell-data.html#cb111-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-6"><a href="loading-single-cell-data.html#cb111-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data.metrics,<span class="fu">aes</span>(<span class="at">y=</span>compactness,<span class="at">x=</span>tool,<span class="at">fill =</span> tool)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-29-1.png" width="672" />
Comparison of MetaCell2 and SuperCell for separation (higher the better).</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="loading-single-cell-data.html#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data.metrics,<span class="fu">aes</span>(<span class="at">y=</span>separation,<span class="at">x=</span>tool,<span class="at">fill =</span> tool)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-30-1.png" width="672" /></p>
<p>Comparison of MetaCell2 and SuperCell for INV (lower the better).</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="loading-single-cell-data.html#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data.metrics,<span class="fu">aes</span>(<span class="at">y=</span>INV,<span class="at">x=</span>tool,<span class="at">fill =</span> tool)) <span class="sc">+</span> <span class="fu">geom_boxplot</span>()</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-31-1.png" width="672" /></p>

</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="quality-control-with-the-metacellanalysistoolkit-package.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="downstream-analysis-of-metacells.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/23-downstream_continuous.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/23-downstream_continuous.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
