<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.1 MC2 (Python) | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="3.1 MC2 (Python) | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.1 MC2 (Python) | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-07-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="constructing-metacells-for-discrete-data.html"/>
<link rel="next" href="seacells-python.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="book-structure.html"><a href="book-structure.html"><i class="fa fa-check"></i><b>1.1</b> Book structure</a></li>
<li class="chapter" data-level="1.2" data-path="installation-and-requirements.html"><a href="installation-and-requirements.html"><i class="fa fa-check"></i><b>1.2</b> Installation and requirements</a></li>
<li class="chapter" data-level="1.3" data-path="render-book.html"><a href="render-book.html"><i class="fa fa-check"></i><b>1.3</b> Render book</a></li>
<li class="chapter" data-level="1.4" data-path="get-data.html"><a href="get-data.html"><i class="fa fa-check"></i><b>1.4</b> Get data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a>
<ul>
<li class="chapter" data-level="2.1" data-path="MC2.html"><a href="MC2.html"><i class="fa fa-check"></i><b>2.1</b> Metacell (MC2)</a></li>
<li class="chapter" data-level="2.2" data-path="SuperCell.html"><a href="SuperCell.html"><i class="fa fa-check"></i><b>2.2</b> SuperCell</a></li>
<li class="chapter" data-level="2.3" data-path="SEACells.html"><a href="SEACells.html"><i class="fa fa-check"></i><b>2.3</b> SEACells</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="constructing-metacells-for-discrete-data.html"><a href="constructing-metacells-for-discrete-data.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells (for ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="mc2-python.html"><a href="mc2-python.html"><i class="fa fa-check"></i><b>3.1</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="mc2-python.html"><a href="mc2-python.html#data-loading"><i class="fa fa-check"></i><b>3.1.1</b> Data loading</a></li>
<li class="chapter" data-level="3.1.2" data-path="mc2-python.html"><a href="mc2-python.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.3" data-path="mc2-python.html"><a href="mc2-python.html#building-metacells"><i class="fa fa-check"></i><b>3.1.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.4" data-path="mc2-python.html"><a href="mc2-python.html#visualize-metacells"><i class="fa fa-check"></i><b>3.1.4</b> Visualize metacells</a></li>
<li class="chapter" data-level="3.1.5" data-path="mc2-python.html"><a href="mc2-python.html#metacell-qc"><i class="fa fa-check"></i><b>3.1.5</b> Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="seacells-python.html"><a href="seacells-python.html"><i class="fa fa-check"></i><b>3.2</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="seacells-python.html"><a href="seacells-python.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.1</b> Data loading</a></li>
<li class="chapter" data-level="3.2.2" data-path="seacells-python.html"><a href="seacells-python.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.3" data-path="seacells-python.html"><a href="seacells-python.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.4" data-path="seacells-python.html"><a href="seacells-python.html#visualize-metacells-1"><i class="fa fa-check"></i><b>3.2.4</b> Visualize metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="seacells-python.html"><a href="seacells-python.html#metacell-qc-1"><i class="fa fa-check"></i><b>3.2.5</b> Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="supercell-r.html"><a href="supercell-r.html"><i class="fa fa-check"></i><b>3.3</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="supercell-r.html"><a href="supercell-r.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.1</b> Data loading</a></li>
<li class="chapter" data-level="3.3.2" data-path="supercell-r.html"><a href="supercell-r.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.3" data-path="supercell-r.html"><a href="supercell-r.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.4" data-path="supercell-r.html"><a href="supercell-r.html#visualize-metacells-2"><i class="fa fa-check"></i><b>3.3.4</b> Visualize metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="supercell-r.html"><a href="supercell-r.html#metacell-qc-2"><i class="fa fa-check"></i><b>3.3.5</b> Metacell QC</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><a href="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><i class="fa fa-check"></i><b>4</b> Downstream analysis of metacells (for a discrete dataset)</a>
<ul>
<li class="chapter" data-level="4.1" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html"><i class="fa fa-check"></i><b>4.1</b> Standard analysis (R)</a></li>
<li class="chapter" data-level="4.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>4.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction"><i class="fa fa-check"></i><b>4.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="4.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering"><i class="fa fa-check"></i><b>4.2.2</b> Clustering</a></li>
<li class="chapter" data-level="4.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis"><i class="fa fa-check"></i><b>4.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>4.3</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#grn"><i class="fa fa-check"></i><b>4.3.1</b> GRN</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>4.4</b> Sample-weighted analysis</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="command-line-tool-for-mc-construction.html"><a href="command-line-tool-for-mc-construction.html"><i class="fa fa-check"></i><b>5</b> Command line tool for MC construction</a></li>
<li class="chapter" data-level="6" data-path="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><a href="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><i class="fa fa-check"></i><b>6</b> Comparison of metacells usage for a discrete and continuous data</a>
<ul>
<li class="chapter" data-level="6.1" data-path="continuous-metacells-have-lower-purity.html"><a href="continuous-metacells-have-lower-purity.html"><i class="fa fa-check"></i><b>6.1</b> Continuous metacells have lower purity</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-integration-with-metacells.html"><a href="data-integration-with-metacells.html"><i class="fa fa-check"></i><b>7</b> Data integration with metacells</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mc2-python" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> MC2 (Python)<a href="mc2-python.html#mc2-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this section, we construct metacells using <a href="https://github.com/tanaylab/metacells">Metacell-2 (MC2)</a>. The code is adapted from the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">author’s tutorial</a>. For more information on the method, please refer to the section 1 of chapter 2.</p>
<div id="importing-python-packages" class="section level4 unnumbered hasAnchor">
<h4>Importing python packages<a href="mc2-python.html#importing-python-packages" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To run Metacell-2, the following python packages need to be imported:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="mc2-python.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb8-2"><a href="mc2-python.html#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb8-3"><a href="mc2-python.html#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb8-4"><a href="mc2-python.html#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb8-5"><a href="mc2-python.html#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb8-6"><a href="mc2-python.html#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb8-7"><a href="mc2-python.html#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb8-8"><a href="mc2-python.html#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="mc2-python.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb9-2"><a href="mc2-python.html#cb9-2" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>)</span>
<span id="cb9-3"><a href="mc2-python.html#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span></code></pre></div>
<p>If you don’t have these packages installed, please refer to the section 2 of chapter 1.</p>
</div>
<div id="data-loading" class="section level3 hasAnchor" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Data loading<a href="mc2-python.html#data-loading" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will run Metacell-2 on a single-cell dataset composed of XX peripheral blood mononuclear cells (PBMCs). Please follow the section 4 from Chapter 1 to retrieve these data from the scanpy package and save the data in the following file: “data/3k_pbmc/singlecell_anndata_filtered.h5ad”.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="mc2-python.html#cb10-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;MC2&quot;</span></span>
<span id="cb10-2"><a href="mc2-python.html#cb10-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">&quot;3k_pbmc&quot;</span></span>
<span id="cb10-3"><a href="mc2-python.html#cb10-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(<span class="st">&quot;data&quot;</span>, proj_name, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span></code></pre></div>
<p>We initialize the name of the anndata (in the unstructured annotations) object using the <code>ut.set_name</code> function from the Metacells package.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb11-1"><a href="mc2-python.html#cb11-1" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(ad, proj_name)</span></code></pre></div>
</div>
<div id="filtering-steps" class="section level3 hasAnchor" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Filtering steps<a href="mc2-python.html#filtering-steps" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before building the metacells the Metacell-2 authors recommend to filter the single-cell data in two-steps (See <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">original vignette</a>).
A first filtering step consists in filtering genes based on biological knowledge (<em>e.g.</em> mitochrondrial genes) or based on their expression levels. The latter genes include genes with zero expression or low expression levels, “noisy lonely genes” (<em>i.e.</em>, genes with high expression levels but no correlation with any other gene).
A second filtering step consists in filtering cells based on their UMI counts.</p>
<div id="gene-filtering" class="section level4 unnumbered hasAnchor">
<h4>Gene filtering<a href="mc2-python.html#gene-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In section XX form Chapter XX, we pre-processed the raw scRNA-Seq data and excluded genes with low expression as well as mitochondrial genes. In the following code chunk, we identify additional genes to filter using the <code>analyze_clean_genes</code> and <code>pick_clean_genes</code> functions from the Metacells package, which returns three sets of genes: i) the known-to-be-excluded genes defined by the user as an array of gene names or gene names patterns, ii) the properly sampled genes, and iii) the “noisy lonely genes”.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb12-1"><a href="mc2-python.html#cb12-1" aria-hidden="true" tabindex="-1"></a>excluded_gene_names <span class="op">=</span> [] <span class="co"># for example, [&#39;IGHMBP2&#39;, &#39;IGLL1&#39;, &#39;IGLL5&#39;, &#39;IGLON5&#39;, &#39;NEAT1&#39;, &#39;TMSB10&#39;, &#39;TMSB4X&#39;]</span></span>
<span id="cb12-2"><a href="mc2-python.html#cb12-2" aria-hidden="true" tabindex="-1"></a>excluded_gene_patterns <span class="op">=</span> [<span class="st">&#39;MT-.*&#39;</span>]</span>
<span id="cb12-3"><a href="mc2-python.html#cb12-3" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_genes(ad,</span>
<span id="cb12-4"><a href="mc2-python.html#cb12-4" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_names<span class="op">=</span>excluded_gene_names,</span>
<span id="cb12-5"><a href="mc2-python.html#cb12-5" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_patterns<span class="op">=</span>excluded_gene_patterns,</span>
<span id="cb12-6"><a href="mc2-python.html#cb12-6" aria-hidden="true" tabindex="-1"></a>                          random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb12-7"><a href="mc2-python.html#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[properly_sampled_gene]: 16579 true (50.64%) out of 32738 bools</span></span>
<span id="cb12-8"><a href="mc2-python.html#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[excluded_gene]: 13 true (0.03971%) out of 32738 bools</span></span>
<span id="cb12-9"><a href="mc2-python.html#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[noisy_lonely_gene]: 0 true (0%) out of 32738 bools</span></span>
<span id="cb12-10"><a href="mc2-python.html#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="mc2-python.html#cb12-11" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_genes(ad)</span>
<span id="cb12-12"><a href="mc2-python.html#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[clean_gene]: 16566 true (50.6%) out of 32738 bools</span></span></code></pre></div>
</div>
<div id="cell-filtering" class="section level4 unnumbered hasAnchor">
<h4>Cell filtering<a href="mc2-python.html#cell-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The first round of cell cleaning implies filtering out cell with very low and very hight UMI content (<code>properly_sampled_min_cell_total</code>, <code>properly_sampled_max_cell_total</code> parameters).<br />
The second round includes cell filtering based on UMI counts in excluded genes (<code>properly_sampled_max_excluded_genes_fraction</code> parameter).
Since our dataset have been pre-filtered, very lenient cutoffs will be used in this tutorial (<code>properly_sampled_min_cell_total</code>, <code>properly_sampled_max_cell_total</code> and <code>properly_sampled_max_excluded_genes_fraction</code>) such that all the cells are kept for the metacell construction.
The following code chunk defines these parameters. To adapt them to your datasets, we advise you to explore the distributions of total UMI counts and UMI counts in excluded genes, as recommended and descrided in the Metacell-2 <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">original vignette</a>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb13-1"><a href="mc2-python.html#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">### The first round of cell cleaning (high/low UMIs)</span></span>
<span id="cb13-2"><a href="mc2-python.html#cb13-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_min_cell_total <span class="op">=</span> <span class="dv">200</span> </span>
<span id="cb13-3"><a href="mc2-python.html#cb13-3" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_cell_total <span class="op">=</span> <span class="dv">10000</span> </span>
<span id="cb13-4"><a href="mc2-python.html#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">## The second round of cell cleaning (content of excluded genes, e.g., mito-genes)</span></span>
<span id="cb13-5"><a href="mc2-python.html#cb13-5" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_excluded_genes_fraction <span class="op">=</span> <span class="fl">0.25</span></span></code></pre></div>
<p>The set of cells to be filtered is defined using the <code>analyze_clean_cells</code> and <code>pick_clean_cells</code> functions from the Metacell-2 package.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="mc2-python.html#cb14-1" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_cells(</span>
<span id="cb14-2"><a href="mc2-python.html#cb14-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb14-3"><a href="mc2-python.html#cb14-3" aria-hidden="true" tabindex="-1"></a>    properly_sampled_min_cell_total <span class="op">=</span> properly_sampled_min_cell_total,</span>
<span id="cb14-4"><a href="mc2-python.html#cb14-4" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_cell_total <span class="op">=</span> properly_sampled_max_cell_total,</span>
<span id="cb14-5"><a href="mc2-python.html#cb14-5" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_excluded_genes_fraction <span class="op">=</span> properly_sampled_max_excluded_genes_fraction</span>
<span id="cb14-6"><a href="mc2-python.html#cb14-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-7"><a href="mc2-python.html#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[properly_sampled_cell]: 2638 true (100%) out of 2638 bools</span></span>
<span id="cb14-8"><a href="mc2-python.html#cb14-8" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_cells(ad)</span>
<span id="cb14-9"><a href="mc2-python.html#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[clean_cell]: 2638 true (100%) out of 2638 bools</span></span></code></pre></div>
<p>After performing the two-step filtering (genes and cells), the “cleaned” data can be extracted using the <code>pl.extract_clean_data</code> prior to metacells construction.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="mc2-python.html#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract clean dataset (with filtered cells and genes)</span></span>
<span id="cb15-2"><a href="mc2-python.html#cb15-2" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> mc.pl.extract_clean_data(ad)</span>
<span id="cb15-3"><a href="mc2-python.html#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[full_cell_index]: 2638 int64s</span></span>
<span id="cb15-4"><a href="mc2-python.html#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[full_gene_index]: 16566 int64s</span></span></code></pre></div>
</div>
</div>
<div id="building-metacells" class="section level3 hasAnchor" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Building metacells<a href="mc2-python.html#building-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="estimate-target_metacell_size-gamma" class="section level4 unnumbered hasAnchor">
<h4>Estimate target_metacell_size (gamma)<a href="mc2-python.html#estimate-target_metacell_size-gamma" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In contrast to the SuperCell and SEACells, Metacell-2 does not allow to explicitly obtain metacell data at a user-defined graining level.
Instead, to vary the graining level, we have to vary the <code>target_metacell_size</code> parameter, that is <code>160000</code> by default. Here we provide a chunk to calibrate this value to reach a desired graining level.
Please, increase or decrease the <code>scale</code> parameter if the obtained graining level (<code>gamma_obtained</code>) is lower or larger than the requested one (<code>gamma</code>).</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb16-1"><a href="mc2-python.html#cb16-1" aria-hidden="true" tabindex="-1"></a>gamma   <span class="op">=</span> <span class="dv">50</span> <span class="co"># graining level</span></span>
<span id="cb16-2"><a href="mc2-python.html#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&#39;The requested graining level is </span><span class="sc">{</span>gamma<span class="sc">}</span><span class="ss">, lets estimate the target_metacell_size that should result in such graining level.&#39;</span>)</span>
<span id="cb16-3"><a href="mc2-python.html#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The requested graining level is 50, lets estimate the target_metacell_size that should result in such graining level.</span></span>
<span id="cb16-4"><a href="mc2-python.html#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="mc2-python.html#cb16-5" aria-hidden="true" tabindex="-1"></a>scale <span class="op">=</span> <span class="dv">2</span> <span class="co"># increase or decrease if the obtained graining level (`gamma_obtained`) is significantly &gt; or &lt; then the requested one `gamma`</span></span>
<span id="cb16-6"><a href="mc2-python.html#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="mc2-python.html#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co"># estimated mean UMI content in downsampled data</span></span>
<span id="cb16-8"><a href="mc2-python.html#cb16-8" aria-hidden="true" tabindex="-1"></a>total_umis_of_cells <span class="op">=</span> mc.ut.get_o_numpy(ad, name<span class="op">=</span><span class="st">&#39;__x__&#39;</span>, <span class="bu">sum</span><span class="op">=</span><span class="va">True</span>)</span>
<span id="cb16-9"><a href="mc2-python.html#cb16-9" aria-hidden="true" tabindex="-1"></a>est_downsample_UMI <span class="op">=</span> np.quantile(np.array(total_umis_of_cells), <span class="fl">0.05</span>)</span>
<span id="cb16-10"><a href="mc2-python.html#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="mc2-python.html#cb16-11" aria-hidden="true" tabindex="-1"></a>target_metacell_size <span class="op">=</span> <span class="bu">int</span>(est_downsample_UMI <span class="op">*</span> gamma <span class="op">*</span> scale)</span>
<span id="cb16-12"><a href="mc2-python.html#cb16-12" aria-hidden="true" tabindex="-1"></a>target_metacell_size</span>
<span id="cb16-13"><a href="mc2-python.html#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 98900</span></span></code></pre></div>
</div>
<div id="metacells-identification-using-the-divide-and-conquer-approach" class="section level4 unnumbered hasAnchor">
<h4>Metacells identification using the divide and conquer approach<a href="mc2-python.html#metacells-identification-using-the-divide-and-conquer-approach" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Metacell-2 uses its own feature selection approach (<em>i.e.</em>, selection of genes used to build metacells).
Additionally, users can explicitly specify which features should be used by providing two arguments:<br />
i) <code>feature_gene_names</code>, <em>i.e.</em>, genes that have to be included and, ii) <code>forbidden_gene_names</code>, <em>i.e.</em>, genes that have to be excluded.
Examples of forbidden are cell cycle genes which reflect true biological signal but could bias the metacells construction. The Metacell-2 <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">original vignette</a> provides additional information to identify forbidden genes.
The previous arguments for feature selection are taken as input by the <code>pl.divide_and_conquer_pipeline</code> function which builds the metacells.
Note that by default all cores of the system will be used for the metacells construction. More information are available on the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">original vignette</a> to change this behavior and adapt the number of cores (see <code>ut.set_processors_count</code>) or the number processed in parallel (see <code>pl.set_max_parallel_piles</code>).</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb17-1"><a href="mc2-python.html#cb17-1" aria-hidden="true" tabindex="-1"></a>mc.pl.divide_and_conquer_pipeline(</span>
<span id="cb17-2"><a href="mc2-python.html#cb17-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb17-3"><a href="mc2-python.html#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#feature_gene_names   = feature_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb17-4"><a href="mc2-python.html#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#forbidden_gene_names = forbidden_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb17-5"><a href="mc2-python.html#cb17-5" aria-hidden="true" tabindex="-1"></a>    target_metacell_size <span class="op">=</span> target_metacell_size,</span>
<span id="cb17-6"><a href="mc2-python.html#cb17-6" aria-hidden="true" tabindex="-1"></a>    random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb17-7"><a href="mc2-python.html#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb17-8"><a href="mc2-python.html#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene_module]: 16566 int32 elements with all outliers (100%)</span></span>
<span id="cb17-9"><a href="mc2-python.html#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cells_rare_gene_module]: 2638 int32 elements with all outliers (100%)</span></span>
<span id="cb17-10"><a href="mc2-python.html#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[rare_cell]: 0 true (0%) out of 2638 bools</span></span>
<span id="cb17-11"><a href="mc2-python.html#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.layers[downsampled]: csr_matrix 2638 X 16566 float32s (1218892 &gt; 0)</span></span>
<span id="cb17-12"><a href="mc2-python.html#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[downsample_samples]: 989</span></span>
<span id="cb17-13"><a href="mc2-python.html#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_top3_gene]: 552 true (3.332%) out of 16566 bools</span></span>
<span id="cb17-14"><a href="mc2-python.html#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_total_gene]: 4519 true (27.28%) out of 16566 bools</span></span>
<span id="cb17-15"><a href="mc2-python.html#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_relative_variance_gene]: 3027 true (18.27%) out of 16566 bools</span></span>
<span id="cb17-16"><a href="mc2-python.html#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[feature_gene]: 293 true (1.769%) out of 16566 bools</span></span>
<span id="cb17-17"><a href="mc2-python.html#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obsp[obs_similarity]: ndarray 2638 X 2638 float32s</span></span>
<span id="cb17-18"><a href="mc2-python.html#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obsp[obs_outgoing_weights]: csr_matrix 2638 X 2638 float32s (117211 &gt; 0)</span></span>
<span id="cb17-19"><a href="mc2-python.html#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[seed]: 496 outliers (18.8%) out of 2638 int32 elements with 61 groups with mean size 35.11</span></span>
<span id="cb17-20"><a href="mc2-python.html#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[candidate]: 6 outliers (0.2274%) out of 2638 int32 elements with 67 groups with mean size 39.28</span></span>
<span id="cb17-21"><a href="mc2-python.html#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[gene_deviant_votes]: 942 positive (5.686%) out of 16566 int32s</span></span>
<span id="cb17-22"><a href="mc2-python.html#cb17-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cell_deviant_votes]: 889 positive (33.7%) out of 2638 int32s</span></span>
<span id="cb17-23"><a href="mc2-python.html#cb17-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[dissolved]: 10 true (0.3791%) out of 2638 bools</span></span>
<span id="cb17-24"><a href="mc2-python.html#cb17-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell]: 905 outliers (34.31%) out of 2638 int64 elements with 66 groups with mean size 26.26</span></span>
<span id="cb17-25"><a href="mc2-python.html#cb17-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[outlier]: 905 true (34.31%) out of 2638 bools</span></span>
<span id="cb17-26"><a href="mc2-python.html#cb17-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[pre_directs]: 0</span></span>
<span id="cb17-27"><a href="mc2-python.html#cb17-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[directs]: 1</span></span>
<span id="cb17-28"><a href="mc2-python.html#cb17-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_high_total_gene]: * &lt;- 0</span></span>
<span id="cb17-29"><a href="mc2-python.html#cb17-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_total_gene]: 4519 positive (27.28%) out of 16566 int32s</span></span>
<span id="cb17-30"><a href="mc2-python.html#cb17-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_high_relative_variance_gene]: * &lt;- 0</span></span>
<span id="cb17-31"><a href="mc2-python.html#cb17-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_relative_variance_gene]: 3027 positive (18.27%) out of 16566 int32s</span></span>
<span id="cb17-32"><a href="mc2-python.html#cb17-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[forbidden_gene]: * &lt;- False</span></span>
<span id="cb17-33"><a href="mc2-python.html#cb17-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_feature_gene]: * &lt;- 0</span></span>
<span id="cb17-34"><a href="mc2-python.html#cb17-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[feature_gene]: 293 positive (1.769%) out of 16566 int32s</span></span>
<span id="cb17-35"><a href="mc2-python.html#cb17-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_gene_deviant_votes]: * &lt;- 0</span></span>
<span id="cb17-36"><a href="mc2-python.html#cb17-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_cell_directs]: * &lt;- 0</span></span>
<span id="cb17-37"><a href="mc2-python.html#cb17-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cell_directs]: * &lt;- 0</span></span>
<span id="cb17-38"><a href="mc2-python.html#cb17-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_pile]: * &lt;- -1</span></span>
<span id="cb17-39"><a href="mc2-python.html#cb17-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pile]: * &lt;- 0</span></span>
<span id="cb17-40"><a href="mc2-python.html#cb17-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_candidate]: * &lt;- -1</span></span>
<span id="cb17-41"><a href="mc2-python.html#cb17-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_cell_deviant_votes]: * &lt;- 0</span></span>
<span id="cb17-42"><a href="mc2-python.html#cb17-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_dissolved]: * &lt;- False</span></span>
<span id="cb17-43"><a href="mc2-python.html#cb17-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_metacell]: * &lt;- -1</span></span>
<span id="cb17-44"><a href="mc2-python.html#cb17-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-45"><a href="mc2-python.html#cb17-45" aria-hidden="true" tabindex="-1"></a>ad.obs.metacell.head</span>
<span id="cb17-46"><a href="mc2-python.html#cb17-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;bound method NDFrame.head of index</span></span>
<span id="cb17-47"><a href="mc2-python.html#cb17-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATACAACCAC-1    36</span></span>
<span id="cb17-48"><a href="mc2-python.html#cb17-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGAGCTAC-1    14</span></span>
<span id="cb17-49"><a href="mc2-python.html#cb17-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGATCAGC-1    26</span></span>
<span id="cb17-50"><a href="mc2-python.html#cb17-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGCTTCCG-1    62</span></span>
<span id="cb17-51"><a href="mc2-python.html#cb17-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGTATGCG-1    -1</span></span>
<span id="cb17-52"><a href="mc2-python.html#cb17-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     ..</span></span>
<span id="cb17-53"><a href="mc2-python.html#cb17-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCGAACTCTCAT-1    62</span></span>
<span id="cb17-54"><a href="mc2-python.html#cb17-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTGAGGCA-1    14</span></span>
<span id="cb17-55"><a href="mc2-python.html#cb17-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTTCCTCG-1    -1</span></span>
<span id="cb17-56"><a href="mc2-python.html#cb17-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGAGAGGC-1    28</span></span>
<span id="cb17-57"><a href="mc2-python.html#cb17-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGCCTCAC-1    27</span></span>
<span id="cb17-58"><a href="mc2-python.html#cb17-58" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Name: metacell, Length: 2638, dtype: int64&gt;</span></span></code></pre></div>
</div>
<div id="retrieve-aggregated-metacell-data" class="section level4 unnumbered hasAnchor">
<h4>Retrieve aggregated metacell data<a href="mc2-python.html#retrieve-aggregated-metacell-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The <code>pl.divide_and_conquer_pipeline</code> function associate each cell to a metacell or defines the cell as outlier. These assignments are found in the <code>obs</code> layer of the anndata object
The function <code>pl.collect_metacells</code> should be used to subsequently retrieve an anndata object containing the data at the metacells level instead of the single-cell level.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb18-1"><a href="mc2-python.html#cb18-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-2"><a href="mc2-python.html#cb18-2" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> mc.pl.collect_metacells(ad, name<span class="op">=</span><span class="st">&#39;metacells&#39;</span>)</span>
<span id="cb18-3"><a href="mc2-python.html#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[excluded_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb18-4"><a href="mc2-python.html#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[clean_gene]: 16566 true (100%) out of 16566 bools</span></span>
<span id="cb18-5"><a href="mc2-python.html#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[forbidden_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb18-6"><a href="mc2-python.html#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[pre_feature_gene]: 0 positive (0%) out of 16566 int32s</span></span>
<span id="cb18-7"><a href="mc2-python.html#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[feature_gene]: 293 positive (1.769%) out of 16566 int32s</span></span>
<span id="cb18-8"><a href="mc2-python.html#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[pile]: [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ]</span></span>
<span id="cb18-9"><a href="mc2-python.html#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[candidate]: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, -2, 35 ]</span></span>
<span id="cb18-10"><a href="mc2-python.html#cb18-10" aria-hidden="true" tabindex="-1"></a>mc_ad.shape</span>
<span id="cb18-11"><a href="mc2-python.html#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; (66, 16566)</span></span></code></pre></div>
<p><strong>Comparing the obtained and requested graining level</strong></p>
<p>In the following code chunk, we estimate whether a deviation of the obtained gamma from the requested gamma is acceptable. If not, we suggest to increase or decrease the <code>scale</code> parameter to approach the desired graining level.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb19-1"><a href="mc2-python.html#cb19-1" aria-hidden="true" tabindex="-1"></a>gamma_obtained <span class="op">=</span> ad.shape[<span class="dv">0</span>]<span class="op">/</span>mc_ad.shape[<span class="dv">0</span>]</span>
<span id="cb19-2"><a href="mc2-python.html#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gamma_obtained)</span>
<span id="cb19-3"><a href="mc2-python.html#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 39.96969696969697</span></span>
<span id="cb19-4"><a href="mc2-python.html#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="mc2-python.html#cb19-5" aria-hidden="true" tabindex="-1"></a>gamma_dev <span class="op">=</span> (gamma_obtained <span class="op">-</span> gamma)<span class="op">/</span>gamma</span>
<span id="cb19-6"><a href="mc2-python.html#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">abs</span>(gamma_dev) <span class="op">&lt;</span> <span class="fl">0.3</span>: </span>
<span id="cb19-7"><a href="mc2-python.html#cb19-7" aria-hidden="true" tabindex="-1"></a>    gamma_dev <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb19-8"><a href="mc2-python.html#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="mc2-python.html#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> gamma_dev <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb19-10"><a href="mc2-python.html#cb19-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Increase `target_metacell_size` parameter by increasing `scale` and re-run metacell divide_and_conquer_pipeline() to get larger graining level&quot;</span>)</span>
<span id="cb19-11"><a href="mc2-python.html#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb19-12"><a href="mc2-python.html#cb19-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Deacrease `target_metacell_size` parameter by decreasing `scale` and re-run metacell divide_and_conquer_pipeline() to get smaller graining level&quot;</span>)</span>
<span id="cb19-13"><a href="mc2-python.html#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb19-14"><a href="mc2-python.html#cb19-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size` &quot;</span>)</span>
<span id="cb19-15"><a href="mc2-python.html#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size`</span></span></code></pre></div>
<p>If the obtained graining level is not acceptable and you updated <code>scale</code> parameter according to the suggestion, do not forget to re-run chunk <a href="#chunk:mc2-divide-n-conquer"><strong>??</strong></a> <a href="#aggregate-mc2"><strong>??</strong></a>.</p>
</div>
</div>
<div id="visualize-metacells" class="section level3 hasAnchor" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> Visualize metacells<a href="mc2-python.html#visualize-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The following code chunk adds a columns named <code>membership</code> and containing the single_cell assignments to the obs attribute in the anndata object containing the raw data.
This annotation will be used in the mc_QC package to compute metacells quality metrics. We also save the single-cell metadata in the metacell anndata object.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="mc2-python.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a membership -- index of metacells to which single cells belong to </span></span>
<span id="cb20-2"><a href="mc2-python.html#cb20-2" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&#39;membership&#39;</span>] <span class="op">=</span> [<span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">0</span> <span class="cf">else</span> np.nan <span class="cf">for</span> i <span class="kw">in</span> ad.obs.metacell] </span>
<span id="cb20-3"><a href="mc2-python.html#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="mc2-python.html#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Save single-cell metadata (i.e., `raw.obs` dataframe) in the metacell adata object</span></span>
<span id="cb20-5"><a href="mc2-python.html#cb20-5" aria-hidden="true" tabindex="-1"></a>mc_ad.uns <span class="op">=</span> ad.uns.copy()</span>
<span id="cb20-6"><a href="mc2-python.html#cb20-6" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.obs&#39;</span>] <span class="op">=</span> ad.obs.copy()</span>
<span id="cb20-7"><a href="mc2-python.html#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="mc2-python.html#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the requested gamma</span></span>
<span id="cb20-9"><a href="mc2-python.html#cb20-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;gamma&#39;</span>] <span class="op">=</span> gamma</span></code></pre></div>
<div id="compute-latent-space-for-metacell-qc" class="section level4 unnumbered hasAnchor">
<h4>Compute latent space for metacell QC<a href="mc2-python.html#compute-latent-space-for-metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To visualize the metacells, we can project the metacells on the single-cell UMAP representation. To run UMAP, we will generate in the next code chunk a lower-dimentional embedding of the data, so far not needed since the MC2 methods builds metacells from gene expression data and not from latent space.
Also, note that some of the QC metrics (e.g., <strong>compactness</strong> and <strong>separation</strong>), that we will compute in the next section of this tutorial, are computed from this latent space.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb21-1"><a href="mc2-python.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save count as a separate layer</span></span>
<span id="cb21-2"><a href="mc2-python.html#cb21-2" aria-hidden="true" tabindex="-1"></a>ad.layers[<span class="st">&#39;counts&#39;</span>] <span class="op">=</span> ad.X</span>
<span id="cb21-3"><a href="mc2-python.html#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="mc2-python.html#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the counts to &quot;.raw&quot; attribute of the anndata since it is necessary for downstream analysis</span></span>
<span id="cb21-5"><a href="mc2-python.html#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This step should be performed after filtering </span></span>
<span id="cb21-6"><a href="mc2-python.html#cb21-6" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(ad.layers[<span class="st">&#39;counts&#39;</span>])</span>
<span id="cb21-7"><a href="mc2-python.html#cb21-7" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> ad.obs_names, ad.var_names</span>
<span id="cb21-8"><a href="mc2-python.html#cb21-8" aria-hidden="true" tabindex="-1"></a>ad.raw <span class="op">=</span> raw_ad</span>
<span id="cb21-9"><a href="mc2-python.html#cb21-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-10"><a href="mc2-python.html#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="mc2-python.html#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize cells, log transform and compute highly variable genes</span></span>
<span id="cb21-12"><a href="mc2-python.html#cb21-12" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(ad)</span>
<span id="cb21-13"><a href="mc2-python.html#cb21-13" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ad)</span>
<span id="cb21-14"><a href="mc2-python.html#cb21-14" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(ad, n_top_genes<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb21-15"><a href="mc2-python.html#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="mc2-python.html#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute principal components - </span></span>
<span id="cb21-17"><a href="mc2-python.html#cb21-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-18"><a href="mc2-python.html#cb21-18" aria-hidden="true" tabindex="-1"></a>n_comp    <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb21-19"><a href="mc2-python.html#cb21-19" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(ad, n_comps<span class="op">=</span>n_comp, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb21-20"><a href="mc2-python.html#cb21-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-21"><a href="mc2-python.html#cb21-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-22"><a href="mc2-python.html#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute UMAP for visualization </span></span>
<span id="cb21-23"><a href="mc2-python.html#cb21-23" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(ad, n_neighbors<span class="op">=</span><span class="dv">10</span>, n_pcs<span class="op">=</span>n_comp)</span>
<span id="cb21-24"><a href="mc2-python.html#cb21-24" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(ad)</span></code></pre></div>
<p>To visualize the metacell projection on the single-cell UMAP, we use the <code>mc_visualize</code> function from the <code>mc_QC</code>, this function was adapted from the <code>plot.plot_2D</code> included in the SEACells package.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="mc2-python.html#cb22-1" aria-hidden="true" tabindex="-1"></a>mc_proj <span class="op">=</span> mc_QC.mc_visualize(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb22-2"><a href="mc2-python.html#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.</span></span>
<span id="cb22-3"><a href="mc2-python.html#cb22-3" aria-hidden="true" tabindex="-1"></a>mc_proj.show()</span>
<span id="cb22-4"><a href="mc2-python.html#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-5"><a href="mc2-python.html#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-6"><a href="mc2-python.html#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-7"><a href="mc2-python.html#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-8"><a href="mc2-python.html#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-9"><a href="mc2-python.html#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-10"><a href="mc2-python.html#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-11"><a href="mc2-python.html#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb22-12"><a href="mc2-python.html#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc2-comp-umap-1.png" width="384" /></p>
</div>
</div>
<div id="metacell-qc" class="section level3 hasAnchor" number="3.1.5">
<h3><span class="header-section-number">3.1.5</span> Metacell QC<a href="mc2-python.html#metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="compute-purity-compactness-and-separation-metrics" class="section level4 unnumbered hasAnchor">
<h4>Compute purity, compactness and separation metrics<a href="mc2-python.html#compute-purity-compactness-and-separation-metrics" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Size distribution</strong></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="mc2-python.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_size = SEACells.plot.plot_SEACell_sizes(ad, bins=20)</span></span>
<span id="cb23-2"><a href="mc2-python.html#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="mc2-python.html#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs = pd.merge(mc_ad.obs, mc_size, left_index=True, right_index=True)</span></span>
<span id="cb23-4"><a href="mc2-python.html#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs</span></span></code></pre></div>
<p>When available, we can use cell annotation to annotate each metacell to the most abundant cell category (<em>e.g.</em> cell type) composing the metacell.
This also allows us to compute metacell purity. If the annotation considered is the cell type, the <strong>purity</strong> of a metacell is the proportion of the most abundant cell type within the metacell [ref SuperCell]</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb24-1"><a href="mc2-python.html#cb24-1" aria-hidden="true" tabindex="-1"></a>mc_purity <span class="op">=</span> mc_QC.purity(ad, annotation_label, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>)</span>
<span id="cb24-2"><a href="mc2-python.html#cb24-2" aria-hidden="true" tabindex="-1"></a>mc_purity.head()</span>
<span id="cb24-3"><a href="mc2-python.html#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                 louvain  louvain_purity</span></span>
<span id="cb24-4"><a href="mc2-python.html#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; membership                             </span></span>
<span id="cb24-5"><a href="mc2-python.html#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1.0         CD8 T cells        0.857143</span></span>
<span id="cb24-6"><a href="mc2-python.html#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2.0         CD4 T cells        0.869565</span></span>
<span id="cb24-7"><a href="mc2-python.html#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3.0         CD4 T cells        0.961538</span></span>
<span id="cb24-8"><a href="mc2-python.html#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4.0             B cells        1.000000</span></span>
<span id="cb24-9"><a href="mc2-python.html#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5.0         CD4 T cells        0.782609</span></span>
<span id="cb24-10"><a href="mc2-python.html#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co"># add purity to metadata</span></span>
<span id="cb24-11"><a href="mc2-python.html#cb24-11" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;purity&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(mc_purity[annotation_label <span class="op">+</span> <span class="st">&quot;_purity&quot;</span>])</span></code></pre></div>
<p>The <strong>compactness</strong> of a metacell is the variance of the components within the metacell [ref SEACells]</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb25-1"><a href="mc2-python.html#cb25-1" aria-hidden="true" tabindex="-1"></a>compactness <span class="op">=</span> mc_QC.compactness(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Compactness_PCA&#39;</span>, n_comp<span class="op">=</span><span class="dv">10</span>)[<span class="st">&#39;Compactness_PCA&#39;</span>]</span>
<span id="cb25-2"><a href="mc2-python.html#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add compactness to metadata</span></span>
<span id="cb25-3"><a href="mc2-python.html#cb25-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;Compactness_PCA&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(compactness)</span></code></pre></div>
<p>The <strong>separation</strong> of a metacell is the distance to the closest metacell [ref SEACells]</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb26-1"><a href="mc2-python.html#cb26-1" aria-hidden="true" tabindex="-1"></a>separation <span class="op">=</span> mc_QC.separation(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Separation_PCA&#39;</span>, n_comp<span class="op">=</span><span class="dv">10</span>)[<span class="st">&#39;Separation_PCA&#39;</span>]</span>
<span id="cb26-2"><a href="mc2-python.html#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add separation to metadata</span></span>
<span id="cb26-3"><a href="mc2-python.html#cb26-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;Separation_PCA&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(separation)</span></code></pre></div>
<!-- The **inner normalized variance (INV)** of a metacell is the mean-normalized variance of gene expression within the metacell [ref MC-2] -->
<!-- ```{python, warning = FALSE, message = FALSE, result = FALSE, eval = FALSE} -->
<!-- mc_INV = mc_QC.mc_inner_normalized_var(ad, MC_label = 'membership') -->
<!-- mc_INV_val = mc_INV.quantile(0.95, axis=1, numeric_only=True) -->
<!-- mc_INV_val = pd.DataFrame(mc_INV_val.transpose()).set_axis(['INV'], axis=1, inplace=False) -->
<!-- # add INV to metadata -->
<!-- mc_ad.obs['INV'] = list(mc_INV_val["INV"]) -->
<!-- ``` -->
<div class="sourceCode" id="cb27"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb27-1"><a href="mc2-python.html#cb27-1" aria-hidden="true" tabindex="-1"></a>ad.uns[<span class="st">&#39;mc_obs&#39;</span>] <span class="op">=</span> mc_ad.obs</span>
<span id="cb27-2"><a href="mc2-python.html#cb27-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb27-3"><a href="mc2-python.html#cb27-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;purity&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb27-4"><a href="mc2-python.html#cb27-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb27-5"><a href="mc2-python.html#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-6"><a href="mc2-python.html#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-7"><a href="mc2-python.html#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-8"><a href="mc2-python.html#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-9"><a href="mc2-python.html#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-10"><a href="mc2-python.html#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-11"><a href="mc2-python.html#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-12"><a href="mc2-python.html#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-13"><a href="mc2-python.html#cb27-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-14"><a href="mc2-python.html#cb27-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-15"><a href="mc2-python.html#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-16"><a href="mc2-python.html#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-17"><a href="mc2-python.html#cb27-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-18"><a href="mc2-python.html#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-19"><a href="mc2-python.html#cb27-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-20"><a href="mc2-python.html#cb27-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-21"><a href="mc2-python.html#cb27-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-22"><a href="mc2-python.html#cb27-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-23"><a href="mc2-python.html#cb27-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-24"><a href="mc2-python.html#cb27-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-25"><a href="mc2-python.html#cb27-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-26"><a href="mc2-python.html#cb27-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-27"><a href="mc2-python.html#cb27-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb27-28"><a href="mc2-python.html#cb27-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-16-3.png" width="384" /></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb28-1"><a href="mc2-python.html#cb28-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb28-2"><a href="mc2-python.html#cb28-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb28-3"><a href="mc2-python.html#cb28-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;Compactness_PCA&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb28-4"><a href="mc2-python.html#cb28-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)   </span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-16-4.png" width="384" /></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb29-1"><a href="mc2-python.html#cb29-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb29-2"><a href="mc2-python.html#cb29-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb29-3"><a href="mc2-python.html#cb29-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;Separation_PCA&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb29-4"><a href="mc2-python.html#cb29-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)   </span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-16-5.png" width="384" /></p>
<div id="save-output" class="section level5 unnumbered hasAnchor">
<h5>Save output<a href="mc2-python.html#save-output" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb30"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb30-1"><a href="mc2-python.html#cb30-1" aria-hidden="true" tabindex="-1"></a>mc_ad.write_h5ad(os.path.join(<span class="st">&#39;./data&#39;</span>, proj_name, <span class="ss">f&#39;metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad&#39;</span>))</span></code></pre></div>
</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="constructing-metacells-for-discrete-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="seacells-python.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/21-MC_construction_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/21-MC_construction_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
