<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.1 MC2 (Python) | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="3.1 MC2 (Python) | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.1 MC2 (Python) | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-07-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="constructing-metacells-for-a-discrete-data.html"/>
<link rel="next" href="supercell-r.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="book-structure.html"><a href="book-structure.html"><i class="fa fa-check"></i><b>1.1</b> Book structure</a></li>
<li class="chapter" data-level="1.2" data-path="installation-and-requirements.html"><a href="installation-and-requirements.html"><i class="fa fa-check"></i><b>1.2</b> Installation and requirements</a></li>
<li class="chapter" data-level="1.3" data-path="render-book.html"><a href="render-book.html"><i class="fa fa-check"></i><b>1.3</b> Render book</a></li>
<li class="chapter" data-level="1.4" data-path="get-data.html"><a href="get-data.html"><i class="fa fa-check"></i><b>1.4</b> Get data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a>
<ul>
<li class="chapter" data-level="2.1" data-path="MC2.html"><a href="MC2.html"><i class="fa fa-check"></i><b>2.1</b> Metacell (MC2)</a></li>
<li class="chapter" data-level="2.2" data-path="SuperCell.html"><a href="SuperCell.html"><i class="fa fa-check"></i><b>2.2</b> SuperCell</a></li>
<li class="chapter" data-level="2.3" data-path="SEACells.html"><a href="SEACells.html"><i class="fa fa-check"></i><b>2.3</b> SEACells</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="constructing-metacells-for-a-discrete-data.html"><a href="constructing-metacells-for-a-discrete-data.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells (for a ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="mc2-python.html"><a href="mc2-python.html"><i class="fa fa-check"></i><b>3.1</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="" data-path="mc2-python.html"><a href="mc2-python.html#running-mc2"><i class="fa fa-check"></i>Running MC2</a></li>
<li class="chapter" data-level="" data-path="mc2-python.html"><a href="mc2-python.html#compute-latent-space-for-metacell-qc"><i class="fa fa-check"></i>Compute latent space for metacell QC</a></li>
<li class="chapter" data-level="" data-path="mc2-python.html"><a href="mc2-python.html#metacell-qc"><i class="fa fa-check"></i>Metacell QC</a></li>
<li class="chapter" data-level="" data-path="mc2-python.html"><a href="mc2-python.html#save-output"><i class="fa fa-check"></i>Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="supercell-r.html"><a href="supercell-r.html"><i class="fa fa-check"></i><b>3.2</b> SuperCell (R)</a></li>
<li class="chapter" data-level="3.3" data-path="seacells-python.html"><a href="seacells-python.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="" data-path="seacells-python.html"><a href="seacells-python.html#running-seacells"><i class="fa fa-check"></i>Running SEACells</a></li>
<li class="chapter" data-level="" data-path="seacells-python.html"><a href="seacells-python.html#metacell-qc-1"><i class="fa fa-check"></i>Metacell QC</a></li>
<li class="chapter" data-level="" data-path="seacells-python.html"><a href="seacells-python.html#save-metacell-object-for-further-analysis"><i class="fa fa-check"></i>Save metacell object for further analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><a href="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><i class="fa fa-check"></i><b>4</b> Downstream analysis of metacells (for a discrete dataset)</a>
<ul>
<li class="chapter" data-level="4.1" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html"><i class="fa fa-check"></i><b>4.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html#dimensionality-reduction"><i class="fa fa-check"></i><b>4.1.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="4.1.2" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html#clustering"><i class="fa fa-check"></i><b>4.1.2</b> Clustering</a></li>
<li class="chapter" data-level="4.1.3" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html#differential-expression-analysis"><i class="fa fa-check"></i><b>4.1.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>4.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>4.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="4.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering-1"><i class="fa fa-check"></i><b>4.2.2</b> Clustering</a></li>
<li class="chapter" data-level="4.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>4.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>4.3</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#grn"><i class="fa fa-check"></i><b>4.3.1</b> GRN</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>4.4</b> Sample-weighted analysis</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="command-line-tool-for-mc-construction.html"><a href="command-line-tool-for-mc-construction.html"><i class="fa fa-check"></i><b>5</b> Command line tool for MC construction</a></li>
<li class="chapter" data-level="6" data-path="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><a href="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><i class="fa fa-check"></i><b>6</b> Comparison of metacells usage for a discrete and continuous data</a>
<ul>
<li class="chapter" data-level="6.1" data-path="continuous-metacells-have-lower-purity.html"><a href="continuous-metacells-have-lower-purity.html"><i class="fa fa-check"></i><b>6.1</b> Continuous metacells have lower purity</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-integration-with-metacells.html"><a href="data-integration-with-metacells.html"><i class="fa fa-check"></i><b>7</b> Data integration with metacells</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mc2-python" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> MC2 (Python)<a href="mc2-python.html#mc2-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here we construct metacells using <a href="https://github.com/tanaylab/metacells">Metacell-2 (MC2)</a>. The code is adapted from the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">author’s tutorial</a>.</p>
<div id="imports" class="section level4 unnumbered hasAnchor">
<h4>Imports<a href="mc2-python.html#imports" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="mc2-python.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb7-2"><a href="mc2-python.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb7-3"><a href="mc2-python.html#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb7-4"><a href="mc2-python.html#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb7-5"><a href="mc2-python.html#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb7-6"><a href="mc2-python.html#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb7-7"><a href="mc2-python.html#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb7-8"><a href="mc2-python.html#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span>
<span id="cb7-9"><a href="mc2-python.html#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="mc2-python.html#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb7-11"><a href="mc2-python.html#cb7-11" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>) </span>
<span id="cb7-12"><a href="mc2-python.html#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span></code></pre></div>
</div>
<div id="parameters" class="section level4 unnumbered hasAnchor">
<h4>Parameters<a href="mc2-python.html#parameters" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb8-1"><a href="mc2-python.html#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="mc2-python.html#cb8-2" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;MC2&quot;</span></span>
<span id="cb8-3"><a href="mc2-python.html#cb8-3" aria-hidden="true" tabindex="-1"></a>gamma   <span class="op">=</span> <span class="dv">50</span> <span class="co"># graining level</span></span>
<span id="cb8-4"><a href="mc2-python.html#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="mc2-python.html#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we can modify dataset </span></span>
<span id="cb8-6"><a href="mc2-python.html#cb8-6" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> [<span class="st">&quot;cell_lines&quot;</span>, <span class="st">&quot;3k_pbmc&quot;</span>][<span class="dv">1</span>]</span>
<span id="cb8-7"><a href="mc2-python.html#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="mc2-python.html#cb8-8" aria-hidden="true" tabindex="-1"></a>annotation_label <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span>:<span class="st">&#39;cell_line&#39;</span>,</span>
<span id="cb8-9"><a href="mc2-python.html#cb8-9" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&#39;3k_pbmc&#39;</span>:<span class="st">&#39;louvain&#39;</span>}[proj_name] <span class="co"># name of annotation field (obs)</span></span></code></pre></div>
</div>
<div id="load-data" class="section level4 unnumbered hasAnchor">
<h4>Load data<a href="mc2-python.html#load-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="mc2-python.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">## here code to download dataset and store it at f&#39;Metacell_tutorial/data/{proj_name}/singlecell_anndata_filtered.h5ad&#39;</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="mc2-python.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load pre-filtered data</span></span>
<span id="cb10-2"><a href="mc2-python.html#cb10-2" aria-hidden="true" tabindex="-1"></a>data_folder <span class="op">=</span> os.path.join(<span class="st">&quot;./data/&quot;</span>, proj_name) <span class="co"># here path to folder with adata</span></span>
<span id="cb10-3"><a href="mc2-python.html#cb10-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(data_folder, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span></code></pre></div>
<p><strong>Setup MC object</strong></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb11-1"><a href="mc2-python.html#cb11-1" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(ad, proj_name)</span></code></pre></div>
</div>
<div id="gene-filtering-according-to-mc2" class="section level4 unnumbered hasAnchor">
<h4>Gene filtering according to MC2<a href="mc2-python.html#gene-filtering-according-to-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb12"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb12-1"><a href="mc2-python.html#cb12-1" aria-hidden="true" tabindex="-1"></a>excluded_gene_names <span class="op">=</span> [] <span class="co"># for example, [&#39;IGHMBP2&#39;, &#39;IGLL1&#39;, &#39;IGLL5&#39;, &#39;IGLON5&#39;, &#39;NEAT1&#39;, &#39;TMSB10&#39;, &#39;TMSB4X&#39;]</span></span>
<span id="cb12-2"><a href="mc2-python.html#cb12-2" aria-hidden="true" tabindex="-1"></a>excluded_gene_patterns <span class="op">=</span> [<span class="st">&#39;MT-.*&#39;</span>]</span>
<span id="cb12-3"><a href="mc2-python.html#cb12-3" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_genes(ad,</span>
<span id="cb12-4"><a href="mc2-python.html#cb12-4" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_names<span class="op">=</span>excluded_gene_names,</span>
<span id="cb12-5"><a href="mc2-python.html#cb12-5" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_patterns<span class="op">=</span>excluded_gene_patterns,</span>
<span id="cb12-6"><a href="mc2-python.html#cb12-6" aria-hidden="true" tabindex="-1"></a>                          random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb12-7"><a href="mc2-python.html#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[properly_sampled_gene]: 16579 true (50.64%) out of 32738 bools</span></span>
<span id="cb12-8"><a href="mc2-python.html#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[excluded_gene]: 13 true (0.03971%) out of 32738 bools</span></span>
<span id="cb12-9"><a href="mc2-python.html#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[noisy_lonely_gene]: 0 true (0%) out of 32738 bools</span></span>
<span id="cb12-10"><a href="mc2-python.html#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="mc2-python.html#cb12-11" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_genes(ad)</span>
<span id="cb12-12"><a href="mc2-python.html#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[clean_gene]: 16566 true (50.6%) out of 32738 bools</span></span></code></pre></div>
</div>
<div id="cell-fintering" class="section level4 unnumbered hasAnchor">
<h4>Cell fintering<a href="mc2-python.html#cell-fintering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Since our data is pre-processed and low-quality cells have been already filtered out, this step is not applicable to our data, but we keep this chunk so that you can apply cell filtering to a newly generated dataset.</p>
<p>The first round of cell cleaning usually implies filltering out cell with very low and very hight UMI content. The second round includes cell filtering based on mitochondrial and/or ribosomal content. We will skip both steps as our data have been pre-filtered and will use very lenient cutoffs (<code>properly_sampled_min_cell_total</code>, <code>properly_sampled_max_cell_total</code> and <code>properly_sampled_max_excluded_genes_fraction</code>) such that all the cells are kept for the metacell construction.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb13-1"><a href="mc2-python.html#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">### The first round (high/low UMIs)</span></span>
<span id="cb13-2"><a href="mc2-python.html#cb13-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_min_cell_total <span class="op">=</span> <span class="dv">200</span> <span class="co"># setup for the dataset that will be used </span></span>
<span id="cb13-3"><a href="mc2-python.html#cb13-3" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_cell_total <span class="op">=</span> <span class="dv">10000</span> <span class="co"># setup for the dataset that will be used </span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="mc2-python.html#cb14-1" aria-hidden="true" tabindex="-1"></a>total_umis_of_cells <span class="op">=</span> mc.ut.get_o_numpy(ad, name<span class="op">=</span><span class="st">&#39;__x__&#39;</span>, <span class="bu">sum</span><span class="op">=</span><span class="va">True</span>)</span>
<span id="cb14-2"><a href="mc2-python.html#cb14-2" aria-hidden="true" tabindex="-1"></a>df_umis <span class="op">=</span> pd.DataFrame(total_umis_of_cells, columns <span class="op">=</span> [<span class="st">&#39;total_UMIs&#39;</span>])</span>
<span id="cb14-3"><a href="mc2-python.html#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="mc2-python.html#cb14-4" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb14-5"><a href="mc2-python.html#cb14-5" aria-hidden="true" tabindex="-1"></a>sns.histplot(data<span class="op">=</span>df_umis, x<span class="op">=</span><span class="st">&quot;total_UMIs&quot;</span>)</span>
<span id="cb14-6"><a href="mc2-python.html#cb14-6" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">&#39;UMIs&#39;</span>)</span>
<span id="cb14-7"><a href="mc2-python.html#cb14-7" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">&#39;Density&#39;</span>)</span>
<span id="cb14-8"><a href="mc2-python.html#cb14-8" aria-hidden="true" tabindex="-1"></a>plt.axvline(x<span class="op">=</span>properly_sampled_min_cell_total, color<span class="op">=</span><span class="st">&#39;darkgreen&#39;</span>)</span>
<span id="cb14-9"><a href="mc2-python.html#cb14-9" aria-hidden="true" tabindex="-1"></a>plt.axvline(x<span class="op">=</span>properly_sampled_max_cell_total, color<span class="op">=</span><span class="st">&#39;crimson&#39;</span>)</span>
<span id="cb14-10"><a href="mc2-python.html#cb14-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc-filtering-run-first-1.png" width="672" /></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="mc2-python.html#cb15-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-2"><a href="mc2-python.html#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="mc2-python.html#cb15-3" aria-hidden="true" tabindex="-1"></a>too_small_cells_count <span class="op">=</span> <span class="bu">sum</span>(total_umis_of_cells <span class="op">&lt;</span> properly_sampled_min_cell_total)</span>
<span id="cb15-4"><a href="mc2-python.html#cb15-4" aria-hidden="true" tabindex="-1"></a>too_large_cells_count <span class="op">=</span> <span class="bu">sum</span>(total_umis_of_cells <span class="op">&gt;</span> properly_sampled_max_cell_total)</span>
<span id="cb15-5"><a href="mc2-python.html#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="mc2-python.html#cb15-6" aria-hidden="true" tabindex="-1"></a>too_small_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_small_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb15-7"><a href="mc2-python.html#cb15-7" aria-hidden="true" tabindex="-1"></a>too_large_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_large_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb15-8"><a href="mc2-python.html#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="mc2-python.html#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Will exclude %s (%.2f%%) cells with less than %s UMIs&quot;</span></span>
<span id="cb15-10"><a href="mc2-python.html#cb15-10" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_small_cells_count,</span>
<span id="cb15-11"><a href="mc2-python.html#cb15-11" aria-hidden="true" tabindex="-1"></a>         too_small_cells_percent,</span>
<span id="cb15-12"><a href="mc2-python.html#cb15-12" aria-hidden="true" tabindex="-1"></a>         properly_sampled_min_cell_total))</span>
<span id="cb15-13"><a href="mc2-python.html#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Will exclude 0 (0.00%) cells with less than 200 UMIs</span></span>
<span id="cb15-14"><a href="mc2-python.html#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Will exclude %s (%.2f%%) cells with more than %s UMIs&quot;</span></span>
<span id="cb15-15"><a href="mc2-python.html#cb15-15" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_large_cells_count,</span>
<span id="cb15-16"><a href="mc2-python.html#cb15-16" aria-hidden="true" tabindex="-1"></a>         too_large_cells_percent,</span>
<span id="cb15-17"><a href="mc2-python.html#cb15-17" aria-hidden="true" tabindex="-1"></a>         properly_sampled_max_cell_total))</span>
<span id="cb15-18"><a href="mc2-python.html#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Will exclude 0 (0.00%) cells with more than 10000 UMIs</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb16-1"><a href="mc2-python.html#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The second round (content of non-clean genes, e.g., mito-genes)</span></span>
<span id="cb16-2"><a href="mc2-python.html#cb16-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_excluded_genes_fraction <span class="op">=</span> <span class="fl">0.25</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb17-1"><a href="mc2-python.html#cb17-1" aria-hidden="true" tabindex="-1"></a>excluded_genes_data <span class="op">=</span> mc.tl.filter_data(ad, var_masks<span class="op">=</span>[<span class="st">&#39;~clean_gene&#39;</span>])[<span class="dv">0</span>]</span>
<span id="cb17-2"><a href="mc2-python.html#cb17-2" aria-hidden="true" tabindex="-1"></a>excluded_umis_of_cells <span class="op">=</span> mc.ut.get_o_numpy(excluded_genes_data, name<span class="op">=</span><span class="st">&#39;__x__&#39;</span>, <span class="bu">sum</span><span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-3"><a href="mc2-python.html#cb17-3" aria-hidden="true" tabindex="-1"></a>excluded_fraction_of_umis_of_cells <span class="op">=</span> excluded_umis_of_cells <span class="op">/</span> total_umis_of_cells</span>
<span id="cb17-4"><a href="mc2-python.html#cb17-4" aria-hidden="true" tabindex="-1"></a>df_umis <span class="op">=</span> pd.DataFrame(excluded_fraction_of_umis_of_cells, columns <span class="op">=</span> [<span class="st">&#39;frac_excl_genes_UMIs&#39;</span>])</span>
<span id="cb17-5"><a href="mc2-python.html#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="mc2-python.html#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="mc2-python.html#cb17-7" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb17-8"><a href="mc2-python.html#cb17-8" aria-hidden="true" tabindex="-1"></a>sns.histplot(df_umis, x <span class="op">=</span> <span class="st">&#39;frac_excl_genes_UMIs&#39;</span>)</span>
<span id="cb17-9"><a href="mc2-python.html#cb17-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">&#39;Fraction of excluded gene UMIs&#39;</span>)</span>
<span id="cb17-10"><a href="mc2-python.html#cb17-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">&#39;Density&#39;</span>)</span>
<span id="cb17-11"><a href="mc2-python.html#cb17-11" aria-hidden="true" tabindex="-1"></a>plt.axvline(x<span class="op">=</span>properly_sampled_max_excluded_genes_fraction, color<span class="op">=</span><span class="st">&#39;crimson&#39;</span>)</span>
<span id="cb17-12"><a href="mc2-python.html#cb17-12" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc-cell-filtering-run-second-1.png" width="672" /></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb18-1"><a href="mc2-python.html#cb18-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-2"><a href="mc2-python.html#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="mc2-python.html#cb18-3" aria-hidden="true" tabindex="-1"></a>too_excluded_cells_count <span class="op">=</span> <span class="bu">sum</span>(excluded_fraction_of_umis_of_cells <span class="op">&gt;</span> properly_sampled_max_excluded_genes_fraction)</span>
<span id="cb18-4"><a href="mc2-python.html#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="mc2-python.html#cb18-5" aria-hidden="true" tabindex="-1"></a>too_excluded_cells_percent <span class="op">=</span> <span class="fl">100.0</span> <span class="op">*</span> too_excluded_cells_count <span class="op">/</span> <span class="bu">len</span>(total_umis_of_cells)</span>
<span id="cb18-6"><a href="mc2-python.html#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="mc2-python.html#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&quot;Will exclude %s (%.2f%%) cells with more than %.2f%% excluded gene UMIs&quot;</span></span>
<span id="cb18-8"><a href="mc2-python.html#cb18-8" aria-hidden="true" tabindex="-1"></a>      <span class="op">%</span> (too_excluded_cells_count,</span>
<span id="cb18-9"><a href="mc2-python.html#cb18-9" aria-hidden="true" tabindex="-1"></a>         too_excluded_cells_percent,</span>
<span id="cb18-10"><a href="mc2-python.html#cb18-10" aria-hidden="true" tabindex="-1"></a>         <span class="fl">100.0</span> <span class="op">*</span> properly_sampled_max_excluded_genes_fraction))</span>
<span id="cb18-11"><a href="mc2-python.html#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Will exclude 0 (0.00%) cells with more than 25.00% excluded gene UMIs</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb19-1"><a href="mc2-python.html#cb19-1" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_cells(</span>
<span id="cb19-2"><a href="mc2-python.html#cb19-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb19-3"><a href="mc2-python.html#cb19-3" aria-hidden="true" tabindex="-1"></a>    properly_sampled_min_cell_total<span class="op">=</span>properly_sampled_min_cell_total,</span>
<span id="cb19-4"><a href="mc2-python.html#cb19-4" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_cell_total<span class="op">=</span>properly_sampled_max_cell_total,</span>
<span id="cb19-5"><a href="mc2-python.html#cb19-5" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_excluded_genes_fraction<span class="op">=</span>properly_sampled_max_excluded_genes_fraction</span>
<span id="cb19-6"><a href="mc2-python.html#cb19-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb19-7"><a href="mc2-python.html#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[properly_sampled_cell]: 2638 true (100%) out of 2638 bools</span></span>
<span id="cb19-8"><a href="mc2-python.html#cb19-8" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_cells(ad)</span>
<span id="cb19-9"><a href="mc2-python.html#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[clean_cell]: 2638 true (100%) out of 2638 bools</span></span>
<span id="cb19-10"><a href="mc2-python.html#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="mc2-python.html#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract clean dataset (with filtered cells and genes)</span></span>
<span id="cb19-12"><a href="mc2-python.html#cb19-12" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> mc.pl.extract_clean_data(ad)</span>
<span id="cb19-13"><a href="mc2-python.html#cb19-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[full_cell_index]: 2638 int64s</span></span>
<span id="cb19-14"><a href="mc2-python.html#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[full_gene_index]: 16566 int64s</span></span></code></pre></div>
</div>
<div id="running-mc2" class="section level3 unnumbered hasAnchor">
<h3>Running MC2<a href="mc2-python.html#running-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Metacell-2 uses its own feature selection approach (i.e., selection of genes used to build metacells). Additionally, we can explicitly specify which features to use by providing two arguments: feature_gene_names - genes that have to be used forbidden_gene_names - genes to exclude.
In contrast to the SuperCell and SEACells, Metacell-2 does not allow to explicitly obtain metacell data at a user-defined graining level. Instead, to vary graining level, we have to vary a <code>target_metacell_size</code> parameter, that is <code>160000</code> by default. Here we provide a chunk to calibrate this value to get a desired graining level. Please, increase or decrease scale if the obtained graining level <code>gamma_obtained</code> is lower or larger than the requested one (gamma).</p>
<p><strong>Estimate target_metacell_size (gamma)</strong></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="mc2-python.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&#39;The requested graining level is </span><span class="sc">{</span>gamma<span class="sc">}</span><span class="ss">, lets estimate the target_metacell_size that should result in such graining level.&#39;</span>)</span>
<span id="cb20-2"><a href="mc2-python.html#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The requested graining level is 50, lets estimate the target_metacell_size that should result in such graining level.</span></span>
<span id="cb20-3"><a href="mc2-python.html#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="mc2-python.html#cb20-4" aria-hidden="true" tabindex="-1"></a>scale <span class="op">=</span> <span class="dv">2</span> <span class="co"># increase or decrease if the obtained graining level (`gamma_obtained`) is significantly &gt; or &lt; then the requested one `gamma`</span></span>
<span id="cb20-5"><a href="mc2-python.html#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="mc2-python.html#cb20-6" aria-hidden="true" tabindex="-1"></a>N_c <span class="op">=</span> ad.shape[<span class="dv">0</span>]</span>
<span id="cb20-7"><a href="mc2-python.html#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="mc2-python.html#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="co"># estimated mean UMI content in downsampled data</span></span>
<span id="cb20-9"><a href="mc2-python.html#cb20-9" aria-hidden="true" tabindex="-1"></a>est_downsample_UMI <span class="op">=</span> np.quantile(np.array(total_umis_of_cells), <span class="fl">0.05</span>)</span>
<span id="cb20-10"><a href="mc2-python.html#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="mc2-python.html#cb20-11" aria-hidden="true" tabindex="-1"></a>target_metacell_size <span class="op">=</span> <span class="bu">int</span>(est_downsample_UMI <span class="op">*</span> gamma <span class="op">*</span> scale)</span>
<span id="cb20-12"><a href="mc2-python.html#cb20-12" aria-hidden="true" tabindex="-1"></a>target_metacell_size</span>
<span id="cb20-13"><a href="mc2-python.html#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 101455</span></span></code></pre></div>
<div id="aggregate-mc2" class="section level4 unnumbered hasAnchor">
<h4>Aggregate metacells<a href="mc2-python.html#aggregate-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb21"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb21-1"><a href="mc2-python.html#cb21-1" aria-hidden="true" tabindex="-1"></a>mc.pl.divide_and_conquer_pipeline(</span>
<span id="cb21-2"><a href="mc2-python.html#cb21-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb21-3"><a href="mc2-python.html#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#feature_gene_names   = feature_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb21-4"><a href="mc2-python.html#cb21-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#forbidden_gene_names = forbidden_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb21-5"><a href="mc2-python.html#cb21-5" aria-hidden="true" tabindex="-1"></a>    target_metacell_size <span class="op">=</span> target_metacell_size,</span>
<span id="cb21-6"><a href="mc2-python.html#cb21-6" aria-hidden="true" tabindex="-1"></a>    random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb21-7"><a href="mc2-python.html#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb21-8"><a href="mc2-python.html#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene_module]: 16566 int32 elements with all outliers (100%)</span></span>
<span id="cb21-9"><a href="mc2-python.html#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cells_rare_gene_module]: 2638 int32 elements with all outliers (100%)</span></span>
<span id="cb21-10"><a href="mc2-python.html#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[rare_cell]: 0 true (0%) out of 2638 bools</span></span>
<span id="cb21-11"><a href="mc2-python.html#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.layers[downsampled]: csr_matrix 2638 X 16566 float32s (1218892 &gt; 0)</span></span>
<span id="cb21-12"><a href="mc2-python.html#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[downsample_samples]: 989</span></span>
<span id="cb21-13"><a href="mc2-python.html#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_top3_gene]: 552 true (3.332%) out of 16566 bools</span></span>
<span id="cb21-14"><a href="mc2-python.html#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_total_gene]: 4519 true (27.28%) out of 16566 bools</span></span>
<span id="cb21-15"><a href="mc2-python.html#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_relative_variance_gene]: 3027 true (18.27%) out of 16566 bools</span></span>
<span id="cb21-16"><a href="mc2-python.html#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[feature_gene]: 293 true (1.769%) out of 16566 bools</span></span>
<span id="cb21-17"><a href="mc2-python.html#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obsp[obs_similarity]: ndarray 2638 X 2638 float32s</span></span>
<span id="cb21-18"><a href="mc2-python.html#cb21-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obsp[obs_outgoing_weights]: csr_matrix 2638 X 2638 float32s (119842 &gt; 0)</span></span>
<span id="cb21-19"><a href="mc2-python.html#cb21-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[seed]: 491 outliers (18.61%) out of 2638 int32 elements with 60 groups with mean size 35.78</span></span>
<span id="cb21-20"><a href="mc2-python.html#cb21-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[candidate]: 0 outliers (0%) out of 2638 int32 elements with 63 groups with mean size 41.87</span></span>
<span id="cb21-21"><a href="mc2-python.html#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[gene_deviant_votes]: 936 positive (5.65%) out of 16566 int32s</span></span>
<span id="cb21-22"><a href="mc2-python.html#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cell_deviant_votes]: 899 positive (34.08%) out of 2638 int32s</span></span>
<span id="cb21-23"><a href="mc2-python.html#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[dissolved]: 10 true (0.3791%) out of 2638 bools</span></span>
<span id="cb21-24"><a href="mc2-python.html#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell]: 909 outliers (34.46%) out of 2638 int64 elements with 62 groups with mean size 27.89</span></span>
<span id="cb21-25"><a href="mc2-python.html#cb21-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[outlier]: 909 true (34.46%) out of 2638 bools</span></span>
<span id="cb21-26"><a href="mc2-python.html#cb21-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[pre_directs]: 0</span></span>
<span id="cb21-27"><a href="mc2-python.html#cb21-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.uns[directs]: 1</span></span>
<span id="cb21-28"><a href="mc2-python.html#cb21-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_high_total_gene]: * &lt;- 0</span></span>
<span id="cb21-29"><a href="mc2-python.html#cb21-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_total_gene]: 4519 positive (27.28%) out of 16566 int32s</span></span>
<span id="cb21-30"><a href="mc2-python.html#cb21-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_high_relative_variance_gene]: * &lt;- 0</span></span>
<span id="cb21-31"><a href="mc2-python.html#cb21-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[high_relative_variance_gene]: 3027 positive (18.27%) out of 16566 int32s</span></span>
<span id="cb21-32"><a href="mc2-python.html#cb21-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[forbidden_gene]: * &lt;- False</span></span>
<span id="cb21-33"><a href="mc2-python.html#cb21-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_feature_gene]: * &lt;- 0</span></span>
<span id="cb21-34"><a href="mc2-python.html#cb21-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[feature_gene]: 293 positive (1.769%) out of 16566 int32s</span></span>
<span id="cb21-35"><a href="mc2-python.html#cb21-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[pre_gene_deviant_votes]: * &lt;- 0</span></span>
<span id="cb21-36"><a href="mc2-python.html#cb21-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_cell_directs]: * &lt;- 0</span></span>
<span id="cb21-37"><a href="mc2-python.html#cb21-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cell_directs]: * &lt;- 0</span></span>
<span id="cb21-38"><a href="mc2-python.html#cb21-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_pile]: * &lt;- -1</span></span>
<span id="cb21-39"><a href="mc2-python.html#cb21-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pile]: * &lt;- 0</span></span>
<span id="cb21-40"><a href="mc2-python.html#cb21-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_candidate]: * &lt;- -1</span></span>
<span id="cb21-41"><a href="mc2-python.html#cb21-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_cell_deviant_votes]: * &lt;- 0</span></span>
<span id="cb21-42"><a href="mc2-python.html#cb21-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_dissolved]: * &lt;- False</span></span>
<span id="cb21-43"><a href="mc2-python.html#cb21-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[pre_metacell]: * &lt;- -1</span></span>
<span id="cb21-44"><a href="mc2-python.html#cb21-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-45"><a href="mc2-python.html#cb21-45" aria-hidden="true" tabindex="-1"></a><span class="co">## make anndata of metacells</span></span>
<span id="cb21-46"><a href="mc2-python.html#cb21-46" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> mc.pl.collect_metacells(ad, name<span class="op">=</span><span class="st">&#39;metacells&#39;</span>)</span>
<span id="cb21-47"><a href="mc2-python.html#cb21-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[excluded_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb21-48"><a href="mc2-python.html#cb21-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[clean_gene]: 16566 true (100%) out of 16566 bools</span></span>
<span id="cb21-49"><a href="mc2-python.html#cb21-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[forbidden_gene]: 0 true (0%) out of 16566 bools</span></span>
<span id="cb21-50"><a href="mc2-python.html#cb21-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[pre_feature_gene]: 0 positive (0%) out of 16566 int32s</span></span>
<span id="cb21-51"><a href="mc2-python.html#cb21-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[feature_gene]: 293 positive (1.769%) out of 16566 int32s</span></span>
<span id="cb21-52"><a href="mc2-python.html#cb21-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[pile]: [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ]</span></span>
<span id="cb21-53"><a href="mc2-python.html#cb21-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[candidate]: [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 15 ]</span></span></code></pre></div>
<p>Here we estimate whether a deviation of the obtained gamma is acceptable, and if not, suggest to increase or decrease <code>scale</code> parameter to get better graining level.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="mc2-python.html#cb22-1" aria-hidden="true" tabindex="-1"></a>gamma_obtained <span class="op">=</span> ad.shape[<span class="dv">0</span>]<span class="op">/</span>mc_ad.shape[<span class="dv">0</span>]</span>
<span id="cb22-2"><a href="mc2-python.html#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gamma_obtained)</span>
<span id="cb22-3"><a href="mc2-python.html#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 42.54838709677419</span></span>
<span id="cb22-4"><a href="mc2-python.html#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="mc2-python.html#cb22-5" aria-hidden="true" tabindex="-1"></a>gamma_dev <span class="op">=</span> (gamma_obtained <span class="op">-</span> gamma)<span class="op">/</span>gamma</span>
<span id="cb22-6"><a href="mc2-python.html#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">abs</span>(gamma_dev) <span class="op">&lt;</span> <span class="fl">0.3</span>: </span>
<span id="cb22-7"><a href="mc2-python.html#cb22-7" aria-hidden="true" tabindex="-1"></a>    gamma_dev <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb22-8"><a href="mc2-python.html#cb22-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-9"><a href="mc2-python.html#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> gamma_dev <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb22-10"><a href="mc2-python.html#cb22-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Increase `target_metacell_size` parameter by increasing `scale` and re-run metacell divide_and_conquer_pipeline() to get larger graining level&quot;</span>)</span>
<span id="cb22-11"><a href="mc2-python.html#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb22-12"><a href="mc2-python.html#cb22-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Deacrease `target_metacell_size` parameter by decreasing `scale` and re-run metacell divide_and_conquer_pipeline() to get smaller graining level&quot;</span>)</span>
<span id="cb22-13"><a href="mc2-python.html#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb22-14"><a href="mc2-python.html#cb22-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size` &quot;</span>)</span>
<span id="cb22-15"><a href="mc2-python.html#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size`</span></span></code></pre></div>
<p>If the obtained graining level is not acceptable and you updated <code>scale</code> parameter according to suggestion, do not forget to re-run chunk <a href="#chunk:mc2-divide-n-conquer"><strong>??</strong></a> <a href="mc2-python.html#aggregate-mc2"><strong>??</strong></a></p>
</div>
<div id="visualize-metacells" class="section level4 unnumbered hasAnchor">
<h4>Visualize metacells<a href="mc2-python.html#visualize-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="mc2-python.html#cb23-1" aria-hidden="true" tabindex="-1"></a>mc.pl.compute_umap_by_features(mc_ad, max_top_feature_genes<span class="op">=</span><span class="dv">1000</span>,</span>
<span id="cb23-2"><a href="mc2-python.html#cb23-2" aria-hidden="true" tabindex="-1"></a>                               min_dist<span class="op">=</span><span class="fl">2.0</span>, random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb23-3"><a href="mc2-python.html#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[top_feature_gene]: 293 true (1.769%) out of 16566 bools</span></span>
<span id="cb23-4"><a href="mc2-python.html#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[obs_balanced_ranks]: 762 nonzero (19.82%) out of 3844 elements</span></span>
<span id="cb23-5"><a href="mc2-python.html#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[obs_pruned_ranks]: 257 nonzero (6.686%) out of 3844 elements</span></span>
<span id="cb23-6"><a href="mc2-python.html#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[obs_outgoing_weights]: 257 nonzero (6.686%) out of 3844 elements</span></span>
<span id="cb23-7"><a href="mc2-python.html#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obsp[umap_distances]: csr_matrix 62 X 62 float32s (3782 &gt; 0)</span></span>
<span id="cb23-8"><a href="mc2-python.html#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /Users/mariiabilous/Documents/PhD/UNIL/R/Metacell_tutorial/my_env_mc2/lib/python3.8/site-packages/umap/umap_.py:1356: RuntimeWarning: divide by zero encountered in power</span></span>
<span id="cb23-9"><a href="mc2-python.html#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   return 1.0 / (1.0 + a * x ** (2 * b))</span></span>
<span id="cb23-10"><a href="mc2-python.html#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; /Users/mariiabilous/Documents/PhD/UNIL/R/Metacell_tutorial/my_env_mc2/lib/python3.8/site-packages/umap/umap_.py:1780: UserWarning: using precomputed metric; inverse_transform will be unavailable</span></span>
<span id="cb23-11"><a href="mc2-python.html#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   warn(&quot;using precomputed metric; inverse_transform will be unavailable&quot;)</span></span>
<span id="cb23-12"><a href="mc2-python.html#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[umap_x]: [ 1.6564819, 10.523616, 0.29960603, 12.346935, -18.181883, 10.824747, 9.3235655, 11.606001, 8.378801, 11.439771, -18.370312, 13.707805, -16.662771, 13.427194, 10.35691, 0.69431365, -16.371231, 8.772759, 8.827683, -15.460641, 10.793309, 4.4988446, -17.57337, 7.6907706, 14.141677, 8.847828, 10.136032, 8.403874, 9.716653, -19.049778, 3.3346436, 8.916034, -17.734402, -19.183643, 2.4466102, 9.458938, 9.972965, 14.90282, 14.809836, 8.051339, 11.39458, 13.204389, 9.030105, 13.247052, -19.935547, 12.671885, 13.710647, 11.188181, -16.07956, 12.613102, 12.006663, 2.273494, 10.232938, -16.919983, 12.528366, 14.898915, 11.630459, -13.0429125, -15.82163, -14.578497, -14.365959, -13.944745 ]</span></span>
<span id="cb23-13"><a href="mc2-python.html#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[umap_y]: [ 7.753808, 5.679683, 8.05517, 3.7745268, 17.191807, 12.818527, 6.2553945, 17.103827, 4.138062, 11.021334, 20.70879, 13.585729, 20.566952, 12.627911, 18.061642, 6.2272234, 15.916855, 18.031816, 7.2213125, 19.455956, 14.636021, 8.450331, 15.314596, 6.2872787, 6.581571, 14.994367, 3.1421099, 8.773476, 4.545281, 15.859472, 7.0959816, 13.212819, 19.351728, 18.600061, 6.080823, 10.117547, 16.1381, 11.50319, 7.828978, 16.442448, 6.698089, 11.034152, 11.458855, 5.4966006, 17.60186, 14.4407425, 8.617028, 8.6199465, 18.102512, 9.883782, 12.338453, 9.131331, 8.351595, 17.561296, 7.718665, 10.21264, 4.989835, 14.793287, 13.923492, 15.697431, 13.76865, 16.49492 ]</span></span>
<span id="cb23-14"><a href="mc2-python.html#cb23-14" aria-hidden="true" tabindex="-1"></a>                               </span>
<span id="cb23-15"><a href="mc2-python.html#cb23-15" aria-hidden="true" tabindex="-1"></a>umap_x <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_x&#39;</span>)</span>
<span id="cb23-16"><a href="mc2-python.html#cb23-16" aria-hidden="true" tabindex="-1"></a>umap_y <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_y&#39;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb24-1"><a href="mc2-python.html#cb24-1" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb24-2"><a href="mc2-python.html#cb24-2" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(x<span class="op">=</span>umap_x, y<span class="op">=</span>umap_y)</span>
<span id="cb24-3"><a href="mc2-python.html#cb24-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc2-plot-umap-1.png" width="672" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb25-1"><a href="mc2-python.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a membership -- index of metacell single cell belongs to </span></span>
<span id="cb25-2"><a href="mc2-python.html#cb25-2" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&#39;membership&#39;</span>] <span class="op">=</span> [<span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">0</span> <span class="cf">else</span> np.nan <span class="cf">for</span> i <span class="kw">in</span> ad.obs.metacell] </span>
<span id="cb25-3"><a href="mc2-python.html#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="mc2-python.html#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Save single-cell metadata (i.e., `raw.obs` dataframe) in the metacell adata object</span></span>
<span id="cb25-5"><a href="mc2-python.html#cb25-5" aria-hidden="true" tabindex="-1"></a>mc_ad.uns <span class="op">=</span> ad.uns.copy()</span>
<span id="cb25-6"><a href="mc2-python.html#cb25-6" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.obs&#39;</span>] <span class="op">=</span> ad.obs.copy()</span>
<span id="cb25-7"><a href="mc2-python.html#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="mc2-python.html#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the requested gamma</span></span>
<span id="cb25-9"><a href="mc2-python.html#cb25-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;gamma&#39;</span>] <span class="op">=</span> gamma</span></code></pre></div>
</div>
</div>
<div id="compute-latent-space-for-metacell-qc" class="section level3 unnumbered hasAnchor">
<h3>Compute latent space for metacell QC<a href="mc2-python.html#compute-latent-space-for-metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>MC2 builds metacells from gene expression data, not from latent space. Some of QC metrics (e.g., <strong>compactness</strong> and <strong>separation</strong>) are computed from the latent space. Thus, to compute those metrics, we need to compute latent space.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb26-1"><a href="mc2-python.html#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save count as a separate layer</span></span>
<span id="cb26-2"><a href="mc2-python.html#cb26-2" aria-hidden="true" tabindex="-1"></a>ad.layers[<span class="st">&#39;counts&#39;</span>] <span class="op">=</span> ad.X</span>
<span id="cb26-3"><a href="mc2-python.html#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="mc2-python.html#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the counts to &quot;.raw&quot; attribute of the anndata since it is necessary for downstream analysis</span></span>
<span id="cb26-5"><a href="mc2-python.html#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This step should be performed after filtering </span></span>
<span id="cb26-6"><a href="mc2-python.html#cb26-6" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(ad.layers[<span class="st">&#39;counts&#39;</span>])</span>
<span id="cb26-7"><a href="mc2-python.html#cb26-7" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> ad.obs_names, ad.var_names</span>
<span id="cb26-8"><a href="mc2-python.html#cb26-8" aria-hidden="true" tabindex="-1"></a>ad.raw <span class="op">=</span> raw_ad</span>
<span id="cb26-9"><a href="mc2-python.html#cb26-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-10"><a href="mc2-python.html#cb26-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-11"><a href="mc2-python.html#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize cells, log transform and compute highly variable genes</span></span>
<span id="cb26-12"><a href="mc2-python.html#cb26-12" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(ad)</span>
<span id="cb26-13"><a href="mc2-python.html#cb26-13" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ad)</span>
<span id="cb26-14"><a href="mc2-python.html#cb26-14" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(ad, n_top_genes<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb26-15"><a href="mc2-python.html#cb26-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-16"><a href="mc2-python.html#cb26-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute principal components - </span></span>
<span id="cb26-17"><a href="mc2-python.html#cb26-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-18"><a href="mc2-python.html#cb26-18" aria-hidden="true" tabindex="-1"></a>n_comp    <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb26-19"><a href="mc2-python.html#cb26-19" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(ad, n_comps<span class="op">=</span>n_comp, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb26-20"><a href="mc2-python.html#cb26-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-21"><a href="mc2-python.html#cb26-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-22"><a href="mc2-python.html#cb26-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute UMAP for visualization </span></span>
<span id="cb26-23"><a href="mc2-python.html#cb26-23" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(ad, n_neighbors<span class="op">=</span><span class="dv">10</span>, n_pcs<span class="op">=</span>n_comp)</span>
<span id="cb26-24"><a href="mc2-python.html#cb26-24" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(ad)</span></code></pre></div>
</div>
<div id="metacell-qc" class="section level3 unnumbered hasAnchor">
<h3>Metacell QC<a href="mc2-python.html#metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="save-output" class="section level3 unnumbered hasAnchor">
<h3>Save output<a href="mc2-python.html#save-output" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="constructing-metacells-for-a-discrete-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="supercell-r.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/21-MC_construction_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/21-MC_construction_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
