<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.1 MC2 (Python) | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="5.1 MC2 (Python) | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.1 MC2 (Python) | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-07-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="MC-continuous.html"/>
<link rel="next" href="supercell-r-1.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="book-structure.html"><a href="book-structure.html"><i class="fa fa-check"></i><b>1.1</b> Book structure</a></li>
<li class="chapter" data-level="1.2" data-path="installation-and-requirements.html"><a href="installation-and-requirements.html"><i class="fa fa-check"></i><b>1.2</b> Installation and requirements</a></li>
<li class="chapter" data-level="1.3" data-path="render-book.html"><a href="render-book.html"><i class="fa fa-check"></i><b>1.3</b> Render book</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a>
<ul>
<li class="chapter" data-level="2.1" data-path="MC2.html"><a href="MC2.html"><i class="fa fa-check"></i><b>2.1</b> Metacell (MC2)</a></li>
<li class="chapter" data-level="2.2" data-path="SuperCell.html"><a href="SuperCell.html"><i class="fa fa-check"></i><b>2.2</b> SuperCell</a></li>
<li class="chapter" data-level="2.3" data-path="SEACells.html"><a href="SEACells.html"><i class="fa fa-check"></i><b>2.3</b> SEACells</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="constructing-metacells-for-a-discrete-data.html"><a href="constructing-metacells-for-a-discrete-data.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells (for a ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="mc2-python.html"><a href="mc2-python.html"><i class="fa fa-check"></i><b>3.1</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="" data-path="mc2-python.html"><a href="mc2-python.html#running-mc2"><i class="fa fa-check"></i>Running MC2</a></li>
<li class="chapter" data-level="" data-path="mc2-python.html"><a href="mc2-python.html#compute-latent-space-for-metacell-qc"><i class="fa fa-check"></i>Compute latent space for metacell QC</a></li>
<li class="chapter" data-level="" data-path="mc2-python.html"><a href="mc2-python.html#metacell-qc"><i class="fa fa-check"></i>Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="supercell-r.html"><a href="supercell-r.html"><i class="fa fa-check"></i><b>3.2</b> SuperCell (R)</a></li>
<li class="chapter" data-level="3.3" data-path="seacells-python.html"><a href="seacells-python.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="" data-path="seacells-python.html"><a href="seacells-python.html#running-seacells"><i class="fa fa-check"></i>Running SEACells</a></li>
<li class="chapter" data-level="" data-path="seacells-python.html"><a href="seacells-python.html#metacell-qc-1"><i class="fa fa-check"></i>Metacell QC</a></li>
<li class="chapter" data-level="" data-path="seacells-python.html"><a href="seacells-python.html#save-metacell-object-for-further-analysis"><i class="fa fa-check"></i>Save metacell object for further analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><a href="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><i class="fa fa-check"></i><b>4</b> Downstream analysis of metacells (for a discrete dataset)</a>
<ul>
<li class="chapter" data-level="4.1" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html"><i class="fa fa-check"></i><b>4.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html#dimensionality-reduction"><i class="fa fa-check"></i><b>4.1.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="4.1.2" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html#clustering"><i class="fa fa-check"></i><b>4.1.2</b> Clustering</a></li>
<li class="chapter" data-level="4.1.3" data-path="standard-analysis-r.html"><a href="standard-analysis-r.html#differential-expression-analysis"><i class="fa fa-check"></i><b>4.1.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>4.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>4.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="4.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering-1"><i class="fa fa-check"></i><b>4.2.2</b> Clustering</a></li>
<li class="chapter" data-level="4.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>4.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>4.3</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#grn"><i class="fa fa-check"></i><b>4.3.1</b> GRN</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>4.4</b> Sample-weighted analysis</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="MC-continuous.html"><a href="MC-continuous.html"><i class="fa fa-check"></i><b>5</b> Construction of metacells (for ‘continuous’ data)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="mc2-python-1.html"><a href="mc2-python-1.html"><i class="fa fa-check"></i><b>5.1</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="" data-path="mc2-python-1.html"><a href="mc2-python-1.html#running-mc2-1"><i class="fa fa-check"></i>Running MC2</a></li>
<li class="chapter" data-level="" data-path="mc2-python-1.html"><a href="mc2-python-1.html#compute-latent-space-for-metacell-qc-1"><i class="fa fa-check"></i>Compute latent space for metacell QC</a></li>
<li class="chapter" data-level="" data-path="mc2-python-1.html"><a href="mc2-python-1.html#metacell-qc-2"><i class="fa fa-check"></i>Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="supercell-r-1.html"><a href="supercell-r-1.html"><i class="fa fa-check"></i><b>5.2</b> SuperCell (R)</a></li>
<li class="chapter" data-level="5.3" data-path="seacells-python-1.html"><a href="seacells-python-1.html"><i class="fa fa-check"></i><b>5.3</b> SEACells (Python)</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="downstream-analysis-of-metacells-for-a-continuous-dataset.html"><a href="downstream-analysis-of-metacells-for-a-continuous-dataset.html"><i class="fa fa-check"></i><b>6</b> Downstream analysis of metacells (for a continuous dataset)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="standard-analysis-r-1.html"><a href="standard-analysis-r-1.html"><i class="fa fa-check"></i><b>6.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="standard-analysis-r-1.html"><a href="standard-analysis-r-1.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>6.1.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.2" data-path="standard-analysis-r-1.html"><a href="standard-analysis-r-1.html#clustering-2"><i class="fa fa-check"></i><b>6.1.2</b> Clustering</a></li>
<li class="chapter" data-level="6.1.3" data-path="standard-analysis-r-1.html"><a href="standard-analysis-r-1.html#differential-expression-analysis-2"><i class="fa fa-check"></i><b>6.1.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="standard-analysis-Py-cont.html"><a href="standard-analysis-Py-cont.html"><i class="fa fa-check"></i><b>6.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="standard-analysis-Py-cont.html"><a href="standard-analysis-Py-cont.html#dimensionality-reduction-3"><i class="fa fa-check"></i><b>6.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.2.2" data-path="standard-analysis-Py-cont.html"><a href="standard-analysis-Py-cont.html#clustering-3"><i class="fa fa-check"></i><b>6.2.2</b> Clustering</a></li>
<li class="chapter" data-level="6.2.3" data-path="standard-analysis-Py-cont.html"><a href="standard-analysis-Py-cont.html#differential-expression-analysis-3"><i class="fa fa-check"></i><b>6.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="weighted-analysis-cont.html"><a href="weighted-analysis-cont.html"><i class="fa fa-check"></i><b>6.3</b> Sample-weighted analysis</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><a href="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><i class="fa fa-check"></i><b>7</b> Comparison of metacells usage for a discrete and continuous data</a>
<ul>
<li class="chapter" data-level="7.1" data-path="continuous-metacells-have-lower-purity.html"><a href="continuous-metacells-have-lower-purity.html"><i class="fa fa-check"></i><b>7.1</b> Continuous metacells have lower purity</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-integration-with-metacells.html"><a href="data-integration-with-metacells.html"><i class="fa fa-check"></i><b>8</b> Data integration with metacells</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mc2-python-1" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> MC2 (Python)<a href="mc2-python-1.html#mc2-python-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>This is a template, not working code</strong></p>
<p>Here we construct metacells using <a href="https://github.com/tanaylab/metacells">Metacell-2 (MC2)</a>. The code is adapted from the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">author’s tutorial</a>.</p>
<div id="imports-2" class="section level4 unnumbered hasAnchor">
<h4>Imports<a href="mc2-python-1.html#imports-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb47"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb47-1"><a href="mc2-python-1.html#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb47-2"><a href="mc2-python-1.html#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb47-3"><a href="mc2-python-1.html#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb47-4"><a href="mc2-python-1.html#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb47-5"><a href="mc2-python-1.html#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb47-6"><a href="mc2-python-1.html#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb47-7"><a href="mc2-python-1.html#cb47-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb47-8"><a href="mc2-python-1.html#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span>
<span id="cb47-9"><a href="mc2-python-1.html#cb47-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-10"><a href="mc2-python-1.html#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb47-11"><a href="mc2-python-1.html#cb47-11" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>) </span>
<span id="cb47-12"><a href="mc2-python-1.html#cb47-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span></code></pre></div>
</div>
<div id="parameters-1" class="section level4 unnumbered hasAnchor">
<h4>Parameters<a href="mc2-python-1.html#parameters-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb48"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb48-1"><a href="mc2-python-1.html#cb48-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-2"><a href="mc2-python-1.html#cb48-2" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;MC2&quot;</span></span>
<span id="cb48-3"><a href="mc2-python-1.html#cb48-3" aria-hidden="true" tabindex="-1"></a>gamma   <span class="op">=</span> <span class="dv">50</span> <span class="co"># graining level</span></span>
<span id="cb48-4"><a href="mc2-python-1.html#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="mc2-python-1.html#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we can modify dataset </span></span>
<span id="cb48-6"><a href="mc2-python-1.html#cb48-6" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">&quot;CD34&quot;</span></span>
<span id="cb48-7"><a href="mc2-python-1.html#cb48-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-8"><a href="mc2-python-1.html#cb48-8" aria-hidden="true" tabindex="-1"></a>annotation_label <span class="op">=</span> <span class="st">&quot;cell_type&quot;</span> <span class="co"># name of annotation field (obs)</span></span></code></pre></div>
</div>
<div id="load-data-2" class="section level4 unnumbered hasAnchor">
<h4>Load data<a href="mc2-python-1.html#load-data-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Setup MC object</strong></p>
<div class="sourceCode" id="cb49"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb49-1"><a href="mc2-python-1.html#cb49-1" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(ad, proj_name)</span></code></pre></div>
</div>
<div id="gene-filtering-according-to-mc2-1" class="section level4 unnumbered hasAnchor">
<h4>Gene filtering according to MC2<a href="mc2-python-1.html#gene-filtering-according-to-mc2-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb50"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb50-1"><a href="mc2-python-1.html#cb50-1" aria-hidden="true" tabindex="-1"></a>excluded_gene_names <span class="op">=</span> [] <span class="co"># for example, [&#39;IGHMBP2&#39;, &#39;IGLL1&#39;, &#39;IGLL5&#39;, &#39;IGLON5&#39;, &#39;NEAT1&#39;, &#39;TMSB10&#39;, &#39;TMSB4X&#39;]</span></span>
<span id="cb50-2"><a href="mc2-python-1.html#cb50-2" aria-hidden="true" tabindex="-1"></a>excluded_gene_patterns <span class="op">=</span> [<span class="st">&#39;MT-.*&#39;</span>]</span>
<span id="cb50-3"><a href="mc2-python-1.html#cb50-3" aria-hidden="true" tabindex="-1"></a>mc.pl.analyze_clean_genes(ad,</span>
<span id="cb50-4"><a href="mc2-python-1.html#cb50-4" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_names<span class="op">=</span>excluded_gene_names,</span>
<span id="cb50-5"><a href="mc2-python-1.html#cb50-5" aria-hidden="true" tabindex="-1"></a>                          excluded_gene_patterns<span class="op">=</span>excluded_gene_patterns,</span>
<span id="cb50-6"><a href="mc2-python-1.html#cb50-6" aria-hidden="true" tabindex="-1"></a>                          random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb50-7"><a href="mc2-python-1.html#cb50-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-8"><a href="mc2-python-1.html#cb50-8" aria-hidden="true" tabindex="-1"></a>mc.pl.pick_clean_genes(ad)</span></code></pre></div>
</div>
<div id="cell-fintering-1" class="section level4 unnumbered hasAnchor">
<h4>Cell fintering<a href="mc2-python-1.html#cell-fintering-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Since our data is pre-processed and low-quality cells have been already filtered out, this step is not applicable to our data, but we keep this chunk so that you can apply cell filtering to a newly generated dataset.</p>
<p>The first round of cell cleaning usually implies filltering out cell with very low and very hight UMI content. The second round includes cell filtering based on mitochondrial and/or ribosomal content. We will skip both steps as our data have been pre-filtered and will use very lenient cutoffs (<code>properly_sampled_min_cell_total</code>, <code>properly_sampled_max_cell_total</code> and <code>properly_sampled_max_excluded_genes_fraction</code>) such that all the cells are kept for the metacell construction.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb51-1"><a href="mc2-python-1.html#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">### The first round (high/low UMIs)</span></span>
<span id="cb51-2"><a href="mc2-python-1.html#cb51-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_min_cell_total <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span> : <span class="dv">5000</span>, <span class="st">&#39;3k_pbmc&#39;</span>: <span class="dv">200</span>}[proj_name] <span class="co"># setup for the dataset that will be used </span></span>
<span id="cb51-3"><a href="mc2-python-1.html#cb51-3" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_cell_total <span class="op">=</span> {<span class="st">&#39;cell_lines&#39;</span> : <span class="dv">110000</span>, <span class="st">&#39;3k_pbmc&#39;</span>: <span class="dv">10000</span>}[proj_name] <span class="co"># setup for the dataset that will be used </span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb52-1"><a href="mc2-python-1.html#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">## The second round (content of non-clean genes, e.g., mito-genes)</span></span>
<span id="cb52-2"><a href="mc2-python-1.html#cb52-2" aria-hidden="true" tabindex="-1"></a>properly_sampled_max_excluded_genes_fraction <span class="op">=</span> <span class="fl">0.25</span></span></code></pre></div>
</div>
<div id="running-mc2-1" class="section level3 unnumbered hasAnchor">
<h3>Running MC2<a href="mc2-python-1.html#running-mc2-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Metacell-2 uses its own feature selection approach (i.e., selection of genes used to build metacells). Additionally, we can explicitly specify which features to use by providing two arguments: feature_gene_names - genes that have to be used forbidden_gene_names - genes to exclude.
In contrast to the SuperCell and SEACells, Metacell-2 does not allow to explicitly obtain metacell data at a user-defined graining level. Instead, to vary graining level, we have to vary a <code>target_metacell_size</code> parameter, that is <code>160000</code> by default. Here we provide a chunk to calibrate this value to get a desired graining level. Please, increase or decrease scale if the obtained graining level <code>gamma_obtained</code> is lower or larger than the requested one (gamma).</p>
<p><strong>Estimate target_metacell_size (gamma)</strong></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb53-1"><a href="mc2-python-1.html#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f&#39;The requested graining level is </span><span class="sc">{</span>gamma<span class="sc">}</span><span class="ss">, lets estimate the target_metacell_size that should result in such graining level.&#39;</span>)</span>
<span id="cb53-2"><a href="mc2-python-1.html#cb53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-3"><a href="mc2-python-1.html#cb53-3" aria-hidden="true" tabindex="-1"></a>scale <span class="op">=</span> <span class="dv">2</span> <span class="co"># increase or decrease if the obtained graining level (`gamma_obtained`) is significantly &gt; or &lt; then the requested one `gamma`</span></span>
<span id="cb53-4"><a href="mc2-python-1.html#cb53-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-5"><a href="mc2-python-1.html#cb53-5" aria-hidden="true" tabindex="-1"></a>N_c <span class="op">=</span> ad.shape[<span class="dv">0</span>]</span>
<span id="cb53-6"><a href="mc2-python-1.html#cb53-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-7"><a href="mc2-python-1.html#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co"># estimated mean UMI content in downsampled data</span></span>
<span id="cb53-8"><a href="mc2-python-1.html#cb53-8" aria-hidden="true" tabindex="-1"></a>est_downsample_UMI <span class="op">=</span> np.quantile(np.array(total_umis_of_cells), <span class="fl">0.05</span>)</span>
<span id="cb53-9"><a href="mc2-python-1.html#cb53-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-10"><a href="mc2-python-1.html#cb53-10" aria-hidden="true" tabindex="-1"></a>target_metacell_size <span class="op">=</span> <span class="bu">int</span>(est_downsample_UMI <span class="op">*</span> gamma <span class="op">*</span> scale)</span>
<span id="cb53-11"><a href="mc2-python-1.html#cb53-11" aria-hidden="true" tabindex="-1"></a>target_metacell_size</span></code></pre></div>
<div id="aggregate-mc2" class="section level4 unnumbered hasAnchor">
<h4>Aggregate metacells<a href="mc2-python.html#aggregate-mc2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb54"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb54-1"><a href="mc2-python-1.html#cb54-1" aria-hidden="true" tabindex="-1"></a>mc.pl.divide_and_conquer_pipeline(</span>
<span id="cb54-2"><a href="mc2-python-1.html#cb54-2" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb54-3"><a href="mc2-python-1.html#cb54-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#feature_gene_names   = feature_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb54-4"><a href="mc2-python-1.html#cb54-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#forbidden_gene_names = forbidden_gene_names, # comment this line to allow Metacell2 selecting features</span></span>
<span id="cb54-5"><a href="mc2-python-1.html#cb54-5" aria-hidden="true" tabindex="-1"></a>    target_metacell_size <span class="op">=</span> target_metacell_size,</span>
<span id="cb54-6"><a href="mc2-python-1.html#cb54-6" aria-hidden="true" tabindex="-1"></a>    random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb54-7"><a href="mc2-python-1.html#cb54-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-8"><a href="mc2-python-1.html#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="co">## make anndata of metacells</span></span>
<span id="cb54-9"><a href="mc2-python-1.html#cb54-9" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> mc.pl.collect_metacells(ad, name<span class="op">=</span><span class="st">&#39;metacells&#39;</span>)</span></code></pre></div>
<p>Here we estimate whether a deviation of the obtained gamma is acceptable, and if not, suggest to increase or decrease <code>scale</code> parameter to get better graining level.</p>
<p>If the obtained graining level is not acceptable and you updated <code>scale</code> parameter according to suggestion, do not forget to re-run chunk <a href="#chunk:mc2-divide-n-conquer"><strong>??</strong></a> <a href="mc2-python.html#aggregate-mc2"><strong>??</strong></a></p>
</div>
<div id="visualize-metacells-1" class="section level4 unnumbered hasAnchor">
<h4>Visualize metacells<a href="mc2-python-1.html#visualize-metacells-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb55"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb55-1"><a href="mc2-python-1.html#cb55-1" aria-hidden="true" tabindex="-1"></a>mc.pl.compute_umap_by_features(mc_ad, max_top_feature_genes<span class="op">=</span><span class="dv">1000</span>,</span>
<span id="cb55-2"><a href="mc2-python-1.html#cb55-2" aria-hidden="true" tabindex="-1"></a>                               min_dist<span class="op">=</span><span class="fl">2.0</span>, random_seed<span class="op">=</span><span class="dv">123456</span>)</span>
<span id="cb55-3"><a href="mc2-python-1.html#cb55-3" aria-hidden="true" tabindex="-1"></a>                               </span>
<span id="cb55-4"><a href="mc2-python-1.html#cb55-4" aria-hidden="true" tabindex="-1"></a>umap_x <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_x&#39;</span>)</span>
<span id="cb55-5"><a href="mc2-python-1.html#cb55-5" aria-hidden="true" tabindex="-1"></a>umap_y <span class="op">=</span> mc.ut.get_o_numpy(mc_ad, <span class="st">&#39;umap_y&#39;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb56-1"><a href="mc2-python-1.html#cb56-1" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb56-2"><a href="mc2-python-1.html#cb56-2" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(x<span class="op">=</span>umap_x, y<span class="op">=</span>umap_y)</span>
<span id="cb56-3"><a href="mc2-python-1.html#cb56-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb57-1"><a href="mc2-python-1.html#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a membership -- index of metacell single cell belongs to </span></span>
<span id="cb57-2"><a href="mc2-python-1.html#cb57-2" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&#39;membership&#39;</span>] <span class="op">=</span> [<span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">0</span> <span class="cf">else</span> np.nan <span class="cf">for</span> i <span class="kw">in</span> ad.obs.metacell] </span>
<span id="cb57-3"><a href="mc2-python-1.html#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="mc2-python-1.html#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Save single-cell metadata (i.e., `raw.obs` dataframe) in the metacell adata object</span></span>
<span id="cb57-5"><a href="mc2-python-1.html#cb57-5" aria-hidden="true" tabindex="-1"></a>mc_ad.uns <span class="op">=</span> ad.uns.copy()</span>
<span id="cb57-6"><a href="mc2-python-1.html#cb57-6" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.obs&#39;</span>] <span class="op">=</span> ad.obs.copy()</span>
<span id="cb57-7"><a href="mc2-python-1.html#cb57-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-8"><a href="mc2-python-1.html#cb57-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the requested gamma</span></span>
<span id="cb57-9"><a href="mc2-python-1.html#cb57-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;gamma&#39;</span>] <span class="op">=</span> gamma</span></code></pre></div>
</div>
</div>
<div id="compute-latent-space-for-metacell-qc-1" class="section level3 unnumbered hasAnchor">
<h3>Compute latent space for metacell QC<a href="mc2-python-1.html#compute-latent-space-for-metacell-qc-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="metacell-qc-2" class="section level3 unnumbered hasAnchor">
<h3>Metacell QC<a href="mc2-python-1.html#metacell-qc-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="MC-continuous.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="supercell-r-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/31-MC_construction_continuous.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/31-MC_construction_continuous.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
