<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.2 MC2 (Python) | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="3.2 MC2 (Python) | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.2 MC2 (Python) | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-08-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="supercell-r.html"/>
<link rel="next" href="seacells-python.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="book-structure.html"><a href="book-structure.html"><i class="fa fa-check"></i><b>1.1</b> Book structure</a></li>
<li class="chapter" data-level="1.2" data-path="installation-and-requirements.html"><a href="installation-and-requirements.html"><i class="fa fa-check"></i><b>1.2</b> Installation and requirements</a></li>
<li class="chapter" data-level="1.3" data-path="render-book.html"><a href="render-book.html"><i class="fa fa-check"></i><b>1.3</b> Render book</a></li>
<li class="chapter" data-level="1.4" data-path="get-data.html"><a href="get-data.html"><i class="fa fa-check"></i><b>1.4</b> Get data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a>
<ul>
<li class="chapter" data-level="2.1" data-path="MC2.html"><a href="MC2.html"><i class="fa fa-check"></i><b>2.1</b> Metacell (MC2)</a></li>
<li class="chapter" data-level="2.2" data-path="SuperCell.html"><a href="SuperCell.html"><i class="fa fa-check"></i><b>2.2</b> SuperCell</a></li>
<li class="chapter" data-level="2.3" data-path="SEACells.html"><a href="SEACells.html"><i class="fa fa-check"></i><b>2.3</b> SEACells</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="constructing-metacells-for-discrete-data.html"><a href="constructing-metacells-for-discrete-data.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells (for ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="supercell-r.html"><a href="supercell-r.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="supercell-r.html"><a href="supercell-r.html#data-loading"><i class="fa fa-check"></i><b>3.1.1</b> Data loading</a></li>
<li class="chapter" data-level="3.1.2" data-path="supercell-r.html"><a href="supercell-r.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.3" data-path="supercell-r.html"><a href="supercell-r.html#building-metacells"><i class="fa fa-check"></i><b>3.1.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.4" data-path="supercell-r.html"><a href="supercell-r.html#visualize-the-supercell-graph"><i class="fa fa-check"></i><b>3.1.4</b> Visualize the SuperCell graph</a></li>
<li class="chapter" data-level="3.1.5" data-path="supercell-r.html"><a href="supercell-r.html#metacell-qc"><i class="fa fa-check"></i><b>3.1.5</b> Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="mc2-python.html"><a href="mc2-python.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="mc2-python.html"><a href="mc2-python.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.1</b> Data loading</a></li>
<li class="chapter" data-level="3.2.2" data-path="mc2-python.html"><a href="mc2-python.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.3" data-path="mc2-python.html"><a href="mc2-python.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.4" data-path="mc2-python.html"><a href="mc2-python.html#visualize-metacells"><i class="fa fa-check"></i><b>3.2.4</b> Visualize metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="mc2-python.html"><a href="mc2-python.html#metacell-qc-1"><i class="fa fa-check"></i><b>3.2.5</b> Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="seacells-python.html"><a href="seacells-python.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="seacells-python.html"><a href="seacells-python.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.1</b> Data loading</a></li>
<li class="chapter" data-level="3.3.2" data-path="seacells-python.html"><a href="seacells-python.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.3" data-path="seacells-python.html"><a href="seacells-python.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.4" data-path="seacells-python.html"><a href="seacells-python.html#visualize-metacells-1"><i class="fa fa-check"></i><b>3.3.4</b> Visualize metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="seacells-python.html"><a href="seacells-python.html#metacell-qc-2"><i class="fa fa-check"></i><b>3.3.5</b> Metacell QC</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="command-line-tool-for-mc-construction.html"><a href="command-line-tool-for-mc-construction.html"><i class="fa fa-check"></i><b>4</b> Command line tool for MC construction</a></li>
<li class="chapter" data-level="5" data-path="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><a href="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><i class="fa fa-check"></i><b>5</b> Downstream analysis of metacells (for a discrete dataset)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>5.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>5.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>5.1.3</b> Clustering</a></li>
<li class="chapter" data-level="5.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>5.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="5.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#plot-the-expression-of-some-markers"><i class="fa fa-check"></i><b>5.1.5</b> Plot the expression of some markers</a></li>
<li class="chapter" data-level="5.1.6" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#plot-gene-gene-correlation-at-single-cell-and-metacell-levels"><i class="fa fa-check"></i><b>5.1.6</b> Plot gene-gene correlation at single-cell and metacell levels</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>5.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>5.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering-1"><i class="fa fa-check"></i><b>5.2.2</b> Clustering</a></li>
<li class="chapter" data-level="5.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>5.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>5.3</b> Sample-weighted analysis</a></li>
<li class="chapter" data-level="5.4" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>5.4</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#GRN-analysis"><i class="fa fa-check"></i><b>5.4.1</b> GRN</a></li>
<li class="chapter" data-level="5.4.2" data-path="advanced-analysis.html"><a href="advanced-analysis.html#integration"><i class="fa fa-check"></i><b>5.4.2</b> Integration of metacells</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><a href="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><i class="fa fa-check"></i><b>6</b> Comparison of metacells usage for a discrete and continuous data</a>
<ul>
<li class="chapter" data-level="6.1" data-path="continuous-metacells-have-lower-purity.html"><a href="continuous-metacells-have-lower-purity.html"><i class="fa fa-check"></i><b>6.1</b> Continuous metacells have lower purity</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mc2-python" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> MC2 (Python)<a href="mc2-python.html#mc2-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this section, we construct metacells using <a href="https://github.com/tanaylab/metacells">Metacell-2 (MC2)</a>. The code is adapted from the <a href="https://metacells.readthedocs.io/en/latest/Metacells_Vignette.html">author’s tutorial</a>. For more information on the method, please refer to the section 1 of chapter 2.</p>
<div id="importing-python-packages" class="section level4 unnumbered hasAnchor">
<h4>Importing python packages<a href="mc2-python.html#importing-python-packages" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To run Metacell-2, the following python packages need to be imported:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb34-1"><a href="mc2-python.html#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb34-2"><a href="mc2-python.html#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb34-3"><a href="mc2-python.html#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb34-4"><a href="mc2-python.html#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb34-5"><a href="mc2-python.html#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb34-6"><a href="mc2-python.html#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb34-7"><a href="mc2-python.html#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb34-8"><a href="mc2-python.html#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb35-1"><a href="mc2-python.html#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb35-2"><a href="mc2-python.html#cb35-2" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>)</span>
<span id="cb35-3"><a href="mc2-python.html#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span></code></pre></div>
<p>If you don’t have these packages installed, please refer to the section 2 of chapter 1.</p>
</div>
<div id="data-loading-1" class="section level3 hasAnchor" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Data loading<a href="mc2-python.html#data-loading-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will run Metacell-2 (MC2) on a single-cell dataset composed of XX peripheral blood mononuclear cells (PBMCs). Please follow the section 4 from Chapter 1 to retrieve these data from the scanpy package and save the data in the following file: “data/3k_pbmc/singlecell_anndata_filtered.h5ad”.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb36-1"><a href="mc2-python.html#cb36-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;MC2&quot;</span></span>
<span id="cb36-2"><a href="mc2-python.html#cb36-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">&quot;3k_pbmc&quot;</span></span>
<span id="cb36-3"><a href="mc2-python.html#cb36-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(<span class="st">&quot;data&quot;</span>, proj_name, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span>
<span id="cb36-4"><a href="mc2-python.html#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="mc2-python.html#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="co"># MC_tool = &quot;MC2&quot;</span></span>
<span id="cb36-6"><a href="mc2-python.html#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="co"># proj_name = &quot;CD34&quot;</span></span>
<span id="cb36-7"><a href="mc2-python.html#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ad = sc.read(os.path.join(&quot;data&quot;, &quot;CD34&quot;, &quot;cd34_multiome_rna.h5ad&quot;))</span></span></code></pre></div>
<p>We initialize the name of the anndata (in the unstructured annotations) object using the <code>mc.ut.set_name()</code> function from the MC2 package.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb37-1"><a href="mc2-python.html#cb37-1" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(ad, proj_name)</span></code></pre></div>
</div>
<div id="filtering-steps-1" class="section level3 hasAnchor" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Filtering steps<a href="mc2-python.html#filtering-steps-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>MC2 requires that standard filtering steps such as doublet filtering is performed outside of the MC2 framework.
In addition to standard data filtering steps, the MC2 package proposes functions to filter the single-cell data at the gene and at the cell level (See <a href="https://tanaylab.github.io/metacells-vignettes/one-pass.html">original vignette</a>).
At the gene level, the filtering steps consist in excluding genes based on biological knowledge (<em>e.g.</em> mitochrondrial genes) as well as based on their expression levels.
The latter genes include genes with zero expression or low expression levels and “bursty lonely genes” (<em>i.e.</em>, genes with high expression levels but no correlation with any other gene).
At the cell level, filtering is performed based on cells UMI counts.</p>
<div id="gene-filtering" class="section level4 unnumbered hasAnchor">
<h4>Gene filtering<a href="mc2-python.html#gene-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In section XX form Chapter XX, we pre-processed the raw scRNA-Seq data and excluded genes with low expression as well as mitochondrial genes.
In the following code chunk, we exclude additional genes using the <code>mc.pl.exclude_genes()</code>function from the MC2 package.
Based on the authors vignette, we provide a minimal list of genes to exclude, <em>i.e.</em>, sex-specific and non-coding genes. To complete this list of genes, an iterative approach can be used following the guidelines of the MC2 authors in a <a href="https://tanaylab.github.io/metacells-vignettes/iterative.html">second vignette</a>.
The <code>mc.pl.exclude_genes()</code>function will filter out: i) the known-to-be-excluded genes defined by the user as gene names or gene names patterns (<code>EXCLUDED_GENE_NAMES</code> and <code>EXCLUDED_GENE_PATTERNS</code> parameters respectively),
and ii) the “bursty lonely genes”.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb38-1"><a href="mc2-python.html#cb38-1" aria-hidden="true" tabindex="-1"></a>EXCLUDED_GENE_NAMES <span class="op">=</span> [<span class="st">&quot;XIST&quot;</span>, <span class="st">&quot;MALAT1&quot;</span>, <span class="st">&quot;NEAT1&quot;</span>] </span>
<span id="cb38-2"><a href="mc2-python.html#cb38-2" aria-hidden="true" tabindex="-1"></a>EXCLUDED_GENE_PATTERNS <span class="op">=</span> [<span class="st">&#39;MT-.*&#39;</span>]</span>
<span id="cb38-3"><a href="mc2-python.html#cb38-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-4"><a href="mc2-python.html#cb38-4" aria-hidden="true" tabindex="-1"></a>mc.pl.exclude_genes(</span>
<span id="cb38-5"><a href="mc2-python.html#cb38-5" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb38-6"><a href="mc2-python.html#cb38-6" aria-hidden="true" tabindex="-1"></a>    excluded_gene_names<span class="op">=</span>EXCLUDED_GENE_NAMES,</span>
<span id="cb38-7"><a href="mc2-python.html#cb38-7" aria-hidden="true" tabindex="-1"></a>    excluded_gene_patterns<span class="op">=</span>EXCLUDED_GENE_PATTERNS,</span>
<span id="cb38-8"><a href="mc2-python.html#cb38-8" aria-hidden="true" tabindex="-1"></a>    random_seed<span class="op">=</span><span class="dv">123456</span></span>
<span id="cb38-9"><a href="mc2-python.html#cb38-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-10"><a href="mc2-python.html#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[bursty_lonely_gene]: 0 true (0%) out of 32738 bools</span></span>
<span id="cb38-11"><a href="mc2-python.html#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[properly_sampled_gene]: 16579 true (50.64%) out of 32738 bools</span></span>
<span id="cb38-12"><a href="mc2-python.html#cb38-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[excluded_gene]: 16174 true (49.4%) out of 32738 bools</span></span></code></pre></div>
</div>
<div id="cell-filtering-based-on-umis-counts" class="section level4 unnumbered hasAnchor">
<h4>Cell filtering based on UMIs counts<a href="mc2-python.html#cell-filtering-based-on-umis-counts" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In the MC2 framework, cells with very low and very high UMI content will be filtered out (<code>PROPERLY_SAMPLED_MIN_CELL_TOTAL</code>, <code>PROPERLY_SAMPLED_MAX_CELL_TOTAL</code> variables defining thresholds in the next code chunk).<br />
Also, cell filtering based on UMI counts in excluded genes is also performed(<code>PROPERLY_SAMPLED_MAX_EXCLUDED_GENES_FRACTION</code> variable).
Since our dataset has been pre-filtered, very lenient cutoffs will be used in this tutorial.
The following code chunk defines these parameters. To adapt them to your datasets, we advise you to explore the distributions of total UMI counts and UMI counts in excluded genes, as recommended and described in the MC2 <a href="https://tanaylab.github.io/metacells-vignettes/one-pass.html">original vignette</a>.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb39-1"><a href="mc2-python.html#cb39-1" aria-hidden="true" tabindex="-1"></a>PROPERLY_SAMPLED_MIN_CELL_TOTAL <span class="op">=</span> <span class="dv">200</span> </span>
<span id="cb39-2"><a href="mc2-python.html#cb39-2" aria-hidden="true" tabindex="-1"></a>PROPERLY_SAMPLED_MAX_CELL_TOTAL <span class="op">=</span> <span class="dv">10000</span> </span>
<span id="cb39-3"><a href="mc2-python.html#cb39-3" aria-hidden="true" tabindex="-1"></a>PROPERLY_SAMPLED_MAX_EXCLUDED_GENES_FRACTION <span class="op">=</span> <span class="fl">0.25</span></span></code></pre></div>
<p>The number of UMIs in excluded genes is computed using the <code>mc.tl.compute_excluded_gene_umis()</code> function and cells are filtered out using the <code>mc.pl.exclude_cells()</code> function.
Additional cells can be filtered out by adding a cell description columns in the <code>obs</code> data frame in the anndata oject. This annotation should be a boolean indicating whether the cell should filtered out or not.
The name of this column should be provided to the <code>mc.pl.exclude_cells()</code> function via the <code>additional_cells_masks</code> parameter.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb40-1"><a href="mc2-python.html#cb40-1" aria-hidden="true" tabindex="-1"></a>mc.tl.compute_excluded_gene_umis(ad)</span>
<span id="cb40-2"><a href="mc2-python.html#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="mc2-python.html#cb40-3" aria-hidden="true" tabindex="-1"></a>mc.pl.exclude_cells(</span>
<span id="cb40-4"><a href="mc2-python.html#cb40-4" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb40-5"><a href="mc2-python.html#cb40-5" aria-hidden="true" tabindex="-1"></a>    properly_sampled_min_cell_total<span class="op">=</span>PROPERLY_SAMPLED_MIN_CELL_TOTAL,</span>
<span id="cb40-6"><a href="mc2-python.html#cb40-6" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_cell_total<span class="op">=</span>PROPERLY_SAMPLED_MAX_CELL_TOTAL,</span>
<span id="cb40-7"><a href="mc2-python.html#cb40-7" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_excluded_genes_fraction<span class="op">=</span>PROPERLY_SAMPLED_MAX_EXCLUDED_GENES_FRACTION <span class="co"># ,</span></span>
<span id="cb40-8"><a href="mc2-python.html#cb40-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># additional_cells_masks=[&quot;|doublet_cell&quot;]</span></span>
<span id="cb40-9"><a href="mc2-python.html#cb40-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-10"><a href="mc2-python.html#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[properly_sampled_cell]: 2638 true (100%) out of 2638 bools</span></span>
<span id="cb40-11"><a href="mc2-python.html#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[excluded_cell]: 0 true (0%) out of 2638 bools</span></span></code></pre></div>
<p>After performing the two-step filtering (genes and cells), the “cleaned” data can be extracted using the <code>mc.pl.extract_clean_data()</code> function.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb41-1"><a href="mc2-python.html#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract clean dataset (with filtered cells and genes)</span></span>
<span id="cb41-2"><a href="mc2-python.html#cb41-2" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> mc.pl.extract_clean_data(ad)</span>
<span id="cb41-3"><a href="mc2-python.html#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[full_cell_index]: 2638 int32s</span></span>
<span id="cb41-4"><a href="mc2-python.html#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[full_gene_index]: 16564 int32s</span></span></code></pre></div>
</div>
</div>
<div id="building-metacells-1" class="section level3 hasAnchor" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Building metacells<a href="mc2-python.html#building-metacells-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="defining-lateral-genes" class="section level4 unnumbered hasAnchor">
<h4>Defining lateral genes<a href="mc2-python.html#defining-lateral-genes" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To build metacells, we need to define lateral genes, which are genes with strong biological signal which is independent of cell-state, <em>e.g.</em> cell-cycle genes.
These genes will be ignored for computing cells similarity and build metacells
but will be considered to define outlier cells (<em>i.e.</em>, expression levels of lateral genes should be consistent within metacells).
In the following chunk, we consider a minimal list of lateral genes including cell-cycle and ribosomal genes and mark them in the MC2 object using the <code>mc.pl.mark_lateral_genes()</code> function.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb42-1"><a href="mc2-python.html#cb42-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-2"><a href="mc2-python.html#cb42-2" aria-hidden="true" tabindex="-1"></a>LATERAL_GENE_NAMES <span class="op">=</span> [</span>
<span id="cb42-3"><a href="mc2-python.html#cb42-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;AURKA&quot;</span>, <span class="st">&quot;MCM3&quot;</span>, <span class="st">&quot;MCM4&quot;</span>, <span class="st">&quot;MCM7&quot;</span>, <span class="st">&quot;MKI67&quot;</span>, <span class="st">&quot;PCNA&quot;</span>, <span class="st">&quot;RRM2&quot;</span>, <span class="st">&quot;SMC4&quot;</span>, <span class="st">&quot;TPX2&quot;</span>,  <span class="co"># Cell-cycle</span></span>
<span id="cb42-4"><a href="mc2-python.html#cb42-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;FOS&quot;</span>, <span class="st">&quot;HSP90AB1&quot;</span>, <span class="st">&quot;TXN&quot;</span>,                                                  <span class="co"># Stress</span></span>
<span id="cb42-5"><a href="mc2-python.html#cb42-5" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb42-6"><a href="mc2-python.html#cb42-6" aria-hidden="true" tabindex="-1"></a>LATERAL_GENE_PATTERNS <span class="op">=</span> [<span class="st">&quot;RP[LS].*&quot;</span>]  <span class="co"># Ribosomal</span></span>
<span id="cb42-7"><a href="mc2-python.html#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="mc2-python.html#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="co"># This will mark as &quot;lateral_gene&quot; any genes that match the above, if they exist in the clean dataset.</span></span>
<span id="cb42-9"><a href="mc2-python.html#cb42-9" aria-hidden="true" tabindex="-1"></a>mc.pl.mark_lateral_genes(</span>
<span id="cb42-10"><a href="mc2-python.html#cb42-10" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb42-11"><a href="mc2-python.html#cb42-11" aria-hidden="true" tabindex="-1"></a>    lateral_gene_names<span class="op">=</span>LATERAL_GENE_NAMES,</span>
<span id="cb42-12"><a href="mc2-python.html#cb42-12" aria-hidden="true" tabindex="-1"></a>    lateral_gene_patterns<span class="op">=</span>LATERAL_GENE_PATTERNS,</span>
<span id="cb42-13"><a href="mc2-python.html#cb42-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-14"><a href="mc2-python.html#cb42-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[lateral_gene]: 111 true (0.6701%) out of 16564 bools</span></span></code></pre></div>
<p>Some genes have higher variances that expected which could lead to false positive outlier identification.
Users can mark those genes as <em>noisy genes</em> using the <code>mc.pl.mark_noisy_genes()</code> function.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb43-1"><a href="mc2-python.html#cb43-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-2"><a href="mc2-python.html#cb43-2" aria-hidden="true" tabindex="-1"></a>NOISY_GENE_NAMES <span class="op">=</span> [</span>
<span id="cb43-3"><a href="mc2-python.html#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;CCL3&quot;</span>, <span class="st">&quot;CCL4&quot;</span>, <span class="st">&quot;CCL5&quot;</span>, <span class="st">&quot;CXCL8&quot;</span>, <span class="st">&quot;DUSP1&quot;</span>, <span class="st">&quot;FOS&quot;</span>, <span class="st">&quot;G0S2&quot;</span>, <span class="st">&quot;HBB&quot;</span>, <span class="st">&quot;HIST1H4C&quot;</span>, <span class="st">&quot;IER2&quot;</span>, <span class="st">&quot;IGKC&quot;</span>,</span>
<span id="cb43-4"><a href="mc2-python.html#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;IGLC2&quot;</span>, <span class="st">&quot;JUN&quot;</span>, <span class="st">&quot;JUNB&quot;</span>, <span class="st">&quot;KLRB1&quot;</span>, <span class="st">&quot;MT2A&quot;</span>, <span class="st">&quot;RPS26&quot;</span>, <span class="st">&quot;RPS4Y1&quot;</span>, <span class="st">&quot;TRBC1&quot;</span>, <span class="st">&quot;TUBA1B&quot;</span>, <span class="st">&quot;TUBB&quot;</span></span>
<span id="cb43-5"><a href="mc2-python.html#cb43-5" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb43-6"><a href="mc2-python.html#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="co"># This will mark as &quot;noisy_gene&quot; any genes that match the above, if they exist in the clean dataset.</span></span>
<span id="cb43-7"><a href="mc2-python.html#cb43-7" aria-hidden="true" tabindex="-1"></a>mc.pl.mark_noisy_genes(ad, noisy_gene_names<span class="op">=</span>NOISY_GENE_NAMES)</span>
<span id="cb43-8"><a href="mc2-python.html#cb43-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[noisy_gene]: 17 true (0.1026%) out of 16564 bools</span></span></code></pre></div>
<p>To extend this list of lateral genes, users can use the `` function to identify genes that are highly correlated with the predefined lateral genes.</p>
</div>
<div id="estimate-target_metacell_size-gamma" class="section level4 unnumbered hasAnchor">
<h4>Estimate target_metacell_size (gamma)<a href="mc2-python.html#estimate-target_metacell_size-gamma" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>By default, MC2 will build metacells with a size of 96 cells per metacells.
Users can vary the <code>target_metacell_size</code> parameter to reach a desired graining level.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb44-1"><a href="mc2-python.html#cb44-1" aria-hidden="true" tabindex="-1"></a>gamma   <span class="op">=</span> <span class="dv">50</span> <span class="co"># graining level</span></span>
<span id="cb44-2"><a href="mc2-python.html#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="mc2-python.html#cb44-3" aria-hidden="true" tabindex="-1"></a>target_metacell_size <span class="op">=</span> <span class="bu">round</span>(ad.shape[<span class="dv">0</span>]<span class="op">/</span>gamma)</span>
<span id="cb44-4"><a href="mc2-python.html#cb44-4" aria-hidden="true" tabindex="-1"></a>target_metacell_size</span>
<span id="cb44-5"><a href="mc2-python.html#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 53</span></span></code></pre></div>
</div>
<div id="metacells-identification-using-the-divide-and-conquer-approach" class="section level4 unnumbered hasAnchor">
<h4>Metacells identification using the divide and conquer approach<a href="mc2-python.html#metacells-identification-using-the-divide-and-conquer-approach" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The construction of metacells by MC2 is performed using the <code>mc.pl.divide_and_conquer_pipeline()</code> function.
Note that by default all cores of the system will be used for the metacells construction.
To change this behavior and adapt the number of cores the MC2 authors propose to use the <code>mc.pl.guess_max_parallel_piles()</code> and <code>mc.pl.set_max_parallel_piles()</code> functions to adapt the number of processed in parallel depending on the available memory.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb45-1"><a href="mc2-python.html#cb45-1" aria-hidden="true" tabindex="-1"></a>max_parallel_piles <span class="op">=</span> mc.pl.guess_max_parallel_piles(ad)</span>
<span id="cb45-2"><a href="mc2-python.html#cb45-2" aria-hidden="true" tabindex="-1"></a>mc.pl.set_max_parallel_piles(max_parallel_piles)</span>
<span id="cb45-3"><a href="mc2-python.html#cb45-3" aria-hidden="true" tabindex="-1"></a>mc.pl.divide_and_conquer_pipeline(</span>
<span id="cb45-4"><a href="mc2-python.html#cb45-4" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb45-5"><a href="mc2-python.html#cb45-5" aria-hidden="true" tabindex="-1"></a>    target_metacell_size <span class="op">=</span> target_metacell_size,</span>
<span id="cb45-6"><a href="mc2-python.html#cb45-6" aria-hidden="true" tabindex="-1"></a>    random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb45-7"><a href="mc2-python.html#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[selected_gene]: * -&gt; False</span></span>
<span id="cb45-8"><a href="mc2-python.html#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb45-9"><a href="mc2-python.html#cb45-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene_module]: 16564 int32 elements with all outliers (100%)</span></span>
<span id="cb45-10"><a href="mc2-python.html#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cells_rare_gene_module]: 2638 int32 elements with all outliers (100%)</span></span>
<span id="cb45-11"><a href="mc2-python.html#cb45-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[rare_cell]: 0 true (0%) out of 2638 bools</span></span>
<span id="cb45-12"><a href="mc2-python.html#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[selected_gene]: 295 true (1.781%) out of 16564 bools</span></span>
<span id="cb45-13"><a href="mc2-python.html#cb45-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell]: 2638 int32s</span></span>
<span id="cb45-14"><a href="mc2-python.html#cb45-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[dissolved]: 14 true (0.5307%) out of 2638 bools</span></span>
<span id="cb45-15"><a href="mc2-python.html#cb45-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell_level]: 2638 int32s</span></span>
<span id="cb45-16"><a href="mc2-python.html#cb45-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-17"><a href="mc2-python.html#cb45-17" aria-hidden="true" tabindex="-1"></a>ad.obs.metacell.head</span>
<span id="cb45-18"><a href="mc2-python.html#cb45-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;bound method NDFrame.head of index</span></span>
<span id="cb45-19"><a href="mc2-python.html#cb45-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATACAACCAC-1    30</span></span>
<span id="cb45-20"><a href="mc2-python.html#cb45-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGAGCTAC-1    31</span></span>
<span id="cb45-21"><a href="mc2-python.html#cb45-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGATCAGC-1    49</span></span>
<span id="cb45-22"><a href="mc2-python.html#cb45-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGCTTCCG-1    13</span></span>
<span id="cb45-23"><a href="mc2-python.html#cb45-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGTATGCG-1    -1</span></span>
<span id="cb45-24"><a href="mc2-python.html#cb45-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     ..</span></span>
<span id="cb45-25"><a href="mc2-python.html#cb45-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCGAACTCTCAT-1    43</span></span>
<span id="cb45-26"><a href="mc2-python.html#cb45-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTGAGGCA-1    25</span></span>
<span id="cb45-27"><a href="mc2-python.html#cb45-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTTCCTCG-1    57</span></span>
<span id="cb45-28"><a href="mc2-python.html#cb45-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGAGAGGC-1    16</span></span>
<span id="cb45-29"><a href="mc2-python.html#cb45-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGCCTCAC-1    49</span></span>
<span id="cb45-30"><a href="mc2-python.html#cb45-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Name: metacell, Length: 2638, dtype: int32&gt;</span></span></code></pre></div>
</div>
<div id="retrieve-aggregated-metacell-data-1" class="section level4 unnumbered hasAnchor">
<h4>Retrieve aggregated metacell data<a href="mc2-python.html#retrieve-aggregated-metacell-data-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The <code>mc.pl.divide_and_conquer_pipeline()</code> function associates each cell to a metacell or defines the cell as outlier. These assignments are found in the <code>obs</code> layer of the anndata object
The function <code>mc.pl.collect_metacells()</code> should be used to subsequently retrieve an anndata object containing the data at the metacells level instead of the single-cell level.
This function will store in the <code>X</code> data matrix of the anndata object a matrix of gene fraction, <em>i.e.</em>, the sum of all gene levels in a metacell sums to 1
and it will store the total UMIs per gene per metacell in the layer <code>total_umis</code>.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb46-1"><a href="mc2-python.html#cb46-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-2"><a href="mc2-python.html#cb46-2" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> mc.pl.collect_metacells(ad, name<span class="op">=</span><span class="st">&#39;metacells&#39;</span>, random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb46-3"><a href="mc2-python.html#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[grouped]: 62 int64s</span></span>
<span id="cb46-4"><a href="mc2-python.html#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[total_umis]: 62 float64s</span></span>
<span id="cb46-5"><a href="mc2-python.html#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.layers[total_umis]: ndarray 62 X 16564 float32s</span></span>
<span id="cb46-6"><a href="mc2-python.html#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[__zeros_downsample_umis]: 62 int64s</span></span>
<span id="cb46-7"><a href="mc2-python.html#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.layers[zeros]: ndarray 62 X 16564 int32s</span></span>
<span id="cb46-8"><a href="mc2-python.html#cb46-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell_name]: 2638 &lt;U8s</span></span>
<span id="cb46-9"><a href="mc2-python.html#cb46-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[gene_ids]: 16564 objects</span></span>
<span id="cb46-10"><a href="mc2-python.html#cb46-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[bursty_lonely_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb46-11"><a href="mc2-python.html#cb46-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[properly_sampled_gene]: 16564 true (100%) out of 16564 bools</span></span>
<span id="cb46-12"><a href="mc2-python.html#cb46-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[excluded_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb46-13"><a href="mc2-python.html#cb46-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[full_gene_index]: 16564 int32s</span></span>
<span id="cb46-14"><a href="mc2-python.html#cb46-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[lateral_gene]: 111 true (0.6701%) out of 16564 bools</span></span>
<span id="cb46-15"><a href="mc2-python.html#cb46-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[noisy_gene]: 17 true (0.1026%) out of 16564 bools</span></span>
<span id="cb46-16"><a href="mc2-python.html#cb46-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[selected_gene]: 295 true (1.781%) out of 16564 bools</span></span>
<span id="cb46-17"><a href="mc2-python.html#cb46-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[rare_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb46-18"><a href="mc2-python.html#cb46-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[rare_gene_module]: 16564 int32s</span></span>
<span id="cb46-19"><a href="mc2-python.html#cb46-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[metacells_rare_gene_module]: 62 int32s</span></span>
<span id="cb46-20"><a href="mc2-python.html#cb46-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[rare_metacell]: 0 true (0%) out of 62 bools</span></span>
<span id="cb46-21"><a href="mc2-python.html#cb46-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.uns[outliers]: 158</span></span>
<span id="cb46-22"><a href="mc2-python.html#cb46-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.uns[metacells_algorithm]: metacells.0.9.0</span></span>
<span id="cb46-23"><a href="mc2-python.html#cb46-23" aria-hidden="true" tabindex="-1"></a>mc_ad.shape</span>
<span id="cb46-24"><a href="mc2-python.html#cb46-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; (62, 16564)</span></span>
<span id="cb46-25"><a href="mc2-python.html#cb46-25" aria-hidden="true" tabindex="-1"></a>mc_ad.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)[:<span class="dv">5</span>] </span>
<span id="cb46-26"><a href="mc2-python.html#cb46-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; matrix([[1.],</span></span>
<span id="cb46-27"><a href="mc2-python.html#cb46-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [1.],</span></span>
<span id="cb46-28"><a href="mc2-python.html#cb46-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [1.],</span></span>
<span id="cb46-29"><a href="mc2-python.html#cb46-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [1.],</span></span>
<span id="cb46-30"><a href="mc2-python.html#cb46-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [1.]], dtype=float32)</span></span>
<span id="cb46-31"><a href="mc2-python.html#cb46-31" aria-hidden="true" tabindex="-1"></a>mc_ad.layers[<span class="st">&#39;total_umis&#39;</span>]</span>
<span id="cb46-32"><a href="mc2-python.html#cb46-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; array([[0., 0., 0., ..., 0., 1., 0.],</span></span>
<span id="cb46-33"><a href="mc2-python.html#cb46-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 6., 1.],</span></span>
<span id="cb46-34"><a href="mc2-python.html#cb46-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 2., 1.],</span></span>
<span id="cb46-35"><a href="mc2-python.html#cb46-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        ...,</span></span>
<span id="cb46-36"><a href="mc2-python.html#cb46-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 3., 0.],</span></span>
<span id="cb46-37"><a href="mc2-python.html#cb46-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 1., 1.],</span></span>
<span id="cb46-38"><a href="mc2-python.html#cb46-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 2., 0.]], dtype=float32)</span></span></code></pre></div>
<p><strong>Comparing the obtained and requested graining level</strong></p>
<p>In the following code chunk, we estimate whether a deviation of the obtained gamma from the requested gamma is acceptable. If not, we suggest to increase or decrease the <code>target_metacell_size</code> parameter to approach the desired graining level.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb47-1"><a href="mc2-python.html#cb47-1" aria-hidden="true" tabindex="-1"></a>gamma_obtained <span class="op">=</span> ad.shape[<span class="dv">0</span>]<span class="op">/</span>mc_ad.shape[<span class="dv">0</span>]</span>
<span id="cb47-2"><a href="mc2-python.html#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gamma_obtained)</span>
<span id="cb47-3"><a href="mc2-python.html#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 42.54838709677419</span></span>
<span id="cb47-4"><a href="mc2-python.html#cb47-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-5"><a href="mc2-python.html#cb47-5" aria-hidden="true" tabindex="-1"></a>gamma_dev <span class="op">=</span> (gamma_obtained <span class="op">-</span> gamma)<span class="op">/</span>gamma</span>
<span id="cb47-6"><a href="mc2-python.html#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">abs</span>(gamma_dev) <span class="op">&lt;</span> <span class="fl">0.3</span>: </span>
<span id="cb47-7"><a href="mc2-python.html#cb47-7" aria-hidden="true" tabindex="-1"></a>    gamma_dev <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb47-8"><a href="mc2-python.html#cb47-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-9"><a href="mc2-python.html#cb47-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> gamma_dev <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb47-10"><a href="mc2-python.html#cb47-10" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Increase `target_metacell_size` parameter by increasing `scale` and re-run metacell divide_and_conquer_pipeline() to get larger graining level&quot;</span>)</span>
<span id="cb47-11"><a href="mc2-python.html#cb47-11" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb47-12"><a href="mc2-python.html#cb47-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;Deacrease `target_metacell_size` parameter by decreasing `scale` and re-run metacell divide_and_conquer_pipeline() to get smaller graining level&quot;</span>)</span>
<span id="cb47-13"><a href="mc2-python.html#cb47-13" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> gamma_dev <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb47-14"><a href="mc2-python.html#cb47-14" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">&quot;The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size` &quot;</span>)</span>
<span id="cb47-15"><a href="mc2-python.html#cb47-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The obtained graining level is acceptable, no need to re-run the metacell divide_and_conquer_pipeline() with a new `target_metacell_size`</span></span></code></pre></div>
</div>
</div>
<div id="visualize-metacells" class="section level3 hasAnchor" number="3.2.4">
<h3><span class="header-section-number">3.2.4</span> Visualize metacells<a href="mc2-python.html#visualize-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If single-cell annotations are available in the original single-cell anndata object. We can transfer these annotations to the metacell anndata object
using the <code>mc.tl.convey_obs_to_group()</code> function which will associate each metacell to the most frequent annotation (categorical) or averaged annotation (continuous) across the single-cells composing the metacell
(use of the <code>mc.ut.most_frequent</code> and <code>np.mean</code> respectively in the <code>mode</code> paratemer).</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb48-1"><a href="mc2-python.html#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign a single value for each metacell based on the cells.</span></span>
<span id="cb48-2"><a href="mc2-python.html#cb48-2" aria-hidden="true" tabindex="-1"></a>mc.tl.convey_obs_to_group(</span>
<span id="cb48-3"><a href="mc2-python.html#cb48-3" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>ad, gdata<span class="op">=</span>mc_ad,</span>
<span id="cb48-4"><a href="mc2-python.html#cb48-4" aria-hidden="true" tabindex="-1"></a>    property_name<span class="op">=</span><span class="st">&quot;louvain&quot;</span>, to_property_name<span class="op">=</span><span class="st">&quot;annotation&quot;</span>,</span>
<span id="cb48-5"><a href="mc2-python.html#cb48-5" aria-hidden="true" tabindex="-1"></a>    method<span class="op">=</span>mc.ut.most_frequent  <span class="co"># This is the default, for categorical data</span></span>
<span id="cb48-6"><a href="mc2-python.html#cb48-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb48-7"><a href="mc2-python.html#cb48-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation]: 62 &lt;U17s</span></span>
<span id="cb48-8"><a href="mc2-python.html#cb48-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-9"><a href="mc2-python.html#cb48-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the fraction of cells with each possible value in each metacell:</span></span>
<span id="cb48-10"><a href="mc2-python.html#cb48-10" aria-hidden="true" tabindex="-1"></a>mc.tl.convey_obs_fractions_to_group(  </span>
<span id="cb48-11"><a href="mc2-python.html#cb48-11" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>ad, gdata<span class="op">=</span>mc_ad,</span>
<span id="cb48-12"><a href="mc2-python.html#cb48-12" aria-hidden="true" tabindex="-1"></a>    property_name<span class="op">=</span><span class="st">&quot;louvain&quot;</span>, to_property_name<span class="op">=</span><span class="st">&quot;annotation&quot;</span></span>
<span id="cb48-13"><a href="mc2-python.html#cb48-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb48-14"><a href="mc2-python.html#cb48-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_B cells]: 62 float64s</span></span>
<span id="cb48-15"><a href="mc2-python.html#cb48-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_CD14+ Monocytes]: 62 float64s</span></span>
<span id="cb48-16"><a href="mc2-python.html#cb48-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_CD4 T cells]: 62 float64s</span></span>
<span id="cb48-17"><a href="mc2-python.html#cb48-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_CD8 T cells]: 62 float64s</span></span>
<span id="cb48-18"><a href="mc2-python.html#cb48-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_Dendritic cells]: 62 float64s</span></span>
<span id="cb48-19"><a href="mc2-python.html#cb48-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_FCGR3A+ Monocytes]: 62 float64s</span></span>
<span id="cb48-20"><a href="mc2-python.html#cb48-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_Megakaryocytes]: 62 float64s</span></span>
<span id="cb48-21"><a href="mc2-python.html#cb48-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[annotation_fraction_of_NK cells]: 62 float64s</span></span></code></pre></div>
<p>The following code chunk adds a columns named <code>membership</code> and containing the single_cell assignments to the obs attribute in the anndata object containing the raw data.
This annotation will be used in the mc_QC package to compute metacells quality metrics. We also save the single-cell metadata in the metacell anndata object.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb49-1"><a href="mc2-python.html#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a membership -- index of metacells to which single cells belong to </span></span>
<span id="cb49-2"><a href="mc2-python.html#cb49-2" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&#39;membership&#39;</span>] <span class="op">=</span> [<span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">0</span> <span class="cf">else</span> np.nan <span class="cf">for</span> i <span class="kw">in</span> ad.obs.metacell] </span>
<span id="cb49-3"><a href="mc2-python.html#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="mc2-python.html#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Save single-cell metadata (i.e., `raw.obs` dataframe) in the metacell adata object</span></span>
<span id="cb49-5"><a href="mc2-python.html#cb49-5" aria-hidden="true" tabindex="-1"></a>mc_ad.uns <span class="op">=</span> ad.uns.copy()</span>
<span id="cb49-6"><a href="mc2-python.html#cb49-6" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.obs&#39;</span>] <span class="op">=</span> ad.obs.copy()</span>
<span id="cb49-7"><a href="mc2-python.html#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="mc2-python.html#cb49-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save the requested gamma</span></span>
<span id="cb49-9"><a href="mc2-python.html#cb49-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;gamma&#39;</span>] <span class="op">=</span> gamma</span></code></pre></div>
<div id="compute-latent-space-for-metacell-qc" class="section level4 unnumbered hasAnchor">
<h4>Compute latent space for metacell QC<a href="mc2-python.html#compute-latent-space-for-metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To visualize the metacells, we can project the metacells on the single-cell UMAP representation. To run UMAP, we will generate in the next code chunk a lower-dimentional embedding of the data, so far not needed since the MC2 methods builds metacells from gene expression data and not from latent space.
Also, note that some of the QC metrics (e.g., <strong>compactness</strong> and <strong>separation</strong>), that we will compute in the next section of this tutorial, are computed from this latent space.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb50-1"><a href="mc2-python.html#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save count as a separate layer</span></span>
<span id="cb50-2"><a href="mc2-python.html#cb50-2" aria-hidden="true" tabindex="-1"></a>ad.layers[<span class="st">&#39;counts&#39;</span>] <span class="op">=</span> ad.X</span>
<span id="cb50-3"><a href="mc2-python.html#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="mc2-python.html#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the counts to &quot;.raw&quot; attribute of the anndata since it is necessary for downstream analysis</span></span>
<span id="cb50-5"><a href="mc2-python.html#cb50-5" aria-hidden="true" tabindex="-1"></a><span class="co"># This step should be performed after filtering </span></span>
<span id="cb50-6"><a href="mc2-python.html#cb50-6" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(ad.layers[<span class="st">&#39;counts&#39;</span>])</span>
<span id="cb50-7"><a href="mc2-python.html#cb50-7" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> ad.obs_names, ad.var_names</span>
<span id="cb50-8"><a href="mc2-python.html#cb50-8" aria-hidden="true" tabindex="-1"></a>ad.raw <span class="op">=</span> raw_ad</span>
<span id="cb50-9"><a href="mc2-python.html#cb50-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-10"><a href="mc2-python.html#cb50-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-11"><a href="mc2-python.html#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize cells, log transform and compute highly variable genes</span></span>
<span id="cb50-12"><a href="mc2-python.html#cb50-12" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(ad)</span>
<span id="cb50-13"><a href="mc2-python.html#cb50-13" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ad)</span>
<span id="cb50-14"><a href="mc2-python.html#cb50-14" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(ad, n_top_genes<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb50-15"><a href="mc2-python.html#cb50-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-16"><a href="mc2-python.html#cb50-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute principal components - </span></span>
<span id="cb50-17"><a href="mc2-python.html#cb50-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-18"><a href="mc2-python.html#cb50-18" aria-hidden="true" tabindex="-1"></a>n_comp    <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb50-19"><a href="mc2-python.html#cb50-19" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(ad, n_comps<span class="op">=</span>n_comp, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb50-20"><a href="mc2-python.html#cb50-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-21"><a href="mc2-python.html#cb50-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-22"><a href="mc2-python.html#cb50-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute UMAP for visualization </span></span>
<span id="cb50-23"><a href="mc2-python.html#cb50-23" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(ad, n_neighbors<span class="op">=</span><span class="dv">10</span>, n_pcs<span class="op">=</span>n_comp)</span>
<span id="cb50-24"><a href="mc2-python.html#cb50-24" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(ad)</span></code></pre></div>
<p>To visualize the metacell projection on the single-cell UMAP, we use the <code>mc_visualize</code> function from the <code>mc_QC</code>, this function was adapted from the <code>plot.plot_2D</code> included in the SEACells package.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb51-1"><a href="mc2-python.html#cb51-1" aria-hidden="true" tabindex="-1"></a>mc_proj <span class="op">=</span> mc_QC.mc_visualize(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb51-2"><a href="mc2-python.html#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.</span></span>
<span id="cb51-3"><a href="mc2-python.html#cb51-3" aria-hidden="true" tabindex="-1"></a>mc_proj.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/mc2-comp-umap-1.png" width="384" /></p>
</div>
</div>
<div id="metacell-qc-1" class="section level3 hasAnchor" number="3.2.5">
<h3><span class="header-section-number">3.2.5</span> Metacell QC<a href="mc2-python.html#metacell-qc-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="compute-purity-compactness-and-separation-metrics-1" class="section level4 unnumbered hasAnchor">
<h4>Compute purity, compactness and separation metrics<a href="mc2-python.html#compute-purity-compactness-and-separation-metrics-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Size distribution</strong></p>
<div class="sourceCode" id="cb52"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb52-1"><a href="mc2-python.html#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_size = SEACells.plot.plot_SEACell_sizes(ad, bins=20)</span></span>
<span id="cb52-2"><a href="mc2-python.html#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="mc2-python.html#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs = pd.merge(mc_ad.obs, mc_size, left_index=True, right_index=True)</span></span>
<span id="cb52-4"><a href="mc2-python.html#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs</span></span></code></pre></div>
<p>When available, we can use cell annotation to annotate each metacell to the most abundant cell category (<em>e.g.</em> cell type) composing the metacell.
This also allows us to compute metacell purity. If the annotation considered is the cell type, the <strong>purity</strong> of a metacell is the proportion of the most abundant cell type within the metacell [ref SuperCell]</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb53-1"><a href="mc2-python.html#cb53-1" aria-hidden="true" tabindex="-1"></a>mc_purity <span class="op">=</span> mc_QC.purity(ad, annotation_label, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>)</span>
<span id="cb53-2"><a href="mc2-python.html#cb53-2" aria-hidden="true" tabindex="-1"></a>mc_purity.head()</span>
<span id="cb53-3"><a href="mc2-python.html#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     louvain  louvain_purity</span></span>
<span id="cb53-4"><a href="mc2-python.html#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; membership                                 </span></span>
<span id="cb53-5"><a href="mc2-python.html#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1.0         CD14+ Monocytes        1.000000</span></span>
<span id="cb53-6"><a href="mc2-python.html#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2.0             CD8 T cells        0.884058</span></span>
<span id="cb53-7"><a href="mc2-python.html#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3.0             CD4 T cells        0.897436</span></span>
<span id="cb53-8"><a href="mc2-python.html#cb53-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4.0         CD14+ Monocytes        1.000000</span></span>
<span id="cb53-9"><a href="mc2-python.html#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5.0             CD4 T cells        1.000000</span></span>
<span id="cb53-10"><a href="mc2-python.html#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="co"># add purity to metadata</span></span>
<span id="cb53-11"><a href="mc2-python.html#cb53-11" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;purity&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(mc_purity[annotation_label <span class="op">+</span> <span class="st">&quot;_purity&quot;</span>])</span></code></pre></div>
<p>The <strong>compactness</strong> of a metacell is the variance of the components within the metacell [ref SEACells]</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb54-1"><a href="mc2-python.html#cb54-1" aria-hidden="true" tabindex="-1"></a>compactness <span class="op">=</span> mc_QC.compactness(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Compactness_PCA&#39;</span>, n_comp<span class="op">=</span><span class="dv">10</span>)[<span class="st">&#39;Compactness_PCA&#39;</span>]</span>
<span id="cb54-2"><a href="mc2-python.html#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add compactness to metadata</span></span>
<span id="cb54-3"><a href="mc2-python.html#cb54-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;Compactness_PCA&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(compactness)</span></code></pre></div>
<p>The <strong>separation</strong> of a metacell is the distance to the closest metacell [ref SEACells]</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb55-1"><a href="mc2-python.html#cb55-1" aria-hidden="true" tabindex="-1"></a>separation <span class="op">=</span> mc_QC.separation(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Separation_PCA&#39;</span>, n_comp<span class="op">=</span><span class="dv">10</span>)[<span class="st">&#39;Separation_PCA&#39;</span>]</span>
<span id="cb55-2"><a href="mc2-python.html#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add separation to metadata</span></span>
<span id="cb55-3"><a href="mc2-python.html#cb55-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;Separation_PCA&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(separation)</span></code></pre></div>
<p>The <strong>inner normalized variance (INV)</strong> of a metacell is the mean-normalized variance of gene expression within the metacell [ref MC-2]</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb56-1"><a href="mc2-python.html#cb56-1" aria-hidden="true" tabindex="-1"></a>mc_INV <span class="op">=</span> mc_QC.mc_inner_normalized_var(ad<span class="op">=</span>ad, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>)</span>
<span id="cb56-2"><a href="mc2-python.html#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="mc2-python.html#cb56-3" aria-hidden="true" tabindex="-1"></a>mc_INV_val <span class="op">=</span> mc_INV.quantile(<span class="fl">0.95</span>, axis<span class="op">=</span><span class="dv">1</span>, numeric_only<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb56-4"><a href="mc2-python.html#cb56-4" aria-hidden="true" tabindex="-1"></a>mc_INV_val <span class="op">=</span> pd.DataFrame(mc_INV_val.transpose()).set_axis([<span class="st">&#39;INV&#39;</span>], axis<span class="op">=</span><span class="dv">1</span>, copy<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb56-5"><a href="mc2-python.html#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add INV to metadata</span></span>
<span id="cb56-6"><a href="mc2-python.html#cb56-6" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;INV&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(mc_INV_val[<span class="st">&quot;INV&quot;</span>])</span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb57-1"><a href="mc2-python.html#cb57-1" aria-hidden="true" tabindex="-1"></a>ad.uns[<span class="st">&#39;mc_obs&#39;</span>] <span class="op">=</span> mc_ad.obs</span>
<span id="cb57-2"><a href="mc2-python.html#cb57-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb57-3"><a href="mc2-python.html#cb57-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;purity&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb57-4"><a href="mc2-python.html#cb57-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-26-3.png" width="384" /></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb58-1"><a href="mc2-python.html#cb58-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb58-2"><a href="mc2-python.html#cb58-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb58-3"><a href="mc2-python.html#cb58-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;Compactness_PCA&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb58-4"><a href="mc2-python.html#cb58-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)   </span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-26-4.png" width="384" /></p>
<div class="sourceCode" id="cb59"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb59-1"><a href="mc2-python.html#cb59-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb59-2"><a href="mc2-python.html#cb59-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb59-3"><a href="mc2-python.html#cb59-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;Separation_PCA&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb59-4"><a href="mc2-python.html#cb59-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)   </span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-26-5.png" width="384" /></p>
<div class="sourceCode" id="cb60"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb60-1"><a href="mc2-python.html#cb60-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb60-2"><a href="mc2-python.html#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="co"># mc_QC.mc_visualize_continuous(ad, key=&#39;X_umap&#39;, group_by_name=&#39;membership&#39;,</span></span>
<span id="cb60-3"><a href="mc2-python.html#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="co">#              colour_sc_name=&#39;louvain&#39;,  colour_mc_name=&#39;INV&#39;, colour_metacells=True,</span></span>
<span id="cb60-4"><a href="mc2-python.html#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co">#              legend_sc=None, legend_mc=&#39;auto&#39;, metacell_size=30)</span></span></code></pre></div>
<div id="save-output-1" class="section level5 unnumbered hasAnchor">
<h5>Save output<a href="mc2-python.html#save-output-1" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb61"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb61-1"><a href="mc2-python.html#cb61-1" aria-hidden="true" tabindex="-1"></a>mc_ad.write_h5ad(os.path.join(<span class="st">&#39;./data&#39;</span>, proj_name, <span class="ss">f&#39;metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad&#39;</span>))</span></code></pre></div>
</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="supercell-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="seacells-python.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/21-MC_construction_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/21-MC_construction_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
