<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 3 Constructing metacells | Metacell Analysis Tutorial</title>
<meta name="author" content="Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller">
<meta name="description" content="In this chapter, we will demonstrate metacell construction using three different methods: SuperCell in R, MetaCell-2 (MC2) and SEACells in Pyhton. For this, we will first use a dataset of PBMCs...">
<meta name="generator" content="bookdown 0.36 with bs4_book()">
<meta property="og:title" content="Chapter 3 Constructing metacells | Metacell Analysis Tutorial">
<meta property="og:type" content="book">
<meta property="og:url" content="https://GfellerLab.github.io/MetacellAnalysisTutorial/Metacell-construction-chapter.html">
<meta property="og:description" content="In this chapter, we will demonstrate metacell construction using three different methods: SuperCell in R, MetaCell-2 (MC2) and SEACells in Pyhton. For this, we will first use a dataset of PBMCs...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 3 Constructing metacells | Metacell Analysis Tutorial">
<meta name="twitter:description" content="In this chapter, we will demonstrate metacell construction using three different methods: SuperCell in R, MetaCell-2 (MC2) and SEACells in Pyhton. For this, we will first use a dataset of PBMCs...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Metacell Analysis Tutorial</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">This tutorial</a></li>
<li><a class="" href="requirements.html"><span class="header-section-number">1</span> Requirements</a></li>
<li><a class="" href="the-metacell-concept.html"><span class="header-section-number">2</span> The metacell concept</a></li>
<li><a class="active" href="Metacell-construction-chapter.html"><span class="header-section-number">3</span> Constructing metacells</a></li>
<li><a class="" href="QCs.html"><span class="header-section-number">4</span> Metacells QCs</a></li>
<li><a class="" href="command-line.html"><span class="header-section-number">5</span> Metacell Analysis Toolkit (MCAT)</a></li>
<li><a class="" href="downstream-analysis-of-metacells.html"><span class="header-section-number">6</span> Downstream analysis of metacells</a></li>
<li><a class="" href="integration.html"><span class="header-section-number">7</span> Integration of metacells</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/GfellerLab/MetacellAnalysisTutorial">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="Metacell-construction-chapter" class="section level1" number="3">
<h1>
<span class="header-section-number">3</span> Constructing metacells<a class="anchor" aria-label="anchor" href="#Metacell-construction-chapter"><i class="fas fa-link"></i></a>
</h1>
<p>In this chapter, we will demonstrate metacell construction using three different methods: SuperCell in R, MetaCell-2 (MC2) and SEACells in Pyhton.</p>
<p>For this, we will first use a dataset of PBMCs from study. This dataset contains around 3K cells which is an example of a dataset with well defined cell types.
For an example of more continuous data, see chapter <a href="#MC-continuous"><strong>??</strong></a>.</p>
<div id="SuperCell-construction" class="section level2" number="3.1">
<h2>
<span class="header-section-number">3.1</span> SuperCell (R)<a class="anchor" aria-label="anchor" href="#SuperCell-construction"><i class="fas fa-link"></i></a>
</h2>
<p>In this section, we construct metacells using the R package <a href="https://github.com/GfellerLab/SuperCell">SuperCell</a>.</p>
<div id="method" class="section level3" number="3.1.1">
<h3>
<span class="header-section-number">3.1.1</span> Method<a class="anchor" aria-label="anchor" href="#method"><i class="fas fa-link"></i></a>
</h3>
<p>The SuperCell method first reduces the gene expression space using principal component analysis (PCA) and computes euclidean distances based on the reduced space.
Using the euclidean distances, a single-cell kNN graph is built and metacells are identified by applying the walktrap community detection algorithm.
The number of metacells obtained can be chosen by the user by defining the graining level parameter.</p>
<p>The code provided in this section is adapted from the <a href="https://github.com/GfellerLab/SuperCell/blob/master/README.Rmd">author’s github documentation</a>.
For more information on the method, please refer to our review <span class="citation">(<a href="#ref-Review" role="doc-biblioref"><strong>Review?</strong></a>)</span> and the original paper <span class="citation">(<a href="references.html#ref-SuperCell" role="doc-biblioref">Bilous et al. 2022</a>)</span>.</p>
<div id="importing-r-packages" class="section level4 unnumbered">
<h4>Importing R packages<a class="anchor" aria-label="anchor" href="#importing-r-packages"><i class="fas fa-link"></i></a>
</h4>
<p>To run SuperCell, the following R package needs to be imported:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">'SuperCell'</span><span class="op">)</span> <span class="op">==</span> <span class="st">""</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"GfellerLab/SuperCell"</span>, force <span class="op">=</span> <span class="cn">TRUE</span>, upgrade <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SuperCell</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div id="data-loading" class="section level3" number="3.1.2">
<h3>
<span class="header-section-number">3.1.2</span> Data loading<a class="anchor" aria-label="anchor" href="#data-loading"><i class="fas fa-link"></i></a>
</h3>
<p>We will run SuperCell on a single-cell dataset composed of 2638 peripheral blood mononuclear cells (PBMCs) available in the scanpy package.
Please follow the section <a href="requirements.html#PBMC-data">1.2</a> to retrieve these data from the scanpy package, preprocess and save the data in the following file: “data/3k_pbmc/singlecell_seurat_filtered.rds”.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">proj_name</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "3k_pbmc"</span></span>
<span><span class="va">celltype_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"CD14+ Monocytes"</span>    <span class="op">=</span> <span class="st">"#E69F00"</span>,  <span class="co"># orange</span></span>
<span>  <span class="st">"B cells"</span>            <span class="op">=</span> <span class="st">"#56B4E9"</span>,  <span class="co"># sky blue</span></span>
<span>  <span class="st">"CD4 T cells"</span>        <span class="op">=</span> <span class="st">"#009E73"</span>,  <span class="co"># bluish green</span></span>
<span>  <span class="st">"NK cells"</span>           <span class="op">=</span> <span class="st">"#F0E442"</span>,  <span class="co"># yellow</span></span>
<span>  <span class="st">"CD8 T cells"</span>        <span class="op">=</span> <span class="st">"#0072B2"</span>,  <span class="co"># blue</span></span>
<span>  <span class="st">"FCGR3A+ Monocytes"</span>  <span class="op">=</span> <span class="st">"#D55E00"</span>,  <span class="co"># vermillion</span></span>
<span>  <span class="st">"Dendritic cells"</span>    <span class="op">=</span> <span class="st">"#CC79A7"</span>,  <span class="co"># reddish purple</span></span>
<span>  <span class="st">"Megakaryocytes"</span>     <span class="op">=</span> <span class="st">"#000000"</span>   <span class="co"># black</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/"</span>, <span class="va">proj_name</span>, <span class="st">"/singlecell_seurat_filtered.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="filtering-steps" class="section level3" number="3.1.3">
<h3>
<span class="header-section-number">3.1.3</span> Filtering steps<a class="anchor" aria-label="anchor" href="#filtering-steps"><i class="fas fa-link"></i></a>
</h3>
<p>In this tutorial, the data have been pre-filtered and SuperCell does not require further filtering steps.</p>
</div>
<div id="building-metacells" class="section level3" number="3.1.4">
<h3>
<span class="header-section-number">3.1.4</span> Building metacells<a class="anchor" aria-label="anchor" href="#building-metacells"><i class="fas fa-link"></i></a>
</h3>
<p>Metacells construction using SuperCell requires one main input, <em>i.e.</em> a matrix of log-normalized gene expression data which will be used to compute PCA to subsequently build a knn graph for metacells identification.
Important optional inputs are:
(i) the graining level (<code>gamma</code> parameter),
(ii) the number of neighbors to consider for the knn graph (<code>k.knn</code> parameter),
(iii) the number of principal components to use to generate the knn graph (<code>n.pc</code> parameter),
and (iv) the number of most variable genes to consider for PCA (<code>n.var.genes</code> parameter).</p>
<div id="data-pre-processing" class="section level4 unnumbered">
<h4>Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"><i class="fas fa-link"></i></a>
</h4>
<p>SuperCell builds its knn graph based on Euclidean distances defined in the PCA space.
PCA computation is performed on the log-normalized gene expression data in the <code>SCimplify</code> SuperCell function.
In the following code chunk, we use Seurat to normalize and visualize the data:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning this package</span></span>
<span><span class="co">#&gt; will retire shortly. Please refer to R-spatial evolution reports on</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html for details.</span></span>
<span><span class="co">#&gt; This package is now running under evolution status 0</span></span>
<span><span class="co">#&gt; Attaching SeuratObject</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">sc_data</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">sc_data</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">sc_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">sc_data</span>, npcs <span class="op">=</span> <span class="fl">50</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">sc_data</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>, n.neighbors <span class="op">=</span> <span class="fl">15</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">UMAPPlot</a></span><span class="op">(</span><span class="va">sc_data</span>, group.by <span class="op">=</span> <span class="va">annotation_label</span>, cols <span class="op">=</span> <span class="va">celltype_colors</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="21-MC_construction_discrete_files/figure-html/supercell-data-processing-1.png" width="672"></div>
</div>
<div id="setting-up-supercell-parameters" class="section level4 unnumbered">
<h4>Setting up SuperCell parameters<a class="anchor" aria-label="anchor" href="#setting-up-supercell-parameters"><i class="fas fa-link"></i></a>
</h4>
<p>In this tutorial, we will run SuperCell using the 30 first principal components resulting from the PCA.
We chose a graining level of 25 and a number of neighbors of 15 for the knn step.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gamma</span> <span class="op">=</span> <span class="fl">10</span> <span class="co"># the requested graining level.</span></span>
<span><span class="va">k_knn</span> <span class="op">=</span> <span class="fl">15</span> <span class="co"># the number of neighbors considered to build the knn network.</span></span>
<span><span class="va">nb_var_genes</span> <span class="op">=</span> <span class="fl">2000</span> <span class="co"># number of the top variable genes to use for dimensionality reduction </span></span>
<span><span class="va">nb_pc</span> <span class="op">=</span> <span class="fl">50</span> <span class="co"># the number of principal components to use.   </span></span></code></pre></div>
</div>
<div id="metacells-identification" class="section level4 unnumbered">
<h4>Metacells identification<a class="anchor" aria-label="anchor" href="#metacells-identification"><i class="fas fa-link"></i></a>
</h4>
<p>The metacells are identified using the <code>SCimplify</code> function from the SuperCell package.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC</span> <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/SCimplify.html">SCimplify</a></span><span class="op">(</span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">sc_data</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,  <span class="co"># single-cell log-normalized gene expression data</span></span>
<span>                           k.knn <span class="op">=</span> <span class="va">k_knn</span>,</span>
<span>                           gamma <span class="op">=</span> <span class="va">gamma</span>,</span>
<span>                           <span class="co"># n.var.genes = nb_var_genes,  </span></span>
<span>                           n.pc <span class="op">=</span> <span class="va">nb_pc</span>,</span>
<span>                           genes.use <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span><span class="va">sc_data</span><span class="op">)</span></span>
<span>                           <span class="op">)</span></span></code></pre></div>
<p><code>SCimplify</code> returns a list containing the following main elements:
(i) the single-cell assignments to metacells (<code>membership</code>),
(ii) the metacell sizes (<code>supercell_size</code>),
(iii) the single-cell graph (<code>graph.singlecell</code>),
(iv) the metacell graph (<code>graph.supercells</code>),
(v) the genes used for metacell identification (<code>genes.use</code>).</p>
</div>
<div id="retrieve-aggregated-metacell-data" class="section level4 unnumbered">
<h4>Retrieve aggregated metacell data<a class="anchor" aria-label="anchor" href="#retrieve-aggregated-metacell-data"><i class="fas fa-link"></i></a>
</h4>
<p>The <code><a href="https://rdrr.io/pkg/SuperCell/man/supercell_GE.html">supercell_GE()</a></code> function can be used to generate a metacell counts matrix (aggregation of gene expression across all cells belonging to each metacell).
Two modes can be used for single-cell aggregation, <em>i.e.</em> averaging of log-normalized gene expression or summing up raw counts (using the <code>mode</code> parameter).
Note that we provide raw counts for the aggregation in this tutorial to match the aggregation steps using PC2 and SEAcells (see <a href="Metacell-construction-chapter.html#MC2-construction">3.2</a> and <a href="Metacell-construction-chapter.html#SEACells-construction">3.3</a>).
Data normalization will thus be needed for downstream analyses on the metacell counts matrix.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC.GE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_GE.html">supercell_GE</a></span><span class="op">(</span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">sc_data</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>,</span>
<span>                      <span class="va">MC</span><span class="op">$</span><span class="va">membership</span>,</span>
<span>                      mode <span class="op">=</span>  <span class="st">"sum"</span></span>
<span>                      <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">MC.GE</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] 32738   264</span></span></code></pre></div>
</div>
</div>
<div id="annotate-metacells-using-available-annotations" class="section level3" number="3.1.5">
<h3>
<span class="header-section-number">3.1.5</span> Annotate metacells (using available annotations)<a class="anchor" aria-label="anchor" href="#annotate-metacells-using-available-annotations"><i class="fas fa-link"></i></a>
</h3>
<p>We can assign each metacell to a particular annotation using the <code><a href="https://rdrr.io/pkg/SuperCell/man/supercell_assign.html">supercell_assign()</a></code> function.
By default, this function assigns each metacell to a cluster with the largest Jaccard coefficient to avoid biases towards very rare or very abundant clusters.
Alternatively, assignment can be performed using <code>relative</code> (<code>method = "relative"</code>, may cause biases towards very small populations) or <code>absolute</code> (<code>method = "absolute"</code>, may cause biases towards large populations) abundance.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">annotation_label</span><span class="op">)</span></span>
<span><span class="co">#&gt;   3k_pbmc </span></span>
<span><span class="co">#&gt; "louvain"</span></span>
<span><span class="va">MC</span><span class="op">$</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_assign.html">supercell_assign</a></span><span class="op">(</span>clusters <span class="op">=</span> <span class="va">sc_data</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_label</span><span class="op">]</span>, <span class="co"># single-cell annotation</span></span>
<span>                                  supercell_membership <span class="op">=</span> <span class="va">MC</span><span class="op">$</span><span class="va">membership</span>, <span class="co"># single-cell assignment to metacells</span></span>
<span>                                  method <span class="op">=</span> <span class="st">"absolute"</span></span>
<span>                                  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">MC</span><span class="op">$</span><span class="va">annotation</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 1                 2                 3                 4 </span></span>
<span><span class="co">#&gt;         "B cells" "CD14+ Monocytes"     "CD4 T cells"        "NK cells" </span></span>
<span><span class="co">#&gt;                 5                 6 </span></span>
<span><span class="co">#&gt;     "CD4 T cells"     "CD4 T cells"</span></span></code></pre></div>
<p>The SuperCell package provides the <code>supercell_plot</code> function to visualize the metacell network (igraph object where number of nodes corresponds to number of metacells),
which is stored in the <code>MC</code> list in <code>graph.supercells</code>.
The metacells can be colored with respect to a vector of annotation.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_plot.html">supercell_plot</a></span><span class="op">(</span></span>
<span>  <span class="va">MC</span><span class="op">$</span><span class="va">graph.supercells</span>, </span>
<span>  group <span class="op">=</span> <span class="va">MC</span><span class="op">$</span><span class="va">annotation</span>, </span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  alpha <span class="op">=</span> <span class="op">-</span><span class="va">pi</span><span class="op">/</span><span class="fl">2</span>,</span>
<span>  main  <span class="op">=</span> <span class="st">"Metacells colored by cell line assignment"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="21-MC_construction_discrete_files/figure-html/SuperCell-graph-1.png" width="672"></div>
</div>
<div id="save-output" class="section level3" number="3.1.6">
<h3>
<span class="header-section-number">3.1.6</span> Save output<a class="anchor" aria-label="anchor" href="#save-output"><i class="fas fa-link"></i></a>
</h3>
<p>For future downstream analyses in R (section <a href="downstream-analysis-of-metacells.html#standard-analysis-R">6.1</a>), metacell counts can be saved in a Seurat object.
Here we also save in the Seurat object the PCA components and genes used in SCimplify for future QC analysis (See <a href="QCs.html#QCs">4</a>).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">MC.GE</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">MC.GE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">MC.GE</span>, </span>
<span>                                meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">MC</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>                                <span class="op">)</span></span>
<span><span class="va">MC.seurat</span><span class="op">[[</span><span class="va">annotation_label</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">MC</span><span class="op">$</span><span class="va">annotation</span></span>
<span></span>
<span><span class="co"># save single-cell membership to metacells in the MC.seurat object</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">MC</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span>, membership <span class="op">=</span> <span class="va">MC</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">var_features</span> <span class="op">&lt;-</span> <span class="va">MC</span><span class="op">$</span><span class="va">genes.use</span> </span>
<span></span>
<span><span class="co"># Save the PCA components and genes used in SCimplify  </span></span>
<span><span class="va">PCA.res</span> <span class="op">&lt;-</span> <span class="fu">irlba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/irlba/man/irlba.html">irlba</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">sc_data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">[</span><span class="va">MC</span><span class="op">$</span><span class="va">genes.use</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span>, nv <span class="op">=</span> <span class="va">nb_pc</span><span class="op">)</span></span>
<span><span class="va">pca.x</span> <span class="op">&lt;-</span> <span class="va">PCA.res</span><span class="op">$</span><span class="va">u</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">PCA.res</span><span class="op">$</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">pca.x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sc_data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">sc.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateDimReducObject.html">CreateDimReducObject</a></span><span class="op">(</span></span>
<span>  embeddings <span class="op">=</span> <span class="va">pca.x</span>,</span>
<span>  loadings <span class="op">=</span> <span class="va">PCA.res</span><span class="op">$</span><span class="va">v</span>,</span>
<span>  key <span class="op">=</span> <span class="st">"PC_"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Saving metacell object for the "</span>, <span class="va">proj_name</span>, <span class="st">" dataset using "</span>, <span class="va">MC_tool</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Saving metacell object for the 3k_pbmc dataset using SuperCell"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">MC.seurat</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">proj_name</span>, <span class="st">'/metacell_'</span>, <span class="va">MC_tool</span>,<span class="st">'.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can also use the <code><a href="https://rdrr.io/pkg/SuperCell/man/supercell_2_Seurat.html">supercell_2_Seurat()</a></code> function from the SuperCell package.
This function takes as inputs the metacell count matrix (output of the SuperCell <code><a href="https://rdrr.io/pkg/SuperCell/man/supercell_GE.html">supercell_GE()</a></code> function) and the output of the SuperCell <code><a href="https://rdrr.io/pkg/SuperCell/man/SCimplify.html">SCimplify()</a></code> function
to output a Seurat object containing normalized metacells gene expression data as well as the first (<code>N.comp</code>) principal components of PCA performed internally using user defined set of genes (by default the genes used for metacells constructions).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_2_Seurat.html">supercell_2_Seurat</a></span><span class="op">(</span></span>
<span>  SC.GE <span class="op">=</span> <span class="va">MC.GE</span>,</span>
<span>  SC <span class="op">=</span> <span class="va">MC</span>,</span>
<span>  fields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"annotation"</span>, <span class="st">"supercell_size"</span><span class="op">)</span>, <span class="co"># elements of MC to save as metacell metadata </span></span>
<span>  var.genes <span class="op">=</span> <span class="va">MC</span><span class="op">$</span><span class="va">genes.use</span>,</span>
<span>  N.comp <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">MC.seurat</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">proj_name</span>, <span class="st">'/metacell_'</span>, <span class="va">MC_tool</span>,<span class="st">'.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For future downstream analyses in python (section <a href="#standard-analysis-Py"><strong>??</strong></a>), metacell counts can be saved in an Anndata object:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC.seurat.ad</span> <span class="op">&lt;-</span> <span class="fu">anndata</span><span class="fu">::</span><span class="fu"><a href="https://anndata.dynverse.org/reference/AnnData.html">AnnData</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">MC.seurat</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  obs <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">@</span><span class="va">meta.data</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">anndata</span><span class="fu">::</span><span class="fu"><a href="https://anndata.dynverse.org/reference/write_h5ad.html">write_h5ad</a></span><span class="op">(</span>anndata <span class="op">=</span> <span class="va">MC.seurat.ad</span>, filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">proj_name</span>, <span class="st">'/metacell_'</span>, <span class="va">MC_tool</span>,<span class="st">'.h5ad'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>#&gt;            used  (Mb) gc trigger  (Mb) max used  (Mb)
#&gt; Ncells  3474466 185.6    5447132 291.0  5447132 291.0
#&gt; Vcells 20531196 156.7   65218418 497.6 65217828 497.6</code></pre>
</div>
</div>
<div id="MC2-construction" class="section level2" number="3.2">
<h2>
<span class="header-section-number">3.2</span> MC2 (Python)<a class="anchor" aria-label="anchor" href="#MC2-construction"><i class="fas fa-link"></i></a>
</h2>
<p>In this section, we construct metacells using <a href="https://github.com/tanaylab/metacells">Metacell-2 (MC2)</a> implemented in Python.</p>
<div id="method-1" class="section level3" number="3.2.1">
<h3>
<span class="header-section-number">3.2.1</span> Method<a class="anchor" aria-label="anchor" href="#method-1"><i class="fas fa-link"></i></a>
</h3>
<p>Metacell-2 (MC2) is a python tool to construct metacells and is the updated version of the MetaCell algorithm, which introduced the concept of metacell.
MC2 applies a two-phase divide-and-conquer approach. Cells are randomly divided into piles of ~10k cells and
initial metacells are built applying a MetaCell-like approach per pile, i.e. based on a single-cell kNN graph built
from log-normalized counts using a set of highly variable genes.
Then, transcriptionally similar metacells are grouped into metagroup piles for the identification of final metacells and outliers identification.
Note that prior to metacell identification, the MC2 framework recommends gene filtering steps. The choice of the genes used
by the method is of high importance to assure good quality of the metacells.</p>
<p>The code provided in this section is adapted from the <a href="https://tanaylab.github.io/metacells-vignettes/one-pass.html">author’s tutorial</a>.
For more information on the method, please refer to our review <span class="citation">(<a href="#ref-Review" role="doc-biblioref"><strong>Review?</strong></a>)</span> and the original paper <span class="citation">(<a href="references.html#ref-MC2" role="doc-biblioref">Ben-Kiki et al. 2022</a>)</span>.</p>
<div id="importing-python-packages" class="section level4 unnumbered">
<h4>Importing python packages<a class="anchor" aria-label="anchor" href="#importing-python-packages"><i class="fas fa-link"></i></a>
</h4>
<p>To run MC2, the following python packages need to be imported:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb26-1"><a href="Metacell-construction-chapter.html#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb26-2"><a href="Metacell-construction-chapter.html#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb26-3"><a href="Metacell-construction-chapter.html#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb26-4"><a href="Metacell-construction-chapter.html#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb26-5"><a href="Metacell-construction-chapter.html#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb26-6"><a href="Metacell-construction-chapter.html#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb26-7"><a href="Metacell-construction-chapter.html#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb26-8"><a href="Metacell-construction-chapter.html#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> metacells <span class="im">as</span> mc</span></code></pre></div>
<p>If you don’t have these packages installed, please refer to the section <a href="requirements.html#installations">1.1</a>.</p>
</div>
</div>
<div id="data-loading-1" class="section level3" number="3.2.2">
<h3>
<span class="header-section-number">3.2.2</span> Data loading<a class="anchor" aria-label="anchor" href="#data-loading-1"><i class="fas fa-link"></i></a>
</h3>
<p>We will run Metacell-2 (MC2) on a single-cell dataset composed of around 3000 peripheral blood mononuclear cells (PBMCs).
Please follow the section <a href="requirements.html#PBMC-data">1.2</a> to retrieve these data from the scanpy package and save the data in the following file: “data/3k_pbmc/singlecell_anndata_filtered.h5ad”.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb27-1"><a href="Metacell-construction-chapter.html#cb27-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">"MC2"</span></span>
<span id="cb27-2"><a href="Metacell-construction-chapter.html#cb27-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">"3k_pbmc"</span></span>
<span id="cb27-3"><a href="Metacell-construction-chapter.html#cb27-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(<span class="st">"data"</span>, proj_name, <span class="st">"singlecell_anndata_filtered.h5ad"</span>))</span>
<span id="cb27-4"><a href="Metacell-construction-chapter.html#cb27-4" aria-hidden="true" tabindex="-1"></a>ad.X <span class="op">=</span> ad.raw.X.copy()</span></code></pre></div>
<p>We initialize the name of the anndata (in the unstructured annotations) object using the <code>mc.ut.set_name()</code> function from the MC2 package.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb28-1"><a href="Metacell-construction-chapter.html#cb28-1" aria-hidden="true" tabindex="-1"></a>mc.ut.set_name(ad, proj_name)</span></code></pre></div>
</div>
<div id="filtering-steps-1" class="section level3" number="3.2.3">
<h3>
<span class="header-section-number">3.2.3</span> Filtering steps<a class="anchor" aria-label="anchor" href="#filtering-steps-1"><i class="fas fa-link"></i></a>
</h3>
<p>MC2 requires that standard filtering steps such as doublet filtering are performed outside of the MC2 framework.
In addition to standard data filtering steps, the MC2 package proposes functions to filter the single-cell data at the gene and at the cell level (See <a href="https://tanaylab.github.io/metacells-vignettes/one-pass.html">author’s vignette</a>).
At the gene level, the filtering steps consist in excluding genes based on biological knowledge (<em>e.g.</em> mitochrondrial genes) as well as based on their expression levels.
The latter genes include genes with zero expression or low expression levels and “bursty lonely genes” (<em>i.e.</em>, genes with high expression levels but no correlation with any other gene).
At the cell level, filtering is performed based on cells UMI counts.</p>
<div id="gene-filtering" class="section level4 unnumbered">
<h4>Gene filtering<a class="anchor" aria-label="anchor" href="#gene-filtering"><i class="fas fa-link"></i></a>
</h4>
<p>In the following code chunk, we exclude genes using the <code>mc.pl.exclude_genes()</code>function from the MC2 package.
Based on the authors vignette, we consider a minimal list of genes to exclude, <em>i.e.</em>, sex-specific and non-coding genes as well as the mitochondrial genes.
To complete this list of genes, an iterative approach can be used following the guidelines of the MC2 authors in a <a href="https://tanaylab.github.io/metacells-vignettes/iterative.html">second vignette</a>.
The <code>mc.pl.exclude_genes()</code> function will filter out:
i) the known-to-be-excluded genes defined by the user as gene names or gene names patterns (<code>EXCLUDED_GENE_NAMES</code> and <code>EXCLUDED_GENE_PATTERNS</code> parameters respectively),
and ii) the “bursty lonely genes”.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb29-1"><a href="Metacell-construction-chapter.html#cb29-1" aria-hidden="true" tabindex="-1"></a>EXCLUDED_GENE_NAMES <span class="op">=</span> [<span class="st">"XIST"</span>, <span class="st">"MALAT1"</span>, <span class="st">"NEAT1"</span>] </span>
<span id="cb29-2"><a href="Metacell-construction-chapter.html#cb29-2" aria-hidden="true" tabindex="-1"></a>EXCLUDED_GENE_PATTERNS <span class="op">=</span> [<span class="st">'MT-.*'</span>]</span>
<span id="cb29-3"><a href="Metacell-construction-chapter.html#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="Metacell-construction-chapter.html#cb29-4" aria-hidden="true" tabindex="-1"></a>mc.pl.exclude_genes(</span>
<span id="cb29-5"><a href="Metacell-construction-chapter.html#cb29-5" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb29-6"><a href="Metacell-construction-chapter.html#cb29-6" aria-hidden="true" tabindex="-1"></a>    excluded_gene_names<span class="op">=</span>EXCLUDED_GENE_NAMES,</span>
<span id="cb29-7"><a href="Metacell-construction-chapter.html#cb29-7" aria-hidden="true" tabindex="-1"></a>    excluded_gene_patterns<span class="op">=</span>EXCLUDED_GENE_PATTERNS,</span>
<span id="cb29-8"><a href="Metacell-construction-chapter.html#cb29-8" aria-hidden="true" tabindex="-1"></a>    random_seed<span class="op">=</span><span class="dv">123456</span></span>
<span id="cb29-9"><a href="Metacell-construction-chapter.html#cb29-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb29-10"><a href="Metacell-construction-chapter.html#cb29-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[bursty_lonely_gene]: 0 true (0%) out of 32738 bools</span></span>
<span id="cb29-11"><a href="Metacell-construction-chapter.html#cb29-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[properly_sampled_gene]: 16579 true (50.64%) out of 32738 bools</span></span>
<span id="cb29-12"><a href="Metacell-construction-chapter.html#cb29-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.var[excluded_gene]: 16174 true (49.4%) out of 32738 bools</span></span></code></pre></div>
</div>
<div id="cell-filtering-based-on-umis-counts" class="section level4 unnumbered">
<h4>Cell filtering based on UMIs counts<a class="anchor" aria-label="anchor" href="#cell-filtering-based-on-umis-counts"><i class="fas fa-link"></i></a>
</h4>
<p>In the MC2 framework, cells with very low and very high UMI content will be filtered out (<code>PROPERLY_SAMPLED_MIN_CELL_TOTAL</code>, <code>PROPERLY_SAMPLED_MAX_CELL_TOTAL</code> variables defining thresholds in the next code chunk).<br>
Also, cell filtering based on UMI counts in excluded genes is also performed(<code>PROPERLY_SAMPLED_MAX_EXCLUDED_GENES_FRACTION</code> variable).
Since our dataset has been pre-filtered, very lenient cutoffs will be used in this tutorial.
The following code chunk defines these parameters.
To adapt them to your datasets, we advise you to explore the distributions of total UMI counts and UMI counts in excluded genes, as recommended and described in the MC2 <a href="https://tanaylab.github.io/metacells-vignettes/one-pass.html">original vignette</a>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb30-1"><a href="Metacell-construction-chapter.html#cb30-1" aria-hidden="true" tabindex="-1"></a>PROPERLY_SAMPLED_MIN_CELL_TOTAL <span class="op">=</span> <span class="dv">200</span> </span>
<span id="cb30-2"><a href="Metacell-construction-chapter.html#cb30-2" aria-hidden="true" tabindex="-1"></a>PROPERLY_SAMPLED_MAX_CELL_TOTAL <span class="op">=</span> <span class="dv">10000</span> </span>
<span id="cb30-3"><a href="Metacell-construction-chapter.html#cb30-3" aria-hidden="true" tabindex="-1"></a>PROPERLY_SAMPLED_MAX_EXCLUDED_GENES_FRACTION <span class="op">=</span> <span class="fl">0.25</span></span></code></pre></div>
<p>The number of UMIs in excluded genes is computed using the <code>mc.tl.compute_excluded_gene_umis()</code> function and cells are filtered out using the <code>mc.pl.exclude_cells()</code> function.
Additional cells can be filtered out by adding a cell description columns in the <code>obs</code> data frame in the anndata oject. This annotation should be a boolean indicating whether the cell should filtered out or not.
The name of this column should be provided to the <code>mc.pl.exclude_cells()</code> function via the <code>additional_cells_masks</code> parameter.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb31-1"><a href="Metacell-construction-chapter.html#cb31-1" aria-hidden="true" tabindex="-1"></a>mc.tl.compute_excluded_gene_umis(ad)</span>
<span id="cb31-2"><a href="Metacell-construction-chapter.html#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="Metacell-construction-chapter.html#cb31-3" aria-hidden="true" tabindex="-1"></a>mc.pl.exclude_cells(</span>
<span id="cb31-4"><a href="Metacell-construction-chapter.html#cb31-4" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb31-5"><a href="Metacell-construction-chapter.html#cb31-5" aria-hidden="true" tabindex="-1"></a>    properly_sampled_min_cell_total<span class="op">=</span>PROPERLY_SAMPLED_MIN_CELL_TOTAL,</span>
<span id="cb31-6"><a href="Metacell-construction-chapter.html#cb31-6" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_cell_total<span class="op">=</span>PROPERLY_SAMPLED_MAX_CELL_TOTAL,</span>
<span id="cb31-7"><a href="Metacell-construction-chapter.html#cb31-7" aria-hidden="true" tabindex="-1"></a>    properly_sampled_max_excluded_genes_fraction<span class="op">=</span>PROPERLY_SAMPLED_MAX_EXCLUDED_GENES_FRACTION <span class="co"># ,</span></span>
<span id="cb31-8"><a href="Metacell-construction-chapter.html#cb31-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># additional_cells_masks=["|doublet_cell"]</span></span>
<span id="cb31-9"><a href="Metacell-construction-chapter.html#cb31-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-10"><a href="Metacell-construction-chapter.html#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[properly_sampled_cell]: 2638 true (100%) out of 2638 bools</span></span>
<span id="cb31-11"><a href="Metacell-construction-chapter.html#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.obs[excluded_cell]: 0 true (0%) out of 2638 bools</span></span></code></pre></div>
<p>After performing the two-step filtering (genes and cells), the “cleaned” data can be extracted using the <code>mc.pl.extract_clean_data()</code> function.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb32-1"><a href="Metacell-construction-chapter.html#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract clean dataset (with filtered cells and genes)</span></span>
<span id="cb32-2"><a href="Metacell-construction-chapter.html#cb32-2" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> mc.pl.extract_clean_data(ad)</span>
<span id="cb32-3"><a href="Metacell-construction-chapter.html#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[full_cell_index]: 2638 int32s</span></span>
<span id="cb32-4"><a href="Metacell-construction-chapter.html#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[full_gene_index]: 16564 int32s</span></span></code></pre></div>
</div>
</div>
<div id="building-metacells-1" class="section level3" number="3.2.4">
<h3>
<span class="header-section-number">3.2.4</span> Building metacells<a class="anchor" aria-label="anchor" href="#building-metacells-1"><i class="fas fa-link"></i></a>
</h3>
<div id="defining-lateral-genes" class="section level4 unnumbered">
<h4>Defining lateral genes<a class="anchor" aria-label="anchor" href="#defining-lateral-genes"><i class="fas fa-link"></i></a>
</h4>
<p>To build metacells, we need to define lateral genes, which are genes with strong biological signal which is independent of cell-state, <em>e.g.</em> cell-cycle genes.
These genes will be ignored for computing cells similarity and building metacells
but will be considered to define outlier cells (<em>i.e.</em>, expression levels of lateral genes should be consistent within metacells).
In the following chunk, we consider a minimal list of lateral genes (provided by the MC2 authors) including cell-cycle and ribosomal genes and
mark them in the MC2 object using the <code>mc.pl.mark_lateral_genes()</code> function.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb33-1"><a href="Metacell-construction-chapter.html#cb33-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-2"><a href="Metacell-construction-chapter.html#cb33-2" aria-hidden="true" tabindex="-1"></a>LATERAL_GENE_NAMES <span class="op">=</span> [</span>
<span id="cb33-3"><a href="Metacell-construction-chapter.html#cb33-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ACSM3"</span>, <span class="st">"ANP32B"</span>, <span class="st">"APOE"</span>, <span class="st">"AURKA"</span>, <span class="st">"B2M"</span>, <span class="st">"BIRC5"</span>, <span class="st">"BTG2"</span>, <span class="st">"CALM1"</span>, <span class="st">"CD63"</span>, <span class="st">"CD69"</span>, <span class="st">"CDK4"</span>,</span>
<span id="cb33-4"><a href="Metacell-construction-chapter.html#cb33-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CENPF"</span>, <span class="st">"CENPU"</span>, <span class="st">"CENPW"</span>, <span class="st">"CH17-373J23.1"</span>, <span class="st">"CKS1B"</span>, <span class="st">"CKS2"</span>, <span class="st">"COX4I1"</span>, <span class="st">"CXCR4"</span>, <span class="st">"DNAJB1"</span>,</span>
<span id="cb33-5"><a href="Metacell-construction-chapter.html#cb33-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"DONSON"</span>, <span class="st">"DUSP1"</span>, <span class="st">"DUT"</span>, <span class="st">"EEF1A1"</span>, <span class="st">"EEF1B2"</span>, <span class="st">"EIF3E"</span>, <span class="st">"EMP3"</span>, <span class="st">"FKBP4"</span>, <span class="st">"FOS"</span>, <span class="st">"FOSB"</span>, <span class="st">"FTH1"</span>,</span>
<span id="cb33-6"><a href="Metacell-construction-chapter.html#cb33-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"G0S2"</span>, <span class="st">"GGH"</span>, <span class="st">"GLTSCR2"</span>, <span class="st">"GMNN"</span>, <span class="st">"GNB2L1"</span>, <span class="st">"GPR183"</span>, <span class="st">"H2AFZ"</span>, <span class="st">"H3F3B"</span>, <span class="st">"HBM"</span>, <span class="st">"HIST1H1C"</span>,</span>
<span id="cb33-7"><a href="Metacell-construction-chapter.html#cb33-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"HIST1H2AC"</span>, <span class="st">"HIST1H2BG"</span>, <span class="st">"HIST1H4C"</span>, <span class="st">"HLA-A"</span>, <span class="st">"HLA-B"</span>, <span class="st">"HLA-C"</span>, <span class="st">"HLA-DMA"</span>, <span class="st">"HLA-DMB"</span>,</span>
<span id="cb33-8"><a href="Metacell-construction-chapter.html#cb33-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"HLA-DPA1"</span>, <span class="st">"HLA-DPB1"</span>, <span class="st">"HLA-DQA1"</span>, <span class="st">"HLA-DQB1"</span>, <span class="st">"HLA-DRA"</span>, <span class="st">"HLA-DRB1"</span>, <span class="st">"HLA-E"</span>, <span class="st">"HLA-F"</span>, <span class="st">"HMGA1"</span>,</span>
<span id="cb33-9"><a href="Metacell-construction-chapter.html#cb33-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"HMGB1"</span>, <span class="st">"HMGB2"</span>, <span class="st">"HMGB3"</span>, <span class="st">"HMGN2"</span>, <span class="st">"HNRNPAB"</span>, <span class="st">"HSP90AA1"</span>, <span class="st">"HSP90AB1"</span>, <span class="st">"HSPA1A"</span>, <span class="st">"HSPA1B"</span>,</span>
<span id="cb33-10"><a href="Metacell-construction-chapter.html#cb33-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"HSPA6"</span>, <span class="st">"HSPD1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"HSPH1"</span>, <span class="st">"ID2"</span>, <span class="st">"IER2"</span>, <span class="st">"IGHA1"</span>, <span class="st">"IGHA2"</span>, <span class="st">"IGHD"</span>, <span class="st">"IGHG1"</span>, <span class="st">"IGHG2"</span>,</span>
<span id="cb33-11"><a href="Metacell-construction-chapter.html#cb33-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"IGHG3"</span>, <span class="st">"IGHG4"</span>, <span class="st">"IGHM"</span>, <span class="st">"IGKC"</span>, <span class="st">"IGKV1-12"</span>, <span class="st">"IGKV1-39"</span>, <span class="st">"IGKV1-5"</span>, <span class="st">"IGKV3-15"</span>, <span class="st">"IGKV4-1"</span>,</span>
<span id="cb33-12"><a href="Metacell-construction-chapter.html#cb33-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"IGLC2"</span>, <span class="st">"IGLC3"</span>, <span class="st">"IGLC6"</span>, <span class="st">"IGLC7"</span>, <span class="st">"IGLL1"</span>, <span class="st">"IGLL5"</span>, <span class="st">"IGLV2-34"</span>, <span class="st">"JUN"</span>, <span class="st">"JUNB"</span>, <span class="st">"KIAA0101"</span>,</span>
<span id="cb33-13"><a href="Metacell-construction-chapter.html#cb33-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"LEPROTL1"</span>, <span class="st">"LGALS1"</span>, <span class="st">"LINC01206"</span>, <span class="st">"LTB"</span>, <span class="st">"MCM3"</span>, <span class="st">"MCM4"</span>, <span class="st">"MCM7"</span>, <span class="st">"MKI67"</span>, <span class="st">"MT2A"</span>, <span class="st">"MYL12A"</span>,</span>
<span id="cb33-14"><a href="Metacell-construction-chapter.html#cb33-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"MYL6"</span>, <span class="st">"NASP"</span>, <span class="st">"NFKBIA"</span>, <span class="st">"NUSAP1"</span>, <span class="st">"PA2G4"</span>, <span class="st">"PCNA"</span>, <span class="st">"PDLIM1"</span>, <span class="st">"PLK3"</span>, <span class="st">"PPP1R15A"</span>, <span class="st">"PTMA"</span>,</span>
<span id="cb33-15"><a href="Metacell-construction-chapter.html#cb33-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"PTTG1"</span>, <span class="st">"RAN"</span>, <span class="st">"RANBP1"</span>, <span class="st">"RGCC"</span>, <span class="st">"RGS1"</span>, <span class="st">"RGS2"</span>, <span class="st">"RGS3"</span>, <span class="st">"RP11-1143G9.4"</span>, <span class="st">"RP11-160E2.6"</span>,</span>
<span id="cb33-16"><a href="Metacell-construction-chapter.html#cb33-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"RP11-53B5.1"</span>, <span class="st">"RP11-620J15.3"</span>, <span class="st">"RP5-1025A1.3"</span>, <span class="st">"RP5-1171I10.5"</span>, <span class="st">"RPS10"</span>, <span class="st">"RPS10-NUDT3"</span>, <span class="st">"RPS11"</span>,</span>
<span id="cb33-17"><a href="Metacell-construction-chapter.html#cb33-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"RPS12"</span>, <span class="st">"RPS13"</span>, <span class="st">"RPS14"</span>, <span class="st">"RPS15"</span>, <span class="st">"RPS15A"</span>, <span class="st">"RPS16"</span>, <span class="st">"RPS17"</span>, <span class="st">"RPS18"</span>, <span class="st">"RPS19"</span>, <span class="st">"RPS19BP1"</span>,</span>
<span id="cb33-18"><a href="Metacell-construction-chapter.html#cb33-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"RPS2"</span>, <span class="st">"RPS20"</span>, <span class="st">"RPS21"</span>, <span class="st">"RPS23"</span>, <span class="st">"RPS24"</span>, <span class="st">"RPS25"</span>, <span class="st">"RPS26"</span>, <span class="st">"RPS27"</span>, <span class="st">"RPS27A"</span>, <span class="st">"RPS27L"</span>,</span>
<span id="cb33-19"><a href="Metacell-construction-chapter.html#cb33-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"RPS28"</span>, <span class="st">"RPS29"</span>, <span class="st">"RPS3"</span>, <span class="st">"RPS3A"</span>, <span class="st">"RPS4X"</span>, <span class="st">"RPS4Y1"</span>, <span class="st">"RPS4Y2"</span>, <span class="st">"RPS5"</span>, <span class="st">"RPS6"</span>, <span class="st">"RPS6KA1"</span>,</span>
<span id="cb33-20"><a href="Metacell-construction-chapter.html#cb33-20" aria-hidden="true" tabindex="-1"></a>    <span class="st">"RPS6KA2"</span>, <span class="st">"RPS6KA2-AS1"</span>, <span class="st">"RPS6KA3"</span>, <span class="st">"RPS6KA4"</span>, <span class="st">"RPS6KA5"</span>, <span class="st">"RPS6KA6"</span>, <span class="st">"RPS6KB1"</span>, <span class="st">"RPS6KB2"</span>,</span>
<span id="cb33-21"><a href="Metacell-construction-chapter.html#cb33-21" aria-hidden="true" tabindex="-1"></a>    <span class="st">"RPS6KC1"</span>, <span class="st">"RPS6KL1"</span>, <span class="st">"RPS7"</span>, <span class="st">"RPS8"</span>, <span class="st">"RPS9"</span>, <span class="st">"RPSA"</span>, <span class="st">"RRM2"</span>, <span class="st">"SMC4"</span>, <span class="st">"SRGN"</span>, <span class="st">"SRSF7"</span>, <span class="st">"STMN1"</span>,</span>
<span id="cb33-22"><a href="Metacell-construction-chapter.html#cb33-22" aria-hidden="true" tabindex="-1"></a>    <span class="st">"TK1"</span>, <span class="st">"TMSB4X"</span>, <span class="st">"TOP2A"</span>, <span class="st">"TPX2"</span>, <span class="st">"TSC22D3"</span>, <span class="st">"TUBA1A"</span>, <span class="st">"TUBA1B"</span>, <span class="st">"TUBB"</span>, <span class="st">"TUBB4B"</span>, <span class="st">"TXN"</span>, <span class="st">"TYMS"</span>,</span>
<span id="cb33-23"><a href="Metacell-construction-chapter.html#cb33-23" aria-hidden="true" tabindex="-1"></a>    <span class="st">"UBA52"</span>, <span class="st">"UBC"</span>, <span class="st">"UBE2C"</span>, <span class="st">"UHRF1"</span>, <span class="st">"YBX1"</span>, <span class="st">"YPEL5"</span>, <span class="st">"ZFP36"</span>, <span class="st">"ZWINT"</span></span>
<span id="cb33-24"><a href="Metacell-construction-chapter.html#cb33-24" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb33-25"><a href="Metacell-construction-chapter.html#cb33-25" aria-hidden="true" tabindex="-1"></a>LATERAL_GENE_PATTERNS <span class="op">=</span> [<span class="st">"RP[LS].*"</span>]  <span class="co"># Ribosomal</span></span>
<span id="cb33-26"><a href="Metacell-construction-chapter.html#cb33-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-27"><a href="Metacell-construction-chapter.html#cb33-27" aria-hidden="true" tabindex="-1"></a><span class="co"># This will mark as "lateral_gene" any genes that match the above, if they exist in the clean dataset.</span></span>
<span id="cb33-28"><a href="Metacell-construction-chapter.html#cb33-28" aria-hidden="true" tabindex="-1"></a>mc.pl.mark_lateral_genes(</span>
<span id="cb33-29"><a href="Metacell-construction-chapter.html#cb33-29" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb33-30"><a href="Metacell-construction-chapter.html#cb33-30" aria-hidden="true" tabindex="-1"></a>    lateral_gene_names<span class="op">=</span>LATERAL_GENE_NAMES,</span>
<span id="cb33-31"><a href="Metacell-construction-chapter.html#cb33-31" aria-hidden="true" tabindex="-1"></a>    lateral_gene_patterns<span class="op">=</span>LATERAL_GENE_PATTERNS,</span>
<span id="cb33-32"><a href="Metacell-construction-chapter.html#cb33-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-33"><a href="Metacell-construction-chapter.html#cb33-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[lateral_gene]: 225 true (1.358%) out of 16564 bools</span></span></code></pre></div>
<p>Some genes have higher variances than expected which could lead to false positive outlier identification.
Users can mark those genes as <em>noisy genes</em> using the <code>mc.pl.mark_noisy_genes()</code> function.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb34-1"><a href="Metacell-construction-chapter.html#cb34-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-2"><a href="Metacell-construction-chapter.html#cb34-2" aria-hidden="true" tabindex="-1"></a>NOISY_GENE_NAMES <span class="op">=</span> [</span>
<span id="cb34-3"><a href="Metacell-construction-chapter.html#cb34-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CCL3"</span>, <span class="st">"CCL4"</span>, <span class="st">"CCL5"</span>, <span class="st">"CXCL8"</span>, <span class="st">"DUSP1"</span>, <span class="st">"FOS"</span>, <span class="st">"G0S2"</span>, <span class="st">"HBB"</span>, <span class="st">"HIST1H4C"</span>, <span class="st">"IER2"</span>, <span class="st">"IGKC"</span>,</span>
<span id="cb34-4"><a href="Metacell-construction-chapter.html#cb34-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"IGLC2"</span>, <span class="st">"JUN"</span>, <span class="st">"JUNB"</span>, <span class="st">"KLRB1"</span>, <span class="st">"MT2A"</span>, <span class="st">"RPS26"</span>, <span class="st">"RPS4Y1"</span>, <span class="st">"TRBC1"</span>, <span class="st">"TUBA1B"</span>, <span class="st">"TUBB"</span></span>
<span id="cb34-5"><a href="Metacell-construction-chapter.html#cb34-5" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb34-6"><a href="Metacell-construction-chapter.html#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="co"># This will mark as "noisy_gene" any genes that match the above, if they exist in the clean dataset.</span></span>
<span id="cb34-7"><a href="Metacell-construction-chapter.html#cb34-7" aria-hidden="true" tabindex="-1"></a>mc.pl.mark_noisy_genes(ad, noisy_gene_names<span class="op">=</span>NOISY_GENE_NAMES)</span>
<span id="cb34-8"><a href="Metacell-construction-chapter.html#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[noisy_gene]: 17 true (0.1026%) out of 16564 bools</span></span></code></pre></div>
<p>To extend this list of lateral genes, users can use the <code>relate_to_lateral_genes</code> function to identify genes that are highly correlated with the predefined lateral genes.
The use of this function is described in <a href="https://tanaylab.github.io/metacells-vignettes/iterative.html">the vignette from the MC2 authors</a>.</p>
</div>
<div id="define-target_metacell_size-graining-level" class="section level4 unnumbered">
<h4>Define target_metacell_size (graining level)<a class="anchor" aria-label="anchor" href="#define-target_metacell_size-graining-level"><i class="fas fa-link"></i></a>
</h4>
<p>By default, MC2 will build metacells with a size of 96 cells per metacells.
Users can vary the <code>target_metacell_size</code> parameter to reach a desired graining level.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb35-1"><a href="Metacell-construction-chapter.html#cb35-1" aria-hidden="true" tabindex="-1"></a>gamma <span class="op">=</span> <span class="dv">25</span> </span>
<span id="cb35-2"><a href="Metacell-construction-chapter.html#cb35-2" aria-hidden="true" tabindex="-1"></a>target_metacell_size <span class="op">=</span> gamma</span></code></pre></div>
</div>
<div id="metacells-identification-using-the-divide-and-conquer-approach" class="section level4 unnumbered">
<h4>Metacells identification using the divide and conquer approach<a class="anchor" aria-label="anchor" href="#metacells-identification-using-the-divide-and-conquer-approach"><i class="fas fa-link"></i></a>
</h4>
<p>The construction of metacells by MC2 is performed using the <code>mc.pl.divide_and_conquer_pipeline()</code> function.
Note that by default all cores of the system will be used for the metacells construction.
To change this behavior and adapt the number of cores the MC2 authors propose to use the <code>mc.pl.guess_max_parallel_piles()</code> and <code>mc.pl.set_max_parallel_piles()</code> functions
to adapt the number of processed in parallel depending on the available memory.</p>
<p>The <code>mc.pl.divide_and_conquer_pipeline()</code> function associates each cell to a metacell or defines the cell as outlier.
These assignments are found in the <code>obs</code> layer of the anndata object.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb36-1"><a href="Metacell-construction-chapter.html#cb36-1" aria-hidden="true" tabindex="-1"></a>max_parallel_piles <span class="op">=</span> mc.pl.guess_max_parallel_piles(ad)</span>
<span id="cb36-2"><a href="Metacell-construction-chapter.html#cb36-2" aria-hidden="true" tabindex="-1"></a>mc.pl.set_max_parallel_piles(max_parallel_piles)</span>
<span id="cb36-3"><a href="Metacell-construction-chapter.html#cb36-3" aria-hidden="true" tabindex="-1"></a>mc.pl.divide_and_conquer_pipeline(</span>
<span id="cb36-4"><a href="Metacell-construction-chapter.html#cb36-4" aria-hidden="true" tabindex="-1"></a>    ad,</span>
<span id="cb36-5"><a href="Metacell-construction-chapter.html#cb36-5" aria-hidden="true" tabindex="-1"></a>    target_metacell_size <span class="op">=</span> target_metacell_size,</span>
<span id="cb36-6"><a href="Metacell-construction-chapter.html#cb36-6" aria-hidden="true" tabindex="-1"></a>    random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb36-7"><a href="Metacell-construction-chapter.html#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[selected_gene]: * -&gt; False</span></span>
<span id="cb36-8"><a href="Metacell-construction-chapter.html#cb36-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb36-9"><a href="Metacell-construction-chapter.html#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[rare_gene_module]: 16564 int32 elements with all outliers (100%)</span></span>
<span id="cb36-10"><a href="Metacell-construction-chapter.html#cb36-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[cells_rare_gene_module]: 2638 int32 elements with all outliers (100%)</span></span>
<span id="cb36-11"><a href="Metacell-construction-chapter.html#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[rare_cell]: 0 true (0%) out of 2638 bools</span></span>
<span id="cb36-12"><a href="Metacell-construction-chapter.html#cb36-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.var[selected_gene]: 256 true (1.546%) out of 16564 bools</span></span>
<span id="cb36-13"><a href="Metacell-construction-chapter.html#cb36-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell]: 2638 int32s</span></span>
<span id="cb36-14"><a href="Metacell-construction-chapter.html#cb36-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[dissolved]: 45 true (1.706%) out of 2638 bools</span></span>
<span id="cb36-15"><a href="Metacell-construction-chapter.html#cb36-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell_level]: 2638 int32s</span></span>
<span id="cb36-16"><a href="Metacell-construction-chapter.html#cb36-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-17"><a href="Metacell-construction-chapter.html#cb36-17" aria-hidden="true" tabindex="-1"></a>ad.obs.metacell.head</span>
<span id="cb36-18"><a href="Metacell-construction-chapter.html#cb36-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;bound method NDFrame.head of index</span></span>
<span id="cb36-19"><a href="Metacell-construction-chapter.html#cb36-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATACAACCAC-1     66</span></span>
<span id="cb36-20"><a href="Metacell-construction-chapter.html#cb36-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGAGCTAC-1     87</span></span>
<span id="cb36-21"><a href="Metacell-construction-chapter.html#cb36-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGATCAGC-1     84</span></span>
<span id="cb36-22"><a href="Metacell-construction-chapter.html#cb36-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGCTTCCG-1     91</span></span>
<span id="cb36-23"><a href="Metacell-construction-chapter.html#cb36-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGTATGCG-1    110</span></span>
<span id="cb36-24"><a href="Metacell-construction-chapter.html#cb36-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                    ... </span></span>
<span id="cb36-25"><a href="Metacell-construction-chapter.html#cb36-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCGAACTCTCAT-1    112</span></span>
<span id="cb36-26"><a href="Metacell-construction-chapter.html#cb36-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTGAGGCA-1     74</span></span>
<span id="cb36-27"><a href="Metacell-construction-chapter.html#cb36-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTTCCTCG-1    113</span></span>
<span id="cb36-28"><a href="Metacell-construction-chapter.html#cb36-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGAGAGGC-1     57</span></span>
<span id="cb36-29"><a href="Metacell-construction-chapter.html#cb36-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGCCTCAC-1     82</span></span>
<span id="cb36-30"><a href="Metacell-construction-chapter.html#cb36-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Name: metacell, Length: 2638, dtype: int32&gt;</span></span></code></pre></div>
<p>The following code chunk adds a columns (named <code>membership</code>) containing the single_cell assignments to the obs attribute in the single-cell anndata object.
The membership information is required to compute metacells quality metrics as shown in chapter <a href="QCs.html#QCs">4</a>.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb37-1"><a href="Metacell-construction-chapter.html#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make a membership -- index of metacells to which single cells belong to </span></span>
<span id="cb37-2"><a href="Metacell-construction-chapter.html#cb37-2" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">'membership'</span>] <span class="op">=</span> [<span class="bu">int</span>(i)<span class="op">+</span><span class="dv">1</span> <span class="cf">if</span> i <span class="op">&gt;=</span> <span class="dv">0</span> <span class="cf">else</span> np.nan <span class="cf">for</span> i <span class="kw">in</span> ad.obs.metacell] </span></code></pre></div>
</div>
<div id="retrieve-aggregated-metacell-data-1" class="section level4 unnumbered">
<h4>Retrieve aggregated metacell data<a class="anchor" aria-label="anchor" href="#retrieve-aggregated-metacell-data-1"><i class="fas fa-link"></i></a>
</h4>
<p>The function <code>mc.pl.collect_metacells()</code> should be used to subsequently retrieve an anndata object containing the data at the metacells level instead of the single-cell level.
This function will store in the <code>X</code> data matrix of the anndata object a matrix of gene fraction (<em>i.e.</em>, the sum of all gene levels in a metacell sums to 1)
and it will store the total UMIs per gene per metacell in the layer <code>total_umis</code>.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb38-1"><a href="Metacell-construction-chapter.html#cb38-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-2"><a href="Metacell-construction-chapter.html#cb38-2" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> mc.pl.collect_metacells(ad, name<span class="op">=</span><span class="st">'metacells'</span>, random_seed <span class="op">=</span> <span class="dv">123456</span>)</span>
<span id="cb38-3"><a href="Metacell-construction-chapter.html#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[grouped]: 122 int64s</span></span>
<span id="cb38-4"><a href="Metacell-construction-chapter.html#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[total_umis]: 122 float64s</span></span>
<span id="cb38-5"><a href="Metacell-construction-chapter.html#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.layers[total_umis]: ndarray 122 X 16564 float32s</span></span>
<span id="cb38-6"><a href="Metacell-construction-chapter.html#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[__zeros_downsample_umis]: 122 int64s</span></span>
<span id="cb38-7"><a href="Metacell-construction-chapter.html#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.layers[zeros]: ndarray 122 X 16564 int32s</span></span>
<span id="cb38-8"><a href="Metacell-construction-chapter.html#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set 3k_pbmc.clean.obs[metacell_name]: 2638 &lt;U8s</span></span>
<span id="cb38-9"><a href="Metacell-construction-chapter.html#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[gene_ids]: 16564 objects</span></span>
<span id="cb38-10"><a href="Metacell-construction-chapter.html#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[bursty_lonely_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb38-11"><a href="Metacell-construction-chapter.html#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[properly_sampled_gene]: 16564 true (100%) out of 16564 bools</span></span>
<span id="cb38-12"><a href="Metacell-construction-chapter.html#cb38-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[excluded_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb38-13"><a href="Metacell-construction-chapter.html#cb38-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[full_gene_index]: 16564 int32s</span></span>
<span id="cb38-14"><a href="Metacell-construction-chapter.html#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[lateral_gene]: 225 true (1.358%) out of 16564 bools</span></span>
<span id="cb38-15"><a href="Metacell-construction-chapter.html#cb38-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[noisy_gene]: 17 true (0.1026%) out of 16564 bools</span></span>
<span id="cb38-16"><a href="Metacell-construction-chapter.html#cb38-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[selected_gene]: 256 true (1.546%) out of 16564 bools</span></span>
<span id="cb38-17"><a href="Metacell-construction-chapter.html#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[rare_gene]: 0 true (0%) out of 16564 bools</span></span>
<span id="cb38-18"><a href="Metacell-construction-chapter.html#cb38-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.var[rare_gene_module]: 16564 int32s</span></span>
<span id="cb38-19"><a href="Metacell-construction-chapter.html#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[metacells_rare_gene_module]: 122 int32s</span></span>
<span id="cb38-20"><a href="Metacell-construction-chapter.html#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[rare_metacell]: 0 true (0%) out of 122 bools</span></span>
<span id="cb38-21"><a href="Metacell-construction-chapter.html#cb38-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.uns[outliers]: 161</span></span>
<span id="cb38-22"><a href="Metacell-construction-chapter.html#cb38-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.uns[metacells_algorithm]: metacells.0.9.0</span></span>
<span id="cb38-23"><a href="Metacell-construction-chapter.html#cb38-23" aria-hidden="true" tabindex="-1"></a>mc_ad.shape</span>
<span id="cb38-24"><a href="Metacell-construction-chapter.html#cb38-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; (122, 16564)</span></span>
<span id="cb38-25"><a href="Metacell-construction-chapter.html#cb38-25" aria-hidden="true" tabindex="-1"></a>mc_ad.X.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)[:<span class="dv">5</span>] </span>
<span id="cb38-26"><a href="Metacell-construction-chapter.html#cb38-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; matrix([[1.        ],</span></span>
<span id="cb38-27"><a href="Metacell-construction-chapter.html#cb38-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [1.        ],</span></span>
<span id="cb38-28"><a href="Metacell-construction-chapter.html#cb38-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [1.        ],</span></span>
<span id="cb38-29"><a href="Metacell-construction-chapter.html#cb38-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [1.        ],</span></span>
<span id="cb38-30"><a href="Metacell-construction-chapter.html#cb38-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         [0.99999994]], dtype=float32)</span></span>
<span id="cb38-31"><a href="Metacell-construction-chapter.html#cb38-31" aria-hidden="true" tabindex="-1"></a>mc_ad.layers[<span class="st">'total_umis'</span>]</span>
<span id="cb38-32"><a href="Metacell-construction-chapter.html#cb38-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; array([[1., 0., 0., ..., 0., 0., 0.],</span></span>
<span id="cb38-33"><a href="Metacell-construction-chapter.html#cb38-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 1., 0.],</span></span>
<span id="cb38-34"><a href="Metacell-construction-chapter.html#cb38-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 0., 2.],</span></span>
<span id="cb38-35"><a href="Metacell-construction-chapter.html#cb38-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        ...,</span></span>
<span id="cb38-36"><a href="Metacell-construction-chapter.html#cb38-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 2., 3.],</span></span>
<span id="cb38-37"><a href="Metacell-construction-chapter.html#cb38-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [0., 0., 0., ..., 0., 0., 0.],</span></span>
<span id="cb38-38"><a href="Metacell-construction-chapter.html#cb38-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        [1., 0., 0., ..., 0., 1., 0.]], dtype=float32)</span></span></code></pre></div>
</div>
</div>
<div id="annotate-metacells-using-available-annotations-1" class="section level3" number="3.2.5">
<h3>
<span class="header-section-number">3.2.5</span> Annotate metacells (using available annotations)<a class="anchor" aria-label="anchor" href="#annotate-metacells-using-available-annotations-1"><i class="fas fa-link"></i></a>
</h3>
<p>If single-cell annotations are available in the original single-cell anndata object. We can transfer these annotations to the metacell anndata object
using the <code>mc.tl.convey_obs_to_group()</code> function which will associate each metacell to the most frequent annotation (categorical) or
averaged annotation (continuous) across the single-cells composing the metacell
(use of the <code>mc.ut.most_frequent</code> and <code>np.mean</code> respectively in the <code>mode</code> paratemer).</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb39-1"><a href="Metacell-construction-chapter.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign a single value for each metacell based on the cells.</span></span>
<span id="cb39-2"><a href="Metacell-construction-chapter.html#cb39-2" aria-hidden="true" tabindex="-1"></a>mc.tl.convey_obs_to_group(</span>
<span id="cb39-3"><a href="Metacell-construction-chapter.html#cb39-3" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>ad, gdata<span class="op">=</span>mc_ad,</span>
<span id="cb39-4"><a href="Metacell-construction-chapter.html#cb39-4" aria-hidden="true" tabindex="-1"></a>    property_name<span class="op">=</span>annotation_label, to_property_name<span class="op">=</span>annotation_label,</span>
<span id="cb39-5"><a href="Metacell-construction-chapter.html#cb39-5" aria-hidden="true" tabindex="-1"></a>    method<span class="op">=</span>mc.ut.most_frequent  <span class="co"># This is the default, for categorical data</span></span>
<span id="cb39-6"><a href="Metacell-construction-chapter.html#cb39-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb39-7"><a href="Metacell-construction-chapter.html#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain]: 122 &lt;U17s</span></span>
<span id="cb39-8"><a href="Metacell-construction-chapter.html#cb39-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-9"><a href="Metacell-construction-chapter.html#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the fraction of cells with each possible value in each metacell:</span></span>
<span id="cb39-10"><a href="Metacell-construction-chapter.html#cb39-10" aria-hidden="true" tabindex="-1"></a>mc.tl.convey_obs_fractions_to_group(  </span>
<span id="cb39-11"><a href="Metacell-construction-chapter.html#cb39-11" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>ad, gdata<span class="op">=</span>mc_ad,</span>
<span id="cb39-12"><a href="Metacell-construction-chapter.html#cb39-12" aria-hidden="true" tabindex="-1"></a>    property_name<span class="op">=</span>annotation_label, to_property_name<span class="op">=</span>annotation_label</span>
<span id="cb39-13"><a href="Metacell-construction-chapter.html#cb39-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb39-14"><a href="Metacell-construction-chapter.html#cb39-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_B cells]: 122 float64s</span></span>
<span id="cb39-15"><a href="Metacell-construction-chapter.html#cb39-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_CD14+ Monocytes]: 122 float64s</span></span>
<span id="cb39-16"><a href="Metacell-construction-chapter.html#cb39-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_CD4 T cells]: 122 float64s</span></span>
<span id="cb39-17"><a href="Metacell-construction-chapter.html#cb39-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_CD8 T cells]: 122 float64s</span></span>
<span id="cb39-18"><a href="Metacell-construction-chapter.html#cb39-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_Dendritic cells]: 122 float64s</span></span>
<span id="cb39-19"><a href="Metacell-construction-chapter.html#cb39-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_FCGR3A+ Monocytes]: 122 float64s</span></span>
<span id="cb39-20"><a href="Metacell-construction-chapter.html#cb39-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_Megakaryocytes]: 122 float64s</span></span>
<span id="cb39-21"><a href="Metacell-construction-chapter.html#cb39-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; set metacells.obs[louvain_fraction_of_NK cells]: 122 float64s</span></span></code></pre></div>
</div>
<div id="save-output-1" class="section level3" number="3.2.6">
<h3>
<span class="header-section-number">3.2.6</span> Save output<a class="anchor" aria-label="anchor" href="#save-output-1"><i class="fas fa-link"></i></a>
</h3>
<p>For future downstream analyses in python (section <a href="#standard-analysis-Py"><strong>??</strong></a>), we save the metacell counts in an Anndata object:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb40-1"><a href="Metacell-construction-chapter.html#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Save single-cell metadata (i.e., `raw.obs` dataframe) in the metacell adata object</span></span>
<span id="cb40-2"><a href="Metacell-construction-chapter.html#cb40-2" aria-hidden="true" tabindex="-1"></a>mc_ad.uns <span class="op">=</span> ad.uns.copy()</span>
<span id="cb40-3"><a href="Metacell-construction-chapter.html#cb40-3" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">'sc.obs'</span>] <span class="op">=</span> ad.obs.copy()</span>
<span id="cb40-4"><a href="Metacell-construction-chapter.html#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="Metacell-construction-chapter.html#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># save the requested gamma</span></span>
<span id="cb40-6"><a href="Metacell-construction-chapter.html#cb40-6" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">'gamma'</span>] <span class="op">=</span> gamma</span>
<span id="cb40-7"><a href="Metacell-construction-chapter.html#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="Metacell-construction-chapter.html#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save metacell size</span></span>
<span id="cb40-9"><a href="Metacell-construction-chapter.html#cb40-9" aria-hidden="true" tabindex="-1"></a>mc_ad.obs.rename(columns<span class="op">=</span>{<span class="st">'grouped'</span>:<span class="st">'size'</span>}, inplace <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb40-10"><a href="Metacell-construction-chapter.html#cb40-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-11"><a href="Metacell-construction-chapter.html#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Saving metacell object for the "</span><span class="op">+</span> proj_name<span class="op">+</span> <span class="st">" dataset using "</span><span class="op">+</span> MC_tool)</span>
<span id="cb40-12"><a href="Metacell-construction-chapter.html#cb40-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Saving metacell object for the 3k_pbmc dataset using MC2</span></span>
<span id="cb40-13"><a href="Metacell-construction-chapter.html#cb40-13" aria-hidden="true" tabindex="-1"></a>mc_ad.write_h5ad(os.path.join(<span class="st">'./data'</span>, proj_name, <span class="ss">f'metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad'</span>))</span></code></pre></div>
<p>For future QCs and downstream analyses in R (section <a href="downstream-analysis-of-metacells.html#standard-analysis-R">6.1</a>), we save the metacell counts in a Seurat object:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://anndata.dynverse.org">anndata</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/">reticulate</a></span><span class="op">)</span></span>
<span><span class="va">adata_mc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://anndata.dynverse.org/reference/read_h5ad.html">read_h5ad</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/"</span>, <span class="va">py</span><span class="op">$</span><span class="va">proj_name</span>, <span class="st">"/metacell_MC2.h5ad"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save counts and metadata in a Seurat object</span></span>
<span><span class="va">countMatrix</span> <span class="op">&lt;-</span>  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">adata_mc</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">adata_mc</span><span class="op">$</span><span class="va">obs_names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">adata_mc</span><span class="op">$</span><span class="va">var_names</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="va">countMatrix</span>, <span class="st">'CsparseMatrix'</span><span class="op">)</span>, meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">adata_mc</span><span class="op">$</span><span class="va">obs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Feature names cannot have underscores ('_'), replacing with dashes</span></span>
<span><span class="co">#&gt; ('-')</span></span>
<span><span class="co">#&gt; Warning: Invalid name supplied, making object name syntactically valid. New</span></span>
<span><span class="co">#&gt; object name is</span></span>
<span><span class="co">#&gt; sizetotal_umisX__zeros_downsample_umismetacells_rare_gene_modulerare_metacelllouvainlouvain_fraction_of_B.cellslouvain_fraction_of_CD14..Monocyteslouvain_fraction_of_CD4.T.cellslouvain_fraction_of_CD8.T.cellslouvain_fraction_of_Dendritic.cellslouvain_fraction_of_FCGR3A..Monocyteslouvain_fraction_of_Megakaryocyteslouvain_fraction_of_NK.cells;</span></span>
<span><span class="co">#&gt; see ?make.names for more details on syntax validity</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">[[</span><span class="st">"var_features"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">adata_mc</span><span class="op">$</span><span class="va">var</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">adata_mc</span><span class="op">$</span><span class="va">var</span><span class="op">$</span><span class="va">selected_gene</span> <span class="op">==</span> <span class="cn">T</span><span class="op">)</span><span class="op">]</span> </span>
<span></span>
<span><span class="co"># Save membership in misc</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span> <span class="op">&lt;-</span> <span class="va">py</span><span class="op">$</span><span class="va">ad</span><span class="op">$</span><span class="va">obs</span><span class="op">[</span><span class="st">'membership'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">MC.seurat</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">py</span><span class="op">$</span><span class="va">proj_name</span>, <span class="st">'/metacell_MC2.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div id="SEACells-construction" class="section level2" number="3.3">
<h2>
<span class="header-section-number">3.3</span> SEACells (Python)<a class="anchor" aria-label="anchor" href="#SEACells-construction"><i class="fas fa-link"></i></a>
</h2>
<p>In this section, we construct metacells using <a href="https://github.com/dpeerlab/SEACells">SEACells</a>.</p>
<div id="method-2" class="section level3" number="3.3.1">
<h3>
<span class="header-section-number">3.3.1</span> Method<a class="anchor" aria-label="anchor" href="#method-2"><i class="fas fa-link"></i></a>
</h3>
<p>The SEAcells method builds a single-cell kNN graph from the Euclidean distance in the principal component space (SVD for scATAC-seq) space.
Distances in the graph are transformed to affinity by applying an adaptive Gaussian kernel.
The affinity matrix is then decomposed into archetypes (linear combination of cells) and membership matrices (cells as a linear combination of archetypes).
Single cells are assigned to a given metacell based on the maximum membership value of the corresponding archetype.</p>
<p>The code provided in this section is adapted from the <a href="https://github.com/dpeerlab/SEACells/blob/main/notebooks/SEACell_computation.ipynb">author’s jupyter notebook</a>.
For more information on the method, please refer to our review <span class="citation">(<a href="#ref-Review" role="doc-biblioref"><strong>Review?</strong></a>)</span> and the original paper <span class="citation">(<a href="references.html#ref-SEACells" role="doc-biblioref">Persad et al. 2023</a>)</span>.</p>
<div id="importing-python-packages-1" class="section level4 unnumbered">
<h4>Importing python packages<a class="anchor" aria-label="anchor" href="#importing-python-packages-1"><i class="fas fa-link"></i></a>
</h4>
<p>To run the SEACells, the following python packages need to be imported:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb42-1"><a href="Metacell-construction-chapter.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb42-2"><a href="Metacell-construction-chapter.html#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb42-3"><a href="Metacell-construction-chapter.html#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb42-4"><a href="Metacell-construction-chapter.html#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> SEACells</span>
<span id="cb42-5"><a href="Metacell-construction-chapter.html#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random</span></code></pre></div>
<p>If you don’t have these packages installed, please refer to the section <a href="requirements.html#installations">1.1</a>.</p>
</div>
</div>
<div id="data-loading-2" class="section level3" number="3.3.2">
<h3>
<span class="header-section-number">3.3.2</span> Data loading<a class="anchor" aria-label="anchor" href="#data-loading-2"><i class="fas fa-link"></i></a>
</h3>
<p>Similarly to SuperCell and MC2, we will run SEACells on the single-cell dataset composed of around 3000 peripheral blood mononuclear cells (PBMCs).
Please follow the section <a href="requirements.html#PBMC-data">1.2</a> to retrieve these data from the scanpy package and save the data in the following file: “data/3k_pbmc/singlecell_anndata_filtered.h5ad”.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb43-1"><a href="Metacell-construction-chapter.html#cb43-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">"SEACells"</span></span>
<span id="cb43-2"><a href="Metacell-construction-chapter.html#cb43-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">"3k_pbmc"</span></span>
<span id="cb43-3"><a href="Metacell-construction-chapter.html#cb43-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(<span class="st">"data"</span>, proj_name, <span class="st">"singlecell_anndata_filtered.h5ad"</span>))</span></code></pre></div>
</div>
<div id="filtering-steps-2" class="section level3" number="3.3.3">
<h3>
<span class="header-section-number">3.3.3</span> Filtering steps<a class="anchor" aria-label="anchor" href="#filtering-steps-2"><i class="fas fa-link"></i></a>
</h3>
<p>In this tutorial, the data have been pre-filterd and SEACells does not perform additionnal filtering.</p>
</div>
<div id="building-metacells-2" class="section level3" number="3.3.4">
<h3>
<span class="header-section-number">3.3.4</span> Building metacells<a class="anchor" aria-label="anchor" href="#building-metacells-2"><i class="fas fa-link"></i></a>
</h3>
<p>Metacells construction using SEACells requires 2 main inputs: i) an anndata object (<code>build_kernel_on</code> parameter), and
ii) a key indicating which matrix in the <code>obsm</code> attribute of the anndata object should be considered to compute the kernel needed for archetypal analysis (<code>build_kernel_on</code> parameter).
Important optional inputs are: the number of metacells to identify (<code>n_SEACells</code> parameter), which is used as input of the archetypal analysis,
ii) the number of neighbors to consider for the knn graph (<code>n_neighbors</code> parameter).</p>
<div id="data-pre-processing-1" class="section level4 unnumbered">
<h4>Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing-1"><i class="fas fa-link"></i></a>
</h4>
<p>The following code chunk saves the raw counts of the filtered data in the raw attribute of the anndata object.
The raw counts will be used later for metacells aggregation.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb44-1"><a href="Metacell-construction-chapter.html#cb44-1" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(ad.X)</span>
<span id="cb44-2"><a href="Metacell-construction-chapter.html#cb44-2" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> ad.obs_names, ad.var_names</span>
<span id="cb44-3"><a href="Metacell-construction-chapter.html#cb44-3" aria-hidden="true" tabindex="-1"></a>ad.raw <span class="op">=</span> raw_ad</span></code></pre></div>
<p>To build the kernel for archetypal analysis, SEACells requires a lower-dimensionnal embedding of the data (for example using PCA for scRNA-Seq data or SVD for scATAC-Seq data).
In the next code chunk, we follow standard pre-processing steps prior to PCA computation, <em>i.e.</em>, data normalization, log transformation, identification of highly variable genes.
PCA components are saved in the <code>obsm</code> attribute of the anndata object.</p>
<p>To pre-process the single-cell data, we are using standard pre-processing for single-cell RNA-seq data using Scanpy. For more information, see <a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">the Scanpy tutorial</a>.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb45-1"><a href="Metacell-construction-chapter.html#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize cells, log transform and compute highly variable genes</span></span>
<span id="cb45-2"><a href="Metacell-construction-chapter.html#cb45-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(ad, <span class="dv">10000</span>)</span>
<span id="cb45-3"><a href="Metacell-construction-chapter.html#cb45-3" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ad)</span>
<span id="cb45-4"><a href="Metacell-construction-chapter.html#cb45-4" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(ad, n_top_genes<span class="op">=</span><span class="dv">2000</span>)</span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb46-1"><a href="Metacell-construction-chapter.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute principal components - </span></span>
<span id="cb46-2"><a href="Metacell-construction-chapter.html#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="Metacell-construction-chapter.html#cb46-3" aria-hidden="true" tabindex="-1"></a>n_comp    <span class="op">=</span> <span class="dv">30</span></span>
<span id="cb46-4"><a href="Metacell-construction-chapter.html#cb46-4" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(ad, n_comps<span class="op">=</span>n_comp, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb46-5"><a href="Metacell-construction-chapter.html#cb46-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-6"><a href="Metacell-construction-chapter.html#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute UMAP for visualization</span></span>
<span id="cb46-7"><a href="Metacell-construction-chapter.html#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we use 30 components to be consistent with our main tutorial, but fill free to explore other number of principal components to use </span></span>
<span id="cb46-8"><a href="Metacell-construction-chapter.html#cb46-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-9"><a href="Metacell-construction-chapter.html#cb46-9" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(ad, n_neighbors<span class="op">=</span><span class="dv">15</span>, n_pcs<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb46-10"><a href="Metacell-construction-chapter.html#cb46-10" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(ad)</span></code></pre></div>
</div>
<div id="setting-up-seacells-parameters" class="section level4 unnumbered">
<h4>Setting up SEACells parameters<a class="anchor" aria-label="anchor" href="#setting-up-seacells-parameters"><i class="fas fa-link"></i></a>
</h4>
<p>In this tutorial, we will use in the SEACells model the 30 first principal components resulting from the PCA to build the knn graph which will be used to compute the kernel.
The number of neighbors to considered for the knn graph can be fixed using the <code>n_neighbors</code> parameter (here 15).<br>
As mentioned previously, users should provide as input the number of metacells required (<code>n_SEACells</code> parameter). This number can be defined as the ratio between the number of single cells and the desired graining level (<code>gamma</code> parameter in the following code chunk).
In this example, we choose a graining level of 25.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb47-1"><a href="Metacell-construction-chapter.html#cb47-1" aria-hidden="true" tabindex="-1"></a>build_kernel_on <span class="op">=</span> <span class="st">'X_pca'</span> <span class="co"># key in ad.obsm to use for computing metacells</span></span>
<span id="cb47-2"><a href="Metacell-construction-chapter.html#cb47-2" aria-hidden="true" tabindex="-1"></a>n_waypoint_eigs <span class="op">=</span> <span class="dv">30</span>      <span class="co"># Number of eigenvalues to consider when initializing metacells</span></span>
<span id="cb47-3"><a href="Metacell-construction-chapter.html#cb47-3" aria-hidden="true" tabindex="-1"></a>n_neighbors <span class="op">=</span> <span class="dv">30</span> <span class="co"># Number of neighbors used for graph construction </span></span>
<span id="cb47-4"><a href="Metacell-construction-chapter.html#cb47-4" aria-hidden="true" tabindex="-1"></a>gamma <span class="op">=</span> <span class="dv">25</span>   <span class="co"># the requested graining level</span></span>
<span id="cb47-5"><a href="Metacell-construction-chapter.html#cb47-5" aria-hidden="true" tabindex="-1"></a>n_SEACells <span class="op">=</span> <span class="bu">int</span>(ad.shape[<span class="dv">0</span>]<span class="op">/</span>gamma) <span class="co"># the requested number of metacells  </span></span></code></pre></div>
</div>
<div id="initializing-the-seacells-model" class="section level4 unnumbered">
<h4>Initializing the SEACells model<a class="anchor" aria-label="anchor" href="#initializing-the-seacells-model"><i class="fas fa-link"></i></a>
</h4>
<p>The SEACells model is initialized with the previously defined parameters using the <code>SEACells.core.SEACells</code> function.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb48-1"><a href="Metacell-construction-chapter.html#cb48-1" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> SEACells.core.SEACells(ad,</span>
<span id="cb48-2"><a href="Metacell-construction-chapter.html#cb48-2" aria-hidden="true" tabindex="-1"></a>                  build_kernel_on <span class="op">=</span> build_kernel_on,</span>
<span id="cb48-3"><a href="Metacell-construction-chapter.html#cb48-3" aria-hidden="true" tabindex="-1"></a>                  n_SEACells <span class="op">=</span> n_SEACells,</span>
<span id="cb48-4"><a href="Metacell-construction-chapter.html#cb48-4" aria-hidden="true" tabindex="-1"></a>                  n_waypoint_eigs <span class="op">=</span> n_waypoint_eigs,</span>
<span id="cb48-5"><a href="Metacell-construction-chapter.html#cb48-5" aria-hidden="true" tabindex="-1"></a>                  n_neighbors <span class="op">=</span> n_neighbors,</span>
<span id="cb48-6"><a href="Metacell-construction-chapter.html#cb48-6" aria-hidden="true" tabindex="-1"></a>                  convergence_epsilon <span class="op">=</span> <span class="fl">1e-5</span>,</span>
<span id="cb48-7"><a href="Metacell-construction-chapter.html#cb48-7" aria-hidden="true" tabindex="-1"></a>                  verbose <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb48-8"><a href="Metacell-construction-chapter.html#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Welcome to SEACells!</span></span></code></pre></div>
<p>Kernel computation is performed using the <code>mconstruct_kernel_matrix</code> function.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb49-1"><a href="Metacell-construction-chapter.html#cb49-1" aria-hidden="true" tabindex="-1"></a>model.construct_kernel_matrix()</span>
<span id="cb49-2"><a href="Metacell-construction-chapter.html#cb49-2" aria-hidden="true" tabindex="-1"></a>M <span class="op">=</span> model.kernel_matrix</span></code></pre></div>
<p>Metacells are initialized using the <code>initialize_archetypes</code> function.<br>
The SEACells archetypes initialization is based on cells sampling and thus is stochastic. User can fix a seed for reproducible results.
To check that the archetypes are evenly spread, users can visualize them using the <code>plot.plot_initialization</code> function.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb50-1"><a href="Metacell-construction-chapter.html#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed for reproducibility</span></span>
<span id="cb50-2"><a href="Metacell-construction-chapter.html#cb50-2" aria-hidden="true" tabindex="-1"></a>random.seed(<span class="dv">123</span>)</span>
<span id="cb50-3"><a href="Metacell-construction-chapter.html#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="Metacell-construction-chapter.html#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize archetypes</span></span>
<span id="cb50-5"><a href="Metacell-construction-chapter.html#cb50-5" aria-hidden="true" tabindex="-1"></a>model.initialize_archetypes()</span>
<span id="cb50-6"><a href="Metacell-construction-chapter.html#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Building kernel on X_pca</span></span>
<span id="cb50-7"><a href="Metacell-construction-chapter.html#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing diffusion components from X_pca for waypoint initialization ... </span></span>
<span id="cb50-8"><a href="Metacell-construction-chapter.html#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb50-9"><a href="Metacell-construction-chapter.html#cb50-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Sampling waypoints ...</span></span>
<span id="cb50-10"><a href="Metacell-construction-chapter.html#cb50-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb50-11"><a href="Metacell-construction-chapter.html#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Selecting 78 cells from waypoint initialization.</span></span>
<span id="cb50-12"><a href="Metacell-construction-chapter.html#cb50-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Initializing residual matrix using greedy column selection</span></span>
<span id="cb50-13"><a href="Metacell-construction-chapter.html#cb50-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Initializing f and g...</span></span>
<span id="cb50-14"><a href="Metacell-construction-chapter.html#cb50-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Selecting 27 cells from greedy initialization.</span></span>
<span id="cb50-15"><a href="Metacell-construction-chapter.html#cb50-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb50-16"><a href="Metacell-construction-chapter.html#cb50-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb50-17"><a href="Metacell-construction-chapter.html#cb50-17" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span><span class="op">%|</span>          <span class="op">|</span> <span class="dv">0</span><span class="op">/</span><span class="dv">37</span> [<span class="dv">00</span>:<span class="dv">00</span><span class="op">&lt;</span>?, ?it<span class="op">/</span>s]</span>
<span id="cb50-18"><a href="Metacell-construction-chapter.html#cb50-18" aria-hidden="true" tabindex="-1"></a> <span class="dv">59</span><span class="op">%|</span><span class="co">#####9    | 22/37 [00:00&lt;00:00, 213.26it/s]</span></span>
<span id="cb50-19"><a href="Metacell-construction-chapter.html#cb50-19" aria-hidden="true" tabindex="-1"></a><span class="dv">100</span><span class="op">%|</span><span class="co">##########| 37/37 [00:00&lt;00:00, 217.95it/s]</span></span>
<span id="cb50-20"><a href="Metacell-construction-chapter.html#cb50-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the initialization </span></span>
<span id="cb50-21"><a href="Metacell-construction-chapter.html#cb50-21" aria-hidden="true" tabindex="-1"></a>SEACells.plot.plot_initialization(ad, model, plot_basis<span class="op">=</span><span class="st">'X_umap'</span>) </span></code></pre></div>
<div class="inline-figure"><img src="21-MC_construction_discrete_files/figure-html/seacells-model-init-1.png" width="672"></div>
</div>
<div id="fitting-the-seacells-model-to-identify-metacells" class="section level4 unnumbered">
<h4>Fitting the SEACells model to identify metacells<a class="anchor" aria-label="anchor" href="#fitting-the-seacells-model-to-identify-metacells"><i class="fas fa-link"></i></a>
</h4>
<p>The identification of the archetypes is an iterative process. In this example, we fixed the minimum and maximum number of iteration to 10 and 50 respectively.
We then check the model convergence using the <code>plot_convergence</code> function.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb51-1"><a href="Metacell-construction-chapter.html#cb51-1" aria-hidden="true" tabindex="-1"></a>model.fit(min_iter <span class="op">=</span> <span class="dv">10</span>, max_iter <span class="op">=</span> <span class="dv">100</span>)</span>
<span id="cb51-2"><a href="Metacell-construction-chapter.html#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Randomly initialized A matrix.</span></span>
<span id="cb51-3"><a href="Metacell-construction-chapter.html#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Setting convergence threshold at 0.00109</span></span>
<span id="cb51-4"><a href="Metacell-construction-chapter.html#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 1.</span></span>
<span id="cb51-5"><a href="Metacell-construction-chapter.html#cb51-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 1.</span></span>
<span id="cb51-6"><a href="Metacell-construction-chapter.html#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 10.</span></span>
<span id="cb51-7"><a href="Metacell-construction-chapter.html#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 10.</span></span>
<span id="cb51-8"><a href="Metacell-construction-chapter.html#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 20.</span></span>
<span id="cb51-9"><a href="Metacell-construction-chapter.html#cb51-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 20.</span></span>
<span id="cb51-10"><a href="Metacell-construction-chapter.html#cb51-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 30.</span></span>
<span id="cb51-11"><a href="Metacell-construction-chapter.html#cb51-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 30.</span></span>
<span id="cb51-12"><a href="Metacell-construction-chapter.html#cb51-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 40.</span></span>
<span id="cb51-13"><a href="Metacell-construction-chapter.html#cb51-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 40.</span></span>
<span id="cb51-14"><a href="Metacell-construction-chapter.html#cb51-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Converged after 43 iterations.</span></span>
<span id="cb51-15"><a href="Metacell-construction-chapter.html#cb51-15" aria-hidden="true" tabindex="-1"></a>model.plot_convergence()</span></code></pre></div>
<div class="inline-figure"><img src="21-MC_construction_discrete_files/figure-html/seacells-model-fit-3.png" width="672"></div>
<p>Once the final archetypes have been identified, we can assign each single-cell to one metacell (hard assignments).
These assignments (<code>membership</code>) can be retrieved using the <code>get_hard_assignments</code> function or extracted from the anndata object using <code>ad.obs["SEACell"]</code>.
In this tutorial, we will only consider hard assignments. However, the SEACells package also provides the option to retrieve soft assignments (multiple weighted assignments for each cell) using the <code>get_soft_assignments</code> function.
For more details on the soft assignments, please refer to the <a href="https://www.nature.com/articles/s41587-023-01716-9#Sec11">SEACell paper</a> and the original <a href="https://github.com/dpeerlab/SEACells/blob/main/notebooks/SEACell_computation.ipynb">author’s jupyter notebook</a>.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb52-1"><a href="Metacell-construction-chapter.html#cb52-1" aria-hidden="true" tabindex="-1"></a>membership <span class="op">=</span> model.get_hard_assignments()</span>
<span id="cb52-2"><a href="Metacell-construction-chapter.html#cb52-2" aria-hidden="true" tabindex="-1"></a>membership.head</span>
<span id="cb52-3"><a href="Metacell-construction-chapter.html#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;bound method NDFrame.head of                      SEACell</span></span>
<span id="cb52-4"><a href="Metacell-construction-chapter.html#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; index                       </span></span>
<span id="cb52-5"><a href="Metacell-construction-chapter.html#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATACAACCAC-1  SEACell-26</span></span>
<span id="cb52-6"><a href="Metacell-construction-chapter.html#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGAGCTAC-1  SEACell-89</span></span>
<span id="cb52-7"><a href="Metacell-construction-chapter.html#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGATCAGC-1  SEACell-33</span></span>
<span id="cb52-8"><a href="Metacell-construction-chapter.html#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGCTTCCG-1  SEACell-84</span></span>
<span id="cb52-9"><a href="Metacell-construction-chapter.html#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGTATGCG-1  SEACell-34</span></span>
<span id="cb52-10"><a href="Metacell-construction-chapter.html#cb52-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ...                      ...</span></span>
<span id="cb52-11"><a href="Metacell-construction-chapter.html#cb52-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCGAACTCTCAT-1  SEACell-92</span></span>
<span id="cb52-12"><a href="Metacell-construction-chapter.html#cb52-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTGAGGCA-1  SEACell-21</span></span>
<span id="cb52-13"><a href="Metacell-construction-chapter.html#cb52-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTTCCTCG-1  SEACell-19</span></span>
<span id="cb52-14"><a href="Metacell-construction-chapter.html#cb52-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGAGAGGC-1  SEACell-73</span></span>
<span id="cb52-15"><a href="Metacell-construction-chapter.html#cb52-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGCCTCAC-1  SEACell-72</span></span>
<span id="cb52-16"><a href="Metacell-construction-chapter.html#cb52-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb52-17"><a href="Metacell-construction-chapter.html#cb52-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2638 rows x 1 columns]&gt;</span></span>
<span id="cb52-18"><a href="Metacell-construction-chapter.html#cb52-18" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">"SEACell"</span>].head</span>
<span id="cb52-19"><a href="Metacell-construction-chapter.html#cb52-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;bound method NDFrame.head of index</span></span>
<span id="cb52-20"><a href="Metacell-construction-chapter.html#cb52-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATACAACCAC-1    SEACell-26</span></span>
<span id="cb52-21"><a href="Metacell-construction-chapter.html#cb52-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGAGCTAC-1    SEACell-89</span></span>
<span id="cb52-22"><a href="Metacell-construction-chapter.html#cb52-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGATCAGC-1    SEACell-33</span></span>
<span id="cb52-23"><a href="Metacell-construction-chapter.html#cb52-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGCTTCCG-1    SEACell-84</span></span>
<span id="cb52-24"><a href="Metacell-construction-chapter.html#cb52-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGTATGCG-1    SEACell-34</span></span>
<span id="cb52-25"><a href="Metacell-construction-chapter.html#cb52-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                        ...    </span></span>
<span id="cb52-26"><a href="Metacell-construction-chapter.html#cb52-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCGAACTCTCAT-1    SEACell-92</span></span>
<span id="cb52-27"><a href="Metacell-construction-chapter.html#cb52-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTGAGGCA-1    SEACell-21</span></span>
<span id="cb52-28"><a href="Metacell-construction-chapter.html#cb52-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTTCCTCG-1    SEACell-19</span></span>
<span id="cb52-29"><a href="Metacell-construction-chapter.html#cb52-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGAGAGGC-1    SEACell-73</span></span>
<span id="cb52-30"><a href="Metacell-construction-chapter.html#cb52-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGCCTCAC-1    SEACell-72</span></span>
<span id="cb52-31"><a href="Metacell-construction-chapter.html#cb52-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Name: SEACell, Length: 2638, dtype: object&gt;</span></span></code></pre></div>
</div>
<div id="retrieve-aggregated-metacell-data-2" class="section level4 unnumbered">
<h4>Retrieve aggregated metacell data<a class="anchor" aria-label="anchor" href="#retrieve-aggregated-metacell-data-2"><i class="fas fa-link"></i></a>
</h4>
<p>The <code>core.summarize_by_SEACell</code> function can be used to generate a metacell count matrix (aggregation of counts across all cells belonging to each metacell).</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb53-1"><a href="Metacell-construction-chapter.html#cb53-1" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> SEACells.core.summarize_by_SEACell(ad, SEACells_label<span class="op">=</span><span class="st">'SEACell'</span>, summarize_layer<span class="op">=</span><span class="st">'raw'</span>, celltype_label<span class="op">=</span>annotation_label)</span>
<span id="cb53-2"><a href="Metacell-construction-chapter.html#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb53-3"><a href="Metacell-construction-chapter.html#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span><span class="op">%|</span>          <span class="op">|</span> <span class="dv">0</span><span class="op">/</span><span class="dv">105</span> [<span class="dv">00</span>:<span class="dv">00</span><span class="op">&lt;</span>?, ?it<span class="op">/</span>s]</span>
<span id="cb53-4"><a href="Metacell-construction-chapter.html#cb53-4" aria-hidden="true" tabindex="-1"></a> <span class="dv">59</span><span class="op">%|</span><span class="co">#####9    | 62/105 [00:00&lt;00:00, 615.29it/s]</span></span>
<span id="cb53-5"><a href="Metacell-construction-chapter.html#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="dv">100</span><span class="op">%|</span><span class="co">##########| 105/105 [00:00&lt;00:00, 634.76it/s]</span></span></code></pre></div>
</div>
<div id="annotate-metacells" class="section level4 unnumbered">
<h4>Annotate metacells<a class="anchor" aria-label="anchor" href="#annotate-metacells"><i class="fas fa-link"></i></a>
</h4>
<p>Note that providing an annotation to the <code>celltype_label</code> parameter in the <code>SEACells.core.summarize_by_SEACell</code> function
allowed us to annotate the metacells to the most common cell type in each metacell.</p>
</div>
</div>
<div id="visualize-metacells" class="section level3" number="3.3.5">
<h3>
<span class="header-section-number">3.3.5</span> Visualize metacells<a class="anchor" aria-label="anchor" href="#visualize-metacells"><i class="fas fa-link"></i></a>
</h3>
<p>To visualize the metacells, we can project the metacells on the single-cell UMAP representation using the <code>plot.plot_2D</code> included in the SEACells package.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb54-1"><a href="Metacell-construction-chapter.html#cb54-1" aria-hidden="true" tabindex="-1"></a>SEACells.plot.plot_2D(ad, key<span class="op">=</span><span class="st">'X_umap'</span>, colour_metacells<span class="op">=</span><span class="va">True</span>)</span></code></pre></div>
<div class="inline-figure"><img src="21-MC_construction_discrete_files/figure-html/seacells-umap-5.png" width="480"></div>
<!-- The following code chunk adds a columns named `membership` and containing the single_cell assignments to the obs attribute in the anndata object containing the raw data. -->
<!-- This annotation will be used to compute metacells quality metrics. -->
<!-- ```{python seacells-membership, cache = TO_CACHE, warning = FALSE, message = FALSE} -->
<!-- # make `membership` numeric -->
<!-- d = {x: int(i)+1 for i, x in enumerate(mc_ad.obs_names)} -->
<!-- ad.obs.merge(membership) -->
<!-- ad.obs['membership'] = [d[x] for x in membership.SEACell] -->
<!-- ``` -->
<div id="save-output-2" class="section level5 unnumbered">
<h5>Save output<a class="anchor" aria-label="anchor" href="#save-output-2"><i class="fas fa-link"></i></a>
</h5>
<p>For future downstream analyses in python (section <a href="#standard-analysis-Py"><strong>??</strong></a>), we save the metacell counts in an Anndata object:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb55-1"><a href="Metacell-construction-chapter.html#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Saving metacell object for the "</span><span class="op">+</span> proj_name<span class="op">+</span> <span class="st">" dataset using "</span><span class="op">+</span> MC_tool)</span>
<span id="cb55-2"><a href="Metacell-construction-chapter.html#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Saving metacell object for the 3k_pbmc dataset using SEACells</span></span>
<span id="cb55-3"><a href="Metacell-construction-chapter.html#cb55-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-4"><a href="Metacell-construction-chapter.html#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Save metacell sizes </span></span>
<span id="cb55-5"><a href="Metacell-construction-chapter.html#cb55-5" aria-hidden="true" tabindex="-1"></a>label_df <span class="op">=</span> ad.obs[[<span class="st">'SEACell'</span>]].reset_index()</span>
<span id="cb55-6"><a href="Metacell-construction-chapter.html#cb55-6" aria-hidden="true" tabindex="-1"></a>mc_ad.obs <span class="op">=</span> mc_ad.obs.join(pd.DataFrame(label_df.groupby(<span class="st">'SEACell'</span>).count().iloc[:, <span class="dv">0</span>]).rename(columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'size'</span>}))</span>
<span id="cb55-7"><a href="Metacell-construction-chapter.html#cb55-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-8"><a href="Metacell-construction-chapter.html#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save pca used to compute metacells</span></span>
<span id="cb55-9"><a href="Metacell-construction-chapter.html#cb55-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">'var_features'</span>]<span class="op">=</span>ad.var_names[ad.var.highly_variable].tolist()</span>
<span id="cb55-10"><a href="Metacell-construction-chapter.html#cb55-10" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">'sc.pca'</span>]<span class="op">=</span>ad.obsm[<span class="st">'X_pca'</span>] </span>
<span id="cb55-11"><a href="Metacell-construction-chapter.html#cb55-11" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">'sc.umap'</span>]<span class="op">=</span>ad.obsm[<span class="st">'X_umap'</span>] </span>
<span id="cb55-12"><a href="Metacell-construction-chapter.html#cb55-12" aria-hidden="true" tabindex="-1"></a>mc_ad.write_h5ad(os.path.join(<span class="st">'./data'</span>, proj_name, <span class="ss">f'metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad'</span>))</span></code></pre></div>
<p>For future downstream analyses in R (section <a href="downstream-analysis-of-metacells.html#standard-analysis-R">6.1</a>), we save the metacell counts in a Seurat object:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://anndata.dynverse.org">anndata</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/">reticulate</a></span><span class="op">)</span></span>
<span><span class="va">adata_mc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://anndata.dynverse.org/reference/read_h5ad.html">read_h5ad</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/"</span>, <span class="va">py</span><span class="op">$</span><span class="va">proj_name</span>, <span class="st">"/metacell_SEACells.h5ad"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save counts and metadata in a Seurat object</span></span>
<span><span class="va">countMatrix</span> <span class="op">&lt;-</span>  <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">t</span><span class="op">(</span><span class="va">adata_mc</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">adata_mc</span><span class="op">$</span><span class="va">obs_names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">adata_mc</span><span class="op">$</span><span class="va">var_names</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="va">countMatrix</span>, <span class="st">'CsparseMatrix'</span><span class="op">)</span>, meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">adata_mc</span><span class="op">$</span><span class="va">obs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Feature names cannot have underscores ('_'), replacing with dashes</span></span>
<span><span class="co">#&gt; ('-')</span></span>
<span><span class="co"># MC.seurat@misc[["sc.pca"]] &lt;- adata_mc$uns$sc.pca</span></span>
<span><span class="co"># MC.seurat@misc[["sc.umap"]] &lt;- adata_mc$uns$sc.umap</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">[[</span><span class="st">"var_features"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">adata_mc</span><span class="op">$</span><span class="va">uns</span><span class="op">$</span><span class="va">var_features</span> </span>
<span><span class="va">pca.res</span> <span class="op">&lt;-</span> <span class="va">adata_mc</span><span class="op">$</span><span class="va">uns</span><span class="op">$</span><span class="va">sc.pca</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">pca.res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">py</span><span class="op">$</span><span class="va">ad</span><span class="op">$</span><span class="va">obs</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">sc.pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateDimReducObject.html">CreateDimReducObject</a></span><span class="op">(</span></span>
<span>  embeddings <span class="op">=</span> <span class="va">pca.res</span>,</span>
<span>  key <span class="op">=</span> <span class="st">"PC_"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: No columnames present in cell embeddings, setting to 'PC_1:30'</span></span>
<span><span class="co"># Save membership in misc</span></span>
<span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">py</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span>, membership <span class="op">=</span> <span class="va">py</span><span class="op">$</span><span class="va">membership</span><span class="op">$</span><span class="va">SEACell</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">MC.seurat</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">py</span><span class="op">$</span><span class="va">proj_name</span>, <span class="st">'/metacell_SEACells.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<!-- We provide a command line tool allowing users to build metacells using either tool (MC2, SuperCell or SEACells) from a provided dataset. -->
<!-- The command line tool takes multiple parameters as input, *e.g.,* number of neighbors considered in the knn, number of components used, graining level. -->
<!-- which is for example required in a benchmark setting. -->
<!-- ```{bash, commandlines_setup, eval = FALSE} -->
<!-- setwd("MetacellToolkit/") -->
<!-- ``` -->
<!-- ```{bash, SuperCell-command, eval = FALSE} -->
<!-- proj_name="3k_pbmc" -->
<!-- MC_tool="SuperCell" -->
<!-- # input raw adata output adata -->
<!-- Rscript cli/${MC_tool}CL.R -i data/${proj_name}/singlecell_anndata_filtered.h5ad -o data/${proj_name}/${MC_tool}/ -n 50 -f 2000 -k 30 -g 50 -s adata -->
<!-- # input raw adata output seurat -->
<!-- Rscript cli/${MC_tool}CL.R -i data/${proj_name}/singlecell_anndata_filtered.h5ad -o data/${proj_name}/${MC_tool}/ -n 50 -f 2000 -k 30 -g 50 -s seurat -->
<!-- ``` -->
<!-- ```{bash, SEACells-command, eval = FALSE} -->
<!-- proj_name="3k_pbmc" -->
<!-- MC_tool="SEACells" -->
<!-- # input raw adata output adata -->
<!-- Rscript cli/${MC_tool}CL.R -i data/${proj_name}/singlecell_anndata_filtered.h5ad -o data/${proj_name}/${MC_tool}/ -n 50 -f 2000 -k 30 -g 50 -s adata -->
<!-- # input raw adata output seurat -->
<!-- Rscript cli/${MC_tool}CL.R -i data/${proj_name}/singlecell_anndata_filtered.h5ad -o data/${proj_name}/${MC_tool}/ -n 50 -f 2000 -k 30 -g 50 -s seurat -->
<!-- ``` -->

</div>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="the-metacell-concept.html"><span class="header-section-number">2</span> The metacell concept</a></div>
<div class="next"><a href="QCs.html"><span class="header-section-number">4</span> Metacells QCs</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#Metacell-construction-chapter"><span class="header-section-number">3</span> Constructing metacells</a></li>
<li>
<a class="nav-link" href="#SuperCell-construction"><span class="header-section-number">3.1</span> SuperCell (R)</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#method"><span class="header-section-number">3.1.1</span> Method</a></li>
<li><a class="nav-link" href="#data-loading"><span class="header-section-number">3.1.2</span> Data loading</a></li>
<li><a class="nav-link" href="#filtering-steps"><span class="header-section-number">3.1.3</span> Filtering steps</a></li>
<li><a class="nav-link" href="#building-metacells"><span class="header-section-number">3.1.4</span> Building metacells</a></li>
<li><a class="nav-link" href="#annotate-metacells-using-available-annotations"><span class="header-section-number">3.1.5</span> Annotate metacells (using available annotations)</a></li>
<li><a class="nav-link" href="#save-output"><span class="header-section-number">3.1.6</span> Save output</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#MC2-construction"><span class="header-section-number">3.2</span> MC2 (Python)</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#method-1"><span class="header-section-number">3.2.1</span> Method</a></li>
<li><a class="nav-link" href="#data-loading-1"><span class="header-section-number">3.2.2</span> Data loading</a></li>
<li><a class="nav-link" href="#filtering-steps-1"><span class="header-section-number">3.2.3</span> Filtering steps</a></li>
<li><a class="nav-link" href="#building-metacells-1"><span class="header-section-number">3.2.4</span> Building metacells</a></li>
<li><a class="nav-link" href="#annotate-metacells-using-available-annotations-1"><span class="header-section-number">3.2.5</span> Annotate metacells (using available annotations)</a></li>
<li><a class="nav-link" href="#save-output-1"><span class="header-section-number">3.2.6</span> Save output</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#SEACells-construction"><span class="header-section-number">3.3</span> SEACells (Python)</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#method-2"><span class="header-section-number">3.3.1</span> Method</a></li>
<li><a class="nav-link" href="#data-loading-2"><span class="header-section-number">3.3.2</span> Data loading</a></li>
<li><a class="nav-link" href="#filtering-steps-2"><span class="header-section-number">3.3.3</span> Filtering steps</a></li>
<li><a class="nav-link" href="#building-metacells-2"><span class="header-section-number">3.3.4</span> Building metacells</a></li>
<li><a class="nav-link" href="#visualize-metacells"><span class="header-section-number">3.3.5</span> Visualize metacells</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/GfellerLab/MetacellAnalysisTutorial/blob/main/21-MC_construction_discrete.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/GfellerLab/MetacellAnalysisTutorial/edit/main/21-MC_construction_discrete.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Metacell Analysis Tutorial</strong>" was written by Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller. It was last built on 2023-11-21.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
