<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.6 MetaCell2 metacells | Metacell analysis on a continuous dataset</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="5.6 MetaCell2 metacells | Metacell analysis on a continuous dataset" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.6 MetaCell2 metacells | Metacell analysis on a continuous dataset" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Leonard Herault, Aurelie Gabriel, Mariia Bilous, David Gfeller" />


<meta name="date" content="2023-11-21" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="short-downstream-analysis-of-the-metacells.html"/>
<link rel="next" href="quality-control-with-the-metacellanalysistoolkit-package.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>This tutorial</a></li>
<li class="chapter" data-level="1" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>1</b> Requirements</a>
<ul>
<li class="chapter" data-level="1.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>1.1</b> Installations</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="installations.html"><a href="installations.html#using-conda-recommended"><i class="fa fa-check"></i><b>1.1.1</b> Using conda (recommended)</a></li>
<li class="chapter" data-level="1.1.2" data-path="installations.html"><a href="installations.html#without-conda"><i class="fa fa-check"></i><b>1.1.2</b> Without conda</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>1.2</b> Retrieve a discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="1.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>1.3</b> Retrieve a continuous dataset (CD34 dataset)</a></li>
<li class="chapter" data-level="1.4" data-path="HLCA-data.html"><a href="HLCA-data.html"><i class="fa fa-check"></i><b>1.4</b> Retrieve the lung atlas dataset</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="HLCA-data.html"><a href="HLCA-data.html#downloading-the-atlas"><i class="fa fa-check"></i><b>1.4.1</b> Downloading the atlas</a></li>
<li class="chapter" data-level="1.4.2" data-path="HLCA-data.html"><a href="HLCA-data.html#splitting-atlas-by-datasets"><i class="fa fa-check"></i><b>1.4.2</b> Splitting atlas by datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a></li>
<li class="chapter" data-level="3" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells</a>
<ul>
<li class="chapter" data-level="3.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#method"><i class="fa fa-check"></i><b>3.1.1</b> Method</a></li>
<li class="chapter" data-level="3.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>3.1.2</b> Data loading</a></li>
<li class="chapter" data-level="3.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>3.1.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.5" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#annotate-metacells-using-available-annotations"><i class="fa fa-check"></i><b>3.1.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.1.6" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#save-output"><i class="fa fa-check"></i><b>3.1.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="MC2-construction.html"><a href="MC2-construction.html#method-1"><i class="fa fa-check"></i><b>3.2.1</b> Method</a></li>
<li class="chapter" data-level="3.2.2" data-path="MC2-construction.html"><a href="MC2-construction.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.2</b> Data loading</a></li>
<li class="chapter" data-level="3.2.3" data-path="MC2-construction.html"><a href="MC2-construction.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.4" data-path="MC2-construction.html"><a href="MC2-construction.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="MC2-construction.html"><a href="MC2-construction.html#annotate-metacells-using-available-annotations-1"><i class="fa fa-check"></i><b>3.2.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.2.6" data-path="MC2-construction.html"><a href="MC2-construction.html#save-output-1"><i class="fa fa-check"></i><b>3.2.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="SEACells-construction.html"><a href="SEACells-construction.html#method-2"><i class="fa fa-check"></i><b>3.3.1</b> Method</a></li>
<li class="chapter" data-level="3.3.2" data-path="SEACells-construction.html"><a href="SEACells-construction.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.2</b> Data loading</a></li>
<li class="chapter" data-level="3.3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.4" data-path="SEACells-construction.html"><a href="SEACells-construction.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="SEACells-construction.html"><a href="SEACells-construction.html#visualize-metacells"><i class="fa fa-check"></i><b>3.3.5</b> Visualize metacells</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>4</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="4.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>4.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>4.1.1</b> Purity</a></li>
<li class="chapter" data-level="4.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>4.1.2</b> Compactness</a></li>
<li class="chapter" data-level="4.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>4.1.3</b> Separation</a></li>
<li class="chapter" data-level="4.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>4.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>4.2</b> Size distribution</a></li>
<li class="chapter" data-level="4.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>4.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>5</b> Metacell Analysis Toolkit (MCAT)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="setting-up-the-environment.html"><a href="setting-up-the-environment.html"><i class="fa fa-check"></i><b>5.1</b> Setting up the environment</a></li>
<li class="chapter" data-level="5.2" data-path="metacell-building.html"><a href="metacell-building.html"><i class="fa fa-check"></i><b>5.2</b> Metacell building</a></li>
<li class="chapter" data-level="5.3" data-path="building-metacell-with-metacell2-mc2.html"><a href="building-metacell-with-metacell2-mc2.html"><i class="fa fa-check"></i><b>5.3</b> Building metacell with MetaCell2 (MC2)</a></li>
<li class="chapter" data-level="5.4" data-path="building-metacell-with-supercell.html"><a href="building-metacell-with-supercell.html"><i class="fa fa-check"></i><b>5.4</b> Building metacell with SuperCell</a></li>
<li class="chapter" data-level="5.5" data-path="short-downstream-analysis-of-the-metacells.html"><a href="short-downstream-analysis-of-the-metacells.html"><i class="fa fa-check"></i><b>5.5</b> Short downstream analysis of the metacells</a></li>
<li class="chapter" data-level="5.6" data-path="metacell2-metacells.html"><a href="metacell2-metacells.html"><i class="fa fa-check"></i><b>5.6</b> MetaCell2 metacells</a></li>
<li class="chapter" data-level="5.7" data-path="quality-control-with-the-metacellanalysistoolkit-package.html"><a href="quality-control-with-the-metacellanalysistoolkit-package.html"><i class="fa fa-check"></i><b>5.7</b> Quality control with the MetacellAnalysisToolkit package</a></li>
<li class="chapter" data-level="5.8" data-path="loading-single-cell-data.html"><a href="loading-single-cell-data.html"><i class="fa fa-check"></i><b>5.8</b> Loading single-cell data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>6</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>6.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>6.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>6.1.3</b> Clustering</a></li>
<li class="chapter" data-level="6.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="6.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>6.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>6.2</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>6.2.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.2.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>6.2.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.2.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-1"><i class="fa fa-check"></i><b>6.2.3</b> Clustering</a></li>
<li class="chapter" data-level="6.2.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>6.2.4</b> Differential expression analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="integration.html"><a href="integration.html"><i class="fa fa-check"></i><b>7</b> Integration of metacells</a>
<ul>
<li class="chapter" data-level="7.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html"><i class="fa fa-check"></i><b>7.1</b> Unsupervised integration</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#data-loading-3"><i class="fa fa-check"></i><b>7.1.1</b> Data loading</a></li>
<li class="chapter" data-level="7.1.2" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#setting-up-the-environment-1"><i class="fa fa-check"></i><b>7.1.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="7.1.3" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#building-metacell"><i class="fa fa-check"></i><b>7.1.3</b> Building metacell</a></li>
<li class="chapter" data-level="7.1.4" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#loading-metacell-objects"><i class="fa fa-check"></i><b>7.1.4</b> Loading metacell objects</a></li>
<li class="chapter" data-level="7.1.5" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#merging-objects-and-basic-quality-control"><i class="fa fa-check"></i><b>7.1.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="7.1.6" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#unintegrated-analysis"><i class="fa fa-check"></i><b>7.1.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="7.1.7" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#seurat-integration"><i class="fa fa-check"></i><b>7.1.7</b> Seurat integration</a></li>
<li class="chapter" data-level="7.1.8" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#downstream-analysis"><i class="fa fa-check"></i><b>7.1.8</b> Downstream analysis</a></li>
<li class="chapter" data-level="7.1.9" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#conclusion"><i class="fa fa-check"></i><b>7.1.9</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="integration_supervised.html"><a href="integration_supervised.html"><i class="fa fa-check"></i><b>7.2</b> Supervised integration</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="integration_supervised.html"><a href="integration_supervised.html#data-loading-4"><i class="fa fa-check"></i><b>7.2.1</b> Data loading</a></li>
<li class="chapter" data-level="7.2.2" data-path="integration_supervised.html"><a href="integration_supervised.html#setting-up-the-environment-2"><i class="fa fa-check"></i><b>7.2.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="7.2.3" data-path="integration_supervised.html"><a href="integration_supervised.html#constructing-supervised-metacell"><i class="fa fa-check"></i><b>7.2.3</b> Constructing supervised metacell</a></li>
<li class="chapter" data-level="7.2.4" data-path="integration_supervised.html"><a href="integration_supervised.html#load-metacell-objects"><i class="fa fa-check"></i><b>7.2.4</b> Load metacell objects</a></li>
<li class="chapter" data-level="7.2.5" data-path="integration_supervised.html"><a href="integration_supervised.html#merging-objects-and-basic-quality-control-1"><i class="fa fa-check"></i><b>7.2.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="7.2.6" data-path="integration_supervised.html"><a href="integration_supervised.html#unintegrated-analysis-1"><i class="fa fa-check"></i><b>7.2.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="7.2.7" data-path="integration_supervised.html"><a href="integration_supervised.html#stacas-integration"><i class="fa fa-check"></i><b>7.2.7</b> STACAS integration</a></li>
<li class="chapter" data-level="7.2.8" data-path="integration_supervised.html"><a href="integration_supervised.html#comparison-with-unsupervised-analysis"><i class="fa fa-check"></i><b>7.2.8</b> Comparison with unsupervised analysis</a></li>
<li class="chapter" data-level="7.2.9" data-path="integration_supervised.html"><a href="integration_supervised.html#downstream-analysis-1"><i class="fa fa-check"></i><b>7.2.9</b> Downstream analysis</a></li>
<li class="chapter" data-level="7.2.10" data-path="integration_supervised.html"><a href="integration_supervised.html#conclusion-1"><i class="fa fa-check"></i><b>7.2.10</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell analysis on a continuous dataset</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="metacell2-metacells" class="section level2 hasAnchor" number="5.6">
<h2><span class="header-section-number">5.6</span> MetaCell2 metacells<a href="metacell2-metacells.html#metacell2-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We load the object obtained with MCAT</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="metacell2-metacells.html#cb78-1" aria-hidden="true" tabindex="-1"></a>cd34.metacell <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/CD34/MC2/mc_Seurat.rds&quot;</span>)</span>
<span id="cb78-2"><a href="metacell2-metacells.html#cb78-2" aria-hidden="true" tabindex="-1"></a>cd34.metacell</span>
<span id="cb78-3"><a href="metacell2-metacells.html#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb78-4"><a href="metacell2-metacells.html#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12462 features across 140 samples within 1 assay </span></span>
<span id="cb78-5"><a href="metacell2-metacells.html#cb78-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Active assay: RNA (12462 features, 0 variable features)</span></span></code></pre></div>
<p>We Normalize the metacells data, identify the HVGs, and we make a classical dimensionality reduction (first a PCA and then vizualisation with a UMAP).</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="metacell2-metacells.html#cb79-1" aria-hidden="true" tabindex="-1"></a>cd34.metacell <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(cd34.metacell)</span>
<span id="cb79-2"><a href="metacell2-metacells.html#cb79-2" aria-hidden="true" tabindex="-1"></a>cd34.metacell <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cd34.metacell)</span>
<span id="cb79-3"><a href="metacell2-metacells.html#cb79-3" aria-hidden="true" tabindex="-1"></a>cd34.metacell <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cd34.metacell)</span>
<span id="cb79-4"><a href="metacell2-metacells.html#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span id="cb79-5"><a href="metacell2-metacells.html#cb79-5" aria-hidden="true" tabindex="-1"></a>cd34.metacell <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cd34.metacell)</span>
<span id="cb79-6"><a href="metacell2-metacells.html#cb79-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 1 </span></span>
<span id="cb79-7"><a href="metacell2-metacells.html#cb79-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  NUP210, MIS18BP1, LCP1, IQGAP1, P4HB, SLC2A4RG, COLGALT1, GRB2, ARHGEF18, BRI3BP </span></span>
<span id="cb79-8"><a href="metacell2-metacells.html#cb79-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     RBL1, AGTPBP1, GGA2, TTC7A, PTPRE, IL17RA, CLEC12A, SLC4A7, CEP128, ARHGAP26 </span></span>
<span id="cb79-9"><a href="metacell2-metacells.html#cb79-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     LINC01572, UVRAG, LCORL, LRP8, MAP3K1, PDE3B, UBE2E2, GEN1, NSD2, SND1 </span></span>
<span id="cb79-10"><a href="metacell2-metacells.html#cb79-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  RPL34, MEIS1, NKAIN2, CHRM3, LAPTM4B, RPL11, ANGPT1, PRKG1, INPP4B, NRIP1 </span></span>
<span id="cb79-11"><a href="metacell2-metacells.html#cb79-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     RPS2, PDZD2, CRHBP, GPC5, NPR3, PRKG2, ITGA9, CASC15, BEX2, ST8SIA6 </span></span>
<span id="cb79-12"><a href="metacell2-metacells.html#cb79-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     YES1, DAPK1, ITGA9-AS1, MSI2, MED12L, NFATC2, HMGA2, RPS18, SLC8A3, CHRM3-AS2 </span></span>
<span id="cb79-13"><a href="metacell2-metacells.html#cb79-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 2 </span></span>
<span id="cb79-14"><a href="metacell2-metacells.html#cb79-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  TFR2, GATA1, UROD, KLF1, DNPH1, ANK1, TIMM13, PRDX2, XACT, ADAMTS3 </span></span>
<span id="cb79-15"><a href="metacell2-metacells.html#cb79-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     MYC, EPOR, RYR3, TMEM14C, PLIN2, CNRIP1, KEL, CSF1, PTH2R, CYTOR </span></span>
<span id="cb79-16"><a href="metacell2-metacells.html#cb79-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     MPST, P2RX5, SPTBN2, ELOVL6, SLC40A1, CXADR, SLC39A3, ZFPM1, NMNAT3, FADS2 </span></span>
<span id="cb79-17"><a href="metacell2-metacells.html#cb79-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  AKAP13, VIM, CD74, PRKCB, STK17B, CYTH1, HLA-DQB1, DOCK10, SBF2, RASSF2 </span></span>
<span id="cb79-18"><a href="metacell2-metacells.html#cb79-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PRKCE, TSC22D3, PLCB1, CD44, ANKRD44, DST, HDAC9, PDE4B, HLA-DRA, TCF4 </span></span>
<span id="cb79-19"><a href="metacell2-metacells.html#cb79-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CALCRL, IDS, TGFBR2, SLCO3A1, POU2F2, AKNA, NEGR1, YPEL5, FYN, COL24A1 </span></span>
<span id="cb79-20"><a href="metacell2-metacells.html#cb79-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 3 </span></span>
<span id="cb79-21"><a href="metacell2-metacells.html#cb79-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  MYL12A, CARD11, PLP2, CCDC69, LGMN, CTSB, CST3, ARL4C, LIME1, PCED1B </span></span>
<span id="cb79-22"><a href="metacell2-metacells.html#cb79-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CYTH4, CDKN2D, IQSEC1, RNASE6, IGFLR1, CYB561A3, NCF1, PCED1B-AS1, SPIB, FAM160A1 </span></span>
<span id="cb79-23"><a href="metacell2-metacells.html#cb79-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CCDC50, MPEG1, RUBCNL, CYBB, GABARAP, LY96, UGCG, TGFBI, IGF2R, HLA-DMA </span></span>
<span id="cb79-24"><a href="metacell2-metacells.html#cb79-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  ANKRD28, LRMDA, EREG, SNHG25, IRAK3, FNDC3B, PCBP3, CSF3R, RAB44, AZU1 </span></span>
<span id="cb79-25"><a href="metacell2-metacells.html#cb79-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     GPI, MTUS2, SLC22A15, SLC22A4, DENND3, ATP8B4, ELANE, ERLIN1, MARC1, TENT5A </span></span>
<span id="cb79-26"><a href="metacell2-metacells.html#cb79-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     SLC36A4, DSTYK, IL1RAP, RFX8, MS4A3, MFSD10, PKP2, CEP170, RAB27A, HGF </span></span>
<span id="cb79-27"><a href="metacell2-metacells.html#cb79-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 4 </span></span>
<span id="cb79-28"><a href="metacell2-metacells.html#cb79-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  EVI5, TET2, CPQ, FCGRT, SRGN, DENND4A, THEMIS2, RNH1, SUCLG2, PIP5K1B </span></span>
<span id="cb79-29"><a href="metacell2-metacells.html#cb79-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     NR4A2, S100A6, FAM102B, CD4, FGD4, EMP3, CPPED1, EHBP1L1, NID1, AOAH </span></span>
<span id="cb79-30"><a href="metacell2-metacells.html#cb79-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ARHGEF12, PHTF1, GRASP, ZNF385A, PRKCA, RIPOR3, PRKAR2B, SAT1, CHD7, KCNQ1 </span></span>
<span id="cb79-31"><a href="metacell2-metacells.html#cb79-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  SMIM24, IGLL1, PHGDH, SPARC, LGALS3BP, ITM2C, MZB1, CD69, IGFBP7, SPINK2 </span></span>
<span id="cb79-32"><a href="metacell2-metacells.html#cb79-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     LAT2, TMIGD2, C1QTNF4, RPLP1, DOCK1, GYPC, ARMH1, CDK6, ADA, SDK2 </span></span>
<span id="cb79-33"><a href="metacell2-metacells.html#cb79-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     FLT3, PIK3R1, BAALC, PLCB1, SATB1, LRRC26, ARHGAP25, CYFIP2, LNCAROD, PROM1 </span></span>
<span id="cb79-34"><a href="metacell2-metacells.html#cb79-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 5 </span></span>
<span id="cb79-35"><a href="metacell2-metacells.html#cb79-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  CIT, C21orf58, NUF2, KIF15, SPC25, DIAPH3, SMC4, NDC80, CDCA5, RRM2 </span></span>
<span id="cb79-36"><a href="metacell2-metacells.html#cb79-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ESCO2, CDC25C, DEPDC1B, RTKN2, HIST1H4C, MELK, MFGE8, SKA3, KIF20B, NUSAP1 </span></span>
<span id="cb79-37"><a href="metacell2-metacells.html#cb79-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     NCAPG, ESPL1, EXO1, TACC3, RRM1, NCAPG2, PLK4, E2F8, KIF11, CDK1 </span></span>
<span id="cb79-38"><a href="metacell2-metacells.html#cb79-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  SOX4, ACTG1, MIR181A1HG, LINC00173, MED13L, IGFBP7, IL2RG, HCST, SLC43A2, SATB1 </span></span>
<span id="cb79-39"><a href="metacell2-metacells.html#cb79-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ABHD17B, ARPP21, TMSB10, TRPM2, LAT2, DNTT, SLC45A3, MYO1G, S100Z, SMAD7 </span></span>
<span id="cb79-40"><a href="metacell2-metacells.html#cb79-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     REEP5, RPLP1, LINC01934, TNFRSF21, RBMS3, WDR66, MZB1, COBL, CDK6, MOB1B</span></span>
<span id="cb79-41"><a href="metacell2-metacells.html#cb79-41" aria-hidden="true" tabindex="-1"></a>cd34.metacell <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(cd34.metacell,<span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>),<span class="at">min.dist =</span> <span class="fl">0.5</span>)</span>
<span id="cb79-42"><a href="metacell2-metacells.html#cb79-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span id="cb79-43"><a href="metacell2-metacells.html#cb79-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;</span></span>
<span id="cb79-44"><a href="metacell2-metacells.html#cb79-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; This message will be shown once per session</span></span>
<span id="cb79-45"><a href="metacell2-metacells.html#cb79-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:01 UMAP embedding parameters a = 0.583 b = 1.334</span></span>
<span id="cb79-46"><a href="metacell2-metacells.html#cb79-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:01 Read 140 rows and found 50 numeric columns</span></span>
<span id="cb79-47"><a href="metacell2-metacells.html#cb79-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:01 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span id="cb79-48"><a href="metacell2-metacells.html#cb79-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:01 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span id="cb79-49"><a href="metacell2-metacells.html#cb79-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span id="cb79-50"><a href="metacell2-metacells.html#cb79-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span id="cb79-51"><a href="metacell2-metacells.html#cb79-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; **************************************************|</span></span>
<span id="cb79-52"><a href="metacell2-metacells.html#cb79-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:01 Writing NN index file to temp file /tmp/35395114/RtmptFIlEg/file14f2012276370f</span></span>
<span id="cb79-53"><a href="metacell2-metacells.html#cb79-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:01 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span id="cb79-54"><a href="metacell2-metacells.html#cb79-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:01 Annoy recall = 100%</span></span>
<span id="cb79-55"><a href="metacell2-metacells.html#cb79-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:02 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span id="cb79-56"><a href="metacell2-metacells.html#cb79-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:02 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span id="cb79-57"><a href="metacell2-metacells.html#cb79-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:02 Commencing optimization for 500 epochs, with 4956 positive edges</span></span>
<span id="cb79-58"><a href="metacell2-metacells.html#cb79-58" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:02 Optimization finished</span></span></code></pre></div>
<p>Plot the results using Seurat.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="metacell2-metacells.html#cb80-1" aria-hidden="true" tabindex="-1"></a>umap.metacell <span class="ot">&lt;-</span> <span class="fu">UMAPPlot</span>(cd34.metacell,<span class="at">group.by =</span> <span class="st">&quot;celltype&quot;</span>,<span class="at">label =</span> T)</span>
<span id="cb80-2"><a href="metacell2-metacells.html#cb80-2" aria-hidden="true" tabindex="-1"></a>umap.metacell</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>We correctly retrieve the two main differentiation path of CD34+ cells.
In one hand, Megakaryocyte Erythrocyte Progenitor (MEP) fate with GATA2 and then GATA1 transcription factor expression.
On the other, Lymphoid Myeloid Multipotent Progenitor (LMPP) fate with SPI1 transcription factor expression.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="metacell2-metacells.html#cb81-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;CD34&quot;</span>,<span class="st">&quot;GATA2&quot;</span>,<span class="st">&quot;GATA1&quot;</span>,<span class="st">&quot;SPI1&quot;</span>)</span>
<span id="cb81-2"><a href="metacell2-metacells.html#cb81-2" aria-hidden="true" tabindex="-1"></a>marker.metacell.umaps <span class="ot">&lt;-</span> <span class="fu">lapply</span>(genes, <span class="at">FUN =</span> <span class="cf">function</span>(x){</span>
<span id="cb81-3"><a href="metacell2-metacells.html#cb81-3" aria-hidden="true" tabindex="-1"></a>  plot <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(cd34.metacell, <span class="at">features =</span> x)</span>
<span id="cb81-4"><a href="metacell2-metacells.html#cb81-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb81-5"><a href="metacell2-metacells.html#cb81-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-6"><a href="metacell2-metacells.html#cb81-6" aria-hidden="true" tabindex="-1"></a>marker.metacell.umaps</span>
<span id="cb81-7"><a href="metacell2-metacells.html#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [[1]]</span></span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-11-2.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-11-3.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[4]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-11-4.png" width="672" />
When analyzing metacells it’s a good idea to plot their size on these. We can do it with ggplot.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="metacell2-metacells.html#cb85-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(umap.metacell<span class="sc">$</span>data, cd34.metacell<span class="sc">$</span>size)</span>
<span id="cb85-2"><a href="metacell2-metacells.html#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(data)[<span class="fu">length</span>(<span class="fu">colnames</span>(data))] <span class="ot">&lt;-</span> <span class="st">&quot;size&quot;</span></span>
<span id="cb85-3"><a href="metacell2-metacells.html#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data,<span class="fu">aes</span>(<span class="at">x=</span> UMAP_1,<span class="at">y=</span>UMAP_2,<span class="at">color =</span> celltype)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">size=</span>size)) <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="metacell2-metacells.html#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(marker.metacell.umaps,<span class="at">FUN =</span> <span class="cf">function</span>(x){</span>
<span id="cb86-2"><a href="metacell2-metacells.html#cb86-2" aria-hidden="true" tabindex="-1"></a>  data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(x<span class="sc">$</span>data,cd34.metacell<span class="sc">$</span>size)</span>
<span id="cb86-3"><a href="metacell2-metacells.html#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(data)[<span class="fu">length</span>(<span class="fu">colnames</span>(data))] <span class="ot">&lt;-</span> <span class="st">&quot;size&quot;</span></span>
<span id="cb86-4"><a href="metacell2-metacells.html#cb86-4" aria-hidden="true" tabindex="-1"></a>  gene <span class="ot">&lt;-</span> <span class="fu">colnames</span>(data)[<span class="fu">length</span>(<span class="fu">colnames</span>(data))<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb86-5"><a href="metacell2-metacells.html#cb86-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(data,<span class="fu">aes</span>(<span class="at">x=</span> UMAP_1,<span class="at">y=</span>UMAP_2,<span class="at">size=</span>size,<span class="at">colour =</span> .data[[gene]])) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb86-6"><a href="metacell2-metacells.html#cb86-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span>  <span class="fu">c</span>(<span class="st">&quot;lightgrey&quot;</span>, <span class="st">&quot;blue&quot;</span>), </span>
<span id="cb86-7"><a href="metacell2-metacells.html#cb86-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">guide =</span> <span class="st">&quot;colorbar&quot;</span>)  <span class="sc">+</span></span>
<span id="cb86-8"><a href="metacell2-metacells.html#cb86-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(gene)</span>
<span id="cb86-9"><a href="metacell2-metacells.html#cb86-9" aria-hidden="true" tabindex="-1"></a>} )</span>
<span id="cb86-10"><a href="metacell2-metacells.html#cb86-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [[1]]</span></span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-13-2.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-13-3.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[4]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-13-4.png" width="672" />
## SuperCell metacells</p>
<p>We can do the same with the metacells obtained with SuperCell</p>
<p>We load the object obtained with MCAT</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="metacell2-metacells.html#cb90-1" aria-hidden="true" tabindex="-1"></a>cd34.supercell <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/CD34/SuperCell/mc_Seurat.rds&quot;</span>)</span>
<span id="cb90-2"><a href="metacell2-metacells.html#cb90-2" aria-hidden="true" tabindex="-1"></a>cd34.supercell</span>
<span id="cb90-3"><a href="metacell2-metacells.html#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb90-4"><a href="metacell2-metacells.html#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12464 features across 138 samples within 1 assay </span></span>
<span id="cb90-5"><a href="metacell2-metacells.html#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Active assay: RNA (12464 features, 0 variable features)</span></span></code></pre></div>
<p>We Normalize the supercells data, identify the HVGs, and we make a classical dimensionality reduction (first a PCA and then vizualisation with a UMAP).</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="metacell2-metacells.html#cb91-1" aria-hidden="true" tabindex="-1"></a>cd34.supercell <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(cd34.supercell)</span>
<span id="cb91-2"><a href="metacell2-metacells.html#cb91-2" aria-hidden="true" tabindex="-1"></a>cd34.supercell <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cd34.supercell)</span>
<span id="cb91-3"><a href="metacell2-metacells.html#cb91-3" aria-hidden="true" tabindex="-1"></a>cd34.supercell <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cd34.supercell)</span>
<span id="cb91-4"><a href="metacell2-metacells.html#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span id="cb91-5"><a href="metacell2-metacells.html#cb91-5" aria-hidden="true" tabindex="-1"></a>cd34.supercell <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cd34.supercell)</span>
<span id="cb91-6"><a href="metacell2-metacells.html#cb91-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 1 </span></span>
<span id="cb91-7"><a href="metacell2-metacells.html#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  TST, NECTIN1, PCLAF, BLVRB, CD36, S100A6, CCDC26, PPP1R14A, PMP22, ANK1 </span></span>
<span id="cb91-8"><a href="metacell2-metacells.html#cb91-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ECE1, KLF1, KEL, PLIN2, KCNH2, ADAMTS3, EPOR, CKS1B, HBD, CSF1 </span></span>
<span id="cb91-9"><a href="metacell2-metacells.html#cb91-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PRKAR2B, HES6, GATA1, H2AFX, CXADR, APOC1, CENPM, TFR2, APOE, CSF2RB </span></span>
<span id="cb91-10"><a href="metacell2-metacells.html#cb91-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  PRKCH, MIR99AHG, RBPMS, CD109, LINC01122, CRHBP, FAM169A, SCML4, ADGRG6, GUCY1A1 </span></span>
<span id="cb91-11"><a href="metacell2-metacells.html#cb91-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ABLIM1, NRIP1, MEIS1, SGCZ, NBEA, TMEM163, RORA, MAGI2, PRDM16, ROBO4 </span></span>
<span id="cb91-12"><a href="metacell2-metacells.html#cb91-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     FGD5, RABGAP1, RNF220, L3MBTL4, ITGA9, SPTBN1, PTPRM, PROM1, MPPED2, IL12RB2 </span></span>
<span id="cb91-13"><a href="metacell2-metacells.html#cb91-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 2 </span></span>
<span id="cb91-14"><a href="metacell2-metacells.html#cb91-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  PRKCE, GAS7, SPNS3, HDAC9, AFF3, CRYBG1, CYTH1, NEGR1, RUNX2, LRRFIP1 </span></span>
<span id="cb91-15"><a href="metacell2-metacells.html#cb91-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     MAP3K8, PDE4B, RASGEF1B, RAB11FIP1, FLT3, PRAG1, FAM107B, ITGAL, TIMP2, SDK2 </span></span>
<span id="cb91-16"><a href="metacell2-metacells.html#cb91-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     HLA-DMB, PLD4, IRF8, APP, LGALS1, AHNAK, ARL4C, PAG1, ZFP36, TSPOAP1 </span></span>
<span id="cb91-17"><a href="metacell2-metacells.html#cb91-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  STON2, TAL1, ZNF385D, MYH10, ALDH1A1, VPS37B, CDC42BPA, SLC40A1, DNAJC6, ST8SIA6 </span></span>
<span id="cb91-18"><a href="metacell2-metacells.html#cb91-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CYTL1, UROD, PHTF1, TNIK, ABO, NFIA, PIP5K1B, PDZD2, SORBS1, ZNF385D-AS2 </span></span>
<span id="cb91-19"><a href="metacell2-metacells.html#cb91-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     TRIM58, ICAM4, XACT, ZFPM1, GATA1, HNRNPLL, ARHGEF12, TFR2, RYR3, MINPP1 </span></span>
<span id="cb91-20"><a href="metacell2-metacells.html#cb91-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 3 </span></span>
<span id="cb91-21"><a href="metacell2-metacells.html#cb91-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  EXT1, RGS1, CARD11, MAPRE2, P2RY14, TMEM59, LGMN, JCHAIN, CCND3, SETBP1 </span></span>
<span id="cb91-22"><a href="metacell2-metacells.html#cb91-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CYTH4, CD37, IGF2R, ITM2B, HLA-DMA, ADAM19, LIME1, IL18R1, PMEPA1, HLA-DRA </span></span>
<span id="cb91-23"><a href="metacell2-metacells.html#cb91-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     UGCG, HLA-DPA1, RASD1, LINC01226, SLC9A7, CDKN2D, HLA-DPB1, SPIB, FAM160A1, TSBP1 </span></span>
<span id="cb91-24"><a href="metacell2-metacells.html#cb91-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  FNDC3B, RAB44, SLC22A4, AZU1, PLPPR3, RFX8, PLD1, RAB27A, LRMDA, ELANE </span></span>
<span id="cb91-25"><a href="metacell2-metacells.html#cb91-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PCBP3, HGF, CST7, NT5DC3, C16orf74, CTSG, MTUS2, SLC22A15, YBX3, CFD </span></span>
<span id="cb91-26"><a href="metacell2-metacells.html#cb91-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     MPO, IL1RAP, MS4A3, SLC1A3, AFF2, PRTN3, LYST, NR6A1, IGFBP2, CSF3R </span></span>
<span id="cb91-27"><a href="metacell2-metacells.html#cb91-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 4 </span></span>
<span id="cb91-28"><a href="metacell2-metacells.html#cb91-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  IGLL1, PHGDH, LGALS3BP, TSPAN7, SMIM24, PTPRD, NDC80, SKA3, MXD3, BAALC </span></span>
<span id="cb91-29"><a href="metacell2-metacells.html#cb91-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     SPC25, SPARC, CDKN3, FANCD2, KIF2C, FBXO43, ESCO2, RAD51AP1, CDCA5, CDK1 </span></span>
<span id="cb91-30"><a href="metacell2-metacells.html#cb91-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     PRDX1, ARHGAP11B, ITGA6, DEPDC1B, MYLK, BAALC-AS1, TPX2, MREG, MYO5C, SPC24 </span></span>
<span id="cb91-31"><a href="metacell2-metacells.html#cb91-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  KCNQ1, OSCAR, GRASP, ADAM8, ADAP1, FCGRT, FGD4, KDM7A, CTSH, CD86 </span></span>
<span id="cb91-32"><a href="metacell2-metacells.html#cb91-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     NR4A2, CD4, ENTPD1, LYZ, VDR, NAMPT, HCK, CSTA, ITGAX, PER1 </span></span>
<span id="cb91-33"><a href="metacell2-metacells.html#cb91-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     GABARAPL1, MNDA, FAM102B, CFP, ZNF503, TNFAIP2, PPARD, LRP1, NLRP12, RETN </span></span>
<span id="cb91-34"><a href="metacell2-metacells.html#cb91-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 5 </span></span>
<span id="cb91-35"><a href="metacell2-metacells.html#cb91-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  DNTT, LINC00426, GABPB1-AS1, SCN3A, VPREB1, LTB, LINC01237, ARPP21, MME, PCAT14 </span></span>
<span id="cb91-36"><a href="metacell2-metacells.html#cb91-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CD79A, CD96, ERGIC1, CYGB, IL7R, EBF1, LINC00173, CD79B, LINC01934, UMODL1 </span></span>
<span id="cb91-37"><a href="metacell2-metacells.html#cb91-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     SCAI, ATP1A3, LGALS3BP, C5orf56, RETREG1, SLC43A2, SIDT1, LINC01375, SPON1, COBL </span></span>
<span id="cb91-38"><a href="metacell2-metacells.html#cb91-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  AURKB, TOP2A, DEPDC1B, UBE2C, CDCA3, CKAP2L, ASPM, NUSAP1, KIF4A, KIF14 </span></span>
<span id="cb91-39"><a href="metacell2-metacells.html#cb91-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     KIFC1, PRC1, GTSE1, CIT, KIF18B, NUF2, CDK1, KIF15, KIF11, FOXM1 </span></span>
<span id="cb91-40"><a href="metacell2-metacells.html#cb91-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     BIRC5, CCNA2, CDC25C, MKI67, TUBB4B, CDCA8, TPX2, CENPE, KIF23, CDCA2</span></span>
<span id="cb91-41"><a href="metacell2-metacells.html#cb91-41" aria-hidden="true" tabindex="-1"></a>cd34.supercell <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(cd34.supercell,<span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>),<span class="at">min.dist =</span> <span class="fl">0.5</span>)</span>
<span id="cb91-42"><a href="metacell2-metacells.html#cb91-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 UMAP embedding parameters a = 0.583 b = 1.334</span></span>
<span id="cb91-43"><a href="metacell2-metacells.html#cb91-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Read 138 rows and found 50 numeric columns</span></span>
<span id="cb91-44"><a href="metacell2-metacells.html#cb91-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span id="cb91-45"><a href="metacell2-metacells.html#cb91-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span id="cb91-46"><a href="metacell2-metacells.html#cb91-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span id="cb91-47"><a href="metacell2-metacells.html#cb91-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span id="cb91-48"><a href="metacell2-metacells.html#cb91-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; **************************************************|</span></span>
<span id="cb91-49"><a href="metacell2-metacells.html#cb91-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Writing NN index file to temp file /tmp/35395114/RtmptFIlEg/file14f20112bb73de</span></span>
<span id="cb91-50"><a href="metacell2-metacells.html#cb91-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span id="cb91-51"><a href="metacell2-metacells.html#cb91-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Annoy recall = 100%</span></span>
<span id="cb91-52"><a href="metacell2-metacells.html#cb91-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span id="cb91-53"><a href="metacell2-metacells.html#cb91-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span id="cb91-54"><a href="metacell2-metacells.html#cb91-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:08 Commencing optimization for 500 epochs, with 4504 positive edges</span></span>
<span id="cb91-55"><a href="metacell2-metacells.html#cb91-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 12:15:09 Optimization finished</span></span></code></pre></div>
<p>Plot the results using Seurat.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="metacell2-metacells.html#cb92-1" aria-hidden="true" tabindex="-1"></a>umap.supercell <span class="ot">&lt;-</span> <span class="fu">UMAPPlot</span>(cd34.supercell,<span class="at">group.by =</span> <span class="st">&quot;celltype&quot;</span>,<span class="at">label =</span> T)</span>
<span id="cb92-2"><a href="metacell2-metacells.html#cb92-2" aria-hidden="true" tabindex="-1"></a>umap.supercell</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-16-1.png" width="672" />
We correctly retrieve the two main differentiation path of CD34+ cells.
In one hand, Megakaryocyte Erythrocyte Progenitor (MEP) fate with GATA2 and then GATA1 transcription factor expression.
On the other, Lymphoid Myeloid Multipotent Progenitor (LMPP) fate with SPI1 transcription factor expression.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="metacell2-metacells.html#cb93-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;CD34&quot;</span>,<span class="st">&quot;GATA2&quot;</span>,<span class="st">&quot;GATA1&quot;</span>,<span class="st">&quot;SPI1&quot;</span>)</span>
<span id="cb93-2"><a href="metacell2-metacells.html#cb93-2" aria-hidden="true" tabindex="-1"></a>marker.supercell.umaps <span class="ot">&lt;-</span> <span class="fu">lapply</span>(genes, <span class="at">FUN =</span> <span class="cf">function</span>(x){</span>
<span id="cb93-3"><a href="metacell2-metacells.html#cb93-3" aria-hidden="true" tabindex="-1"></a>  plot <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(cd34.supercell, <span class="at">features =</span> x)</span>
<span id="cb93-4"><a href="metacell2-metacells.html#cb93-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb93-5"><a href="metacell2-metacells.html#cb93-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb93-6"><a href="metacell2-metacells.html#cb93-6" aria-hidden="true" tabindex="-1"></a>marker.supercell.umaps</span>
<span id="cb93-7"><a href="metacell2-metacells.html#cb93-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [[1]]</span></span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-17-2.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-17-3.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[4]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-17-4.png" width="672" />
When analyzing supercells it’s a good idea to plot their size on these. We can do it with ggplot.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="metacell2-metacells.html#cb97-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(umap.supercell<span class="sc">$</span>data,cd34.supercell<span class="sc">$</span>size)</span>
<span id="cb97-2"><a href="metacell2-metacells.html#cb97-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(data)[<span class="fu">length</span>(<span class="fu">colnames</span>(data))] <span class="ot">&lt;-</span> <span class="st">&quot;size&quot;</span></span>
<span id="cb97-3"><a href="metacell2-metacells.html#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data,<span class="fu">aes</span>(<span class="at">x=</span> UMAP_1,<span class="at">y=</span>UMAP_2,<span class="at">color =</span> celltype)) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">size=</span>size)) <span class="sc">+</span> <span class="fu">theme_classic</span>()</span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="metacell2-metacells.html#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(marker.supercell.umaps,<span class="at">FUN =</span> <span class="cf">function</span>(x){</span>
<span id="cb98-2"><a href="metacell2-metacells.html#cb98-2" aria-hidden="true" tabindex="-1"></a>  data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(x<span class="sc">$</span>data,cd34.supercell<span class="sc">$</span>size)</span>
<span id="cb98-3"><a href="metacell2-metacells.html#cb98-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(data)[<span class="fu">length</span>(<span class="fu">colnames</span>(data))] <span class="ot">&lt;-</span> <span class="st">&quot;size&quot;</span></span>
<span id="cb98-4"><a href="metacell2-metacells.html#cb98-4" aria-hidden="true" tabindex="-1"></a>  gene <span class="ot">&lt;-</span> <span class="fu">colnames</span>(data)[<span class="fu">length</span>(<span class="fu">colnames</span>(data))<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb98-5"><a href="metacell2-metacells.html#cb98-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(data,<span class="fu">aes</span>(<span class="at">x=</span> UMAP_1,<span class="at">y=</span>UMAP_2,<span class="at">size=</span>size,<span class="at">colour =</span> .data[[gene]])) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb98-6"><a href="metacell2-metacells.html#cb98-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span>  <span class="fu">c</span>(<span class="st">&quot;lightgrey&quot;</span>, <span class="st">&quot;blue&quot;</span>), </span>
<span id="cb98-7"><a href="metacell2-metacells.html#cb98-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">guide =</span> <span class="st">&quot;colorbar&quot;</span>)  <span class="sc">+</span></span>
<span id="cb98-8"><a href="metacell2-metacells.html#cb98-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(gene)</span>
<span id="cb98-9"><a href="metacell2-metacells.html#cb98-9" aria-hidden="true" tabindex="-1"></a>} )</span>
<span id="cb98-10"><a href="metacell2-metacells.html#cb98-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [[1]]</span></span></code></pre></div>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-19-2.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-19-3.png" width="672" /></p>
<pre><code>#&gt; 
#&gt; [[4]]</code></pre>
<p><img src="23-downstream_continuous_files/figure-html/unnamed-chunk-19-4.png" width="672" /></p>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="short-downstream-analysis-of-the-metacells.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="quality-control-with-the-metacellanalysistoolkit-package.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/23-downstream_continuous.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/23-downstream_continuous.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
