<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 4 Metacells QCs | Metacell Analysis Tutorial</title>
<meta name="author" content="Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller">
<meta name="description" content="Different metrics have been proposed in previous metacell studies to evaluate the quality of metacells. We propose a R package called MetacellToolkit, to compute and visualize these metrics. The...">
<meta name="generator" content="bookdown 0.36 with bs4_book()">
<meta property="og:title" content="Chapter 4 Metacells QCs | Metacell Analysis Tutorial">
<meta property="og:type" content="book">
<meta property="og:url" content="https://GfellerLab.github.io/MetacellAnalysisTutorial/QCs.html">
<meta property="og:description" content="Different metrics have been proposed in previous metacell studies to evaluate the quality of metacells. We propose a R package called MetacellToolkit, to compute and visualize these metrics. The...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 4 Metacells QCs | Metacell Analysis Tutorial">
<meta name="twitter:description" content="Different metrics have been proposed in previous metacell studies to evaluate the quality of metacells. We propose a R package called MetacellToolkit, to compute and visualize these metrics. The...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Metacell Analysis Tutorial</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">This tutorial</a></li>
<li><a class="" href="requirements.html"><span class="header-section-number">1</span> Requirements</a></li>
<li><a class="" href="the-metacell-concept.html"><span class="header-section-number">2</span> The metacell concept</a></li>
<li><a class="" href="Metacell-construction-chapter.html"><span class="header-section-number">3</span> Constructing metacells</a></li>
<li><a class="active" href="QCs.html"><span class="header-section-number">4</span> Metacells QCs</a></li>
<li><a class="" href="downstream-analysis.html"><span class="header-section-number">5</span> Downstream analysis of metacells</a></li>
<li><a class="" href="command-line.html"><span class="header-section-number">6</span> Metacell Analysis Toolkit (MCAT)</a></li>
<li><a class="" href="integration.html"><span class="header-section-number">7</span> Integration of metacells</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/GfellerLab/MetacellAnalysisTutorial">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="QCs" class="section level1" number="4">
<h1>
<span class="header-section-number">4</span> Metacells QCs<a class="anchor" aria-label="anchor" href="#QCs"><i class="fas fa-link"></i></a>
</h1>
<p>Different metrics have been proposed in previous metacell studies to evaluate the quality of metacells.
We propose a R package called <em>MetacellToolkit</em>, to compute and visualize these metrics.
The package also provides a function to visualize metacells projected in the single-cell space.</p>
<p>Import packages:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if(system.file(package='MetacellToolkit') == ""){</span></span>
<span><span class="co">#   remotes::install_github("GfellerLab/MetacellAnalysisToolkit@dev_ag", force = TRUE, upgrade = FALSE)</span></span>
<span><span class="co"># } </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">MetacellAnalysisToolkit</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<p>To explore metacells QCs, we need to load:
(i) the single-cell data used to build the metacells and
(ii) the metacell data saved in a Seurat object (see chapter <a href="Metacell-construction-chapter.html#Metacell-construction-chapter">3</a>).</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the single-cell data </span></span>
<span><span class="va">sc_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/"</span>, <span class="va">proj_name</span>, <span class="st">"/singlecell_seurat_filtered.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Load the metacell data </span></span>
<span><span class="va">mc_data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'data/'</span>, <span class="va">proj_name</span>, <span class="st">'/metacell_'</span>, <span class="va">MC_tool</span>,<span class="st">'.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div id="quantitative-metrics" class="section level2" number="4.1">
<h2>
<span class="header-section-number">4.1</span> Quantitative metrics<a class="anchor" aria-label="anchor" href="#quantitative-metrics"><i class="fas fa-link"></i></a>
</h2>
<div id="purity" class="section level3" number="4.1.1">
<h3>
<span class="header-section-number">4.1.1</span> Purity<a class="anchor" aria-label="anchor" href="#purity"><i class="fas fa-link"></i></a>
</h3>
<p>When available, cell annotations can be used to annotate each metacell to the most abundant cell category (<em>e.g.</em> cell type) composing the metacell (see chapter <a href="Metacell-construction-chapter.html#Metacell-construction-chapter">3</a>).
This also allows us to compute metacell purity. If the annotation considered is the cell type, the <strong>purity</strong> of a metacell is the proportion of the most abundant
cell type within the metacell <span class="citation">(<a href="references.html#ref-SuperCell" role="doc-biblioref">Bilous et al. 2022</a>)</span>.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mc_data</span><span class="op">$</span><span class="va">purity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/mc_purity.html">mc_purity</a></span><span class="op">(</span>membership <span class="op">=</span> <span class="va">mc_data</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span><span class="op">$</span><span class="va">membership</span>, annotation <span class="op">=</span> <span class="va">sc_data</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_label</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"purity"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_purity-1.png" width="672"></div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"purity"</span>, split.by <span class="op">=</span> <span class="va">annotation_label</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_purity-2.png" width="672"></div>
</div>
<div id="compactness" class="section level3" number="4.1.2">
<h3>
<span class="header-section-number">4.1.2</span> Compactness<a class="anchor" aria-label="anchor" href="#compactness"><i class="fas fa-link"></i></a>
</h3>
<p>The <strong>compactness</strong> of a metacell is the variance of the components within the metacell <span class="citation">(<a href="references.html#ref-SEACells" role="doc-biblioref">Persad et al. 2023</a>)</span>.
The lower the compactness value the better.</p>
<p>This metric as well as the separation metric are computed based on a low embedding of the single-cell data (e.g., PCA).
Note that it is important to use the embedding used initially to construc the metacells.
In the next chunk, we retrieve the principal components computed for metacell construction
(in chapter <a href="Metacell-construction-chapter.html#Metacell-construction-chapter">3</a> these principal components were saved in the Seurat objects containing the metacell data)
and run UMAP for visualization.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_data</span><span class="op">@</span><span class="va">reductions</span><span class="op">[[</span><span class="st">"pca"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mc_data</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">sc.pca</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">sc_data</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>, n.neighbors <span class="op">=</span> <span class="fl">15</span>, verbose <span class="op">=</span> <span class="cn">F</span>, min.dist <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">UMAPPlot</a></span><span class="op">(</span><span class="va">sc_data</span>, group.by <span class="op">=</span> <span class="va">annotation_label</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/sc-embessing-1.png" width="672"></div>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">membership_df</span> <span class="op">&lt;-</span> <span class="va">mc_data</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span></span>
<span><span class="va">mc_data</span><span class="op">$</span><span class="va">compactness</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/mc_compactness.html">mc_compactness</a></span><span class="op">(</span>cell.membership <span class="op">=</span> <span class="va">membership_df</span>, sc.obj <span class="op">=</span> <span class="va">sc_data</span>,</span>
<span>                                      sc.reduction <span class="op">=</span> <span class="st">"pca"</span>, n.components <span class="op">=</span> <span class="fl">30</span>, diffusion.components <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing compactness ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"compactness"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 15 rows containing non-finite values (`stat_boxplot()`).</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_compactness-1.png" width="672"></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"compactness"</span>, split.by <span class="op">=</span> <span class="va">annotation_label</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 15 rows containing non-finite values (`stat_boxplot()`).</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_compactness-2.png" width="672"></div>
</div>
<div id="separation" class="section level3" number="4.1.3">
<h3>
<span class="header-section-number">4.1.3</span> Separation<a class="anchor" aria-label="anchor" href="#separation"><i class="fas fa-link"></i></a>
</h3>
<p>The <strong>separation</strong> of a metacell is the distance to the closest metacell <span class="citation">(<a href="references.html#ref-SEACells" role="doc-biblioref">Persad et al. 2023</a>)</span>.
The higher the separation value the better.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mc_data</span><span class="op">$</span><span class="va">separation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/mc_separation.html">mc_separation</a></span><span class="op">(</span>cell.membership <span class="op">=</span> <span class="va">membership_df</span>, sc.obj <span class="op">=</span> <span class="va">sc_data</span>, sc.reduction <span class="op">=</span> <span class="st">"pca"</span>, diffusion.components <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing separation ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"separation"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_separation-1.png" width="672"></div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"separation"</span>, split.by <span class="op">=</span> <span class="va">annotation_label</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_separation-2.png" width="672"></div>
<p>Note that compactness and separation metrics are correlated, better compactness results in worse separation and vice versa.
Metacells from dense regions will have better compactness but worse separation,
while metacells from sparse regions will have better separation but worse compactness.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>compactness <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">mc_data</span><span class="op">$</span><span class="va">compactness</span><span class="op">)</span>, separation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">mc_data</span><span class="op">$</span><span class="va">separation</span><span class="op">)</span><span class="op">)</span>, </span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">compactness</span>, y<span class="op">=</span><span class="va">separation</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method<span class="op">=</span><span class="va">lm</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_cor.html">stat_cor</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span>
<span><span class="co">#&gt; Warning: Removed 15 rows containing non-finite values (`stat_smooth()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 15 rows containing non-finite values (`stat_cor()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 15 rows containing missing values (`geom_point()`).</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/comparison_compactness_separation-1.png" width="672"></div>
</div>
<div id="inv" class="section level3" number="4.1.4">
<h3>
<span class="header-section-number">4.1.4</span> INV<a class="anchor" aria-label="anchor" href="#inv"><i class="fas fa-link"></i></a>
</h3>
<p>The <strong>inner normalized variance (INV)</strong> of a metacell is the mean-normalized variance of gene expression within the metacell <span class="citation">(<a href="references.html#ref-MC2" role="doc-biblioref">Ben-Kiki et al. 2022</a>)</span>.<br>
The lower the INV value the better.
Note that it is the only metric that is latent-space independent.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mc_data</span><span class="op">$</span><span class="va">INV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/mc_INV.html">mc_INV</a></span><span class="op">(</span>cell.membership <span class="op">=</span> <span class="va">membership_df</span>, sc.obj <span class="op">=</span> <span class="va">sc_data</span>, group.label <span class="op">=</span> <span class="st">"membership"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Counts and data slots are identical.</span></span>
<span><span class="co">#&gt; Normalizing data ...</span></span>
<span><span class="co">#&gt; Computing INV ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"INV"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 15 rows containing non-finite values (`stat_boxplot()`).</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_INV-1.png" width="672"></div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"INV"</span>, split.by <span class="op">=</span> <span class="va">annotation_label</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 15 rows containing non-finite values (`stat_boxplot()`).</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/compute_INV-2.png" width="672"></div>
</div>
</div>
<div id="size-distribution" class="section level2" number="4.2">
<h2>
<span class="header-section-number">4.2</span> Size distribution<a class="anchor" aria-label="anchor" href="#size-distribution"><i class="fas fa-link"></i></a>
</h2>
<p>The size of a metacell corresponds to the number of single cells it contains.
Having a homogeneous metacell size distribution is ideal for downstream analyses, since larger metacells will express more genes,
which could confound analyses. When heterogeneous size distributions are obtained we recommend weighted downstream analyses as described in section <a href="downstream-analysis.html#weighted-analysis">5.2</a>.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Seurat::VlnPlot(mc_data, features = "size", pt.size = 2)</span></span>
<span><span class="co"># Seurat::VlnPlot(mc_data, features = "size", pt.size = 2, group.by = annotation_label)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">mc_data</span><span class="op">$</span><span class="va">size</span>, main <span class="op">=</span> <span class="st">"Size distribution"</span>, xlab <span class="op">=</span> <span class="st">"Size"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/visualize_size_distribution-1.png" width="672"></div>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"size"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/visualize_size_distribution-2.png" width="672"></div>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/qc_boxplot.html">qc_boxplot</a></span><span class="op">(</span>mc.obj <span class="op">=</span> <span class="va">mc_data</span>, qc.metrics <span class="op">=</span> <span class="st">"size"</span>, split.by <span class="op">=</span> <span class="va">annotation_label</span><span class="op">)</span> </span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/visualize_size_distribution-3.png" width="672"></div>
</div>
<div id="representativeness-of-metacells" class="section level2" number="4.3">
<h2>
<span class="header-section-number">4.3</span> Representativeness of metacells<a class="anchor" aria-label="anchor" href="#representativeness-of-metacells"><i class="fas fa-link"></i></a>
</h2>
<p>To visualize the metacells, we can project the metacells on the single-cell UMAP representation using the <code><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/mc_projection.html">mc_projection()</a></code> function (adapted from the <code>plot.plot_2D()</code> from the SEACells package).
A good metacell partition should reproduce the overall structure of the single-cell data by uniformly representing the latent space.
To use this function we need the data at the single-cell level (or at least an low-dimensional embedding of the data) and the single-cell membership to each the metacell.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/mc_projection.html">mc_projection</a></span><span class="op">(</span></span>
<span>  sc.obj <span class="op">=</span> <span class="va">sc_data</span>, </span>
<span>  mc.obj <span class="op">=</span> <span class="va">mc_data</span>,</span>
<span>  cell.membership <span class="op">=</span> <span class="va">membership_df</span>,</span>
<span>  sc.reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>  sc.label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">annotation_label</span><span class="op">)</span>, <span class="co"># single cells will be colored according the sc.label  </span></span>
<span>  metacell.label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">annotation_label</span><span class="op">)</span> <span class="co"># metacells cell will be colored according the metacell.label</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/visualize_metacells-1.png" width="672"></div>
<p>By default the size of the metacells dots is proportionnal to the size of the metacells.
Metacells can also be colored by a continuous variable such as one of the QC metrics computed in the previous chunks:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MetacellAnalysisToolkit/man/mc_projection.html">mc_projection</a></span><span class="op">(</span></span>
<span>  sc.obj <span class="op">=</span> <span class="va">sc_data</span>, </span>
<span>  mc.obj <span class="op">=</span> <span class="va">mc_data</span>, </span>
<span>  cell.membership <span class="op">=</span> <span class="va">membership_df</span>,</span>
<span>  sc.reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>  sc.label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">annotation_label</span><span class="op">)</span>, <span class="co"># single cells will be colored according the sc.label  </span></span>
<span>  continuous_metric <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"compactness"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="22-Metacells_QCs_files/figure-html/visualize_metacells_continuous-1.png" width="672"></div>

<p>In this tutorial, standard analyses include dimensionality reduction, clustering and differential expression using the <a href="https://satijalab.org/seurat/">Seurat</a> framework.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span> </span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning this package</span></span>
<span><span class="co">#&gt; will retire shortly. Please refer to R-spatial evolution reports on</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html for details.</span></span>
<span><span class="co">#&gt; This package is now running under evolution status 0</span></span>
<span><span class="co">#&gt; Attaching SeuratObject</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div id="load-metacell-seurat-object" class="section level3" number="4.3.1">
<h3>
<span class="header-section-number">4.3.1</span> Load metacell Seurat object<a class="anchor" aria-label="anchor" href="#load-metacell-seurat-object"><i class="fas fa-link"></i></a>
</h3>
<p>We will use Seurat objects containing the metacells counts data and their annotation (<em>e.g.</em> cell-type annotation) and
proceed with standard Seurat downstream analyses.
Seurat objects containing metacells counts data and their annotation were generated at the end of sections <a href="Metacell-construction-chapter.html#SuperCell-construction">3.1</a>
These objects can also be generated using the command line described in chapter <a href="command-line.html#command-line">6</a></p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC_tool</span> <span class="op">=</span> <span class="st">"SuperCell"</span></span>
<span><span class="va">proj_name</span> <span class="op">=</span> <span class="st">"3k_pbmc"</span></span>
<span><span class="va">annotation_column</span> <span class="op">=</span> <span class="st">"louvain"</span></span>
<span><span class="va">celltype_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"CD14+ Monocytes"</span>    <span class="op">=</span> <span class="st">"#E69F00"</span>,  <span class="co"># orange</span></span>
<span>  <span class="st">"B cells"</span>            <span class="op">=</span> <span class="st">"#56B4E9"</span>,  <span class="co"># sky blue</span></span>
<span>  <span class="st">"CD4 T cells"</span>        <span class="op">=</span> <span class="st">"#009E73"</span>,  <span class="co"># bluish green</span></span>
<span>  <span class="st">"NK cells"</span>           <span class="op">=</span> <span class="st">"#F0E442"</span>,  <span class="co"># yellow</span></span>
<span>  <span class="st">"CD8 T cells"</span>        <span class="op">=</span> <span class="st">"#0072B2"</span>,  <span class="co"># blue</span></span>
<span>  <span class="st">"FCGR3A+ Monocytes"</span>  <span class="op">=</span> <span class="st">"#D55E00"</span>,  <span class="co"># vermillion</span></span>
<span>  <span class="st">"Dendritic cells"</span>    <span class="op">=</span> <span class="st">"#CC79A7"</span>,  <span class="co"># reddish purple</span></span>
<span>  <span class="st">"Megakaryocytes"</span>     <span class="op">=</span> <span class="st">"#000000"</span>   <span class="co"># black</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">MC.seurat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'./data/'</span>, <span class="va">proj_name</span>, <span class="st">'/metacell_'</span>, <span class="va">MC_tool</span>,<span class="st">'.rds'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="dimensionality-reduction" class="section level3" number="4.3.2">
<h3>
<span class="header-section-number">4.3.2</span> Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"><i class="fas fa-link"></i></a>
</h3>
<p>As for single-cells, we normalize the raw counts (here aggregated raw counts) and we identify the most variable features in the metacells gene expression data.
Based on these features, we run PCA and use the first principal components to obtain a two dimensionnal representation of the data using UMAP.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">annotation_column</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">MC.seurat</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">MC.seurat</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">MC.seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">MC.seurat</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">F</span>, min.dist <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html">Embeddings</a></span><span class="op">(</span><span class="va">MC.seurat</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">$</span><span class="va">size</span>,</span>
<span>                         cell_type <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_column</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span> <span class="va">UMAP_1</span>, y<span class="op">=</span><span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">cell_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size<span class="op">=</span><span class="va">size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,  <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">celltype_colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p_annot</span></span></code></pre></div>
<div class="inline-figure"><img src="30-downstream-R_files/figure-html/r-mc-dim-reduc-1.png" width="672"></div>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#DimPlot(MC.seurat, reduction = "umap", cols = celltype_colors, pt.size = log1p(MC.seurat$size)) </span></span></code></pre></div>
</div>
<div id="clustering" class="section level3" number="4.3.3">
<h3>
<span class="header-section-number">4.3.3</span> Clustering<a class="anchor" aria-label="anchor" href="#clustering"><i class="fas fa-link"></i></a>
</h3>
<p>We cluster the metacells using Seurat clustering steps and visualize these clusters using UMAP:</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MC.seurat</span><span class="op">$</span><span class="va">SCclustering</span>  <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_cluster.html">supercell_cluster</a></span><span class="op">(</span>D <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span>  <span class="op">)</span>, k <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">$</span><span class="va">clustering</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Embeddings.html">Embeddings</a></span><span class="op">(</span><span class="va">MC.seurat</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">$</span><span class="va">size</span>,</span>
<span>                         cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"C_"</span>,<span class="va">MC.seurat</span><span class="op">$</span><span class="va">SCclustering</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span> <span class="va">UMAP_1</span>, y<span class="op">=</span><span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size<span class="op">=</span><span class="va">size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span><span class="op">(</span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p_cluster</span></span></code></pre></div>
<div class="inline-figure"><img src="30-downstream-R_files/figure-html/r-mc-clustering-1.png" width="672"></div>
</div>
<div id="differential-expression-analysis" class="section level3" number="4.3.4">
<h3>
<span class="header-section-number">4.3.4</span> Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"><i class="fas fa-link"></i></a>
</h3>
<p>We perform diffrential analysis to identify the markers of our cluster 3 as an example using the <code>FindMarkers</code> function.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set idents to metacell annotation </span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">MC.seurat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCclustering"</span></span>
<span></span>
<span><span class="va">cells_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">MC.seurat</span>, ident.1 <span class="op">=</span> <span class="st">"4"</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; For a more efficient implementation of the Wilcoxon Rank Sum Test,</span></span>
<span><span class="co">#&gt; (default method for FindMarkers) please install the limma package</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; install.packages('BiocManager')</span></span>
<span><span class="co">#&gt; BiocManager::install('limma')</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; After installation of limma, Seurat will automatically use the more </span></span>
<span><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">)</span></span>
<span><span class="co">#&gt;                p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span><span class="co">#&gt; LIM2    1.188910e-27  0.4687254 0.467 0.000 3.892253e-23</span></span>
<span><span class="co">#&gt; FASLG   1.281776e-26  0.9488759 0.800 0.036 4.196277e-22</span></span>
<span><span class="co">#&gt; SH2D1B  2.757388e-25  1.4437575 0.933 0.072 9.027137e-21</span></span>
<span><span class="co">#&gt; KIR3DL2 1.480605e-23  1.1403146 0.867 0.068 4.847204e-19</span></span>
<span><span class="co">#&gt; KIR3DL1 3.907898e-23  0.5023381 0.533 0.012 1.279368e-18</span></span>
<span><span class="co">#&gt; PRSS23  6.410015e-22  1.7388849 0.933 0.092 2.098511e-17</span></span></code></pre></div>
<p>We see that the top marker genes for this cluster contain Killer cell immunoglobulin-like receptors (KIRs) genes coding for
transmembrane glycoproteins expressed by natural killer cells.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"KIR3DL2"</span>, <span class="st">"KIR3DL1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">MC.seurat</span>, <span class="va">genes</span>, ncol <span class="op">=</span> <span class="fl">2</span>, pt.size <span class="op">=</span> <span class="fl">0.0</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="30-downstream-R_files/figure-html/r-mc-plot-genes-1.png" width="672"></div>
<p>We can verify the identification of the NK cell cluster by comparing the metacell annotation and the metacell clustering.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_cluster</span> <span class="op">+</span> <span class="va">p_annot</span></span></code></pre></div>
<div class="inline-figure"><img src="30-downstream-R_files/figure-html/unnamed-chunk-2-1.png" width="768"></div>
</div>
<div id="visualize-gene-gene-correlation" class="section level3" number="4.3.5">
<h3>
<span class="header-section-number">4.3.5</span> Visualize gene-gene correlation<a class="anchor" aria-label="anchor" href="#visualize-gene-gene-correlation"><i class="fas fa-link"></i></a>
</h3>
<p>We can use the <code>supercell_GeneGenePlot</code> function from the SuperCell package to visualize the correlation between marker genes of a cell-type:
(i) at the single-cell level and
(ii) at the metacell level.</p>
<p>For that, we load the single-cell data from which the metacells were derived from.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">proj_name</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "3k_pbmc"</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/"</span>, <span class="va">proj_name</span>, <span class="st">"/singlecell_seurat_filtered.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">sc_data</span>, normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span><span class="op">)</span></span></code></pre></div>
<p>We visualize gene-gene correlation at the single-cell level:</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells_markers</span> <span class="op">&lt;-</span> <span class="va">cells_markers</span> <span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">$</span><span class="va">avg_log2FC</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">gene_x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>  </span>
<span><span class="va">gene_y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cells_markers</span><span class="op">)</span><span class="op">[</span><span class="fl">6</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.7</span></span>
<span></span>
<span><span class="va">p.sc</span> <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_GeneGenePlot.html">supercell_GeneGenePlot</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">sc_data</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>  gene_x <span class="op">=</span> <span class="va">gene_x</span>,</span>
<span>  gene_y <span class="op">=</span> <span class="va">gene_y</span>,</span>
<span>  clusters <span class="op">=</span> <span class="va">sc_data</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_column</span><span class="op">]</span>,</span>
<span>  sort.by.corr <span class="op">=</span> <span class="cn">F</span>, </span>
<span>  alpha <span class="op">=</span> <span class="va">alpha</span>, </span>
<span>  color.use <span class="op">=</span> <span class="va">celltype_colors</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p.sc</span><span class="op">$</span><span class="va">p</span></span></code></pre></div>
<div class="inline-figure"><img src="30-downstream-R_files/figure-html/r-sc-plot-genes-1.png" width="672"></div>
<p>We visualize gene-gene correlation at the metacell level:</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p.MC</span> <span class="op">&lt;-</span> <span class="fu">SuperCell</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SuperCell/man/supercell_GeneGenePlot.html">supercell_GeneGenePlot</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">MC.seurat</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>                                          gene_x <span class="op">=</span> <span class="va">gene_x</span>,</span>
<span>                                          gene_y <span class="op">=</span> <span class="va">gene_y</span>,</span>
<span>                                          clusters <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>, <span class="va">annotation_column</span><span class="op">]</span>,</span>
<span>                                          sort.by.corr <span class="op">=</span> <span class="cn">F</span>, supercell_size <span class="op">=</span> <span class="va">MC.seurat</span><span class="op">$</span><span class="va">size</span>,</span>
<span>                                          alpha <span class="op">=</span> <span class="va">alpha</span>,</span>
<span>                                          color.use <span class="op">=</span> <span class="va">celltype_colors</span><span class="op">)</span></span>
<span><span class="va">p.MC</span><span class="op">$</span><span class="va">p</span></span></code></pre></div>
<div class="inline-figure"><img src="30-downstream-R_files/figure-html/r-MC-plot-genes-1.png" width="672"></div>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="Metacell-construction-chapter.html"><span class="header-section-number">3</span> Constructing metacells</a></div>
<div class="next"><a href="downstream-analysis.html"><span class="header-section-number">5</span> Downstream analysis of metacells</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#QCs"><span class="header-section-number">4</span> Metacells QCs</a></li>
<li>
<a class="nav-link" href="#quantitative-metrics"><span class="header-section-number">4.1</span> Quantitative metrics</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#purity"><span class="header-section-number">4.1.1</span> Purity</a></li>
<li><a class="nav-link" href="#compactness"><span class="header-section-number">4.1.2</span> Compactness</a></li>
<li><a class="nav-link" href="#separation"><span class="header-section-number">4.1.3</span> Separation</a></li>
<li><a class="nav-link" href="#inv"><span class="header-section-number">4.1.4</span> INV</a></li>
</ul>
</li>
<li><a class="nav-link" href="#size-distribution"><span class="header-section-number">4.2</span> Size distribution</a></li>
<li>
<a class="nav-link" href="#representativeness-of-metacells"><span class="header-section-number">4.3</span> Representativeness of metacells</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#load-metacell-seurat-object"><span class="header-section-number">4.3.1</span> Load metacell Seurat object</a></li>
<li><a class="nav-link" href="#dimensionality-reduction"><span class="header-section-number">4.3.2</span> Dimensionality reduction</a></li>
<li><a class="nav-link" href="#clustering"><span class="header-section-number">4.3.3</span> Clustering</a></li>
<li><a class="nav-link" href="#differential-expression-analysis"><span class="header-section-number">4.3.4</span> Differential expression analysis</a></li>
<li><a class="nav-link" href="#visualize-gene-gene-correlation"><span class="header-section-number">4.3.5</span> Visualize gene-gene correlation</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/GfellerLab/MetacellAnalysisTutorial/blob/main/22-Metacells_QCs.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/GfellerLab/MetacellAnalysisTutorial/edit/main/22-Metacells_QCs.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Metacell Analysis Tutorial</strong>" was written by Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller. It was last built on 2023-11-22.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
