<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.1 Quantitative metrics | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="5.1 Quantitative metrics | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.1 Quantitative metrics | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-11-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="QCs.html"/>
<link rel="next" href="size-distribution.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> This tutorial</a></li>
<li class="chapter" data-level="2" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>2</b> Requirements</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>2.1</b> Installations</a></li>
<li class="chapter" data-level="2.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>2.2</b> Retrieve the discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="2.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>2.3</b> Retrieve the continuous dataset (CD34 dataset)</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>3</b> The metacell concept</a></li>
<li class="chapter" data-level="4" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>4</b> Constructing metacells (for ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="4.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>4.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#method"><i class="fa fa-check"></i><b>4.1.1</b> Method</a></li>
<li class="chapter" data-level="4.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>4.1.2</b> Data loading</a></li>
<li class="chapter" data-level="4.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>4.1.3</b> Filtering steps</a></li>
<li class="chapter" data-level="4.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>4.1.4</b> Building metacells</a></li>
<li class="chapter" data-level="4.1.5" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#annotate-metacells-using-available-annotations"><i class="fa fa-check"></i><b>4.1.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="4.1.6" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#save-output"><i class="fa fa-check"></i><b>4.1.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>4.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="MC2-construction.html"><a href="MC2-construction.html#method-1"><i class="fa fa-check"></i><b>4.2.1</b> Method</a></li>
<li class="chapter" data-level="4.2.2" data-path="MC2-construction.html"><a href="MC2-construction.html#data-loading-1"><i class="fa fa-check"></i><b>4.2.2</b> Data loading</a></li>
<li class="chapter" data-level="4.2.3" data-path="MC2-construction.html"><a href="MC2-construction.html#filtering-steps-1"><i class="fa fa-check"></i><b>4.2.3</b> Filtering steps</a></li>
<li class="chapter" data-level="4.2.4" data-path="MC2-construction.html"><a href="MC2-construction.html#building-metacells-1"><i class="fa fa-check"></i><b>4.2.4</b> Building metacells</a></li>
<li class="chapter" data-level="4.2.5" data-path="MC2-construction.html"><a href="MC2-construction.html#annotate-metacells-using-available-annotations-1"><i class="fa fa-check"></i><b>4.2.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="4.2.6" data-path="MC2-construction.html"><a href="MC2-construction.html#save-output-1"><i class="fa fa-check"></i><b>4.2.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>4.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="SEACells-construction.html"><a href="SEACells-construction.html#method-2"><i class="fa fa-check"></i><b>4.3.1</b> Method</a></li>
<li class="chapter" data-level="4.3.2" data-path="SEACells-construction.html"><a href="SEACells-construction.html#data-loading-2"><i class="fa fa-check"></i><b>4.3.2</b> Data loading</a></li>
<li class="chapter" data-level="4.3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html#filtering-steps-2"><i class="fa fa-check"></i><b>4.3.3</b> Filtering steps</a></li>
<li class="chapter" data-level="4.3.4" data-path="SEACells-construction.html"><a href="SEACells-construction.html#building-metacells-2"><i class="fa fa-check"></i><b>4.3.4</b> Building metacells</a></li>
<li class="chapter" data-level="4.3.5" data-path="SEACells-construction.html"><a href="SEACells-construction.html#visualize-metacells"><i class="fa fa-check"></i><b>4.3.5</b> Visualize metacells</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>4.4</b> Command line tool</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>5</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>5.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>5.1.1</b> Purity</a></li>
<li class="chapter" data-level="5.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>5.1.2</b> Compactness</a></li>
<li class="chapter" data-level="5.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>5.1.3</b> Separation</a></li>
<li class="chapter" data-level="5.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>5.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>5.2</b> Size distribution</a></li>
<li class="chapter" data-level="5.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>5.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>6</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>6.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>6.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>6.1.3</b> Clustering</a></li>
<li class="chapter" data-level="6.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="6.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>6.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>6.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>6.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering-1"><i class="fa fa-check"></i><b>6.2.2</b> Clustering</a></li>
<li class="chapter" data-level="6.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>6.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>6.3</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>6.3.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.3.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>6.3.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.3.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-2"><i class="fa fa-check"></i><b>6.3.3</b> Clustering</a></li>
<li class="chapter" data-level="6.3.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-2"><i class="fa fa-check"></i><b>6.3.4</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>6.4</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#GRN-analysis"><i class="fa fa-check"></i><b>6.4.1</b> GRN</a></li>
<li class="chapter" data-level="6.4.2" data-path="advanced-analysis.html"><a href="advanced-analysis.html#integration"><i class="fa fa-check"></i><b>6.4.2</b> Integration of metacells</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="quantitative-metrics" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Quantitative metrics<a href="quantitative-metrics.html#quantitative-metrics" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="purity" class="section level3 hasAnchor" number="5.1.1">
<h3><span class="header-section-number">5.1.1</span> Purity<a href="quantitative-metrics.html#purity" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>When available, cell annotations can be used to annotate each metacell to the most abundant cell category (<em>e.g.</em> cell type) composing the metacell (see chapter <a href="Metacell-construction-chapter.html#Metacell-construction-chapter">4</a>).
This also allows us to compute metacell purity. If the annotation considered is the cell type, the <strong>purity</strong> of a metacell is the proportion of the most abundant
cell type within the metacell <span class="citation">(<a href="#ref-SuperCell" role="doc-biblioref"><strong>SuperCell?</strong></a>)</span>.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="quantitative-metrics.html#cb57-1" aria-hidden="true" tabindex="-1"></a>mc_data<span class="sc">$</span>purity <span class="ot">&lt;-</span> <span class="fu">mc_purity</span>(<span class="at">membership =</span> mc_data<span class="sc">@</span>misc<span class="sc">$</span>cell_membership<span class="sc">$</span>membership, <span class="at">annotation =</span> sc_data<span class="sc">@</span>meta.data[, annotation_label])</span>
<span id="cb57-2"><a href="quantitative-metrics.html#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span id="cb57-3"><a href="quantitative-metrics.html#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;purity&quot;</span>)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_purity-1.png" width="672" /></p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="quantitative-metrics.html#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;purity&quot;</span>, <span class="at">split.by =</span> annotation_label)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_purity-2.png" width="672" /></p>
</div>
<div id="compactness" class="section level3 hasAnchor" number="5.1.2">
<h3><span class="header-section-number">5.1.2</span> Compactness<a href="quantitative-metrics.html#compactness" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <strong>compactness</strong> of a metacell is the variance of the components within the metacell <span class="citation">(<a href="#ref-SEACells" role="doc-biblioref"><strong>SEACells?</strong></a>)</span>.
The lower the compactness value the better.</p>
<p>This metric as well as the separation metric are computed based on a low embedding of the single-cell data (e.g., PCA).
Note that it is important to use the embedding used initially to construc the metacells.
In the next chunk, we retrieve the principal components computed for metacell construction
(in chapter <a href="Metacell-construction-chapter.html#Metacell-construction-chapter">4</a> these principal components were saved in the Seurat objects containing the metacell data)
and run UMAP for visualization.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="quantitative-metrics.html#cb59-1" aria-hidden="true" tabindex="-1"></a>sc_data<span class="sc">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]] <span class="ot">&lt;-</span> mc_data<span class="sc">@</span>misc<span class="sc">$</span>sc.pca</span>
<span id="cb59-2"><a href="quantitative-metrics.html#cb59-2" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sc_data, <span class="at">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>), <span class="at">n.neighbors =</span> <span class="dv">15</span>, <span class="at">verbose =</span> F, <span class="at">min.dist =</span> <span class="fl">0.5</span>)</span>
<span id="cb59-3"><a href="quantitative-metrics.html#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span id="cb59-4"><a href="quantitative-metrics.html#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;</span></span>
<span id="cb59-5"><a href="quantitative-metrics.html#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; This message will be shown once per session</span></span>
<span id="cb59-6"><a href="quantitative-metrics.html#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="fu">UMAPPlot</span>(sc_data, <span class="at">group.by =</span> annotation_label, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/sc-embessing-1.png" width="672" /></p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="quantitative-metrics.html#cb60-1" aria-hidden="true" tabindex="-1"></a>membership_df <span class="ot">&lt;-</span> mc_data<span class="sc">@</span>misc<span class="sc">$</span>cell_membership</span>
<span id="cb60-2"><a href="quantitative-metrics.html#cb60-2" aria-hidden="true" tabindex="-1"></a>mc_data<span class="sc">$</span>compactness <span class="ot">&lt;-</span> <span class="fu">mc_compactness</span>(<span class="at">cell.membership =</span> membership_df, <span class="at">sc.obj =</span> sc_data,</span>
<span id="cb60-3"><a href="quantitative-metrics.html#cb60-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">sc.reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">n.components =</span> <span class="dv">30</span>, <span class="at">diffusion.components =</span> T)</span>
<span id="cb60-4"><a href="quantitative-metrics.html#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing compactness ...</span></span>
<span id="cb60-5"><a href="quantitative-metrics.html#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;compactness&quot;</span>)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_compactness-1.png" width="672" /></p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="quantitative-metrics.html#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;compactness&quot;</span>, <span class="at">split.by =</span> annotation_label)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_compactness-2.png" width="672" /></p>
</div>
<div id="separation" class="section level3 hasAnchor" number="5.1.3">
<h3><span class="header-section-number">5.1.3</span> Separation<a href="quantitative-metrics.html#separation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <strong>separation</strong> of a metacell is the distance to the closest metacell <span class="citation">(<a href="#ref-SEACells" role="doc-biblioref"><strong>SEACells?</strong></a>)</span>.
The higher the separation value the better.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="quantitative-metrics.html#cb62-1" aria-hidden="true" tabindex="-1"></a>mc_data<span class="sc">$</span>separation <span class="ot">&lt;-</span> <span class="fu">mc_separation</span>(<span class="at">cell.membership =</span> membership_df, <span class="at">sc.obj =</span> sc_data, <span class="at">sc.reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">diffusion.components =</span> T)</span>
<span id="cb62-2"><a href="quantitative-metrics.html#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing separation ...</span></span>
<span id="cb62-3"><a href="quantitative-metrics.html#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;separation&quot;</span>)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_separation-1.png" width="672" /></p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="quantitative-metrics.html#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;separation&quot;</span>, <span class="at">split.by =</span> annotation_label)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_separation-2.png" width="672" /></p>
<p>Note that compactness and separation metrics are correlated, better compactness results in worse separation and vice versa.
Metacells from dense regions will have better compactness but worse separation,
while metacells from sparse regions will have better separation but worse compactness.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="quantitative-metrics.html#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb64-2"><a href="quantitative-metrics.html#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(<span class="at">compactness =</span> <span class="fu">log</span>(mc_data<span class="sc">$</span>compactness), <span class="at">separation =</span> <span class="fu">log</span>(mc_data<span class="sc">$</span>separation)), </span>
<span id="cb64-3"><a href="quantitative-metrics.html#cb64-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x=</span>compactness, <span class="at">y=</span>separation)) <span class="sc">+</span> </span>
<span id="cb64-4"><a href="quantitative-metrics.html#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb64-5"><a href="quantitative-metrics.html#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm) <span class="sc">+</span> ggpubr<span class="sc">::</span><span class="fu">stat_cor</span>(<span class="at">method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb64-6"><a href="quantitative-metrics.html#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; `geom_smooth()` using formula = &#39;y ~ x&#39;</span></span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/comparison_compactness_separation-1.png" width="672" /></p>
</div>
<div id="inv" class="section level3 hasAnchor" number="5.1.4">
<h3><span class="header-section-number">5.1.4</span> INV<a href="quantitative-metrics.html#inv" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <strong>inner normalized variance (INV)</strong> of a metacell is the mean-normalized variance of gene expression within the metacell <span class="citation">(<a href="#ref-MC2" role="doc-biblioref"><strong>MC2?</strong></a>)</span>.<br />
The lower the INV value the better.
Note that it is the only metric that is latent-space independent.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="quantitative-metrics.html#cb65-1" aria-hidden="true" tabindex="-1"></a>mc_data<span class="sc">$</span>INV <span class="ot">&lt;-</span> <span class="fu">mc_INV</span>(<span class="at">cell.membership =</span> membership_df, <span class="at">sc.obj =</span> sc_data, <span class="at">group.label =</span> <span class="st">&quot;membership&quot;</span>)</span>
<span id="cb65-2"><a href="quantitative-metrics.html#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Counts and data slots are identical.</span></span>
<span id="cb65-3"><a href="quantitative-metrics.html#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Normalizing data ...</span></span>
<span id="cb65-4"><a href="quantitative-metrics.html#cb65-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing INV ...</span></span>
<span id="cb65-5"><a href="quantitative-metrics.html#cb65-5" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;INV&quot;</span>)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_INV-1.png" width="672" /></p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="quantitative-metrics.html#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qc_boxplot</span>(<span class="at">mc.obj =</span> mc_data, <span class="at">qc.metrics =</span> <span class="st">&quot;INV&quot;</span>, <span class="at">split.by =</span> annotation_label)</span></code></pre></div>
<p><img src="22-Metacells_QCs_files/figure-html/compute_INV-2.png" width="672" /></p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="QCs.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="size-distribution.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/22-Metacells_QCs.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/22-Metacells_QCs.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
