<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.3 SEACells (Python) | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="3.3 SEACells (Python) | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.3 SEACells (Python) | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller" />


<meta name="date" content="2023-11-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="MC2-construction.html"/>
<link rel="next" href="command-line.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>This tutorial</a></li>
<li class="chapter" data-level="1" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>1</b> Requirements</a>
<ul>
<li class="chapter" data-level="1.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>1.1</b> Installations</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="installations.html"><a href="installations.html#using-conda-recommended"><i class="fa fa-check"></i><b>1.1.1</b> Using conda (recommended)</a></li>
<li class="chapter" data-level="1.1.2" data-path="installations.html"><a href="installations.html#without-conda"><i class="fa fa-check"></i><b>1.1.2</b> Without conda</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>1.2</b> Retrieve a discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="1.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>1.3</b> Retrieve a continuous dataset (CD34 dataset)</a></li>
<li class="chapter" data-level="1.4" data-path="HLCA-data.html"><a href="HLCA-data.html"><i class="fa fa-check"></i><b>1.4</b> Retrieve the lung atlas dataset</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="HLCA-data.html"><a href="HLCA-data.html#downloading-the-atlas"><i class="fa fa-check"></i><b>1.4.1</b> Downloading the atlas</a></li>
<li class="chapter" data-level="1.4.2" data-path="HLCA-data.html"><a href="HLCA-data.html#splitting-atlas-by-datasets"><i class="fa fa-check"></i><b>1.4.2</b> Splitting atlas by datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a></li>
<li class="chapter" data-level="3" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells</a>
<ul>
<li class="chapter" data-level="3.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#method"><i class="fa fa-check"></i><b>3.1.1</b> Method</a></li>
<li class="chapter" data-level="3.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>3.1.2</b> Data loading</a></li>
<li class="chapter" data-level="3.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>3.1.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.5" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#annotate-metacells-using-available-annotations"><i class="fa fa-check"></i><b>3.1.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.1.6" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#save-output"><i class="fa fa-check"></i><b>3.1.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="MC2-construction.html"><a href="MC2-construction.html#method-1"><i class="fa fa-check"></i><b>3.2.1</b> Method</a></li>
<li class="chapter" data-level="3.2.2" data-path="MC2-construction.html"><a href="MC2-construction.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.2</b> Data loading</a></li>
<li class="chapter" data-level="3.2.3" data-path="MC2-construction.html"><a href="MC2-construction.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.4" data-path="MC2-construction.html"><a href="MC2-construction.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="MC2-construction.html"><a href="MC2-construction.html#annotate-metacells-using-available-annotations-1"><i class="fa fa-check"></i><b>3.2.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.2.6" data-path="MC2-construction.html"><a href="MC2-construction.html#save-output-1"><i class="fa fa-check"></i><b>3.2.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="SEACells-construction.html"><a href="SEACells-construction.html#method-2"><i class="fa fa-check"></i><b>3.3.1</b> Method</a></li>
<li class="chapter" data-level="3.3.2" data-path="SEACells-construction.html"><a href="SEACells-construction.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.2</b> Data loading</a></li>
<li class="chapter" data-level="3.3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.4" data-path="SEACells-construction.html"><a href="SEACells-construction.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="SEACells-construction.html"><a href="SEACells-construction.html#visualize-metacells"><i class="fa fa-check"></i><b>3.3.5</b> Visualize metacells</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>3.4</b> Metacell Analysis Toolkit (MCAT)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>4</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="4.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>4.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>4.1.1</b> Purity</a></li>
<li class="chapter" data-level="4.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>4.1.2</b> Compactness</a></li>
<li class="chapter" data-level="4.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>4.1.3</b> Separation</a></li>
<li class="chapter" data-level="4.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>4.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>4.2</b> Size distribution</a></li>
<li class="chapter" data-level="4.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>4.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>5</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="5.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>5.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>5.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>5.1.3</b> Clustering</a></li>
<li class="chapter" data-level="5.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>5.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="5.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>5.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>5.2</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>5.2.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.2.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>5.2.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.2.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-1"><i class="fa fa-check"></i><b>5.2.3</b> Clustering</a></li>
<li class="chapter" data-level="5.2.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>5.2.4</b> Differential expression analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="integration.html"><a href="integration.html"><i class="fa fa-check"></i><b>6</b> Integration of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html"><i class="fa fa-check"></i><b>6.1</b> Unsupervised integration</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#data-loading-3"><i class="fa fa-check"></i><b>6.1.1</b> Data loading</a></li>
<li class="chapter" data-level="6.1.2" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#setting-up-the-environment"><i class="fa fa-check"></i><b>6.1.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="6.1.3" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#building-metacell"><i class="fa fa-check"></i><b>6.1.3</b> Building metacell</a></li>
<li class="chapter" data-level="6.1.4" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#loading-metacell-objects"><i class="fa fa-check"></i><b>6.1.4</b> Loading metacell objects</a></li>
<li class="chapter" data-level="6.1.5" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#merging-objects-and-basic-quality-control"><i class="fa fa-check"></i><b>6.1.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="6.1.6" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#unintegrated-analysis"><i class="fa fa-check"></i><b>6.1.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="6.1.7" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#seurat-integration"><i class="fa fa-check"></i><b>6.1.7</b> Seurat integration</a></li>
<li class="chapter" data-level="6.1.8" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#downstream-analysis"><i class="fa fa-check"></i><b>6.1.8</b> Downstream analysis</a></li>
<li class="chapter" data-level="6.1.9" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#conclusion"><i class="fa fa-check"></i><b>6.1.9</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="integration_supervised.html"><a href="integration_supervised.html"><i class="fa fa-check"></i><b>6.2</b> Supervised integration</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="integration_supervised.html"><a href="integration_supervised.html#data-loading-4"><i class="fa fa-check"></i><b>6.2.1</b> Data loading</a></li>
<li class="chapter" data-level="6.2.2" data-path="integration_supervised.html"><a href="integration_supervised.html#setting-up-the-environment-1"><i class="fa fa-check"></i><b>6.2.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="6.2.3" data-path="integration_supervised.html"><a href="integration_supervised.html#constructing-supervised-metacell"><i class="fa fa-check"></i><b>6.2.3</b> Constructing supervised metacell</a></li>
<li class="chapter" data-level="6.2.4" data-path="integration_supervised.html"><a href="integration_supervised.html#load-metacell-objects"><i class="fa fa-check"></i><b>6.2.4</b> Load metacell objects</a></li>
<li class="chapter" data-level="6.2.5" data-path="integration_supervised.html"><a href="integration_supervised.html#merging-objects-and-basic-quality-control-1"><i class="fa fa-check"></i><b>6.2.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="6.2.6" data-path="integration_supervised.html"><a href="integration_supervised.html#unintegrated-analysis-1"><i class="fa fa-check"></i><b>6.2.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="6.2.7" data-path="integration_supervised.html"><a href="integration_supervised.html#stacas-integration"><i class="fa fa-check"></i><b>6.2.7</b> STACAS integration</a></li>
<li class="chapter" data-level="6.2.8" data-path="integration_supervised.html"><a href="integration_supervised.html#comparison-with-unsupervised-analysis"><i class="fa fa-check"></i><b>6.2.8</b> Comparison with unsupervised analysis</a></li>
<li class="chapter" data-level="6.2.9" data-path="integration_supervised.html"><a href="integration_supervised.html#downstream-analysis-1"><i class="fa fa-check"></i><b>6.2.9</b> Downstream analysis</a></li>
<li class="chapter" data-level="6.2.10" data-path="integration_supervised.html"><a href="integration_supervised.html#conclusion-1"><i class="fa fa-check"></i><b>6.2.10</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="SEACells-construction" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> SEACells (Python)<a href="SEACells-construction.html#SEACells-construction" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this section, we construct metacells using <a href="https://github.com/dpeerlab/SEACells">SEACells</a>.</p>
<div id="method-2" class="section level3 hasAnchor" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Method<a href="SEACells-construction.html#method-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The SEAcells method builds a single-cell kNN graph from the Euclidean distance in the principal component space (SVD for scATAC-seq) space.
Distances in the graph are transformed to affinity by applying an adaptive Gaussian kernel.
The affinity matrix is then decomposed into archetypes (linear combination of cells) and membership matrices (cells as a linear combination of archetypes).
Single cells are assigned to a given metacell based on the maximum membership value of the corresponding archetype.</p>
<p>The code provided in this section is adapted from the <a href="https://github.com/dpeerlab/SEACells/blob/main/notebooks/SEACell_computation.ipynb">author’s jupyter notebook</a>.
For more information on the method, please refer to our review<span class="citation"><a href="#fn14" class="footnote-ref" id="fnref14"><sup>14</sup></a></span> and the original paper.<span class="citation"><a href="#fn15" class="footnote-ref" id="fnref15"><sup>15</sup></a></span></p>
<div id="importing-python-packages-1" class="section level4 unnumbered hasAnchor">
<h4>Importing python packages<a href="SEACells-construction.html#importing-python-packages-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To run the SEACells, the following python packages need to be imported:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb42-1"><a href="SEACells-construction.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb42-2"><a href="SEACells-construction.html#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb42-3"><a href="SEACells-construction.html#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb42-4"><a href="SEACells-construction.html#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> SEACells</span>
<span id="cb42-5"><a href="SEACells-construction.html#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random</span></code></pre></div>
<p>If you don’t have these packages installed, please refer to the section <a href="installations.html#installations">1.1</a>.</p>
</div>
</div>
<div id="data-loading-2" class="section level3 hasAnchor" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Data loading<a href="SEACells-construction.html#data-loading-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Similarly to SuperCell and MC2, we will run SEACells on the single-cell dataset composed of around 3000 peripheral blood mononuclear cells (PBMCs).
Please follow the section <a href="PBMC-data.html#PBMC-data">1.2</a> to retrieve these data from the scanpy package and save the data in the following file: “data/3k_pbmc/singlecell_anndata_filtered.h5ad”.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb43-1"><a href="SEACells-construction.html#cb43-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="op">=</span> <span class="st">&quot;SEACells&quot;</span></span>
<span id="cb43-2"><a href="SEACells-construction.html#cb43-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="op">=</span> <span class="st">&quot;3k_pbmc&quot;</span></span>
<span id="cb43-3"><a href="SEACells-construction.html#cb43-3" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> sc.read(os.path.join(<span class="st">&quot;data&quot;</span>, proj_name, <span class="st">&quot;singlecell_anndata_filtered.h5ad&quot;</span>))</span></code></pre></div>
</div>
<div id="filtering-steps-2" class="section level3 hasAnchor" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Filtering steps<a href="SEACells-construction.html#filtering-steps-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this tutorial, the data have been pre-filterd and SEACells does not perform additionnal filtering.</p>
</div>
<div id="building-metacells-2" class="section level3 hasAnchor" number="3.3.4">
<h3><span class="header-section-number">3.3.4</span> Building metacells<a href="SEACells-construction.html#building-metacells-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Metacells construction using SEACells requires 2 main inputs: i) an anndata object (<code>build_kernel_on</code> parameter), and
ii) a key indicating which matrix in the <code>obsm</code> attribute of the anndata object should be considered to compute the kernel needed for archetypal analysis (<code>build_kernel_on</code> parameter).
Important optional inputs are: the number of metacells to identify (<code>n_SEACells</code> parameter), which is used as input of the archetypal analysis,
ii) the number of neighbors to consider for the knn graph (<code>n_neighbors</code> parameter).</p>
<div id="data-pre-processing-1" class="section level4 unnumbered hasAnchor">
<h4>Data pre-processing<a href="SEACells-construction.html#data-pre-processing-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The following code chunk saves the raw counts of the filtered data in the raw attribute of the anndata object.
The raw counts will be used later for metacells aggregation.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb44-1"><a href="SEACells-construction.html#cb44-1" aria-hidden="true" tabindex="-1"></a>raw_ad <span class="op">=</span> sc.AnnData(ad.X)</span>
<span id="cb44-2"><a href="SEACells-construction.html#cb44-2" aria-hidden="true" tabindex="-1"></a>raw_ad.obs_names, raw_ad.var_names <span class="op">=</span> ad.obs_names, ad.var_names</span>
<span id="cb44-3"><a href="SEACells-construction.html#cb44-3" aria-hidden="true" tabindex="-1"></a>ad.raw <span class="op">=</span> raw_ad</span></code></pre></div>
<p>To build the kernel for archetypal analysis, SEACells requires a lower-dimensionnal embedding of the data (for example using PCA for scRNA-Seq data or SVD for scATAC-Seq data).
In the next code chunk, we follow standard pre-processing steps prior to PCA computation, <em>i.e.</em>, data normalization, log transformation, identification of highly variable genes.
PCA components are saved in the <code>obsm</code> attribute of the anndata object.</p>
<p>To pre-process the single-cell data, we are using standard pre-processing for single-cell RNA-seq data using Scanpy. For more information, see <a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">the Scanpy tutorial</a>.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb45-1"><a href="SEACells-construction.html#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize cells, log transform and compute highly variable genes</span></span>
<span id="cb45-2"><a href="SEACells-construction.html#cb45-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_per_cell(ad, <span class="dv">10000</span>)</span>
<span id="cb45-3"><a href="SEACells-construction.html#cb45-3" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(ad)</span>
<span id="cb45-4"><a href="SEACells-construction.html#cb45-4" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(ad, n_top_genes<span class="op">=</span><span class="dv">2000</span>)</span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb46-1"><a href="SEACells-construction.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute principal components - </span></span>
<span id="cb46-2"><a href="SEACells-construction.html#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="SEACells-construction.html#cb46-3" aria-hidden="true" tabindex="-1"></a>n_comp    <span class="op">=</span> <span class="dv">30</span></span>
<span id="cb46-4"><a href="SEACells-construction.html#cb46-4" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(ad, n_comps<span class="op">=</span>n_comp, use_highly_variable<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb46-5"><a href="SEACells-construction.html#cb46-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-6"><a href="SEACells-construction.html#cb46-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute UMAP for visualization</span></span>
<span id="cb46-7"><a href="SEACells-construction.html#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Here we use 30 components to be consistent with our main tutorial, but fill free to explore other number of principal components to use </span></span>
<span id="cb46-8"><a href="SEACells-construction.html#cb46-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-9"><a href="SEACells-construction.html#cb46-9" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(ad, n_neighbors<span class="op">=</span><span class="dv">15</span>, n_pcs<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb46-10"><a href="SEACells-construction.html#cb46-10" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(ad)</span></code></pre></div>
</div>
<div id="setting-up-seacells-parameters" class="section level4 unnumbered hasAnchor">
<h4>Setting up SEACells parameters<a href="SEACells-construction.html#setting-up-seacells-parameters" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In this tutorial, we will use in the SEACells model the 30 first principal components resulting from the PCA to build the knn graph which will be used to compute the kernel.
The number of neighbors to considered for the knn graph can be fixed using the <code>n_neighbors</code> parameter (here 15).<br />
As mentioned previously, users should provide as input the number of metacells required (<code>n_SEACells</code> parameter). This number can be defined as the ratio between the number of single cells and the desired graining level (<code>gamma</code> parameter in the following code chunk).
In this example, we choose a graining level of 25.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb47-1"><a href="SEACells-construction.html#cb47-1" aria-hidden="true" tabindex="-1"></a>build_kernel_on <span class="op">=</span> <span class="st">&#39;X_pca&#39;</span> <span class="co"># key in ad.obsm to use for computing metacells</span></span>
<span id="cb47-2"><a href="SEACells-construction.html#cb47-2" aria-hidden="true" tabindex="-1"></a>n_waypoint_eigs <span class="op">=</span> <span class="dv">10</span>      <span class="co"># Number of eigenvalues to consider when initializing metacells</span></span>
<span id="cb47-3"><a href="SEACells-construction.html#cb47-3" aria-hidden="true" tabindex="-1"></a>n_neighbors <span class="op">=</span> <span class="dv">15</span> <span class="co"># Number of neighbors used for graph construction </span></span>
<span id="cb47-4"><a href="SEACells-construction.html#cb47-4" aria-hidden="true" tabindex="-1"></a>gamma <span class="op">=</span> <span class="dv">25</span>   <span class="co"># the requested graining level</span></span>
<span id="cb47-5"><a href="SEACells-construction.html#cb47-5" aria-hidden="true" tabindex="-1"></a>n_SEACells <span class="op">=</span> <span class="bu">int</span>(ad.shape[<span class="dv">0</span>]<span class="op">/</span>gamma) <span class="co"># the requested number of metacells  </span></span></code></pre></div>
</div>
<div id="initializing-the-seacells-model" class="section level4 unnumbered hasAnchor">
<h4>Initializing the SEACells model<a href="SEACells-construction.html#initializing-the-seacells-model" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The SEACells model is initialized with the previously defined parameters using the <code>SEACells.core.SEACells</code> function.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb48-1"><a href="SEACells-construction.html#cb48-1" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> SEACells.core.SEACells(ad,</span>
<span id="cb48-2"><a href="SEACells-construction.html#cb48-2" aria-hidden="true" tabindex="-1"></a>                  build_kernel_on <span class="op">=</span> build_kernel_on,</span>
<span id="cb48-3"><a href="SEACells-construction.html#cb48-3" aria-hidden="true" tabindex="-1"></a>                  n_SEACells <span class="op">=</span> n_SEACells,</span>
<span id="cb48-4"><a href="SEACells-construction.html#cb48-4" aria-hidden="true" tabindex="-1"></a>                  n_waypoint_eigs <span class="op">=</span> n_waypoint_eigs,</span>
<span id="cb48-5"><a href="SEACells-construction.html#cb48-5" aria-hidden="true" tabindex="-1"></a>                  n_neighbors <span class="op">=</span> n_neighbors,</span>
<span id="cb48-6"><a href="SEACells-construction.html#cb48-6" aria-hidden="true" tabindex="-1"></a>                  convergence_epsilon <span class="op">=</span> <span class="fl">1e-5</span>,</span>
<span id="cb48-7"><a href="SEACells-construction.html#cb48-7" aria-hidden="true" tabindex="-1"></a>                  verbose <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb48-8"><a href="SEACells-construction.html#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Welcome to SEACells!</span></span></code></pre></div>
<p>Kernel computation is performed using the <code>mconstruct_kernel_matrix</code> function.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb49-1"><a href="SEACells-construction.html#cb49-1" aria-hidden="true" tabindex="-1"></a>model.construct_kernel_matrix()</span>
<span id="cb49-2"><a href="SEACells-construction.html#cb49-2" aria-hidden="true" tabindex="-1"></a>M <span class="op">=</span> model.kernel_matrix</span></code></pre></div>
<p>Metacells are initialized using the <code>initialize_archetypes</code> function.<br />
The SEACells archetypes initialization is based on cells sampling and thus is stochastic. User can fix a seed for reproducible results.
To check that the archetypes are evenly spread, users can visualize them using the <code>plot.plot_initialization</code> function.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb50-1"><a href="SEACells-construction.html#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set seed for reproducibility</span></span>
<span id="cb50-2"><a href="SEACells-construction.html#cb50-2" aria-hidden="true" tabindex="-1"></a>random.seed(<span class="dv">123</span>)</span>
<span id="cb50-3"><a href="SEACells-construction.html#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="SEACells-construction.html#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize archetypes</span></span>
<span id="cb50-5"><a href="SEACells-construction.html#cb50-5" aria-hidden="true" tabindex="-1"></a>model.initialize_archetypes()</span>
<span id="cb50-6"><a href="SEACells-construction.html#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Building kernel on X_pca</span></span>
<span id="cb50-7"><a href="SEACells-construction.html#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing diffusion components from X_pca for waypoint initialization ... </span></span>
<span id="cb50-8"><a href="SEACells-construction.html#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb50-9"><a href="SEACells-construction.html#cb50-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Sampling waypoints ...</span></span>
<span id="cb50-10"><a href="SEACells-construction.html#cb50-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Done.</span></span>
<span id="cb50-11"><a href="SEACells-construction.html#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Selecting 88 cells from waypoint initialization.</span></span>
<span id="cb50-12"><a href="SEACells-construction.html#cb50-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Initializing residual matrix using greedy column selection</span></span>
<span id="cb50-13"><a href="SEACells-construction.html#cb50-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Initializing f and g...</span></span>
<span id="cb50-14"><a href="SEACells-construction.html#cb50-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Selecting 17 cells from greedy initialization.</span></span>
<span id="cb50-15"><a href="SEACells-construction.html#cb50-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb50-16"><a href="SEACells-construction.html#cb50-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb50-17"><a href="SEACells-construction.html#cb50-17" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span><span class="op">%|</span>          <span class="op">|</span> <span class="dv">0</span><span class="op">/</span><span class="dv">27</span> [<span class="dv">00</span>:<span class="dv">00</span><span class="op">&lt;</span>?, ?it<span class="op">/</span>s]</span>
<span id="cb50-18"><a href="SEACells-construction.html#cb50-18" aria-hidden="true" tabindex="-1"></a><span class="dv">100</span><span class="op">%|</span><span class="co">##########| 27/27 [00:00&lt;00:00, 320.79it/s]</span></span>
<span id="cb50-19"><a href="SEACells-construction.html#cb50-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the initialization </span></span>
<span id="cb50-20"><a href="SEACells-construction.html#cb50-20" aria-hidden="true" tabindex="-1"></a>SEACells.plot.plot_initialization(ad, model, plot_basis<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>) </span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/seacells-model-init-1.png" width="672" /></p>
</div>
<div id="fitting-the-seacells-model-to-identify-metacells" class="section level4 unnumbered hasAnchor">
<h4>Fitting the SEACells model to identify metacells<a href="SEACells-construction.html#fitting-the-seacells-model-to-identify-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The identification of the archetypes is an iterative process. In this example, we fixed the minimum and maximum number of iteration to 10 and 50 respectively.
We then check the model convergence using the <code>plot_convergence</code> function.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb51-1"><a href="SEACells-construction.html#cb51-1" aria-hidden="true" tabindex="-1"></a>model.fit(min_iter <span class="op">=</span> <span class="dv">10</span>, max_iter <span class="op">=</span> <span class="dv">50</span>)</span>
<span id="cb51-2"><a href="SEACells-construction.html#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Randomly initialized A matrix.</span></span>
<span id="cb51-3"><a href="SEACells-construction.html#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Setting convergence threshold at 0.00088</span></span>
<span id="cb51-4"><a href="SEACells-construction.html#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 1.</span></span>
<span id="cb51-5"><a href="SEACells-construction.html#cb51-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 1.</span></span>
<span id="cb51-6"><a href="SEACells-construction.html#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 10.</span></span>
<span id="cb51-7"><a href="SEACells-construction.html#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 10.</span></span>
<span id="cb51-8"><a href="SEACells-construction.html#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 20.</span></span>
<span id="cb51-9"><a href="SEACells-construction.html#cb51-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 20.</span></span>
<span id="cb51-10"><a href="SEACells-construction.html#cb51-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 30.</span></span>
<span id="cb51-11"><a href="SEACells-construction.html#cb51-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 30.</span></span>
<span id="cb51-12"><a href="SEACells-construction.html#cb51-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Starting iteration 40.</span></span>
<span id="cb51-13"><a href="SEACells-construction.html#cb51-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Completed iteration 40.</span></span>
<span id="cb51-14"><a href="SEACells-construction.html#cb51-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Converged after 46 iterations.</span></span>
<span id="cb51-15"><a href="SEACells-construction.html#cb51-15" aria-hidden="true" tabindex="-1"></a>model.plot_convergence()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/seacells-model-fit-3.png" width="672" /></p>
<p>Once the final archetypes have been identified, we can assign each single-cell to one metacell (hard assignments).
These assignments (<code>membership</code>) can be retrieved using the <code>get_hard_assignments</code> function or extracted from the anndata object using <code>ad.obs["SEACell"]</code>.
In this tutorial, we will only consider hard assignments. However, the SEACells package also provides the option to retrieve soft assignments (multiple weighted assignments for each cell) using the <code>get_soft_assignments</code> function.
For more details on the soft assignments, please refer to the <a href="https://www.nature.com/articles/s41587-023-01716-9#Sec11">SEACell paper</a> and the original <a href="https://github.com/dpeerlab/SEACells/blob/main/notebooks/SEACell_computation.ipynb">author’s jupyter notebook</a>.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb52-1"><a href="SEACells-construction.html#cb52-1" aria-hidden="true" tabindex="-1"></a>membership <span class="op">=</span> model.get_hard_assignments()</span>
<span id="cb52-2"><a href="SEACells-construction.html#cb52-2" aria-hidden="true" tabindex="-1"></a>membership.head</span>
<span id="cb52-3"><a href="SEACells-construction.html#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;bound method NDFrame.head of                       SEACell</span></span>
<span id="cb52-4"><a href="SEACells-construction.html#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; index                        </span></span>
<span id="cb52-5"><a href="SEACells-construction.html#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATACAACCAC-1   SEACell-35</span></span>
<span id="cb52-6"><a href="SEACells-construction.html#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGAGCTAC-1   SEACell-69</span></span>
<span id="cb52-7"><a href="SEACells-construction.html#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGATCAGC-1    SEACell-0</span></span>
<span id="cb52-8"><a href="SEACells-construction.html#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGCTTCCG-1   SEACell-24</span></span>
<span id="cb52-9"><a href="SEACells-construction.html#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGTATGCG-1   SEACell-57</span></span>
<span id="cb52-10"><a href="SEACells-construction.html#cb52-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ...                       ...</span></span>
<span id="cb52-11"><a href="SEACells-construction.html#cb52-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCGAACTCTCAT-1    SEACell-1</span></span>
<span id="cb52-12"><a href="SEACells-construction.html#cb52-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTGAGGCA-1  SEACell-101</span></span>
<span id="cb52-13"><a href="SEACells-construction.html#cb52-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTTCCTCG-1   SEACell-30</span></span>
<span id="cb52-14"><a href="SEACells-construction.html#cb52-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGAGAGGC-1   SEACell-40</span></span>
<span id="cb52-15"><a href="SEACells-construction.html#cb52-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGCCTCAC-1   SEACell-42</span></span>
<span id="cb52-16"><a href="SEACells-construction.html#cb52-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb52-17"><a href="SEACells-construction.html#cb52-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2638 rows x 1 columns]&gt;</span></span>
<span id="cb52-18"><a href="SEACells-construction.html#cb52-18" aria-hidden="true" tabindex="-1"></a>ad.obs[<span class="st">&quot;SEACell&quot;</span>].head</span>
<span id="cb52-19"><a href="SEACells-construction.html#cb52-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; &lt;bound method NDFrame.head of index</span></span>
<span id="cb52-20"><a href="SEACells-construction.html#cb52-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATACAACCAC-1     SEACell-35</span></span>
<span id="cb52-21"><a href="SEACells-construction.html#cb52-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGAGCTAC-1     SEACell-69</span></span>
<span id="cb52-22"><a href="SEACells-construction.html#cb52-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACATTGATCAGC-1      SEACell-0</span></span>
<span id="cb52-23"><a href="SEACells-construction.html#cb52-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGCTTCCG-1     SEACell-24</span></span>
<span id="cb52-24"><a href="SEACells-construction.html#cb52-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; AAACCGTGTATGCG-1     SEACell-57</span></span>
<span id="cb52-25"><a href="SEACells-construction.html#cb52-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                        ...     </span></span>
<span id="cb52-26"><a href="SEACells-construction.html#cb52-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCGAACTCTCAT-1      SEACell-1</span></span>
<span id="cb52-27"><a href="SEACells-construction.html#cb52-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTGAGGCA-1    SEACell-101</span></span>
<span id="cb52-28"><a href="SEACells-construction.html#cb52-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTCTACTTCCTCG-1     SEACell-30</span></span>
<span id="cb52-29"><a href="SEACells-construction.html#cb52-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGAGAGGC-1     SEACell-40</span></span>
<span id="cb52-30"><a href="SEACells-construction.html#cb52-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; TTTGCATGCCTCAC-1     SEACell-42</span></span>
<span id="cb52-31"><a href="SEACells-construction.html#cb52-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Name: SEACell, Length: 2638, dtype: object&gt;</span></span></code></pre></div>
</div>
<div id="retrieve-aggregated-metacell-data-2" class="section level4 unnumbered hasAnchor">
<h4>Retrieve aggregated metacell data<a href="SEACells-construction.html#retrieve-aggregated-metacell-data-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The <code>core.summarize_by_SEACell</code> function can be used to generate a metacell count matrix (aggregation of counts across all cells belonging to each metacell).</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb53-1"><a href="SEACells-construction.html#cb53-1" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> SEACells.core.summarize_by_SEACell(ad, SEACells_label<span class="op">=</span><span class="st">&#39;SEACell&#39;</span>, summarize_layer<span class="op">=</span><span class="st">&#39;raw&#39;</span>, celltype_label<span class="op">=</span>annotation_label)</span>
<span id="cb53-2"><a href="SEACells-construction.html#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb53-3"><a href="SEACells-construction.html#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span><span class="op">%|</span>          <span class="op">|</span> <span class="dv">0</span><span class="op">/</span><span class="dv">105</span> [<span class="dv">00</span>:<span class="dv">00</span><span class="op">&lt;</span>?, ?it<span class="op">/</span>s]</span>
<span id="cb53-4"><a href="SEACells-construction.html#cb53-4" aria-hidden="true" tabindex="-1"></a> <span class="dv">56</span><span class="op">%|</span><span class="co">#####6    | 59/105 [00:00&lt;00:00, 587.31it/s]</span></span>
<span id="cb53-5"><a href="SEACells-construction.html#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="dv">100</span><span class="op">%|</span><span class="co">##########| 105/105 [00:00&lt;00:00, 606.52it/s]</span></span></code></pre></div>
</div>
<div id="annotate-metacells" class="section level4 unnumbered hasAnchor">
<h4>Annotate metacells<a href="SEACells-construction.html#annotate-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Note that providing an annotation to the <code>celltype_label</code> parameter in the <code>SEACells.core.summarize_by_SEACell</code> function
allowed us to annotate the metacells to the most common cell type in each metacell.</p>
</div>
</div>
<div id="visualize-metacells" class="section level3 hasAnchor" number="3.3.5">
<h3><span class="header-section-number">3.3.5</span> Visualize metacells<a href="SEACells-construction.html#visualize-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To visualize the metacells, we can project the metacells on the single-cell UMAP representation using the <code>plot.plot_2D</code> included in the SEACells package.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb54-1"><a href="SEACells-construction.html#cb54-1" aria-hidden="true" tabindex="-1"></a>SEACells.plot.plot_2D(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>)</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/seacells-umap-5.png" width="480" /></p>
<!-- The following code chunk adds a columns named `membership` and containing the single_cell assignments to the obs attribute in the anndata object containing the raw data. -->
<!-- This annotation will be used to compute metacells quality metrics. -->
<!-- ```{python seacells-membership, cache = TO_CACHE, warning = FALSE, message = FALSE} -->
<!-- # make `membership` numeric -->
<!-- d = {x: int(i)+1 for i, x in enumerate(mc_ad.obs_names)} -->
<!-- ad.obs.merge(membership) -->
<!-- ad.obs['membership'] = [d[x] for x in membership.SEACell] -->
<!-- ``` -->
<div id="save-output-2" class="section level5 unnumbered hasAnchor">
<h5>Save output<a href="SEACells-construction.html#save-output-2" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>For future downstream analyses in python (section <a href="#standard-analysis-Py"><strong>??</strong></a>), we save the metacell counts in an Anndata object:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb55-1"><a href="SEACells-construction.html#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Saving metacell object for the &quot;</span><span class="op">+</span> proj_name<span class="op">+</span> <span class="st">&quot; dataset using &quot;</span><span class="op">+</span> MC_tool)</span>
<span id="cb55-2"><a href="SEACells-construction.html#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Saving metacell object for the 3k_pbmc dataset using SEACells</span></span>
<span id="cb55-3"><a href="SEACells-construction.html#cb55-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-4"><a href="SEACells-construction.html#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Save metacell sizes </span></span>
<span id="cb55-5"><a href="SEACells-construction.html#cb55-5" aria-hidden="true" tabindex="-1"></a>label_df <span class="op">=</span> ad.obs[[<span class="st">&#39;SEACell&#39;</span>]].reset_index()</span>
<span id="cb55-6"><a href="SEACells-construction.html#cb55-6" aria-hidden="true" tabindex="-1"></a>mc_ad.obs <span class="op">=</span> mc_ad.obs.join(pd.DataFrame(label_df.groupby(<span class="st">&#39;SEACell&#39;</span>).count().iloc[:, <span class="dv">0</span>]).rename(columns<span class="op">=</span>{<span class="st">&#39;index&#39;</span>:<span class="st">&#39;size&#39;</span>}))</span>
<span id="cb55-7"><a href="SEACells-construction.html#cb55-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-8"><a href="SEACells-construction.html#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="co"># save pca used to compute metacells</span></span>
<span id="cb55-9"><a href="SEACells-construction.html#cb55-9" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;var_features&#39;</span>]<span class="op">=</span>ad.var_names[ad.var.highly_variable].tolist()</span>
<span id="cb55-10"><a href="SEACells-construction.html#cb55-10" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.pca&#39;</span>]<span class="op">=</span>ad.obsm[<span class="st">&#39;X_pca&#39;</span>] </span>
<span id="cb55-11"><a href="SEACells-construction.html#cb55-11" aria-hidden="true" tabindex="-1"></a>mc_ad.uns[<span class="st">&#39;sc.umap&#39;</span>]<span class="op">=</span>ad.obsm[<span class="st">&#39;X_umap&#39;</span>] </span>
<span id="cb55-12"><a href="SEACells-construction.html#cb55-12" aria-hidden="true" tabindex="-1"></a>mc_ad.write_h5ad(os.path.join(<span class="st">&#39;./data&#39;</span>, proj_name, <span class="ss">f&#39;metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad&#39;</span>))</span></code></pre></div>
<p>For future downstream analyses in R (section <a href="standard-analysis-R.html#standard-analysis-R">5.1</a>), we save the metacell counts in a Seurat object:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="SEACells-construction.html#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb56-2"><a href="SEACells-construction.html#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(anndata)</span>
<span id="cb56-3"><a href="SEACells-construction.html#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb56-4"><a href="SEACells-construction.html#cb56-4" aria-hidden="true" tabindex="-1"></a>adata_mc <span class="ot">&lt;-</span> <span class="fu">read_h5ad</span>(<span class="fu">paste0</span>(<span class="st">&quot;data/&quot;</span>, py<span class="sc">$</span>proj_name, <span class="st">&quot;/metacell_SEACells.h5ad&quot;</span>))</span>
<span id="cb56-5"><a href="SEACells-construction.html#cb56-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-6"><a href="SEACells-construction.html#cb56-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Save counts and metadata in a Seurat object</span></span>
<span id="cb56-7"><a href="SEACells-construction.html#cb56-7" aria-hidden="true" tabindex="-1"></a>countMatrix <span class="ot">&lt;-</span>  Matrix<span class="sc">::</span><span class="fu">t</span>(adata_mc<span class="sc">$</span>X)</span>
<span id="cb56-8"><a href="SEACells-construction.html#cb56-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(countMatrix) <span class="ot">&lt;-</span> adata_mc<span class="sc">$</span>obs_names</span>
<span id="cb56-9"><a href="SEACells-construction.html#cb56-9" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(countMatrix) <span class="ot">&lt;-</span> adata_mc<span class="sc">$</span>var_names</span>
<span id="cb56-10"><a href="SEACells-construction.html#cb56-10" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> <span class="fu">as</span>(countMatrix, <span class="st">&#39;CsparseMatrix&#39;</span>), <span class="at">meta.data =</span> <span class="fu">as.data.frame</span>(adata_mc<span class="sc">$</span>obs))</span>
<span id="cb56-11"><a href="SEACells-construction.html#cb56-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes</span></span>
<span id="cb56-12"><a href="SEACells-construction.html#cb56-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; (&#39;-&#39;)</span></span>
<span id="cb56-13"><a href="SEACells-construction.html#cb56-13" aria-hidden="true" tabindex="-1"></a><span class="co"># MC.seurat@misc[[&quot;sc.pca&quot;]] &lt;- adata_mc$uns$sc.pca</span></span>
<span id="cb56-14"><a href="SEACells-construction.html#cb56-14" aria-hidden="true" tabindex="-1"></a><span class="co"># MC.seurat@misc[[&quot;sc.umap&quot;]] &lt;- adata_mc$uns$sc.umap</span></span>
<span id="cb56-15"><a href="SEACells-construction.html#cb56-15" aria-hidden="true" tabindex="-1"></a>MC.seurat<span class="sc">@</span>misc[[<span class="st">&quot;var_features&quot;</span>]] <span class="ot">&lt;-</span> adata_mc<span class="sc">$</span>uns<span class="sc">$</span>var_features </span>
<span id="cb56-16"><a href="SEACells-construction.html#cb56-16" aria-hidden="true" tabindex="-1"></a>pca.res <span class="ot">&lt;-</span> adata_mc<span class="sc">$</span>uns<span class="sc">$</span>sc.pca</span>
<span id="cb56-17"><a href="SEACells-construction.html#cb56-17" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pca.res) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(py<span class="sc">$</span>ad<span class="sc">$</span>obs)</span>
<span id="cb56-18"><a href="SEACells-construction.html#cb56-18" aria-hidden="true" tabindex="-1"></a>MC.seurat<span class="sc">@</span>misc<span class="sc">$</span>sc.pca <span class="ot">&lt;-</span> <span class="fu">CreateDimReducObject</span>(</span>
<span id="cb56-19"><a href="SEACells-construction.html#cb56-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">embeddings =</span> pca.res,</span>
<span id="cb56-20"><a href="SEACells-construction.html#cb56-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">key =</span> <span class="st">&quot;PC_&quot;</span>,</span>
<span id="cb56-21"><a href="SEACells-construction.html#cb56-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">&quot;RNA&quot;</span></span>
<span id="cb56-22"><a href="SEACells-construction.html#cb56-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb56-23"><a href="SEACells-construction.html#cb56-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: No columnames present in cell embeddings, setting to &#39;PC_1:30&#39;</span></span>
<span id="cb56-24"><a href="SEACells-construction.html#cb56-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Save membership in misc</span></span>
<span id="cb56-25"><a href="SEACells-construction.html#cb56-25" aria-hidden="true" tabindex="-1"></a>MC.seurat<span class="sc">@</span>misc<span class="sc">$</span>cell_membership <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">row.names =</span> <span class="fu">rownames</span>(py<span class="sc">$</span>membership), <span class="at">membership =</span> py<span class="sc">$</span>membership<span class="sc">$</span>SEACell)</span>
<span id="cb56-26"><a href="SEACells-construction.html#cb56-26" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(MC.seurat, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&#39;./data/&#39;</span>, py<span class="sc">$</span>proj_name, <span class="st">&#39;/metacell_SEACells.rds&#39;</span>))</span></code></pre></div>
</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-SEACells" class="csl-entry">
Persad, Sitara, Zi-Ning Choo, Christine Dien, Noor Sohail, Ignas Masilionis, Ronan Chaligné, Tal Nawy, et al. <span>“<span>SEACells</span> Infers Transcriptional and Epigenomic Cellular States from Single-Cell Genomics Data.”</span> <em>Nature Biotechnology</em>, March 2023. <a href="https://doi.org/10.1038/s41587-023-01716-9">https://doi.org/10.1038/s41587-023-01716-9</a>.
</div>
</div>
<div class="footnotes">
<hr />
<ol start="14">
<li id="fn14"><p><a href="#ref-Review" role="doc-biblioref"><strong>Review?</strong></a><a href="SEACells-construction.html#fnref14" class="footnote-back">↩︎</a></p></li>
<li id="fn15"><p><a href="#ref-SEACells" role="doc-biblioref">Persad et al., <span>“<span>SEACells</span> Infers Transcriptional and Epigenomic Cellular States from Single-Cell Genomics Data”</span></a>.<a href="SEACells-construction.html#fnref15" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="MC2-construction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="command-line.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/21-MC_construction_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/21-MC_construction_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
