<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>6.1 Standard analysis (R) | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="6.1 Standard analysis (R) | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="6.1 Standard analysis (R) | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-11-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="downstream-analysis-of-metacells.html"/>
<link rel="next" href="standard-analysis-Py.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> This tutorial</a></li>
<li class="chapter" data-level="2" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>2</b> Requirements</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>2.1</b> Installations</a></li>
<li class="chapter" data-level="2.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>2.2</b> Retrieve the discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="2.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>2.3</b> Retrieve the continuous dataset (CD34 dataset)</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>3</b> The metacell concept</a></li>
<li class="chapter" data-level="4" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>4</b> Constructing metacells (for ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="4.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>4.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#method"><i class="fa fa-check"></i><b>4.1.1</b> Method</a></li>
<li class="chapter" data-level="4.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>4.1.2</b> Data loading</a></li>
<li class="chapter" data-level="4.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>4.1.3</b> Filtering steps</a></li>
<li class="chapter" data-level="4.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>4.1.4</b> Building metacells</a></li>
<li class="chapter" data-level="4.1.5" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#annotate-metacells-using-available-annotations"><i class="fa fa-check"></i><b>4.1.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="4.1.6" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#save-output"><i class="fa fa-check"></i><b>4.1.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>4.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="MC2-construction.html"><a href="MC2-construction.html#method-1"><i class="fa fa-check"></i><b>4.2.1</b> Method</a></li>
<li class="chapter" data-level="4.2.2" data-path="MC2-construction.html"><a href="MC2-construction.html#data-loading-1"><i class="fa fa-check"></i><b>4.2.2</b> Data loading</a></li>
<li class="chapter" data-level="4.2.3" data-path="MC2-construction.html"><a href="MC2-construction.html#filtering-steps-1"><i class="fa fa-check"></i><b>4.2.3</b> Filtering steps</a></li>
<li class="chapter" data-level="4.2.4" data-path="MC2-construction.html"><a href="MC2-construction.html#building-metacells-1"><i class="fa fa-check"></i><b>4.2.4</b> Building metacells</a></li>
<li class="chapter" data-level="4.2.5" data-path="MC2-construction.html"><a href="MC2-construction.html#annotate-metacells-using-available-annotations-1"><i class="fa fa-check"></i><b>4.2.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="4.2.6" data-path="MC2-construction.html"><a href="MC2-construction.html#save-output-1"><i class="fa fa-check"></i><b>4.2.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>4.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="SEACells-construction.html"><a href="SEACells-construction.html#method-2"><i class="fa fa-check"></i><b>4.3.1</b> Method</a></li>
<li class="chapter" data-level="4.3.2" data-path="SEACells-construction.html"><a href="SEACells-construction.html#data-loading-2"><i class="fa fa-check"></i><b>4.3.2</b> Data loading</a></li>
<li class="chapter" data-level="4.3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html#filtering-steps-2"><i class="fa fa-check"></i><b>4.3.3</b> Filtering steps</a></li>
<li class="chapter" data-level="4.3.4" data-path="SEACells-construction.html"><a href="SEACells-construction.html#building-metacells-2"><i class="fa fa-check"></i><b>4.3.4</b> Building metacells</a></li>
<li class="chapter" data-level="4.3.5" data-path="SEACells-construction.html"><a href="SEACells-construction.html#visualize-metacells"><i class="fa fa-check"></i><b>4.3.5</b> Visualize metacells</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>4.4</b> Command line tool</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>5</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>5.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>5.1.1</b> Purity</a></li>
<li class="chapter" data-level="5.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>5.1.2</b> Compactness</a></li>
<li class="chapter" data-level="5.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>5.1.3</b> Separation</a></li>
<li class="chapter" data-level="5.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>5.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>5.2</b> Size distribution</a></li>
<li class="chapter" data-level="5.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>5.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>6</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>6.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>6.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>6.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>6.1.3</b> Clustering</a></li>
<li class="chapter" data-level="6.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="6.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>6.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>6.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>6.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering-1"><i class="fa fa-check"></i><b>6.2.2</b> Clustering</a></li>
<li class="chapter" data-level="6.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>6.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>6.3</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>6.3.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="6.3.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>6.3.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="6.3.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-2"><i class="fa fa-check"></i><b>6.3.3</b> Clustering</a></li>
<li class="chapter" data-level="6.3.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-2"><i class="fa fa-check"></i><b>6.3.4</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>6.4</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#GRN-analysis"><i class="fa fa-check"></i><b>6.4.1</b> GRN</a></li>
<li class="chapter" data-level="6.4.2" data-path="advanced-analysis.html"><a href="advanced-analysis.html#integration"><i class="fa fa-check"></i><b>6.4.2</b> Integration of metacells</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="standard-analysis-R" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Standard analysis (R)<a href="standard-analysis-R.html#standard-analysis-R" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this tutorial, standard analysis includes dimensionality reduction, clustering and differential expression using the <a href="https://satijalab.org/seurat/">Seurat</a> framework.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="standard-analysis-R.html#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat) </span>
<span id="cb72-2"><a href="standard-analysis-R.html#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Attaching SeuratObject</span></span>
<span id="cb72-3"><a href="standard-analysis-R.html#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb72-4"><a href="standard-analysis-R.html#cb72-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb72-5"><a href="standard-analysis-R.html#cb72-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;dplyr&#39;</span></span>
<span id="cb72-6"><a href="standard-analysis-R.html#cb72-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb72-7"><a href="standard-analysis-R.html#cb72-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb72-8"><a href="standard-analysis-R.html#cb72-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     filter, lag</span></span>
<span id="cb72-9"><a href="standard-analysis-R.html#cb72-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb72-10"><a href="standard-analysis-R.html#cb72-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb72-11"><a href="standard-analysis-R.html#cb72-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span id="cb72-12"><a href="standard-analysis-R.html#cb72-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
<div id="load-metacell-seurat-object" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Load metacell Seurat object<a href="standard-analysis-R.html#load-metacell-seurat-object" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will use Seurat objects containing the metacells counts data and their annotation (<em>e.g.</em> and cell-type annotation) and
proceed with standard Seurat downstream analyses.
Seurat objects containing metacells counts data and their annotation were generated at the end of sections …..
These objects can also be generated using the command line described in chapter …</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="standard-analysis-R.html#cb73-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="ot">=</span> <span class="st">&quot;SuperCell&quot;</span></span>
<span id="cb73-2"><a href="standard-analysis-R.html#cb73-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="ot">=</span> <span class="st">&quot;3k_pbmc&quot;</span></span>
<span id="cb73-3"><a href="standard-analysis-R.html#cb73-3" aria-hidden="true" tabindex="-1"></a>annotation_column <span class="ot">=</span> <span class="st">&quot;louvain&quot;</span></span>
<span id="cb73-4"><a href="standard-analysis-R.html#cb73-4" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">&#39;./data/&#39;</span>, proj_name, <span class="st">&#39;/metacell_&#39;</span>, MC_tool,<span class="st">&#39;.rds&#39;</span>))</span></code></pre></div>
</div>
<div id="dimensionality-reduction" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Dimensionality reduction<a href="standard-analysis-R.html#dimensionality-reduction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As for single-cells, we normalize the raw counts (here aggregated raw counts) and we identify the most variable features in the metacells gene expression data.
Based on these features, we run PCA and use the first principal components to obtain a two dimensionnal representation of the data using UMAP.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="standard-analysis-R.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(MC.seurat) <span class="ot">&lt;-</span> annotation_column</span>
<span id="cb74-2"><a href="standard-analysis-R.html#cb74-2" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(MC.seurat, <span class="at">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>)</span>
<span id="cb74-3"><a href="standard-analysis-R.html#cb74-3" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(MC.seurat, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb74-4"><a href="standard-analysis-R.html#cb74-4" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(MC.seurat)</span>
<span id="cb74-5"><a href="standard-analysis-R.html#cb74-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span id="cb74-6"><a href="standard-analysis-R.html#cb74-6" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(MC.seurat, <span class="at">verbose =</span> F)</span>
<span id="cb74-7"><a href="standard-analysis-R.html#cb74-7" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(MC.seurat, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">verbose =</span> F)</span>
<span id="cb74-8"><a href="standard-analysis-R.html#cb74-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span id="cb74-9"><a href="standard-analysis-R.html#cb74-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;</span></span>
<span id="cb74-10"><a href="standard-analysis-R.html#cb74-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; This message will be shown once per session</span></span>
<span id="cb74-11"><a href="standard-analysis-R.html#cb74-11" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(MC.seurat, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>)</span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-dim-reduc-1.png" width="672" /></p>
</div>
<div id="clustering" class="section level3 hasAnchor" number="6.1.3">
<h3><span class="header-section-number">6.1.3</span> Clustering<a href="standard-analysis-R.html#clustering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We cluster the metacells using Seurat clustering steps and visualize these clusters using UMAP:</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="standard-analysis-R.html#cb75-1" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(MC.seurat, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb75-2"><a href="standard-analysis-R.html#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span id="cb75-3"><a href="standard-analysis-R.html#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing SNN</span></span>
<span id="cb75-4"><a href="standard-analysis-R.html#cb75-4" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(MC.seurat, <span class="at">resolution =</span> <span class="dv">2</span>)</span>
<span id="cb75-5"><a href="standard-analysis-R.html#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span id="cb75-6"><a href="standard-analysis-R.html#cb75-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb75-7"><a href="standard-analysis-R.html#cb75-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of nodes: 106</span></span>
<span id="cb75-8"><a href="standard-analysis-R.html#cb75-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of edges: 2637</span></span>
<span id="cb75-9"><a href="standard-analysis-R.html#cb75-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb75-10"><a href="standard-analysis-R.html#cb75-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span id="cb75-11"><a href="standard-analysis-R.html#cb75-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Maximum modularity in 10 random starts: 0.1672</span></span>
<span id="cb75-12"><a href="standard-analysis-R.html#cb75-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Number of communities: 6</span></span>
<span id="cb75-13"><a href="standard-analysis-R.html#cb75-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span id="cb75-14"><a href="standard-analysis-R.html#cb75-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 singletons identified. 5 final clusters.</span></span>
<span id="cb75-15"><a href="standard-analysis-R.html#cb75-15" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(MC.seurat, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;seurat_clusters&quot;</span>)</span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-clustering-1.png" width="672" /></p>
</div>
<div id="differential-expression-analysis" class="section level3 hasAnchor" number="6.1.4">
<h3><span class="header-section-number">6.1.4</span> Differential expression analysis<a href="standard-analysis-R.html#differential-expression-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="standard-analysis-R.html#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set idents to metacell annotation </span></span>
<span id="cb76-2"><a href="standard-analysis-R.html#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(MC.seurat) <span class="ot">&lt;-</span> annotation_column</span>
<span id="cb76-3"><a href="standard-analysis-R.html#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(MC.seurat) <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">levels</span>(<span class="fu">Idents</span>(MC.seurat)))</span>
<span id="cb76-4"><a href="standard-analysis-R.html#cb76-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-5"><a href="standard-analysis-R.html#cb76-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute upregulated genes in group of metacells (versus other metacells)</span></span>
<span id="cb76-6"><a href="standard-analysis-R.html#cb76-6" aria-hidden="true" tabindex="-1"></a>MC.seurat.all.markers <span class="ot">&lt;-</span>  <span class="fu">FindAllMarkers</span>(</span>
<span id="cb76-7"><a href="standard-analysis-R.html#cb76-7" aria-hidden="true" tabindex="-1"></a>  MC.seurat,</span>
<span id="cb76-8"><a href="standard-analysis-R.html#cb76-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb76-9"><a href="standard-analysis-R.html#cb76-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb76-10"><a href="standard-analysis-R.html#cb76-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>,</span>
<span id="cb76-11"><a href="standard-analysis-R.html#cb76-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">test.use =</span> <span class="st">&quot;t&quot;</span></span>
<span id="cb76-12"><a href="standard-analysis-R.html#cb76-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb76-13"><a href="standard-analysis-R.html#cb76-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster B cells</span></span>
<span id="cb76-14"><a href="standard-analysis-R.html#cb76-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster CD14+ Monocytes</span></span>
<span id="cb76-15"><a href="standard-analysis-R.html#cb76-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster CD4 T cells</span></span>
<span id="cb76-16"><a href="standard-analysis-R.html#cb76-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster CD8 T cells</span></span>
<span id="cb76-17"><a href="standard-analysis-R.html#cb76-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster Dendritic cells</span></span>
<span id="cb76-18"><a href="standard-analysis-R.html#cb76-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster FCGR3A+ Monocytes</span></span>
<span id="cb76-19"><a href="standard-analysis-R.html#cb76-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster Megakaryocytes</span></span>
<span id="cb76-20"><a href="standard-analysis-R.html#cb76-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Calculating cluster NK cells</span></span></code></pre></div>
<p>We select the top markers for each cell-type:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="standard-analysis-R.html#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Top markers (select top markers of each metacell population)</span></span>
<span id="cb77-2"><a href="standard-analysis-R.html#cb77-2" aria-hidden="true" tabindex="-1"></a>MC.seurat.top.markers <span class="ot">&lt;-</span> MC.seurat.all.markers <span class="sc">%&gt;%</span></span>
<span id="cb77-3"><a href="standard-analysis-R.html#cb77-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb77-4"><a href="standard-analysis-R.html#cb77-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">5</span>, <span class="at">order_by =</span> avg_log2FC)</span>
<span id="cb77-5"><a href="standard-analysis-R.html#cb77-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-6"><a href="standard-analysis-R.html#cb77-6" aria-hidden="true" tabindex="-1"></a>MC.seurat.top.markers</span>
<span id="cb77-7"><a href="standard-analysis-R.html#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # A tibble: 35 × 7</span></span>
<span id="cb77-8"><a href="standard-analysis-R.html#cb77-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # Groups:   cluster [7]</span></span>
<span id="cb77-9"><a href="standard-analysis-R.html#cb77-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster         gene  </span></span>
<span id="cb77-10"><a href="standard-analysis-R.html#cb77-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;           &lt;chr&gt; </span></span>
<span id="cb77-11"><a href="standard-analysis-R.html#cb77-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  1 6.34e-11       4.09     1 0.543  2.08e- 6 B cells         CD79A </span></span>
<span id="cb77-12"><a href="standard-analysis-R.html#cb77-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  2 3.56e- 5       3.70     1 0.223  1   e+ 0 B cells         IGLL5 </span></span>
<span id="cb77-13"><a href="standard-analysis-R.html#cb77-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  3 1.60e-12       3.40     1 0.755  5.24e- 8 B cells         CD79B </span></span>
<span id="cb77-14"><a href="standard-analysis-R.html#cb77-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  4 3.70e-19       3.36     1 0.606  1.21e-14 B cells         MS4A1 </span></span>
<span id="cb77-15"><a href="standard-analysis-R.html#cb77-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  5 2.16e- 6       3.12     1 0.383  7.08e- 2 B cells         TCL1A </span></span>
<span id="cb77-16"><a href="standard-analysis-R.html#cb77-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  6 1.28e-23       4.79     1 0.79   4.20e-19 CD14+ Monocytes S100A8</span></span>
<span id="cb77-17"><a href="standard-analysis-R.html#cb77-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  7 4.10e-36       4.51     1 0.926  1.34e-31 CD14+ Monocytes S100A9</span></span>
<span id="cb77-18"><a href="standard-analysis-R.html#cb77-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  8 4.35e-42       3.38     1 1      1.42e-37 CD14+ Monocytes LYZ   </span></span>
<span id="cb77-19"><a href="standard-analysis-R.html#cb77-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  9 2.36e-38       2.90     1 0.58   7.72e-34 CD14+ Monocytes LGALS2</span></span>
<span id="cb77-20"><a href="standard-analysis-R.html#cb77-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 10 2.26e-19       2.79     1 0.395  7.39e-15 CD14+ Monocytes CD14  </span></span>
<span id="cb77-21"><a href="standard-analysis-R.html#cb77-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; # ℹ 25 more rows</span></span></code></pre></div>
<p>We visualize the top 5 markers for the XX cells.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="standard-analysis-R.html#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># genes.to.plot &lt;- MC.seurat.top.markers$gene[MC.seurat.top.markers$cluster == unique(MC.seurat@meta.data[,annotation_column])[1]]</span></span>
<span id="cb78-2"><a href="standard-analysis-R.html#cb78-2" aria-hidden="true" tabindex="-1"></a>genes.to.plot <span class="ot">&lt;-</span> MC.seurat.top.markers<span class="sc">$</span>gene[<span class="fu">c</span>(<span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">20</span>, <span class="dv">5</span>))]</span>
<span id="cb78-3"><a href="standard-analysis-R.html#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(MC.seurat, <span class="at">features =</span> genes.to.plot, <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">pt.size =</span> <span class="fl">0.0</span>)  </span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-plot-genes-1.png" width="672" /></p>
</div>
<div id="visualize-gene-gene-correlation" class="section level3 hasAnchor" number="6.1.5">
<h3><span class="header-section-number">6.1.5</span> Visualize gene-gene correlation<a href="standard-analysis-R.html#visualize-gene-gene-correlation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can use the <code>supercell_GeneGenePlot</code> function from the SuperCell package to visualize the correlation between marker genes of a cell-type:
(i) at the single-cell level and
(ii) at the metacell level.</p>
<p>For that, we load the single-cell data from which the metacells were derived.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="standard-analysis-R.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(proj_name)</span>
<span id="cb79-2"><a href="standard-analysis-R.html#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;3k_pbmc&quot;</span></span>
<span id="cb79-3"><a href="standard-analysis-R.html#cb79-3" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">&quot;data/&quot;</span>, proj_name, <span class="st">&quot;/singlecell_seurat_filtered.rds&quot;</span>))</span>
<span id="cb79-4"><a href="standard-analysis-R.html#cb79-4" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(sc_data, <span class="at">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="standard-analysis-R.html#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SuperCell)</span>
<span id="cb80-2"><a href="standard-analysis-R.html#cb80-2" aria-hidden="true" tabindex="-1"></a>cell_type_markers <span class="ot">&lt;-</span> MC.seurat.top.markers[MC.seurat.top.markers<span class="sc">$</span>cluster <span class="sc">==</span> <span class="st">&quot;CD8 T cells&quot;</span>, ]</span>
<span id="cb80-3"><a href="standard-analysis-R.html#cb80-3" aria-hidden="true" tabindex="-1"></a>gene_x <span class="ot">&lt;-</span> cell_type_markers<span class="sc">$</span>gene[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]  </span>
<span id="cb80-4"><a href="standard-analysis-R.html#cb80-4" aria-hidden="true" tabindex="-1"></a>gene_y <span class="ot">&lt;-</span> cell_type_markers<span class="sc">$</span>gene[<span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>] </span>
<span id="cb80-5"><a href="standard-analysis-R.html#cb80-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-6"><a href="standard-analysis-R.html#cb80-6" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.7</span></span>
<span id="cb80-7"><a href="standard-analysis-R.html#cb80-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-8"><a href="standard-analysis-R.html#cb80-8" aria-hidden="true" tabindex="-1"></a>p.sc <span class="ot">&lt;-</span> <span class="fu">supercell_GeneGenePlot</span>(</span>
<span id="cb80-9"><a href="standard-analysis-R.html#cb80-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">GetAssayData</span>(sc_data, <span class="at">slot =</span> <span class="st">&quot;data&quot;</span>),</span>
<span id="cb80-10"><a href="standard-analysis-R.html#cb80-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_x =</span> gene_x,</span>
<span id="cb80-11"><a href="standard-analysis-R.html#cb80-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_y =</span> gene_y,</span>
<span id="cb80-12"><a href="standard-analysis-R.html#cb80-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">clusters =</span> sc_data<span class="sc">@</span>meta.data[, annotation_column],</span>
<span id="cb80-13"><a href="standard-analysis-R.html#cb80-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">sort.by.corr =</span> F,</span>
<span id="cb80-14"><a href="standard-analysis-R.html#cb80-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">alpha =</span> alpha</span>
<span id="cb80-15"><a href="standard-analysis-R.html#cb80-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb80-16"><a href="standard-analysis-R.html#cb80-16" aria-hidden="true" tabindex="-1"></a>p.sc<span class="sc">$</span>p</span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-sc-plot-genes-1.png" width="672" /></p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="standard-analysis-R.html#cb81-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-2"><a href="standard-analysis-R.html#cb81-2" aria-hidden="true" tabindex="-1"></a>p.MC <span class="ot">&lt;-</span> <span class="fu">supercell_GeneGenePlot</span>(<span class="fu">GetAssayData</span>(MC.seurat, <span class="at">slot =</span> <span class="st">&quot;data&quot;</span>), </span>
<span id="cb81-3"><a href="standard-analysis-R.html#cb81-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">gene_x =</span> gene_x, </span>
<span id="cb81-4"><a href="standard-analysis-R.html#cb81-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">gene_y =</span> gene_y, </span>
<span id="cb81-5"><a href="standard-analysis-R.html#cb81-5" aria-hidden="true" tabindex="-1"></a>                               <span class="at">clusters =</span> MC.seurat<span class="sc">@</span>meta.data[ ,annotation_column],  </span>
<span id="cb81-6"><a href="standard-analysis-R.html#cb81-6" aria-hidden="true" tabindex="-1"></a>                               <span class="at">sort.by.corr =</span> F, </span>
<span id="cb81-7"><a href="standard-analysis-R.html#cb81-7" aria-hidden="true" tabindex="-1"></a>                               <span class="at">alpha =</span> alpha)</span>
<span id="cb81-8"><a href="standard-analysis-R.html#cb81-8" aria-hidden="true" tabindex="-1"></a>p.MC<span class="sc">$</span>p</span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-MC-plot-genes-1.png" width="672" /></p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="downstream-analysis-of-metacells.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="standard-analysis-Py.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/30-MC_analysis_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/30-MC_analysis_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
