<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3.1 SuperCell (R) | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="3.1 SuperCell (R) | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3.1 SuperCell (R) | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Mariia Bilous, Léonard Hérault, Aurélie Gabriel, David Gfeller" />


<meta name="date" content="2023-08-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="constructing-metacells-for-discrete-data.html"/>
<link rel="next" href="mc2-python.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About</a>
<ul>
<li class="chapter" data-level="1.1" data-path="book-structure.html"><a href="book-structure.html"><i class="fa fa-check"></i><b>1.1</b> Book structure</a></li>
<li class="chapter" data-level="1.2" data-path="installation-and-requirements.html"><a href="installation-and-requirements.html"><i class="fa fa-check"></i><b>1.2</b> Installation and requirements</a></li>
<li class="chapter" data-level="1.3" data-path="render-book.html"><a href="render-book.html"><i class="fa fa-check"></i><b>1.3</b> Render book</a></li>
<li class="chapter" data-level="1.4" data-path="get-data.html"><a href="get-data.html"><i class="fa fa-check"></i><b>1.4</b> Get data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a>
<ul>
<li class="chapter" data-level="2.1" data-path="MC2.html"><a href="MC2.html"><i class="fa fa-check"></i><b>2.1</b> Metacell (MC2)</a></li>
<li class="chapter" data-level="2.2" data-path="SuperCell.html"><a href="SuperCell.html"><i class="fa fa-check"></i><b>2.2</b> SuperCell</a></li>
<li class="chapter" data-level="2.3" data-path="SEACells.html"><a href="SEACells.html"><i class="fa fa-check"></i><b>2.3</b> SEACells</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="constructing-metacells-for-discrete-data.html"><a href="constructing-metacells-for-discrete-data.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells (for ‘discrete’ data)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="supercell-r.html"><a href="supercell-r.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="supercell-r.html"><a href="supercell-r.html#data-loading"><i class="fa fa-check"></i><b>3.1.1</b> Data loading</a></li>
<li class="chapter" data-level="3.1.2" data-path="supercell-r.html"><a href="supercell-r.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.3" data-path="supercell-r.html"><a href="supercell-r.html#building-metacells"><i class="fa fa-check"></i><b>3.1.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.4" data-path="supercell-r.html"><a href="supercell-r.html#visualize-the-supercell-graph"><i class="fa fa-check"></i><b>3.1.4</b> Visualize the SuperCell graph</a></li>
<li class="chapter" data-level="3.1.5" data-path="supercell-r.html"><a href="supercell-r.html#metacell-qc"><i class="fa fa-check"></i><b>3.1.5</b> Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="mc2-python.html"><a href="mc2-python.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="mc2-python.html"><a href="mc2-python.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.1</b> Data loading</a></li>
<li class="chapter" data-level="3.2.2" data-path="mc2-python.html"><a href="mc2-python.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.3" data-path="mc2-python.html"><a href="mc2-python.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.4" data-path="mc2-python.html"><a href="mc2-python.html#visualize-metacells"><i class="fa fa-check"></i><b>3.2.4</b> Visualize metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="mc2-python.html"><a href="mc2-python.html#metacell-qc-1"><i class="fa fa-check"></i><b>3.2.5</b> Metacell QC</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="seacells-python.html"><a href="seacells-python.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="seacells-python.html"><a href="seacells-python.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.1</b> Data loading</a></li>
<li class="chapter" data-level="3.3.2" data-path="seacells-python.html"><a href="seacells-python.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.2</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.3" data-path="seacells-python.html"><a href="seacells-python.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.3</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.4" data-path="seacells-python.html"><a href="seacells-python.html#visualize-metacells-1"><i class="fa fa-check"></i><b>3.3.4</b> Visualize metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="seacells-python.html"><a href="seacells-python.html#metacell-qc-2"><i class="fa fa-check"></i><b>3.3.5</b> Metacell QC</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="command-line-tool-for-mc-construction.html"><a href="command-line-tool-for-mc-construction.html"><i class="fa fa-check"></i><b>4</b> Command line tool for MC construction</a></li>
<li class="chapter" data-level="5" data-path="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><a href="downstream-analysis-of-metacells-for-a-discrete-dataset.html"><i class="fa fa-check"></i><b>5</b> Downstream analysis of metacells (for a discrete dataset)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>5.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>5.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>5.1.3</b> Clustering</a></li>
<li class="chapter" data-level="5.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>5.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="5.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#plot-the-expression-of-some-markers"><i class="fa fa-check"></i><b>5.1.5</b> Plot the expression of some markers</a></li>
<li class="chapter" data-level="5.1.6" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#plot-gene-gene-correlation-at-single-cell-and-metacell-levels"><i class="fa fa-check"></i><b>5.1.6</b> Plot gene-gene correlation at single-cell and metacell levels</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html"><i class="fa fa-check"></i><b>5.2</b> Standard analysis (Python)</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>5.2.1</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.2.2" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#clustering-1"><i class="fa fa-check"></i><b>5.2.2</b> Clustering</a></li>
<li class="chapter" data-level="5.2.3" data-path="standard-analysis-Py.html"><a href="standard-analysis-Py.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>5.2.3</b> Differential expression analysis</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>5.3</b> Sample-weighted analysis</a></li>
<li class="chapter" data-level="5.4" data-path="advanced-analysis.html"><a href="advanced-analysis.html"><i class="fa fa-check"></i><b>5.4</b> Advanced analysis</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="advanced-analysis.html"><a href="advanced-analysis.html#GRN-analysis"><i class="fa fa-check"></i><b>5.4.1</b> GRN</a></li>
<li class="chapter" data-level="5.4.2" data-path="advanced-analysis.html"><a href="advanced-analysis.html#integration"><i class="fa fa-check"></i><b>5.4.2</b> Integration of metacells</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><a href="comparison-of-metacells-usage-for-a-discrete-and-continuous-data.html"><i class="fa fa-check"></i><b>6</b> Comparison of metacells usage for a discrete and continuous data</a>
<ul>
<li class="chapter" data-level="6.1" data-path="continuous-metacells-have-lower-purity.html"><a href="continuous-metacells-have-lower-purity.html"><i class="fa fa-check"></i><b>6.1</b> Continuous metacells have lower purity</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="supercell-r" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> SuperCell (R)<a href="supercell-r.html#supercell-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this section, we construct metacells using <a href="https://github.com/GfellerLab/SuperCell">SuperCell</a>. The code is adapted from the <a href="https://github.com/GfellerLab/SuperCell/blob/master/README.Rmd">author’s github documentation</a>.
For more information on the method, please refer to the section <a href="SuperCell.html#SuperCell">2.2</a>.</p>
<div id="importing-r-packages" class="section level4 unnumbered hasAnchor">
<h4>Importing R packages<a href="supercell-r.html#importing-r-packages" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>To run SuperCell, the following python packages need to be imported:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="supercell-r.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">system.file</span>(<span class="at">package=</span><span class="st">&#39;SuperCell&#39;</span>) <span class="sc">==</span> <span class="st">&quot;&quot;</span>){</span>
<span id="cb9-2"><a href="supercell-r.html#cb9-2" aria-hidden="true" tabindex="-1"></a>  remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;GfellerLab/SuperCell&quot;</span>, <span class="at">force =</span> <span class="cn">TRUE</span>, <span class="at">upgrade =</span> <span class="cn">FALSE</span>)</span>
<span id="cb9-3"><a href="supercell-r.html#cb9-3" aria-hidden="true" tabindex="-1"></a>} </span>
<span id="cb9-4"><a href="supercell-r.html#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SuperCell)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb10-1"><a href="supercell-r.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> sys</span>
<span id="cb10-2"><a href="supercell-r.html#cb10-2" aria-hidden="true" tabindex="-1"></a>sys.path.append(<span class="st">&#39;./mc_QC/&#39;</span>)</span>
<span id="cb10-3"><a href="supercell-r.html#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> mc_QC</span>
<span id="cb10-4"><a href="supercell-r.html#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb10-5"><a href="supercell-r.html#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb10-6"><a href="supercell-r.html#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> SEACells</span></code></pre></div>
<p>If you don’t have these packages installed, please refer to the section 2 of chapter 1.</p>
</div>
<div id="data-loading" class="section level3 hasAnchor" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Data loading<a href="supercell-r.html#data-loading" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Similarly to Metacell-2 and SEACells, we will run SuperCell on the single-cell dataset composed of XX peripheral blood mononuclear cells (PBMCs).
Please follow the section 4 from Chapter 1 to retrieve these data from the scanpy package and save the data in the following file: “data/3k_pbmc/singlecell_anndata_filtered.h5ad”.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="supercell-r.html#cb11-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="ot">=</span> <span class="st">&quot;SuperCell&quot;</span></span>
<span id="cb11-2"><a href="supercell-r.html#cb11-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="ot">=</span> <span class="st">&quot;3k_pbmc&quot;</span></span>
<span id="cb11-3"><a href="supercell-r.html#cb11-3" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">&quot;data/&quot;</span>, proj_name, <span class="st">&quot;/singlecell_seurat_filtered.rds&quot;</span>))</span></code></pre></div>
</div>
<div id="filtering-steps" class="section level3 hasAnchor" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Filtering steps<a href="supercell-r.html#filtering-steps" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this tutorial, the data have been pre-filtered in the section XX of chapter XX.</p>
</div>
<div id="building-metacells" class="section level3 hasAnchor" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Building metacells<a href="supercell-r.html#building-metacells" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Metacells construction using SuperCell requires one main inputs, <em>i.e.</em> a matrix of log-normalized gene expression data which will be used to compute PCA to build a knn graph for metacells identification.
Important optional inputs are: i) the graining level (<code>gamma</code> parameter),
ii) the number of neighbors to consider for the knn graph (<code>k.knn</code> parameter).</p>
<div id="data-pre-processing" class="section level4 unnumbered hasAnchor">
<h4>Data pre-processing<a href="supercell-r.html#data-pre-processing" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>SuperCell builds a knn graph based on a lower-dimensional embedding of the data. The computation of this embedding is performed internally in the <code>SCimplify</code> SuperCell function.<br />
In the following code chunk, we normalize the raw counts since SuperCell considers normalized data as input. We then perform dimensionality reduction for data visualization purposes.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="supercell-r.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb12-2"><a href="supercell-r.html#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Attaching SeuratObject</span></span>
<span id="cb12-3"><a href="supercell-r.html#cb12-3" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(sc_data, <span class="at">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>)</span>
<span id="cb12-4"><a href="supercell-r.html#cb12-4" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(sc_data, <span class="at">nfeatures =</span> <span class="dv">1000</span>)</span>
<span id="cb12-5"><a href="supercell-r.html#cb12-5" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(sc_data)</span>
<span id="cb12-6"><a href="supercell-r.html#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span id="cb12-7"><a href="supercell-r.html#cb12-7" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(sc_data, <span class="at">npcs =</span> <span class="dv">10</span>)</span>
<span id="cb12-8"><a href="supercell-r.html#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 1 </span></span>
<span id="cb12-9"><a href="supercell-r.html#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  LTB, CD2, ACAP1, STK17A, CTSW, CCL5, GIMAP5, AQP3, GZMA, CST7 </span></span>
<span id="cb12-10"><a href="supercell-r.html#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     MAL, HOPX, GZMK, NKG7, KLRG1, LYAR, RIC3, PRF1, FAM107B, CCND3 </span></span>
<span id="cb12-11"><a href="supercell-r.html#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CD79A, SRSF7, PASK, PTPN4, GZMH, GPR183, FGFBP2, TIGIT, TCL1A, ARID5B </span></span>
<span id="cb12-12"><a href="supercell-r.html#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  CST3, TYROBP, LST1, AIF1, FTL, LYZ, FTH1, FCN1, S100A9, TYMP </span></span>
<span id="cb12-13"><a href="supercell-r.html#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     FCER1G, CFD, LGALS1, S100A8, LGALS2, CTSS, IFITM3, PSAP, CFP, SAT1 </span></span>
<span id="cb12-14"><a href="supercell-r.html#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     IFI30, COTL1, S100A11, NPC2, LGALS3, GSTP1, NCF2, PYCARD, CDA, GPX1 </span></span>
<span id="cb12-15"><a href="supercell-r.html#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 2 </span></span>
<span id="cb12-16"><a href="supercell-r.html#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  NKG7, CST7, GZMA, PRF1, GZMB, FGFBP2, CTSW, GNLY, CCL4, GZMH </span></span>
<span id="cb12-17"><a href="supercell-r.html#cb12-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CCL5, FCGR3A, XCL2, CLIC3, SRGN, HOPX, S100A4, TTC38, IGFBP7, ID2 </span></span>
<span id="cb12-18"><a href="supercell-r.html#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ANXA1, ACTB, TMSB4X, APOBEC3G, KLRG1, LYAR, CD160, ABI3, HAVCR2, IFITM2 </span></span>
<span id="cb12-19"><a href="supercell-r.html#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  CD79A, MS4A1, HLA-DQA1, HLA-DQB1, TCL1A, HLA-DRA, CD79B, CD74, HLA-DRB1, HLA-DPB1 </span></span>
<span id="cb12-20"><a href="supercell-r.html#cb12-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     HLA-DMA, HLA-DRB5, HLA-DPA1, FCRLA, LTB, HVCN1, BLNK, P2RX5, IRF8, IGLL5 </span></span>
<span id="cb12-21"><a href="supercell-r.html#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     SMIM14, PPP1R14A, C16orf74, MZB1, RP5-887A10.1, BTK, RP11-428G5.5, IL4R, PHACTR1, IGJ </span></span>
<span id="cb12-22"><a href="supercell-r.html#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 3 </span></span>
<span id="cb12-23"><a href="supercell-r.html#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  HLA-DPA1, HLA-DPB1, HLA-DRB5, HLA-DRB1, CD74, HLA-DQB1, HLA-DQA1, CD79B, HLA-DRA, CD79A </span></span>
<span id="cb12-24"><a href="supercell-r.html#cb12-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     MS4A1, RBM3, HLA-DMA, IFITM2, TCL1A, CSNK2B, HVCN1, UBE2L6, YWHAB, IRF8 </span></span>
<span id="cb12-25"><a href="supercell-r.html#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CTSS, PSMA7, UBXN1, S100A4, C1orf162, PYCARD, APOBEC3G, SMIM14, SRSF7, TYROBP </span></span>
<span id="cb12-26"><a href="supercell-r.html#cb12-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  PPBP, PF4, SDPR, SPARC, GNG11, NRGN, HIST1H2AC, GP9, RGS18, TUBB1 </span></span>
<span id="cb12-27"><a href="supercell-r.html#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CLU, AP001189.4, CD9, ITGA2B, PTCRA, CA2, TMEM40, ACRBP, MMD, TREML1 </span></span>
<span id="cb12-28"><a href="supercell-r.html#cb12-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     SEPT5, RUFY1, MYL9, TSC22D1, MPP1, CMTM5, LY6G6F, GP1BA, RP11-367G6.3, CLEC1B </span></span>
<span id="cb12-29"><a href="supercell-r.html#cb12-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 4 </span></span>
<span id="cb12-30"><a href="supercell-r.html#cb12-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  HLA-DQA1, CD79B, CD79A, CD74, HLA-DQB1, HLA-DPB1, MS4A1, HLA-DPA1, HLA-DRB1, HLA-DRA </span></span>
<span id="cb12-31"><a href="supercell-r.html#cb12-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     HLA-DRB5, TCL1A, GZMB, FGFBP2, HLA-DMA, NKG7, PRF1, HVCN1, CST7, GNLY </span></span>
<span id="cb12-32"><a href="supercell-r.html#cb12-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     FCRLA, FCGR3A, GZMH, GZMA, CCL4, IRF8, BLNK, IGLL5, CLIC3, P2RX5 </span></span>
<span id="cb12-33"><a href="supercell-r.html#cb12-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  FYB, MAL, S100A8, AQP3, TMSB4X, CD2, S100A9, S100A4, GIMAP4, RBP7 </span></span>
<span id="cb12-34"><a href="supercell-r.html#cb12-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     GIMAP5, S100A12, LGALS2, ANXA1, FOLR3, LYZ, FCN1, PASK, MS4A6A, S100A11 </span></span>
<span id="cb12-35"><a href="supercell-r.html#cb12-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CORO1B, AIF1, IL8, ATP5H, CRIP2, IL23A, PPA1, LTB, LGALS3BP, ASGR1 </span></span>
<span id="cb12-36"><a href="supercell-r.html#cb12-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PC_ 5 </span></span>
<span id="cb12-37"><a href="supercell-r.html#cb12-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Positive:  S100A8, LGALS2, S100A9, S100A12, RBP7, FGFBP2, FOLR3, MS4A6A, GZMB, CCL4 </span></span>
<span id="cb12-38"><a href="supercell-r.html#cb12-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     CCL3, NKG7, GNLY, CST7, GSTP1, GZMA, PRF1, GZMH, ASGR1, CTSW </span></span>
<span id="cb12-39"><a href="supercell-r.html#cb12-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     LYZ, CCL5, CLIC3, XCL2, TYROBP, IL8, FCN1, GPX1, FCGR1A, TTC38 </span></span>
<span id="cb12-40"><a href="supercell-r.html#cb12-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Negative:  LTB, MS4A7, IFITM2, HN1, LILRB2, WARS, CTD-2006K23.1, AQP3, VMO1, GDI2 </span></span>
<span id="cb12-41"><a href="supercell-r.html#cb12-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     ADA, ANXA5, NAAA, FCGR3A, CORO1B, ABRACL, CD2, PPM1N, ATP5C1, HSP90AA1 </span></span>
<span id="cb12-42"><a href="supercell-r.html#cb12-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;     TIMP1, ARL6IP5, ICAM2, COTL1, C1QA, OAS1, PPA1, MAL, ABI3, TNFSF10</span></span>
<span id="cb12-43"><a href="supercell-r.html#cb12-43" aria-hidden="true" tabindex="-1"></a>sc_data <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(sc_data, <span class="at">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>), <span class="at">n.neighbors =</span> <span class="dv">10</span>)</span>
<span id="cb12-44"><a href="supercell-r.html#cb12-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span id="cb12-45"><a href="supercell-r.html#cb12-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;</span></span>
<span id="cb12-46"><a href="supercell-r.html#cb12-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; This message will be shown once per session</span></span>
<span id="cb12-47"><a href="supercell-r.html#cb12-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span id="cb12-48"><a href="supercell-r.html#cb12-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 Read 2638 rows and found 10 numeric columns</span></span>
<span id="cb12-49"><a href="supercell-r.html#cb12-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 Using Annoy for neighbor search, n_neighbors = 10</span></span>
<span id="cb12-50"><a href="supercell-r.html#cb12-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span id="cb12-51"><a href="supercell-r.html#cb12-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span id="cb12-52"><a href="supercell-r.html#cb12-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span id="cb12-53"><a href="supercell-r.html#cb12-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; **************************************************|</span></span>
<span id="cb12-54"><a href="supercell-r.html#cb12-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 Writing NN index file to temp file /tmp/Rtmp9KacAu/file1a354cb6b86c</span></span>
<span id="cb12-55"><a href="supercell-r.html#cb12-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 Searching Annoy index using 1 thread, search_k = 1000</span></span>
<span id="cb12-56"><a href="supercell-r.html#cb12-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 Annoy recall = 100%</span></span>
<span id="cb12-57"><a href="supercell-r.html#cb12-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:17 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 10</span></span>
<span id="cb12-58"><a href="supercell-r.html#cb12-58" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:18 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span id="cb12-59"><a href="supercell-r.html#cb12-59" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:18 Commencing optimization for 500 epochs, with 34256 positive edges</span></span>
<span id="cb12-60"><a href="supercell-r.html#cb12-60" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 13:16:19 Optimization finished</span></span>
<span id="cb12-61"><a href="supercell-r.html#cb12-61" aria-hidden="true" tabindex="-1"></a><span class="fu">UMAPPlot</span>(sc_data, <span class="at">group.by =</span> <span class="st">&quot;louvain&quot;</span>)</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/supercell-data-processing-1.png" width="672" /></p>
</div>
<div id="setting-up-supercell-parameters" class="section level4 unnumbered hasAnchor">
<h4>Setting up SuperCell parameters<a href="supercell-r.html#setting-up-supercell-parameters" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In this tutorial, we will run SuperCell using the 10 first principal components resulting from the PCA.
We chose a graining level of 50 and a number of neighbors of 15 for the knn step.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="supercell-r.html#cb13-1" aria-hidden="true" tabindex="-1"></a>gamma <span class="ot">=</span> <span class="dv">50</span>   <span class="co"># the requested graining level.</span></span>
<span id="cb13-2"><a href="supercell-r.html#cb13-2" aria-hidden="true" tabindex="-1"></a>k.knn <span class="ot">=</span> <span class="dv">15</span> <span class="co"># the number of neighbors considered to build the knn network.</span></span>
<span id="cb13-3"><a href="supercell-r.html#cb13-3" aria-hidden="true" tabindex="-1"></a>nb.var.genes <span class="ot">=</span> <span class="dv">1000</span> <span class="co"># number of the top variable genes to use for dimensionality reduction </span></span></code></pre></div>
</div>
<div id="metacells-idenification" class="section level4 unnumbered hasAnchor">
<h4>Metacells idenification<a href="supercell-r.html#metacells-idenification" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="supercell-r.html#cb14-1" aria-hidden="true" tabindex="-1"></a>MC <span class="ot">&lt;-</span> SuperCell<span class="sc">::</span><span class="fu">SCimplify</span>(sc_data<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>data,  <span class="co"># single-cell log-normalized gene expression data</span></span>
<span id="cb14-2"><a href="supercell-r.html#cb14-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">k.knn =</span> k.knn, </span>
<span id="cb14-3"><a href="supercell-r.html#cb14-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">gamma =</span> gamma, </span>
<span id="cb14-4"><a href="supercell-r.html#cb14-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">n.var.genes =</span> nb.var.genes </span>
<span id="cb14-5"><a href="supercell-r.html#cb14-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>SuperCell returns a list containing the following main elements:
i) the single-cell assignments to metacells (<code>membership</code>),
ii) …</p>
</div>
<div id="retrieve-aggregated-metacell-data" class="section level4 unnumbered hasAnchor">
<h4>Retrieve aggregated metacell data<a href="supercell-r.html#retrieve-aggregated-metacell-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The <code>supercell_GE()</code> function can be used to generate a metacell count matrix (aggregation of counts across all cells belonging to each metacell).
Two modes can be used for single-cell aggregation, <em>i.e.</em> averaging or summing counts (using the <code>mode</code> parameter).
Note that we provide raw counts for the aggregation similarly to the aggregation steps in MC2 and SEACells. Data normalization will thus be needed for downstream analysis.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="supercell-r.html#cb15-1" aria-hidden="true" tabindex="-1"></a>MC.GE <span class="ot">&lt;-</span> <span class="fu">supercell_GE</span>(sc_data<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>counts, MC<span class="sc">$</span>membership, <span class="at">mode =</span> <span class="st">&quot;sum&quot;</span>) </span>
<span id="cb15-2"><a href="supercell-r.html#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(MC.GE) </span>
<span id="cb15-3"><a href="supercell-r.html#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 32738    53</span></span></code></pre></div>
</div>
</div>
<div id="visualize-the-supercell-graph" class="section level3 hasAnchor" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> Visualize the SuperCell graph<a href="supercell-r.html#visualize-the-supercell-graph" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can assign each metacell to a particular annotation using the <code>supercell_assign()</code> function.
By default, this function assigns each metacell to a cluster with the largest Jaccard coefficient to avoid biases towards very rare or very abundant clusters.
Alternatively, assignment can be performed using <code>relative</code> (may cause biases towards very small populations) or <code>absolute</code> (may cause biases towards large populations) abundance with <code>method = "relative"</code> or <code>method = "absolute"</code>, respectively.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="supercell-r.html#cb16-1" aria-hidden="true" tabindex="-1"></a>MC<span class="sc">$</span>annotation <span class="ot">&lt;-</span> <span class="fu">supercell_assign</span>(<span class="at">clusters =</span> sc_data<span class="sc">@</span>meta.data<span class="sc">$</span>louvain, <span class="co"># single-cell annotation</span></span>
<span id="cb16-2"><a href="supercell-r.html#cb16-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">supercell_membership =</span> MC<span class="sc">$</span>membership, <span class="co"># single-cell assignment to metacells</span></span>
<span id="cb16-3"><a href="supercell-r.html#cb16-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">method =</span> <span class="st">&quot;jaccard&quot;</span>)</span></code></pre></div>
<p>The SuperCell package provides the <code>supercell_plot</code> function to visualize the metacell network, which is stored <code>MC</code> object in <code>graph.supercells</code>.
The metacells can be colored with respect to a vector of annotation.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="supercell-r.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">supercell_plot</span>(</span>
<span id="cb17-2"><a href="supercell-r.html#cb17-2" aria-hidden="true" tabindex="-1"></a>  MC<span class="sc">$</span>graph.supercells, </span>
<span id="cb17-3"><a href="supercell-r.html#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> MC<span class="sc">$</span>annotation, </span>
<span id="cb17-4"><a href="supercell-r.html#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">seed =</span> <span class="dv">1</span>, </span>
<span id="cb17-5"><a href="supercell-r.html#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">alpha =</span> <span class="sc">-</span>pi<span class="sc">/</span><span class="dv">2</span>,</span>
<span id="cb17-6"><a href="supercell-r.html#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">main  =</span> <span class="st">&quot;Metacells colored by cell line assignment&quot;</span></span>
<span id="cb17-7"><a href="supercell-r.html#cb17-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/SuperCell-graph-1.png" width="672" /></p>
<p>To visualize the metacells, we can project the metacells on the single-cell UMAP representation using the <code>mc_QC.mc_visualize()</code> function adapted from the <code>plot.plot_2D()</code> from the SEACells package.
To use this function we need to create an anndata object containing the cell embedding at the single cell level.</p>
<p>We add a column named <code>membership</code> with the single_cell assignments to the obs attribute in the anndata object containing the raw data.
This annotation will be used to visualize single-cell to metacell assignments as well as to compute metacells quality metrics in the next section.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="supercell-r.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(anndata)</span>
<span id="cb18-2"><a href="supercell-r.html#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Create anndata object from R </span></span>
<span id="cb18-3"><a href="supercell-r.html#cb18-3" aria-hidden="true" tabindex="-1"></a>r_ad <span class="ot">&lt;-</span> <span class="fu">AnnData</span>(<span class="at">X =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(sc_data<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>counts),</span>
<span id="cb18-4"><a href="supercell-r.html#cb18-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">obs =</span> sc_data<span class="sc">@</span>meta.data,</span>
<span id="cb18-5"><a href="supercell-r.html#cb18-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">obsm =</span> <span class="fu">list</span>(<span class="at">X_pca =</span> sc_data<span class="sc">@</span>reductions<span class="sc">$</span>pca<span class="sc">@</span>cell.embeddings,</span>
<span id="cb18-6"><a href="supercell-r.html#cb18-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">X_umap =</span> sc_data<span class="sc">@</span>reductions<span class="sc">$</span>umap<span class="sc">@</span>cell.embeddings)</span>
<span id="cb18-7"><a href="supercell-r.html#cb18-7" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb18-8"><a href="supercell-r.html#cb18-8" aria-hidden="true" tabindex="-1"></a>r_ad<span class="sc">$</span>obs[<span class="st">&#39;membership&#39;</span>] <span class="ot">&lt;-</span> MC<span class="sc">$</span>membership </span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb19-1"><a href="supercell-r.html#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save anndata object in python environment </span></span>
<span id="cb19-2"><a href="supercell-r.html#cb19-2" aria-hidden="true" tabindex="-1"></a>ad <span class="op">=</span> r.r_ad </span></code></pre></div>
<p>Here we project metacells on the single-cell UMAP and color cells and metacells using metacells memberships.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb20-1"><a href="supercell-r.html#cb20-1" aria-hidden="true" tabindex="-1"></a>mc_proj <span class="op">=</span> mc_QC.mc_visualize(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, colour_sc_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, colour_mc_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb20-2"><a href="supercell-r.html#cb20-2" aria-hidden="true" tabindex="-1"></a>mc_proj.show()</span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/SuperCell-umap-1.png" width="384" /></p>
</div>
<div id="metacell-qc" class="section level3 hasAnchor" number="3.1.5">
<h3><span class="header-section-number">3.1.5</span> Metacell QC<a href="supercell-r.html#metacell-qc" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="compute-purity-compactness-and-separation-metrics" class="section level4 unnumbered hasAnchor">
<h4>Compute purity, compactness and separation metrics<a href="supercell-r.html#compute-purity-compactness-and-separation-metrics" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="supercell-r.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(anndata)</span>
<span id="cb21-2"><a href="supercell-r.html#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Create anndata object from R</span></span>
<span id="cb21-3"><a href="supercell-r.html#cb21-3" aria-hidden="true" tabindex="-1"></a>r_mc_ad <span class="ot">&lt;-</span> <span class="fu">AnnData</span>(<span class="at">X =</span> Matrix<span class="sc">::</span><span class="fu">t</span>(MC.GE),</span>
<span id="cb21-4"><a href="supercell-r.html#cb21-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">obs =</span> <span class="fu">data.frame</span>(<span class="at">annotation =</span> MC<span class="sc">$</span>annotation))</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb22-1"><a href="supercell-r.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save anndata object in python environment </span></span>
<span id="cb22-2"><a href="supercell-r.html#cb22-2" aria-hidden="true" tabindex="-1"></a>mc_ad <span class="op">=</span> r.r_mc_ad </span></code></pre></div>
<p><strong>Size distribution</strong></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb23-1"><a href="supercell-r.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_size = SEACells.plot.plot_SEACell_sizes(ad, bins=20)</span></span>
<span id="cb23-2"><a href="supercell-r.html#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="supercell-r.html#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs = pd.merge(mc_ad.obs, mc_size, left_index=True, right_index=True)</span></span>
<span id="cb23-4"><a href="supercell-r.html#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co">#mc_ad.obs</span></span></code></pre></div>
<p>When available, we can use cell annotation to annotate each metacell to the most abundant cell category (<em>e.g.</em> cell type) composing the metacell.
This also allows us to compute metacell purity. If the annotation considered is the cell type, the <strong>purity</strong> of a metacell is the proportion of the most abundant cell type within the metacell [ref SuperCell]</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb24-1"><a href="supercell-r.html#cb24-1" aria-hidden="true" tabindex="-1"></a>mc_purity <span class="op">=</span> mc_QC.purity(ad, annotation_label, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>)</span>
<span id="cb24-2"><a href="supercell-r.html#cb24-2" aria-hidden="true" tabindex="-1"></a>mc_purity.head()</span>
<span id="cb24-3"><a href="supercell-r.html#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     louvain  louvain_purity</span></span>
<span id="cb24-4"><a href="supercell-r.html#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; membership                                 </span></span>
<span id="cb24-5"><a href="supercell-r.html#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1.0             CD4 T cells        0.973684</span></span>
<span id="cb24-6"><a href="supercell-r.html#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2.0             CD4 T cells        0.975610</span></span>
<span id="cb24-7"><a href="supercell-r.html#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3.0             CD4 T cells        0.990476</span></span>
<span id="cb24-8"><a href="supercell-r.html#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4.0             CD8 T cells        0.871287</span></span>
<span id="cb24-9"><a href="supercell-r.html#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5.0         CD14+ Monocytes        0.764706</span></span>
<span id="cb24-10"><a href="supercell-r.html#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co"># add purity to metadata</span></span>
<span id="cb24-11"><a href="supercell-r.html#cb24-11" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;purity&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(mc_purity[annotation_label <span class="op">+</span> <span class="st">&quot;_purity&quot;</span>])</span></code></pre></div>
<p>The <strong>compactness</strong> of a metacell is the variance of the components within the metacell [ref SEACells]</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb25-1"><a href="supercell-r.html#cb25-1" aria-hidden="true" tabindex="-1"></a>compactness <span class="op">=</span> mc_QC.compactness(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Compactness_PCA&#39;</span>, n_comp<span class="op">=</span><span class="dv">10</span>)[<span class="st">&#39;Compactness_PCA&#39;</span>]</span>
<span id="cb25-2"><a href="supercell-r.html#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add compactness to metadata</span></span>
<span id="cb25-3"><a href="supercell-r.html#cb25-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;Compactness_PCA&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(compactness)</span></code></pre></div>
<p>The <strong>separation</strong> of a metacell is the distance to the closest metacell [ref SEACells]</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb26-1"><a href="supercell-r.html#cb26-1" aria-hidden="true" tabindex="-1"></a>separation <span class="op">=</span> mc_QC.separation(ad, <span class="st">&#39;X_pca&#39;</span>, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>, DO_DC <span class="op">=</span> <span class="va">False</span>, name <span class="op">=</span> <span class="st">&#39;Separation_PCA&#39;</span>, n_comp<span class="op">=</span><span class="dv">10</span>)[<span class="st">&#39;Separation_PCA&#39;</span>]</span>
<span id="cb26-2"><a href="supercell-r.html#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add separation to metadata</span></span>
<span id="cb26-3"><a href="supercell-r.html#cb26-3" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;Separation_PCA&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(separation)</span></code></pre></div>
<p>The <strong>inner normalized variance (INV)</strong> of a metacell is the mean-normalized variance of gene expression within the metacell [ref MC-2]</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb27-1"><a href="supercell-r.html#cb27-1" aria-hidden="true" tabindex="-1"></a>mc_INV <span class="op">=</span> mc_QC.mc_inner_normalized_var(ad<span class="op">=</span>ad, MC_label <span class="op">=</span> <span class="st">&#39;membership&#39;</span>)</span>
<span id="cb27-2"><a href="supercell-r.html#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="supercell-r.html#cb27-3" aria-hidden="true" tabindex="-1"></a>mc_INV_val <span class="op">=</span> mc_INV.quantile(<span class="fl">0.95</span>, axis<span class="op">=</span><span class="dv">1</span>, numeric_only<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb27-4"><a href="supercell-r.html#cb27-4" aria-hidden="true" tabindex="-1"></a>mc_INV_val <span class="op">=</span> pd.DataFrame(mc_INV_val.transpose()).set_axis([<span class="st">&#39;INV&#39;</span>], axis<span class="op">=</span><span class="dv">1</span>, copy<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb27-5"><a href="supercell-r.html#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add INV to metadata</span></span>
<span id="cb27-6"><a href="supercell-r.html#cb27-6" aria-hidden="true" tabindex="-1"></a>mc_ad.obs[<span class="st">&#39;INV&#39;</span>] <span class="op">=</span> <span class="bu">list</span>(mc_INV_val[<span class="st">&quot;INV&quot;</span>])</span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb28-1"><a href="supercell-r.html#cb28-1" aria-hidden="true" tabindex="-1"></a>ad.uns[<span class="st">&#39;mc_obs&#39;</span>] <span class="op">=</span> mc_ad.obs</span>
<span id="cb28-2"><a href="supercell-r.html#cb28-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb28-3"><a href="supercell-r.html#cb28-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;purity&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb28-4"><a href="supercell-r.html#cb28-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)</span>
<span id="cb28-5"><a href="supercell-r.html#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-6"><a href="supercell-r.html#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-7"><a href="supercell-r.html#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-8"><a href="supercell-r.html#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-9"><a href="supercell-r.html#cb28-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-10"><a href="supercell-r.html#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-11"><a href="supercell-r.html#cb28-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-12"><a href="supercell-r.html#cb28-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-13"><a href="supercell-r.html#cb28-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-14"><a href="supercell-r.html#cb28-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-15"><a href="supercell-r.html#cb28-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-16"><a href="supercell-r.html#cb28-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-17"><a href="supercell-r.html#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-18"><a href="supercell-r.html#cb28-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-19"><a href="supercell-r.html#cb28-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-20"><a href="supercell-r.html#cb28-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-21"><a href="supercell-r.html#cb28-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-22"><a href="supercell-r.html#cb28-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-23"><a href="supercell-r.html#cb28-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-24"><a href="supercell-r.html#cb28-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-25"><a href="supercell-r.html#cb28-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-26"><a href="supercell-r.html#cb28-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-27"><a href="supercell-r.html#cb28-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-28"><a href="supercell-r.html#cb28-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-29"><a href="supercell-r.html#cb28-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-30"><a href="supercell-r.html#cb28-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span>
<span id="cb28-31"><a href="supercell-r.html#cb28-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; findfont: Font family &#39;Bitstream Vera Sans&#39; not found.</span></span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-11-1.png" width="384" /></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb29-1"><a href="supercell-r.html#cb29-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb29-2"><a href="supercell-r.html#cb29-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb29-3"><a href="supercell-r.html#cb29-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;Compactness_PCA&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb29-4"><a href="supercell-r.html#cb29-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)   </span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-11-2.png" width="384" /></p>
<div class="sourceCode" id="cb30"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb30-1"><a href="supercell-r.html#cb30-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb30-2"><a href="supercell-r.html#cb30-2" aria-hidden="true" tabindex="-1"></a>mc_QC.mc_visualize_continuous(ad, key<span class="op">=</span><span class="st">&#39;X_umap&#39;</span>, group_by_name<span class="op">=</span><span class="st">&#39;membership&#39;</span>, </span>
<span id="cb30-3"><a href="supercell-r.html#cb30-3" aria-hidden="true" tabindex="-1"></a>             colour_sc_name<span class="op">=</span><span class="st">&#39;louvain&#39;</span>,  colour_mc_name<span class="op">=</span><span class="st">&#39;Separation_PCA&#39;</span>, colour_metacells<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb30-4"><a href="supercell-r.html#cb30-4" aria-hidden="true" tabindex="-1"></a>             legend_sc<span class="op">=</span><span class="va">None</span>, legend_mc<span class="op">=</span><span class="st">&#39;auto&#39;</span>, metacell_size<span class="op">=</span><span class="dv">30</span>)   </span></code></pre></div>
<p><img src="21-MC_construction_discrete_files/figure-html/unnamed-chunk-11-3.png" width="384" /></p>
<div class="sourceCode" id="cb31"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb31-1"><a href="supercell-r.html#cb31-1" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb31-2"><a href="supercell-r.html#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># mc_QC.mc_visualize_continuous(ad, key=&#39;X_umap&#39;, group_by_name=&#39;membership&#39;,</span></span>
<span id="cb31-3"><a href="supercell-r.html#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co">#              colour_sc_name=&#39;louvain&#39;,  colour_mc_name=&#39;INV&#39;, colour_metacells=True,</span></span>
<span id="cb31-4"><a href="supercell-r.html#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co">#              legend_sc=None, legend_mc=&#39;auto&#39;, metacell_size=30)</span></span></code></pre></div>
<div id="save-output" class="section level5 unnumbered hasAnchor">
<h5>Save output<a href="supercell-r.html#save-output" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>For downstream analyses in python, we save the anndata object containing the metacell counts data:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb32-1"><a href="supercell-r.html#cb32-1" aria-hidden="true" tabindex="-1"></a>mc_ad.write_h5ad(os.path.join(<span class="st">&#39;./data&#39;</span>, proj_name, <span class="ss">f&#39;metacell_</span><span class="sc">{</span>MC_tool<span class="sc">}</span><span class="ss">.h5ad&#39;</span>))</span></code></pre></div>
<p>For downstream analyses in R, we can use the <code>supercell_2_Seurat()</code> function from the SuperCell package.
This function takes as inputs the metacell count matrix (output of the SuperCell <code>supercell_GE()</code> function) and the output of the SuperCell <code>SCimplify()</code> function
and outputs a Seurat object containing normalized metacells gene expression data as well as the first (<code>N.comp</code>) principal components of PCA performed internally using user defined set of genes (by default the genes used for metacells constructions).</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="supercell-r.html#cb33-1" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">supercell_2_Seurat</span>(</span>
<span id="cb33-2"><a href="supercell-r.html#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">SC.GE =</span> MC.GE,</span>
<span id="cb33-3"><a href="supercell-r.html#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">SC =</span> MC,</span>
<span id="cb33-4"><a href="supercell-r.html#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">fields =</span> <span class="fu">c</span>(<span class="st">&quot;annotation&quot;</span>, <span class="st">&quot;purity&quot;</span>, <span class="st">&quot;supercell_size&quot;</span>),</span>
<span id="cb33-5"><a href="supercell-r.html#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">var.genes =</span> MC<span class="sc">$</span>genes.use,</span>
<span id="cb33-6"><a href="supercell-r.html#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">N.comp =</span> <span class="dv">10</span></span>
<span id="cb33-7"><a href="supercell-r.html#cb33-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-8"><a href="supercell-r.html#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Done: NormalizeData&quot;</span></span>
<span id="cb33-9"><a href="supercell-r.html#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Doing: data to normalized data&quot;</span></span>
<span id="cb33-10"><a href="supercell-r.html#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Doing: weighted scaling&quot;</span></span>
<span id="cb33-11"><a href="supercell-r.html#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: 16159 instances of variables with zero scale detected!</span></span>
<span id="cb33-12"><a href="supercell-r.html#cb33-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Done: weighted scaling&quot;</span></span>
<span id="cb33-13"><a href="supercell-r.html#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span id="cb33-14"><a href="supercell-r.html#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Computing SNN</span></span>
<span id="cb33-15"><a href="supercell-r.html#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(MC.seurat, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&#39;./data/&#39;</span>, proj_name, <span class="st">&#39;/metacell_&#39;</span>, MC_tool,<span class="st">&#39;.rds&#39;</span>))</span></code></pre></div>
</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="constructing-metacells-for-discrete-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="mc2-python.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/21-MC_construction_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/21-MC_construction_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
