<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.2 Sample-weighted analysis | Metacell Tutorial</title>
  <meta name="description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="5.2 Sample-weighted analysis | Metacell Tutorial" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is a tutorial about metacell construction and analysis." />
  <meta name="github-repo" content="GfellerLab/Metacell_tutorial" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.2 Sample-weighted analysis | Metacell Tutorial" />
  
  <meta name="twitter:description" content="This is a tutorial about metacell construction and analysis." />
  

<meta name="author" content="Aurélie Gabriel, Léonard Hérault, Mariia Bilous, David Gfeller" />


<meta name="date" content="2023-11-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="standard-analysis-R.html"/>
<link rel="next" href="integration.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Metacell Tutorial </a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>This tutorial</a></li>
<li class="chapter" data-level="1" data-path="requirements.html"><a href="requirements.html"><i class="fa fa-check"></i><b>1</b> Requirements</a>
<ul>
<li class="chapter" data-level="1.1" data-path="installations.html"><a href="installations.html"><i class="fa fa-check"></i><b>1.1</b> Installations</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="installations.html"><a href="installations.html#using-conda-recommended"><i class="fa fa-check"></i><b>1.1.1</b> Using conda (recommended)</a></li>
<li class="chapter" data-level="1.1.2" data-path="installations.html"><a href="installations.html#without-conda"><i class="fa fa-check"></i><b>1.1.2</b> Without conda</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="PBMC-data.html"><a href="PBMC-data.html"><i class="fa fa-check"></i><b>1.2</b> Retrieve a discrete dataset (PBMCs dataset)</a></li>
<li class="chapter" data-level="1.3" data-path="CD34-data.html"><a href="CD34-data.html"><i class="fa fa-check"></i><b>1.3</b> Retrieve a continuous dataset (CD34 dataset)</a></li>
<li class="chapter" data-level="1.4" data-path="HLCA-data.html"><a href="HLCA-data.html"><i class="fa fa-check"></i><b>1.4</b> Retrieve the lung atlas dataset</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="HLCA-data.html"><a href="HLCA-data.html#downloading-the-atlas"><i class="fa fa-check"></i><b>1.4.1</b> Downloading the atlas</a></li>
<li class="chapter" data-level="1.4.2" data-path="HLCA-data.html"><a href="HLCA-data.html#splitting-atlas-by-datasets"><i class="fa fa-check"></i><b>1.4.2</b> Splitting atlas by datasets</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-metacell-concept.html"><a href="the-metacell-concept.html"><i class="fa fa-check"></i><b>2</b> The metacell concept</a></li>
<li class="chapter" data-level="3" data-path="Metacell-construction-chapter.html"><a href="Metacell-construction-chapter.html"><i class="fa fa-check"></i><b>3</b> Constructing metacells</a>
<ul>
<li class="chapter" data-level="3.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html"><i class="fa fa-check"></i><b>3.1</b> SuperCell (R)</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#method"><i class="fa fa-check"></i><b>3.1.1</b> Method</a></li>
<li class="chapter" data-level="3.1.2" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#data-loading"><i class="fa fa-check"></i><b>3.1.2</b> Data loading</a></li>
<li class="chapter" data-level="3.1.3" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#filtering-steps"><i class="fa fa-check"></i><b>3.1.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.1.4" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#building-metacells"><i class="fa fa-check"></i><b>3.1.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.1.5" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#annotate-metacells-using-available-annotations"><i class="fa fa-check"></i><b>3.1.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.1.6" data-path="SuperCell-construction.html"><a href="SuperCell-construction.html#save-output"><i class="fa fa-check"></i><b>3.1.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="MC2-construction.html"><a href="MC2-construction.html"><i class="fa fa-check"></i><b>3.2</b> MC2 (Python)</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="MC2-construction.html"><a href="MC2-construction.html#method-1"><i class="fa fa-check"></i><b>3.2.1</b> Method</a></li>
<li class="chapter" data-level="3.2.2" data-path="MC2-construction.html"><a href="MC2-construction.html#data-loading-1"><i class="fa fa-check"></i><b>3.2.2</b> Data loading</a></li>
<li class="chapter" data-level="3.2.3" data-path="MC2-construction.html"><a href="MC2-construction.html#filtering-steps-1"><i class="fa fa-check"></i><b>3.2.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.2.4" data-path="MC2-construction.html"><a href="MC2-construction.html#building-metacells-1"><i class="fa fa-check"></i><b>3.2.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.2.5" data-path="MC2-construction.html"><a href="MC2-construction.html#annotate-metacells-using-available-annotations-1"><i class="fa fa-check"></i><b>3.2.5</b> Annotate metacells (using available annotations)</a></li>
<li class="chapter" data-level="3.2.6" data-path="MC2-construction.html"><a href="MC2-construction.html#save-output-1"><i class="fa fa-check"></i><b>3.2.6</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html"><i class="fa fa-check"></i><b>3.3</b> SEACells (Python)</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="SEACells-construction.html"><a href="SEACells-construction.html#method-2"><i class="fa fa-check"></i><b>3.3.1</b> Method</a></li>
<li class="chapter" data-level="3.3.2" data-path="SEACells-construction.html"><a href="SEACells-construction.html#data-loading-2"><i class="fa fa-check"></i><b>3.3.2</b> Data loading</a></li>
<li class="chapter" data-level="3.3.3" data-path="SEACells-construction.html"><a href="SEACells-construction.html#filtering-steps-2"><i class="fa fa-check"></i><b>3.3.3</b> Filtering steps</a></li>
<li class="chapter" data-level="3.3.4" data-path="SEACells-construction.html"><a href="SEACells-construction.html#building-metacells-2"><i class="fa fa-check"></i><b>3.3.4</b> Building metacells</a></li>
<li class="chapter" data-level="3.3.5" data-path="SEACells-construction.html"><a href="SEACells-construction.html#visualize-metacells"><i class="fa fa-check"></i><b>3.3.5</b> Visualize metacells</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="command-line.html"><a href="command-line.html"><i class="fa fa-check"></i><b>3.4</b> Metacell Analysis Toolkit (MCAT)</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="QCs.html"><a href="QCs.html"><i class="fa fa-check"></i><b>4</b> Metacells QCs</a>
<ul>
<li class="chapter" data-level="4.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html"><i class="fa fa-check"></i><b>4.1</b> Quantitative metrics</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#purity"><i class="fa fa-check"></i><b>4.1.1</b> Purity</a></li>
<li class="chapter" data-level="4.1.2" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#compactness"><i class="fa fa-check"></i><b>4.1.2</b> Compactness</a></li>
<li class="chapter" data-level="4.1.3" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#separation"><i class="fa fa-check"></i><b>4.1.3</b> Separation</a></li>
<li class="chapter" data-level="4.1.4" data-path="quantitative-metrics.html"><a href="quantitative-metrics.html#inv"><i class="fa fa-check"></i><b>4.1.4</b> INV</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="size-distribution.html"><a href="size-distribution.html"><i class="fa fa-check"></i><b>4.2</b> Size distribution</a></li>
<li class="chapter" data-level="4.3" data-path="representativeness-of-metacells.html"><a href="representativeness-of-metacells.html"><i class="fa fa-check"></i><b>4.3</b> Representativeness of metacells</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="downstream-analysis-of-metacells.html"><a href="downstream-analysis-of-metacells.html"><i class="fa fa-check"></i><b>5</b> Downstream analysis of metacells</a>
<ul>
<li class="chapter" data-level="5.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html"><i class="fa fa-check"></i><b>5.1</b> Standard analysis (R)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#load-metacell-seurat-object"><i class="fa fa-check"></i><b>5.1.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.1.2" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#dimensionality-reduction"><i class="fa fa-check"></i><b>5.1.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.1.3" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#clustering"><i class="fa fa-check"></i><b>5.1.3</b> Clustering</a></li>
<li class="chapter" data-level="5.1.4" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#differential-expression-analysis"><i class="fa fa-check"></i><b>5.1.4</b> Differential expression analysis</a></li>
<li class="chapter" data-level="5.1.5" data-path="standard-analysis-R.html"><a href="standard-analysis-R.html#visualize-gene-gene-correlation"><i class="fa fa-check"></i><b>5.1.5</b> Visualize gene-gene correlation</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html"><i class="fa fa-check"></i><b>5.2</b> Sample-weighted analysis</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="weighted-analysis.html"><a href="weighted-analysis.html#load-metacell-seurat-object-1"><i class="fa fa-check"></i><b>5.2.1</b> Load metacell Seurat object</a></li>
<li class="chapter" data-level="5.2.2" data-path="weighted-analysis.html"><a href="weighted-analysis.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>5.2.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="5.2.3" data-path="weighted-analysis.html"><a href="weighted-analysis.html#clustering-1"><i class="fa fa-check"></i><b>5.2.3</b> Clustering</a></li>
<li class="chapter" data-level="5.2.4" data-path="weighted-analysis.html"><a href="weighted-analysis.html#differential-expression-analysis-1"><i class="fa fa-check"></i><b>5.2.4</b> Differential expression analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="integration.html"><a href="integration.html"><i class="fa fa-check"></i><b>6</b> Integration of metacells</a>
<ul>
<li class="chapter" data-level="6.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html"><i class="fa fa-check"></i><b>6.1</b> Unsupervised integration</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#data-loading-3"><i class="fa fa-check"></i><b>6.1.1</b> Data loading</a></li>
<li class="chapter" data-level="6.1.2" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#setting-up-the-environment"><i class="fa fa-check"></i><b>6.1.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="6.1.3" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#building-metacell"><i class="fa fa-check"></i><b>6.1.3</b> Building metacell</a></li>
<li class="chapter" data-level="6.1.4" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#loading-metacell-objects"><i class="fa fa-check"></i><b>6.1.4</b> Loading metacell objects</a></li>
<li class="chapter" data-level="6.1.5" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#merging-objects-and-basic-quality-control"><i class="fa fa-check"></i><b>6.1.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="6.1.6" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#unintegrated-analysis"><i class="fa fa-check"></i><b>6.1.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="6.1.7" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#seurat-integration"><i class="fa fa-check"></i><b>6.1.7</b> Seurat integration</a></li>
<li class="chapter" data-level="6.1.8" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#downstream-analysis"><i class="fa fa-check"></i><b>6.1.8</b> Downstream analysis</a></li>
<li class="chapter" data-level="6.1.9" data-path="integration_unsupervised.html"><a href="integration_unsupervised.html#conclusion"><i class="fa fa-check"></i><b>6.1.9</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="integration_supervised.html"><a href="integration_supervised.html"><i class="fa fa-check"></i><b>6.2</b> Supervised integration</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="integration_supervised.html"><a href="integration_supervised.html#data-loading-4"><i class="fa fa-check"></i><b>6.2.1</b> Data loading</a></li>
<li class="chapter" data-level="6.2.2" data-path="integration_supervised.html"><a href="integration_supervised.html#setting-up-the-environment-1"><i class="fa fa-check"></i><b>6.2.2</b> Setting up the environment</a></li>
<li class="chapter" data-level="6.2.3" data-path="integration_supervised.html"><a href="integration_supervised.html#constructing-supervised-metacell"><i class="fa fa-check"></i><b>6.2.3</b> Constructing supervised metacell</a></li>
<li class="chapter" data-level="6.2.4" data-path="integration_supervised.html"><a href="integration_supervised.html#load-metacell-objects"><i class="fa fa-check"></i><b>6.2.4</b> Load metacell objects</a></li>
<li class="chapter" data-level="6.2.5" data-path="integration_supervised.html"><a href="integration_supervised.html#merging-objects-and-basic-quality-control-1"><i class="fa fa-check"></i><b>6.2.5</b> Merging objects and basic quality control</a></li>
<li class="chapter" data-level="6.2.6" data-path="integration_supervised.html"><a href="integration_supervised.html#unintegrated-analysis-1"><i class="fa fa-check"></i><b>6.2.6</b> Unintegrated analysis</a></li>
<li class="chapter" data-level="6.2.7" data-path="integration_supervised.html"><a href="integration_supervised.html#stacas-integration"><i class="fa fa-check"></i><b>6.2.7</b> STACAS integration</a></li>
<li class="chapter" data-level="6.2.8" data-path="integration_supervised.html"><a href="integration_supervised.html#comparison-with-unsupervised-analysis"><i class="fa fa-check"></i><b>6.2.8</b> Comparison with unsupervised analysis</a></li>
<li class="chapter" data-level="6.2.9" data-path="integration_supervised.html"><a href="integration_supervised.html#downstream-analysis-1"><i class="fa fa-check"></i><b>6.2.9</b> Downstream analysis</a></li>
<li class="chapter" data-level="6.2.10" data-path="integration_supervised.html"><a href="integration_supervised.html#conclusion-1"><i class="fa fa-check"></i><b>6.2.10</b> Conclusion</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Metacell Tutorial</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="weighted-analysis" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> Sample-weighted analysis<a href="weighted-analysis.html#weighted-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="weighted-analysis.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat) </span>
<span id="cb87-2"><a href="weighted-analysis.html#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb87-3"><a href="weighted-analysis.html#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb87-4"><a href="weighted-analysis.html#cb87-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SuperCell)</span></code></pre></div>
<div id="load-metacell-seurat-object-1" class="section level3 hasAnchor" number="5.2.1">
<h3><span class="header-section-number">5.2.1</span> Load metacell Seurat object<a href="weighted-analysis.html#load-metacell-seurat-object-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will use Seurat objects containing the metacells counts data and their annotation (<em>e.g.</em> and cell-type annotation) and
proceed with downstream analyses considering the size of each metacells.
Seurat objects containing metacells counts data and their annotation were generated at the end of sections <a href="SuperCell-construction.html#SuperCell-construction">3.1</a>
These objects can also be generated using the command line described in chapter <a href="command-line.html#command-line">3.4</a></p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="weighted-analysis.html#cb88-1" aria-hidden="true" tabindex="-1"></a>MC_tool <span class="ot">=</span> <span class="st">&quot;SuperCell&quot;</span></span>
<span id="cb88-2"><a href="weighted-analysis.html#cb88-2" aria-hidden="true" tabindex="-1"></a>proj_name <span class="ot">=</span> <span class="st">&quot;3k_pbmc&quot;</span></span>
<span id="cb88-3"><a href="weighted-analysis.html#cb88-3" aria-hidden="true" tabindex="-1"></a>annotation_column <span class="ot">=</span> <span class="st">&quot;louvain&quot;</span></span>
<span id="cb88-4"><a href="weighted-analysis.html#cb88-4" aria-hidden="true" tabindex="-1"></a>celltype_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb88-5"><a href="weighted-analysis.html#cb88-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;CD14+ Monocytes&quot;</span>    <span class="ot">=</span> <span class="st">&quot;#E69F00&quot;</span>,  <span class="co"># orange</span></span>
<span id="cb88-6"><a href="weighted-analysis.html#cb88-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;B cells&quot;</span>            <span class="ot">=</span> <span class="st">&quot;#56B4E9&quot;</span>,  <span class="co"># sky blue</span></span>
<span id="cb88-7"><a href="weighted-analysis.html#cb88-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;CD4 T cells&quot;</span>        <span class="ot">=</span> <span class="st">&quot;#009E73&quot;</span>,  <span class="co"># bluish green</span></span>
<span id="cb88-8"><a href="weighted-analysis.html#cb88-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;NK cells&quot;</span>           <span class="ot">=</span> <span class="st">&quot;#F0E442&quot;</span>,  <span class="co"># yellow</span></span>
<span id="cb88-9"><a href="weighted-analysis.html#cb88-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;CD8 T cells&quot;</span>        <span class="ot">=</span> <span class="st">&quot;#0072B2&quot;</span>,  <span class="co"># blue</span></span>
<span id="cb88-10"><a href="weighted-analysis.html#cb88-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;FCGR3A+ Monocytes&quot;</span>  <span class="ot">=</span> <span class="st">&quot;#D55E00&quot;</span>,  <span class="co"># vermillion</span></span>
<span id="cb88-11"><a href="weighted-analysis.html#cb88-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Dendritic cells&quot;</span>    <span class="ot">=</span> <span class="st">&quot;#CC79A7&quot;</span>,  <span class="co"># reddish purple</span></span>
<span id="cb88-12"><a href="weighted-analysis.html#cb88-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;Megakaryocytes&quot;</span>     <span class="ot">=</span> <span class="st">&quot;#000000&quot;</span>   <span class="co"># black</span></span>
<span id="cb88-13"><a href="weighted-analysis.html#cb88-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb88-14"><a href="weighted-analysis.html#cb88-14" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">&#39;./data/&#39;</span>, proj_name, <span class="st">&#39;/metacell_&#39;</span>, MC_tool,<span class="st">&#39;.rds&#39;</span>))</span></code></pre></div>
</div>
<div id="dimensionality-reduction-1" class="section level3 hasAnchor" number="5.2.2">
<h3><span class="header-section-number">5.2.2</span> Dimensionality reduction<a href="weighted-analysis.html#dimensionality-reduction-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As for single-cells, we normalize the raw counts (here aggregated raw counts) and we identify the most variable features in the metacells gene expression data.
Based on these features, we run a sample weighted PCA using the function <code>supercell_prcomp</code> from the SuperCell R package
and use the first principal components to obtain a two dimensionnal representation of the data using UMAP.
Using the <code>supercell_DimPlot</code> function from the the SuperCell R package we can visualize the metacells and their sized in UMAP space.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="weighted-analysis.html#cb89-1" aria-hidden="true" tabindex="-1"></a>MC.seurat <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(MC.seurat, <span class="at">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>)</span>
<span id="cb89-2"><a href="weighted-analysis.html#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="weighted-analysis.html#cb89-3" aria-hidden="true" tabindex="-1"></a>MC_list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">N.SC =</span> <span class="fu">ncol</span>(MC.seurat),</span>
<span id="cb89-4"><a href="weighted-analysis.html#cb89-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">supercell_size =</span> MC.seurat<span class="sc">$</span>size)</span>
<span id="cb89-5"><a href="weighted-analysis.html#cb89-5" aria-hidden="true" tabindex="-1"></a>MC_list<span class="sc">$</span>PCA <span class="ot">&lt;-</span> SuperCell<span class="sc">::</span><span class="fu">supercell_prcomp</span>(</span>
<span id="cb89-6"><a href="weighted-analysis.html#cb89-6" aria-hidden="true" tabindex="-1"></a>  Matrix<span class="sc">::</span><span class="fu">t</span>(<span class="fu">GetAssayData</span>(MC.seurat, <span class="at">slot =</span> <span class="st">&quot;data&quot;</span>)),</span>
<span id="cb89-7"><a href="weighted-analysis.html#cb89-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">genes.use =</span> MC.seurat<span class="sc">@</span>misc<span class="sc">$</span>var_features,  <span class="co"># or a new set of HVG can be computed</span></span>
<span id="cb89-8"><a href="weighted-analysis.html#cb89-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">supercell_size =</span> MC_list<span class="sc">$</span>supercell_size, <span class="co"># provide this parameter to run sample-weighted version of PCA,</span></span>
<span id="cb89-9"><a href="weighted-analysis.html#cb89-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">k =</span> <span class="dv">30</span></span>
<span id="cb89-10"><a href="weighted-analysis.html#cb89-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb89-11"><a href="weighted-analysis.html#cb89-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-12"><a href="weighted-analysis.html#cb89-12" aria-hidden="true" tabindex="-1"></a>MC_list<span class="sc">$</span>UMAP <span class="ot">&lt;-</span> <span class="fu">supercell_UMAP</span>(</span>
<span id="cb89-13"><a href="weighted-analysis.html#cb89-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">SC =</span> MC_list,</span>
<span id="cb89-14"><a href="weighted-analysis.html#cb89-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">PCA_name =</span> <span class="st">&quot;PCA&quot;</span>,</span>
<span id="cb89-15"><a href="weighted-analysis.html#cb89-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">n.comp =</span> <span class="dv">30</span>, <span class="at">n_neighbors =</span> <span class="dv">15</span></span>
<span id="cb89-16"><a href="weighted-analysis.html#cb89-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb89-17"><a href="weighted-analysis.html#cb89-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-18"><a href="weighted-analysis.html#cb89-18" aria-hidden="true" tabindex="-1"></a><span class="fu">supercell_DimPlot</span>(<span class="at">SC =</span> MC_list, </span>
<span id="cb89-19"><a href="weighted-analysis.html#cb89-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">groups =</span> MC.seurat<span class="sc">@</span>meta.data[, annotation_column],</span>
<span id="cb89-20"><a href="weighted-analysis.html#cb89-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.name =</span> <span class="st">&quot;UMAP&quot;</span>, </span>
<span id="cb89-21"><a href="weighted-analysis.html#cb89-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">&quot;UMAP of metacells colored by cell type assignment&quot;</span>), <span class="at">color.use =</span> celltype_colors</span>
<span id="cb89-22"><a href="weighted-analysis.html#cb89-22" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-dim-reduc-weighted-1.png" width="672" /></p>
</div>
<div id="clustering-1" class="section level3 hasAnchor" number="5.2.3">
<h3><span class="header-section-number">5.2.3</span> Clustering<a href="weighted-analysis.html#clustering-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We cluster the metacells using the function <code>supercell_cluster</code> from SuperCell R package to perform the clustering step and visualize these clusters in the UMAP space:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="weighted-analysis.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute distance among metacells</span></span>
<span id="cb90-2"><a href="weighted-analysis.html#cb90-2" aria-hidden="true" tabindex="-1"></a>D  <span class="ot">&lt;-</span> <span class="fu">dist</span>(MC_list<span class="sc">$</span>PCA<span class="sc">$</span>x)</span>
<span id="cb90-3"><a href="weighted-analysis.html#cb90-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-4"><a href="weighted-analysis.html#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co"># cluster metacells</span></span>
<span id="cb90-5"><a href="weighted-analysis.html#cb90-5" aria-hidden="true" tabindex="-1"></a>MC_list<span class="sc">$</span>SCclustering  <span class="ot">&lt;-</span> <span class="fu">supercell_cluster</span>(<span class="at">D =</span> D, <span class="at">k =</span> <span class="dv">8</span>, <span class="at">supercell_size =</span> MC_list<span class="sc">$</span>supercell_size)</span>
<span id="cb90-6"><a href="weighted-analysis.html#cb90-6" aria-hidden="true" tabindex="-1"></a>MC.seurat<span class="sc">$</span>SCclustering <span class="ot">&lt;-</span> MC_list<span class="sc">$</span>SCclustering<span class="sc">$</span>clustering</span>
<span id="cb90-7"><a href="weighted-analysis.html#cb90-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-8"><a href="weighted-analysis.html#cb90-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot clustering result</span></span>
<span id="cb90-9"><a href="weighted-analysis.html#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="fu">supercell_DimPlot</span>(</span>
<span id="cb90-10"><a href="weighted-analysis.html#cb90-10" aria-hidden="true" tabindex="-1"></a>  MC_list, </span>
<span id="cb90-11"><a href="weighted-analysis.html#cb90-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">groups =</span> <span class="fu">factor</span>(MC_list<span class="sc">$</span>SCclustering<span class="sc">$</span>clustering),</span>
<span id="cb90-12"><a href="weighted-analysis.html#cb90-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">dim.name =</span> <span class="st">&quot;UMAP&quot;</span>, </span>
<span id="cb90-13"><a href="weighted-analysis.html#cb90-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">&quot;UMAP of metacells colored by metacell clustering&quot;</span>)</span>
<span id="cb90-14"><a href="weighted-analysis.html#cb90-14" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-clustering-weighted-1.png" width="672" /></p>
</div>
<div id="differential-expression-analysis-1" class="section level3 hasAnchor" number="5.2.4">
<h3><span class="header-section-number">5.2.4</span> Differential expression analysis<a href="weighted-analysis.html#differential-expression-analysis-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We perform diffrential analysis to identify the markers of our clusters using the <code>supercell_FindAllMarkers</code> function from the SuperCell package.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="weighted-analysis.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute upregulated genes in each cell line (versus other cells)</span></span>
<span id="cb91-2"><a href="weighted-analysis.html#cb91-2" aria-hidden="true" tabindex="-1"></a>MC.all.markers <span class="ot">&lt;-</span> <span class="fu">supercell_FindAllMarkers</span>(</span>
<span id="cb91-3"><a href="weighted-analysis.html#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">ge =</span> <span class="fu">GetAssayData</span>(MC.seurat, <span class="at">slot =</span> <span class="st">&quot;data&quot;</span>), </span>
<span id="cb91-4"><a href="weighted-analysis.html#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">clusters =</span> MC_list<span class="sc">$</span>SCclustering<span class="sc">$</span>clustering, </span>
<span id="cb91-5"><a href="weighted-analysis.html#cb91-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">supercell_size =</span> MC_list<span class="sc">$</span>supercell_size,</span>
<span id="cb91-6"><a href="weighted-analysis.html#cb91-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">only.pos =</span> <span class="cn">TRUE</span>, </span>
<span id="cb91-7"><a href="weighted-analysis.html#cb91-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.pct =</span> <span class="dv">0</span>, </span>
<span id="cb91-8"><a href="weighted-analysis.html#cb91-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">logfc.threshold =</span> <span class="fl">0.2</span></span>
<span id="cb91-9"><a href="weighted-analysis.html#cb91-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>We select the markers for cluster 4:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="weighted-analysis.html#cb92-1" aria-hidden="true" tabindex="-1"></a>cluster4_markers <span class="ot">&lt;-</span> MC.all.markers[[<span class="dv">4</span>]]</span>
<span id="cb92-2"><a href="weighted-analysis.html#cb92-2" aria-hidden="true" tabindex="-1"></a>MC.top.markers <span class="ot">&lt;-</span> cluster4_markers[<span class="fu">order</span>(cluster4_markers<span class="sc">$</span>logFC, <span class="at">decreasing =</span> T),]</span></code></pre></div>
<p>We visualize the top 5 markers for the cluster 4 and see that the top marker genes for this cluster contain marker genes of natural killer cells such as GZMB and GNLY.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="weighted-analysis.html#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(MC.seurat) <span class="ot">&lt;-</span> <span class="st">&quot;SCclustering&quot;</span></span>
<span id="cb93-2"><a href="weighted-analysis.html#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="co"># genes.to.plot &lt;- MC.seurat.top.markers$gene[MC.seurat.top.markers$cluster == unique(MC.seurat@meta.data[,annotation_column])[1]]</span></span>
<span id="cb93-3"><a href="weighted-analysis.html#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="co"># genes.to.plot &lt;- MC.top.markers$gene[c(seq(1, 20, 5))]</span></span>
<span id="cb93-4"><a href="weighted-analysis.html#cb93-4" aria-hidden="true" tabindex="-1"></a>genes.to.plot <span class="ot">&lt;-</span> <span class="fu">rownames</span>(MC.top.markers)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb93-5"><a href="weighted-analysis.html#cb93-5" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(MC.seurat, <span class="at">features =</span> genes.to.plot, <span class="at">ncol =</span> <span class="dv">5</span>, <span class="at">pt.size =</span> <span class="fl">0.0</span>)  </span>
<span id="cb93-6"><a href="weighted-analysis.html#cb93-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Warning: Groups with fewer than two data points have been dropped.</span></span>
<span id="cb93-7"><a href="weighted-analysis.html#cb93-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span id="cb93-8"><a href="weighted-analysis.html#cb93-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span id="cb93-9"><a href="weighted-analysis.html#cb93-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span>
<span id="cb93-10"><a href="weighted-analysis.html#cb93-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Groups with fewer than two data points have been dropped.</span></span></code></pre></div>
<p><img src="30-MC_analysis_discrete_files/figure-html/r-mc-plot-genes-weighted-1.png" width="672" /></p>

</div>
</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="standard-analysis-R.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="integration.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/edit/master/30-MC_analysis_discrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/GfellerLab/Metacell_tutorial/blob/master/30-MC_analysis_discrete.Rmd",
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
